{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78a56982",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import shutil\n",
    "import importlib\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "from skimage import io\n",
    "import tifffile\n",
    "from scipy.ndimage import median_filter\n",
    "from skimage.util import img_as_ubyte,  img_as_float\n",
    "import skimage\n",
    "from skimage.feature import blob_dog, blob_log, blob_doh\n",
    "from math import sqrt\n",
    "import scipy\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from skimage.filters import unsharp_mask\n",
    "from skimage.restoration import (denoise_tv_chambolle, denoise_bilateral,\n",
    "                                 denoise_wavelet, estimate_sigma)\n",
    "from skimage import color, morphology\n",
    "from skimage.transform import rescale\n",
    "import matplotlib as mpl\n",
    "import lifelines\n",
    "from lifelines import KaplanMeierFitter, CoxPHFitter\n",
    "from lifelines.statistics import multivariate_logrank_test\n",
    "\n",
    "from statannotations.Annotator import Annotator\n",
    "from itertools import combinations\n",
    "import statsmodels\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning) \n",
    "from bokeh.palettes import Colorblind\n",
    "d_colorblind = {'Liver':'#d55e00','Lung':'#0072b2',\n",
    "               'high':'#e69f00','low': '#56b4e9',\n",
    "               'basal-like':'#000000','classical':'#cc79a7',\n",
    "               'high pSUB': '#f0e442','low pSUB':'#d55e00'}\n",
    "# Set Paths\n",
    "codedir = os.getcwd()\n",
    "import util\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fb9d6a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#images: dowload from synapse.org syn51068458\n",
    "# (free account required)\n",
    "rootdir = codedir#'/home/groups/BCC_Chin_Lab/ChinData/Cyclic_Workflow/cmIF_2021-05-03_PDAC'\n",
    "regdir = f'{rootdir}/RegisteredImages'\n",
    "segdir = f'{codedir}/Segmentation'\n",
    "segdiro = f'{rootdir}/Segmentation'\n",
    "\n",
    "# clone mplex_image at https://gitlab.com/engje/mplex_image\n",
    "os.chdir('../..')\n",
    "from mplex_image import preprocess, mpimage #, cmif\n",
    "os.chdir(codedir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ddd69b4",
   "metadata": {},
   "source": [
    "# Table of contents <a name=\"contents\"></a>\n",
    "0. [functions](#func)\n",
    "1. [skimage blobs](#sk)\n",
    "2. [Foci analysis](#focifoci)\n",
    "3. [mIHC analysis](#mihc)\n",
    "4. [Patient metadata](#meta)\n",
    "4. [CPH modeling](#clin)\n",
    "5. [gene expression analysis](#geneexp)\n",
    "6. [TCR analysis](#tcell)\n",
    "7. [GSEA plots](#split)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f1ef3f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import importlib\n",
    "# importlib.reload(util)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "604aaac5",
   "metadata": {},
   "source": [
    "## functions <a name=\"func\"></a> \n",
    "\n",
    "import from util.py\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ba7147b",
   "metadata": {},
   "source": [
    "##  Section 1: skimage blob detection\n",
    "\n",
    "Foci are detected here from images plus segmentation masks\n",
    "\n",
    "**You can skip and use pre-computed foci counts**\n",
    "\n",
    " <a name=\"sk\"></a> \n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f737cb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#intensity data: dowload from synapse.org syn51068458\n",
    "df_mi = pd.read_csv(f'{codedir}/data/20220720_U54-TMA_FilteredMeanIntensity_Link.csv',index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0f1a0ba",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#skimage blob detection\n",
    "#requires images: dowload from synapse.org syn51068458\n",
    "threshold=0.002\n",
    "d_thresh = {'pRPA':1100, 'gH2AX':1100, 'RAD51':1300}\n",
    "os.chdir(regdir)\n",
    "ls_slide = sorted(set(df_mi.slide_scene)) \n",
    "ls_marker = ['pRPA','gH2AX','RAD51',] \n",
    "df_result_all = pd.DataFrame()\n",
    "for s_slide in ls_slide:\n",
    "    print(s_slide)\n",
    "    os.chdir(f'{segdiro}/U54-TMA-9_CellposeSegmentation') #change this to path for your downloaded segmentation masks\n",
    "    s_seg = pd.Series(sorted(os.listdir()),dtype='object')[pd.Series(index=sorted(os.listdir()),dtype='object').index.str.contains(f'{s_slide}_nuc30_NucleiSegmentationBasins')].iloc[0]\n",
    "    label_image = io.imread(s_seg)\n",
    "    os.chdir(f'{regdir}/{s_slide}') # change this to path to your downloaded registered tifs\n",
    "    df_img=mpimage.parse_org()\n",
    "    for idxs, s_marker in enumerate(ls_marker):\n",
    "        intensity_image = io.imread(df_img[df_img.marker==s_marker].index[0])\n",
    "        props = skimage.measure.regionprops_table(label_image, intensity_image=intensity_image, properties=('label','bbox','mean_intensity')) # 'image','intensity_image',\n",
    "        df_props = pd.DataFrame(props,dtype='float').set_index('label')\n",
    "        df_props.columns = [item.replace('-','') for item in df_props.columns]\n",
    "        ls_index = df_props[df_props.mean_intensity>d_thresh[s_marker]].index\n",
    "        print(f'{s_marker} {len(ls_index)}')\n",
    "        for i_cell in ls_index:\n",
    "            se_cell = df_props.loc[i_cell].dropna().astype('int')\n",
    "            image = intensity_image[se_cell.bbox0:se_cell.bbox2,se_cell.bbox1:se_cell.bbox3]\n",
    "            if s_marker == 'pRPA':\n",
    "                blobs, fig = util.get_blobs2(image,min_sigma=0.1,max_sigma=2,threshold=threshold,exclude_border=1)\n",
    "            elif s_marker == 'gH2AX':\n",
    "                blobs, fig = util.get_blobs2(image,min_sigma=1,max_sigma=2,threshold=threshold,exclude_border=1)\n",
    "            else:\n",
    "                blobs, fig = util.get_blobs2(image,min_sigma=0.1,max_sigma=2,threshold=threshold,exclude_border=1)\n",
    "            df_props.loc[i_cell,'blobs'] = len(blobs)\n",
    "            if len(blobs) > 1:\n",
    "                fig.suptitle(s_marker)\n",
    "                plt.tight_layout()\n",
    "                fig.savefig(f'{codedir}/blobs/{s_marker}/{s_slide}_{s_marker}_{i_cell}.png',dpi=200)\n",
    "                plt.close(fig)\n",
    "        df_props.index = [s_slide + '_cell' + str(item) for item in df_props.index]\n",
    "        if len(ls_index)==0:\n",
    "                df_props['blobs'] = np.NaN\n",
    "        if idxs == 0:\n",
    "            df_result = df_props.rename({'blobs':f'{s_marker}_foci'},axis=1)\n",
    "        else:\n",
    "            df_result.loc[df_props.index,f'{s_marker}_foci'] = df_props.blobs\n",
    "    df_result_all = df_result_all.append(df_result)\n",
    "\n",
    "#uncomment to save (this data has been pre-computed and saved in the repo)\n",
    "#df_result_all.loc[:,df_result_all.columns.str.contains('foci')].dropna(how='all').to_csv(f'{codedir}/foci_U54-TMA-9_{\".\".join(ls_marker)}_{threshold}.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b487760",
   "metadata": {},
   "source": [
    "### Section 2: FOCI Analysis <a name=\"focifoci\"></a>\n",
    "\n",
    "Load saved foci for plotting and downstream analysis\n",
    "\n",
    "liver/lung\n",
    "\n",
    "**You don't need to run section 1 to run this**\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1de50de",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_surv = pd.read_csv(f'{codedir}/data/u54_tma_sampleannot_Link.csv')\n",
    "len(df_surv)\n",
    "#df_pORG = pd.read_csv(f'{codedir}/data/GSVA_Scores_Link.csv')#=,index_col='Public_Specimen_ID'\n",
    "#df_pORG = pd.read_csv(f'{codedir}/20230526_GSVA_Scores.csv')\n",
    "df_pORG = pd.read_csv(f'data/20230608_GSVA_Scores.csv')\n",
    "#df_pORG\n",
    "df_primary = df_surv[((df_surv.Tissue=='PDAC') | (df_surv.Tissue=='Intestinal'))].copy()\n",
    "print(len(df_primary))\n",
    "df_primary = df_primary.merge(df_pORG[df_pORG.Group=='GSVA_All'],on='Public_Specimen_ID',how='left')\n",
    "df_primary.set_index(\"Public_Specimen_ID\", inplace = True)\n",
    "len(df_primary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d508cacc",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mapper = df_primary[~df_primary.index.duplicated()]\n",
    "ls_add = ['txi_pORG_Up_42_Genes',\n",
    "          'trim_padj_0.2_pORG_Up_55_Genes',\n",
    "          'trim_padj_0.2_pSUB_Up_100_Genes',\n",
    "        'txi_pSUB_Up_100_Genes']\n",
    "for s_add in ls_add:\n",
    "    d_map = dict(zip(df_mapper.index,df_mapper.loc[:,s_add]))\n",
    "    print(len(d_map))\n",
    "    df_surv[s_add] = df_surv.Public_Specimen_ID.map(d_map)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e3ec3a5",
   "metadata": {},
   "source": [
    "### TMA survival"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59b765f6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "##\n",
    "%matplotlib inline\n",
    "alpha = 0.06\n",
    "s_propo = ''\n",
    "savedir = f'{codedir}'\n",
    "s_time=  'Survival_time'#\n",
    "s_censor='Survival'#\n",
    "s_subtype = ''\n",
    "s_cell = ''\n",
    "s_type_title = ''\n",
    " \n",
    "for s_col in ls_add:\n",
    "    print(s_col)\n",
    "    for cutp in [0.33,0.5,0.66]: #np.round(np.arange(0.25,1,0.25),3):#\n",
    "            print(cutp)\n",
    "            df_km, pvalue = util.single_km(df_surv[~df_surv.Public_Patient_ID.duplicated()],s_cell,s_subtype,s_type_title,s_col,savedir,alpha,cutp, #\n",
    "                                       s_time,s_censor,s_propo)\n",
    "            print(pvalue)\n",
    "    #break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e4366b4",
   "metadata": {},
   "source": [
    "### define high/low pORG\n",
    "\n",
    "based on survival differences on TMA samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7916167",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pri = df_surv[~(df_surv.Public_Patient_ID.duplicated(keep='first')) & ((df_surv.Tissue=='PDAC'))]\n",
    "s_select_porg = 'trim_padj_0.2_pORG_Up_55_Genes'#'txi_pORG_Up_42_Genes'\n",
    "i_pORG_txi = np.quantile(df_pri.loc[:,s_select_porg].dropna(),0.66)#0.66 #0.5\n",
    "print(i_pORG_txi)\n",
    "# df_surv.loc[df_surv['Original_pORG_Up_78_Genes'] >= i_pORG_ori,'pORG_binary_orig'] = 'high'\n",
    "# df_surv.loc[df_surv['Original_pORG_Up_78_Genes'] < i_pORG_ori,'pORG_binary_orig'] = 'low'\n",
    "\n",
    "df_surv.loc[df_surv[s_select_porg] >= i_pORG_txi,'pORG_binary'] = 'high'\n",
    "df_surv.loc[df_surv[s_select_porg] < i_pORG_txi,'pORG_binary'] = 'low'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bff736f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save the 0608\n",
    "#df_surv.drop(['Unnamed: 0','pORG_binary','pORG_Score'],axis=1).to_csv(f'{codedir}/data/u54_tma_sampleannot_Link_new.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "192637e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(df_surv.pORG_binary_orig.value_counts())\n",
    "print(df_surv[(df_surv.Tissue=='PDAC') & (~df_surv.Public_Specimen_ID.duplicated())].pORG_binary.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "899c9d63",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load combined\n",
    "s_sample = '20220721_U54-TMA'#'20220711_U54-TMA' #'20220409_JP-TMAs_IMC-TMAs'\n",
    "s_names = 'Combined' #'unnamed' #\n",
    "s_type = 'PDAC'\n",
    "df_lei = pd.read_csv(f'{codedir}/data/{s_sample}_{s_names}Celltypes_{s_type}_Link.csv',index_col=0)\n",
    "\n",
    "d_patient = dict(zip(df_surv.coor_mplexable,df_surv.Public_Patient_ID))\n",
    "\n",
    "df_lei['Tissue'] = df_lei.Patient.map(dict(zip(df_surv.Public_Patient_ID,df_surv.Tissue)))\n",
    "df_epi = pd.read_csv(f'{codedir}/data/results_20220721_U54-TMA_CellTypeCounts_byPatient_byleidencelltype5_PDAC_Link.csv',index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f19c9989",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ls_marker = ['gH2AX','pRPA','RAD51']\n",
    "threshold=0.002\n",
    "df_foci2 = pd.read_csv(f'{codedir}/data/foci_U54-TMA-9_{\".\".join(ls_marker)}_{threshold}.csv',index_col=0)\n",
    "df_foci2.index = [item.replace('.0','') for item in df_foci2.index]\n",
    "df_foci2 = df_foci2.fillna(0)\n",
    "df_foci2['scene'] = [item.split('_cell')[0] for item in df_foci2.index]\n",
    "df_foci2['Patient'] = df_foci2.scene.map(d_patient)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68276502",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot Ki67 versus pRPA foci\n",
    "s_define = 'pORG_binary'#'pORG_binary_txi'#'pORG_binary_orig'#\n",
    "df_lei['Cohort'] = df_lei.Patient.map(dict(zip(df_surv.Public_Patient_ID,df_surv.Cohort)))\n",
    "df_lei['pORG_binary'] = df_lei.Patient.map(dict(zip(df_surv.Public_Patient_ID,df_surv.loc[:,s_define])))\n",
    "df_lei['Ki67pos'] = np.nan\n",
    "df_lei.loc[df_lei.Ki67>3*256,'Ki67pos'] = 'Ki67+'\n",
    "df_lei.Ki67pos.fillna('Ki67-',inplace=True)\n",
    "ls_foci =['pRPA_foci','gH2AX_foci','RAD51_foci']\n",
    "df_lei_foci = df_lei.merge(df_foci2.loc[:,ls_foci],left_index=True,right_index=True,how='left')\n",
    "df_lei_foci.loc[:,ls_foci] = df_lei_foci.loc[:,ls_foci].fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1274b7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#use df foci sum 2 (lower threshold)\n",
    "ls_index = df_lei[df_lei.leidencelltype5=='epithelial'].index\n",
    "df_foci_sum2 = df_foci2.loc[df_foci2.index.isin(ls_index)].groupby('Patient').sum()\n",
    "for s_marker in ls_marker:\n",
    "    df_foci_sum2[f'log_{s_marker}_foci'] = np.log(df_foci_sum2.loc[:,f'{s_marker}_foci'] + 1)\n",
    "    df_foci_sum2[f'mean_{s_marker}_foci'] = (df_foci_sum2.loc[df_foci_sum2.index.isin(df_epi.index),f'{s_marker}_foci']/df_epi.epithelial).fillna(0)\n",
    "\n",
    "    \n",
    "df_foci_sum2['Public_Patient_ID'] = df_foci_sum2.index\n",
    "df_surv = df_surv.merge(df_foci_sum2,on='Public_Patient_ID',how='left',suffixes=('_1',''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d32c6ef",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "s_group = 'pORG_binary'\n",
    "s_type = ''\n",
    "s_cell = 'epithelial'\n",
    "for s_marker in ['mean_pRPA_foci','mean_RAD51_foci','mean_gH2AX_foci']:\n",
    "    util.categorical_correlation_boxplot(df_surv[~df_surv.Public_Patient_ID.duplicated()],s_group,s_marker,s_type,s_cell,\n",
    "                                         alpha=1.05,s_propo='in',b_ttest=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05d2fab1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#compare Ki67+ versus negative, double violinplots\n",
    "#compare all the combinations, double violinplots\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from statannotations.Annotator import Annotator\n",
    "from itertools import combinations\n",
    "import statsmodels\n",
    "from statsmodels.stats.multicomp import pairwise_tukeyhsd\n",
    "\n",
    "s_compare = 'Ki67'#'all' #\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning) \n",
    "for s_cat in ['pORG_binary','Cohort']:\n",
    "    ls_order = ['Ki67-','Ki67+']\n",
    "    for s_foci in ls_foci:\n",
    "        if s_compare == 'all':\n",
    "            figsize=(3.5,3)\n",
    "        else:\n",
    "            figsize=(2.8,2.8)\n",
    "        fig,ax=plt.subplots(dpi=300,figsize=figsize)\n",
    "        df_both = pd.DataFrame()\n",
    "        order = []\n",
    "        d_pval = {}\n",
    "        for idx, s_cohort in enumerate(df_lei_foci.loc[:,s_cat].dropna().unique()):\n",
    "            print(s_cohort)\n",
    "            df_plot_foci =df_lei_foci[(df_lei_foci.leidencelltype5=='epithelial')\n",
    "                                      & (df_lei_foci.Tissue=='PDAC') & (df_lei_foci.loc[:,s_cat]==s_cohort)]\n",
    "            statistic, pvalue = stats.ttest_ind(df_plot_foci.loc[df_plot_foci.Ki67pos=='Ki67+',s_foci],\n",
    "                                                df_plot_foci.loc[df_plot_foci.Ki67pos=='Ki67-',s_foci],\n",
    "                                               alternative='greater')\n",
    "            for s_order in ls_order:\n",
    "                order.append((s_cohort,s_order))\n",
    "            df_both = pd.concat([df_both,df_plot_foci])\n",
    "            d_pval.update({s_cohort:pvalue})\n",
    "        sns.violinplot(data=df_both,hue='Ki67pos',y=s_foci,x=s_cat,ax=ax,alpha=0.5,linewidth=0.5)\n",
    "        sns.stripplot(data=df_both,hue='Ki67pos',y=s_foci,x=s_cat,s=1,dodge=True,ax=ax,palette='dark',jitter=0.2)\n",
    "        #annotate\n",
    "        if s_compare == 'all':\n",
    "            pairs = list(combinations(order, r=2))\n",
    "            annotator = Annotator(ax, pairs=pairs, data=df_both, y=s_foci,x=s_cat,hue='Ki67pos')\n",
    "            annotator.configure(test=\"t-test_ind\",line_width=1)#,alternative='greater\n",
    "            pvalues = annotator.apply_test().annotations #annotator.apply_and_annotate() # \n",
    "            pvalues = [item.data.pvalue for item in pvalues]     \n",
    "            reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='bonferroni')\n",
    "            formatted_pvalues = [f'p={pvalue:.2}' for pvalue in list(corrected)]\n",
    "            annotator.set_custom_annotations(formatted_pvalues)\n",
    "            annotator.annotate()\n",
    "        else:\n",
    "            pairs = [(order[0],order[1]),(order[2],order[3])]\n",
    "            pvalues = [d_pval[pairs[0][0][0]],d_pval[pairs[1][0][0]]]\n",
    "            reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='bonferroni')\n",
    "            formatted_pvalues = [f'p={pvalue:.2}' for pvalue in list(corrected)]\n",
    "            annotator = Annotator(ax, pairs=pairs, data=df_both, y=s_foci,x=s_cat,hue='Ki67pos')\n",
    "            annotator.set_custom_annotations(formatted_pvalues)\n",
    "            annotator.annotate()\n",
    "        \n",
    "        ax.set_title(f\"{s_foci.replace('_',' ')} vs. {s_cat.split('_')[0]}\", fontsize='x-large')\n",
    "        ax.set_xlabel(s_cat)\n",
    "        ax.set_ylabel(f\"No. {s_foci.replace('_',' ')}\")\n",
    "        h, l = ax.get_legend_handles_labels()\n",
    "        labels =  [f'_{item}' if ind < 2 else item for ind,item in enumerate(l)]\n",
    "        ax.legend(h,labels,title='',fontsize='small',markerscale=.5,bbox_to_anchor=(1.01,0.9))\n",
    "        plt.tight_layout()\n",
    "        fig.savefig(f'{codedir}/violinplot_both_{s_foci}_{s_cohort}_{s_compare}.png')\n",
    "    if s_cat == 'pORG_binary':\n",
    "        df_both.to_csv(f'results_foci_Ki67.csv')\n",
    "        #break\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f068f888",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## make the bargraphs\n",
    "# percent pRPA epithelial cells\n",
    "\n",
    "d_cats = {'pORG_binary':['high','low'],'Cohort':['liver_cohort','lung_cohort']}\n",
    "for s_cat, ls_order in d_cats.items():\n",
    "    for s_foci in ls_foci:\n",
    "        figsize=(2.8,2.5)\n",
    "        fig,ax=plt.subplots(dpi=300,figsize=figsize)\n",
    "        df_both = pd.DataFrame()\n",
    "        d_pval = {}\n",
    "        for idx, s_cohort in enumerate(df_lei_foci.loc[:,s_cat].dropna().unique()):\n",
    "            print(s_cohort)\n",
    "            df_plot_foci =df_lei_foci[(df_lei_foci.leidencelltype5=='epithelial')\n",
    "                                      & (df_lei_foci.Tissue=='PDAC') & (df_lei_foci.loc[:,s_cat]==s_cohort)]\n",
    "            df_both = pd.concat([df_both,df_plot_foci])\n",
    "        df_obs = df_both[df_both.loc[:,s_foci] > 0].loc[:,s_cat]\n",
    "        df_exp = df_both.loc[:,s_cat]\n",
    "        #chi\n",
    "        f_obs = df_obs.value_counts().loc[ls_order]\n",
    "        f_exp = f_obs.sum() * df_exp.value_counts(normalize=True).loc[ls_order]\n",
    "        statistic, pvalue =  stats.chisquare(f_obs, f_exp)\n",
    "        df_obs.value_counts(normalize=True).loc[ls_order].plot(kind='bar',ax=ax,color=['mediumpurple','deepskyblue'])\n",
    "        ax.set_title(f'{s_foci} vs. {s_cat}\\nChi-squared p={pvalue:.3}')\n",
    "        ax.set_ylabel(f'Percent of {s_foci.split(\"_\")[0]}+ Epithelial')\n",
    "        plt.tight_layout()\n",
    "        xtickslocs = ax.get_xticks()\n",
    "        for idx, s_order in enumerate(ls_order):\n",
    "            s = f'{df_obs.value_counts().loc[s_order]}\\nof\\n{df_exp.value_counts().loc[s_order]}'\n",
    "            plt.text(-0.2+idx,0.3,s)\n",
    "        break\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc8f635c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## make the bargraphs\n",
    "# percent pRPA epithelial cells\n",
    "# percent of those that are Ki67+\n",
    "d_cats = {'pORG_binary':['high','low'],'Cohort':['liver_cohort','lung_cohort']}\n",
    "for s_cat, ls_order in d_cats.items():\n",
    "    for s_foci in ls_foci:\n",
    "        figsize=(2.8,2.5)\n",
    "        fig,ax=plt.subplots(dpi=300,figsize=figsize)\n",
    "        df_both = pd.DataFrame()\n",
    "        d_pval = {}\n",
    "        for idx, s_cohort in enumerate(df_lei_foci.loc[:,s_cat].dropna().unique()):\n",
    "            print(s_cohort)\n",
    "            df_plot_foci =df_lei_foci[(df_lei_foci.leidencelltype5=='epithelial') & (df_lei_foci.loc[:,s_foci] > 0)\n",
    "                                      & (df_lei_foci.Tissue=='PDAC') & (df_lei_foci.loc[:,s_cat]==s_cohort)]\n",
    "            df_both = pd.concat([df_both,df_plot_foci])\n",
    "        df_obs = df_both[df_both.loc[:,'Ki67pos']=='Ki67+'].loc[:,s_cat]\n",
    "        df_exp = df_both.loc[:,s_cat]\n",
    "        #chi\n",
    "        f_obs = df_obs.value_counts().loc[ls_order]\n",
    "        f_exp = f_obs.sum() * df_exp.value_counts(normalize=True).loc[ls_order]\n",
    "        statistic, pvalue =  stats.chisquare(f_obs, f_exp)\n",
    "        df_plot = df_obs.value_counts()/df_exp.value_counts()\n",
    "        df_plot.loc[ls_order].plot(kind='bar',ax=ax,color=['mediumpurple','deepskyblue'])\n",
    "        ax.set_title(f'{s_foci} vs. {s_cat}\\nChi-squared p={pvalue:.3}')\n",
    "        ax.set_ylabel(f'Percent of {s_foci.split(\"_\")[0]}+ Epithelial\\nthat are Proliferating')\n",
    "        plt.tight_layout()\n",
    "        xtickslocs = ax.get_xticks()\n",
    "        for idx, s_order in enumerate(ls_order):\n",
    "            s = f'{df_obs.value_counts().loc[s_order]}\\nof\\n{df_exp.value_counts().loc[s_order]}'\n",
    "            plt.text(-0.2+idx,0.15,s)\n",
    "        #break\n",
    "    #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe760931",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## make the boxplots\n",
    "# percent pRPA epithelial cells\n",
    "# percent of those that are Ki67+\n",
    "d_cats = {'pORG_binary':['high','low'],'Cohort':['liver_cohort','lung_cohort']}\n",
    "for s_cat, ls_order in d_cats.items():\n",
    "    for s_foci in ls_foci:\n",
    "        figsize=(2.8,2.5)\n",
    "        #fig,ax=plt.subplots(dpi=300,figsize=figsize)\n",
    "        df_both = pd.DataFrame()\n",
    "        d_pval = {}\n",
    "        for idx, s_cohort in enumerate(df_lei_foci.loc[:,s_cat].dropna().unique()):\n",
    "            print(s_cohort)\n",
    "            df_plot_foci =df_lei_foci[(df_lei_foci.leidencelltype5=='epithelial') & (df_lei_foci.loc[:,s_foci] > 0)\n",
    "                                      & (df_lei_foci.Tissue=='PDAC') & (df_lei_foci.loc[:,s_cat]==s_cohort)]\n",
    "            df_both = pd.concat([df_both,df_plot_foci])\n",
    "        df_sum = df_both.loc[:,[s_cat,'Patient','Ki67pos']].groupby('Patient').value_counts().unstack().reset_index().fillna(0)\n",
    "        df_sum['Frac. Prolif.'] = df_sum.loc[:,'Ki67+']/(df_sum.loc[:,['Ki67+','Ki67-']].sum(axis=1))\n",
    "        util.categorical_correlation_boxplot(df_sum,s_cat,'Frac. Prolif.',s_type='',s_cell=f\"{s_foci.split('_')[0]}+\",\n",
    "                                         alpha=1.05,s_propo='in',b_ttest=True)\n",
    "        #break\n",
    "    #break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be8bfceb",
   "metadata": {},
   "source": [
    "### Section 3: mIHC Analysis <a name=\"mihc\"></a>\n",
    "\n",
    "Stacked barplot\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "693d274e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#survival\n",
    "\n",
    "df_surv = pd.read_csv(f'{codedir}/data/u54_tma_sampleannot_Link_new.csv',index_col=0)\n",
    "df_surv['Sample_ID'] = [item.split('-T')[0][-6::] for item in df_surv.loc[:,'Public_Patient_ID'].fillna('none')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7ce65c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dowload mIHC data from https://www.synapse.org/#!Synapse:syn51078766\n",
    "# combine ROIs into large dataframe and save\n",
    "if not os.path.exists(f'data/20221123_mIHC_LiverLung_Celltypes.csv'):\n",
    "    ls_col = ['Sample_ID','class','Location_Center_X', 'Location_Center_Y', \n",
    "           'GRZB_func', 'KI67_func', 'PD1_func', 'PDL1_func', 'CD163_func',\n",
    "           'CCR2_func', 'HLAII_func', 'EOMES_func','Area']\n",
    "    df_ll =pd.DataFrame()\n",
    "    for s_file in sorted(os.listdir('mIHC_Data')):\n",
    "        df = pd.read_csv(f'mIHC_Data/{s_file}',index_col=0)\n",
    "        s_sample = s_file.split('LiverLungBCC')[1].split('.')[0]\n",
    "        df['Sample_ID'] = s_sample\n",
    "        df.index = [f'{s_sample.split(\"Nuclei_\")[1].replace(\"ROI\",\"_scene\")}_cell{item}' for item in df.index]\n",
    "        df_ll = pd.concat([df_ll,df.loc[:,ls_col]])\n",
    "        break\n",
    "    df_ll['Organ'] = df_ll.Sample_ID.map(dict(zip(df_ll_annot.index,df_ll_annot.loc[:,'Met Site'])))\n",
    "    df_ll['Location'] = df_ll.Sample_ID.map(dict(zip(df_ll_annot.index,df_ll_annot.loc[:,'Location'])))\n",
    "    df_ll['Desc'] = df_ll.Sample_ID.map(dict(zip(df_ll_annot.index,df_ll_annot.loc[:,'Desc'])))\n",
    "    df_ll.to_csv(f'data/20221123_mIHC_LiverLung_Celltypes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a6cbcba",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load data\n",
    "df_ll = pd.read_csv(f'data/20221123_mIHC_LiverLung_Celltypes.csv',index_col=0,low_memory=False)\n",
    "print(len(df_ll))\n",
    "\n",
    "df_ll_ann = pd.read_csv('data/LiverLung_annotations.csv',index_col=0)\n",
    "df_ll_roi = pd.read_csv('data/annotated_LiverLung_perROI.csv',index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db35cd8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add annotation\n",
    "df_ll['Sample_ID_short'] = [item.split('Nuclei_')[1].split('ROI')[0] for item in df_ll.Sample_ID]\n",
    "df_ll['Sample_ID_int'] = [int(item) for item in df_ll.Sample_ID_short]\n",
    "df_ll['ROI'] = [int(item.split('ROI')[1]) for item in df_ll.Sample_ID]\n",
    "df_ll['Cohort'] = df_ll.Sample_ID_int.map(dict(zip(df_ll_ann.index, df_ll_ann.Cohort)))\n",
    "df_ll['Patient'] = df_ll.Sample_ID_int.map(dict(zip(df_ll_ann.index, df_ll_ann.Patient)))\n",
    "df_ll['Sample_ROI'] = df_ll.Sample_ID_int.astype('str') + '_' + df_ll.ROI.astype('str')\n",
    "df_ll_roi['Sample_ROI'] = df_ll_roi.index.astype('str') + '_' + df_ll_roi.ROI.astype('str')\n",
    "df_ll['Location'] = df_ll.Sample_ROI.map(dict(zip(df_ll_roi.Sample_ROI,df_ll_roi.loc[:,'Location'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9a04a6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ll['classII'] = df_ll.loc[:,'class'].replace({'T-regulatory CD4 cells':'T cells',\n",
    "                                                 'CD4 T helper cells':'T cells','CD8 T cells':'T cells'}) #\n",
    "df_ll.rename({'class':'classI'},axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ab5c426",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_color = dict(zip(['ST-00016289', 'ST-00017078', 'ST-00017310', 'ST-00017381',\n",
    "       'ST-00018269', 'ST-00018955', 'ST-00019367', 'ST-00019368',\n",
    "       'ST-00020181'],sns.color_palette('Purples',9)))\n",
    "d_color.update(dict(zip(['ST-00015839', 'ST-00017440', 'ST-00017804'],sns.color_palette('Blues',3))))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a65fb8b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#by location\n",
    "s_define = 'trim_padj_0.2_pORG_Up_55_Genes'#'txi_pORG_Up_42_Genes'#''#\n",
    "s_group = 'Cohort'\n",
    "alpha = 1.05\n",
    "s_column = 'classI'#'classII'#\n",
    "if s_column == 'classI':\n",
    "    ls_mihc = [ 'PanCK+','aSMA+','CD4 T helper cells','CD8 T cells', 'T-regulatory CD4 cells','B cells', 'Granulocytes', 'Monocyte',\n",
    " 'Macrophage', 'Mature DC','Immature DC']\n",
    "elif s_column == 'classII':\n",
    "    ls_mihc = ['T cells', #'B cells', 'Granulocytes', 'Monocyte', 'Macrophage',\n",
    "       #'Mature DC', 'Immature DC', 'Immune Other', 'PanCK+', 'aSMA+','Other Cells'\n",
    "       ]\n",
    "s_patient = 'Public_Patient_ID'\n",
    "\n",
    "for s_loc in ['all','T','B','D']: \n",
    "    if s_loc == 'all':\n",
    "        df_loc = df_ll\n",
    "    else:\n",
    "        df_loc = df_ll[df_ll.Location==s_loc]\n",
    "    print(len(df_loc.Sample_ID_short.unique()))\n",
    "    df_group = (df_loc.groupby(['Patient',s_column]).count().Sample_ID/(df_loc.groupby(['Patient']).count().Sample_ID)).unstack()\n",
    "    df_group[s_patient] = df_group.index\n",
    "    df_group['Cohort'] = df_group.index.map(dict(zip(df_ll_roi.Patient,df_ll_roi.loc[:,'Cohort'])))\n",
    "    df_group = df_group.merge(df_surv.loc[:,['Survival','Survival_time',s_define,s_patient]],on=s_patient)\n",
    "    df_group = df_group[~df_group.loc[:,s_patient].duplicated()]\n",
    "    df_group = df_group[df_group.Cohort!='Liver met'].fillna(0)\n",
    "    ls_order = sorted(df_group.loc[:,s_group].unique())\n",
    "    for s_marker in ls_mihc:\n",
    "        try:\n",
    "            s_high = df_group.loc[:,s_group].unique()[0]\n",
    "            s_low = df_group.loc[:,s_group].unique()[1]\n",
    "        except:\n",
    "            continue\n",
    "        n_high = sum(df_group.loc[:,s_group]==s_high)\n",
    "        n_low = sum(df_group.loc[:,s_group]==s_low)\n",
    "        statistic,pvalue = stats.ttest_ind(df_group.loc[df_group.loc[:,s_group]==s_high,s_marker],\n",
    "                                           df_group.loc[df_group.loc[:,s_group]==s_low,s_marker])\n",
    "        if pvalue <= alpha:\n",
    "            df_group_roi = (df_loc.groupby(['Sample_ROI',s_column]).count().Sample_ID/(df_loc.groupby(['Sample_ROI']).count().Sample_ID)).unstack()\n",
    "            df_group_roi['Cohort'] = df_group_roi.index.map(dict(zip(df_ll_roi.Sample_ROI,df_ll_roi.loc[:,'Cohort'])))\n",
    "            df_group_roi['Patient'] = df_group_roi.index.map(dict(zip(df_ll_roi.Sample_ROI,df_ll_roi.loc[:,'Patient'])))\n",
    "            df_group_roi = df_group_roi[df_group_roi.Cohort.isin(['liver_cohort','lung_cohort'])]\n",
    "            fig, ax = plt.subplots(figsize=(3,3),dpi=300)\n",
    "            sns.boxplot(data=df_group_roi,x=s_group,y=s_marker,showfliers=False,ax=ax,order=[str(item) for item in ls_order],palette=['mediumpurple','deepskyblue'])\n",
    "            sns.stripplot(data=df_group_roi,x=s_group,y=s_marker,ax=ax,hue='Patient',s=3,palette=d_color)\n",
    "            ax.set_ylim(ax.get_ylim()[0],ax.get_ylim()[1])\n",
    "            ax.set_title(f'{s_group} versus\\n {s_marker}\\n p={pvalue:.4f} (n={n_low}, {n_high})')\n",
    "            ax.set_ylabel(f'{s_marker} in {s_loc}')\n",
    "            #ax.set_ylabel(f'{s_marker}')\n",
    "            ax.get_legend().remove()\n",
    "            plt.tight_layout()\n",
    "            #fig.savefig(f'{s_date}/boxplot_mIHC_{s_marker}_versus_{s_group}_in_{s_loc}.png')\n",
    "        break\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e71ca46a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "print(df_group.groupby('Cohort').mean().loc[:,s_define])\n",
    "print(df_group.groupby('Cohort').sem().loc[:,s_define])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f01e0a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax=plt.subplots(dpi=300)\n",
    "df_group.groupby('Cohort').mean().loc[:,ls_mihc].plot(kind='bar',width=.9,stacked=True,ax=ax,colormap='Paired')\n",
    "ax.legend(bbox_to_anchor=(1,.9))\n",
    "ax.set_ylabel('Fraction in Tissue')\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d999eea3",
   "metadata": {},
   "source": [
    "# Section 4 <a name=\"meta\"></a>\n",
    "\n",
    "patient metadata\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f7649de",
   "metadata": {},
   "source": [
    "### Patients in LabKey"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63f160b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load patient vital status\n",
    "#also had stage/ grade/ lymph nodes etc.\n",
    "df_vital = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/FMP_Patients_Nov17_2021.xlsx',sheet_name='Sheet1')\n",
    "\n",
    "#Stage 1 pancreatic cancer means the cancer is not more than 4cm in size and it hasn't spread outside the pancreas. It is split into 1A and 1B.\n",
    "# 1A In TNM staging, this is the same as T1, N0, M0 1B In TNM staging, this is the same as T2, N0, M0\n",
    "\n",
    "# Stage 2A means the cancer is bigger than 4cm but is still within the pancreas. It has not spread to the lymph nodes Open a glossary item or other areas of the body.\n",
    "#2A in TNM staging, this is the same as T3, N0, M0. 2B In TNM staging, this is the same as T1, 2 or 3, N1, M0.\n",
    "\n",
    "#3 Stage 3 can mean that the cancer is any size within the pancreas and has spread to 4 or more nearby lymph nodes Open a glossary item.\n",
    "#In TNM staging, this is the same as T1, 2 or 3, N2, M0.\n",
    "#Or stage 3 can mean the cancer has started to grow outside the pancreas into the major blood vessels nearby. It may or may not have spread into the lymph nodes. It hasn't spread to any other areas of the body.\n",
    "#In TNM staging, this is the same as T4, Any N, M0.\n",
    "\n",
    "#4 Your doctor might call this advanced (metastatic) cancer.\n",
    "#In TNM staging, this is the same as Any T, Any N, M1.\n",
    "\n",
    "#collapse stage\n",
    "d_stage = {'2B - IIB':'II', 'p2A':'II', 'p2B':'II', '4 - IV':'IV', '2A - IIA':'II', 'p3':'III', '1B - IB':'I',\n",
    "    'c1B':'I', 'c2B':'I', 'nan':np.NaN, 'p4':'IV','2B - T1-3, N1, M0':'II','p0':'0','p1B':'I','c1':'I',\n",
    "    'c4':'IV','c3':'III','p1A':'I','pNA':np.NaN,'3 - T4, Any N, M0':'III','c2A':'II','p4B':'IV','c4B':'IV',\n",
    "     '2B - T1, N1, M0 / T2, N1, M0 / T3, N1, M0':'II','pUNK':np.NaN,'p2':'II','p3B':'III','c2':'II','p3A':'III',\n",
    "    '3 - III':'III','99 - Unknown':np.NaN,'1A - IA':'I','c3A':'III','c4A':'IV','p1':'I','c1A':'I','p4A':'IV',\n",
    "    '88 - Not applicable to 7th Edition staging':np.NaN,\n",
    "           '88 - No classification is recommended in 6th Edition':np.NaN,\n",
    "    '2A - T3, N0, M0':'II','4 - Any T, Any N, M1':'IV'}\n",
    "df_vital['Stage'] = df_vital.loc[:,'Stage Grouping _ Dominant'].replace(d_stage) \n",
    "\n",
    "#collapse grade\n",
    "\n",
    "d_grade = {'Grade II  Moderately Diff / Mod Well Diff':'2',\n",
    "       'Grade I   Well Differentiated/Differentiated':'1',\n",
    "       'Cell type not determined; not stated;N/A;Unk; high grade dysplas':np.NaN,\n",
    "       'Grade III Poorly Differentiated':'3', 'nan':np.NaN,'Grade IV Undifferentiated, Anaplastic':'4',\n",
    "         'B-CELL    LYMPHOMA OR LEUKEMIA ONLY':np.NaN}\n",
    "df_vital['Grade'] = df_vital.loc[:,'Grade_Differentiation'].replace(d_grade) \n",
    "\n",
    "#collapse LV invasion\n",
    "d_replace = {'nan':np.NaN, 'LYMPHOVASCULAR INVASION STATED AS NOT PRESENT':'NO',\n",
    "       'LYMPHOVASCULAR INVASION PRESENT/IDENTIFIED':'YES',\n",
    "       'Unknown/Indeterminate':np.NaN, 'NOT APPLICABLE':np.NaN,\n",
    "       'Lymph-vascular Invasion Present/Identified':'YES',\n",
    "       'LYMPHATIC AND SMALL VESSEL INVASION ONLY (L)':'YES',\n",
    "       'BOTH LYMPHATIC AND SMALL VESSEL AND VENOUS (LARGE VESSEL) INVASION':'YES'}\n",
    "df_vital['LV_Invasion'] = df_vital.loc[:,'Lymph_vascular Invasion'].replace(d_replace)\n",
    "\n",
    "# LN positivity\n",
    "df_vital['LN_Pos'] = df_vital.loc[:,'Regional Lymph Nodes Positive'] >= 1\n",
    "df_vital.loc[df_vital.loc[:,'Regional Lymph Nodes Positive'].isna(),'LN_Pos'] = np.NaN\n",
    "df_vital['LN_Pos']  = df_vital.LN_Pos.replace({True:'YES',False:'NO'})\n",
    "\n",
    "df_vital['Survival'] = df_vital.cVitalStatus.replace({'Alive':0,'Dead':1})\n",
    "print(df_vital.Grade.unique())\n",
    "print(df_vital.Stage.unique())\n",
    "print(len(df_vital))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fb38c71",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# kaplan meier\n",
    "# %matplotlib inline\n",
    "# # check prognostic value of clinicopathologiocal variables\n",
    "\n",
    "# ls_vital = ['Stage', 'Grade','LV_Invasion','LN_Pos'] \n",
    "# s_time = 'cDays from Diagnosis to FU'\n",
    "# s_censor = 'Survival'\n",
    "# for s_vital in ls_vital:\n",
    "#     print(s_vital)\n",
    "#     df = df_vital.loc[df_vital.loc[:,'Primary Site _ Major Groups For Staging']=='Pancreas',[s_vital,s_time,s_censor]].dropna(how='any')\n",
    "#     fig, __ = util.km_plot(df,s_vital,s_time,s_censor)\n",
    "#     fig.savefig(f'figures/KM_clinicopath_{s_vital}.png')\n",
    "#     #break\n",
    "\n",
    "#CPH\n",
    "s_time = 'cDays from Diagnosis to FU'\n",
    "#df_vital['Survival_time'] = df_vital.loc[:,s_time]\n",
    "s_censor = 'Survival'\n",
    "ls_vital = ['LV_Invasion','Stage', 'Grade','LN_Pos'] #,'Age'\n",
    "for s_vital in ls_vital:\n",
    "    print(s_vital)\n",
    "    df = df_vital.loc[df_vital.loc[:,'Primary Site _ Major Groups For Staging']=='Pancreas',[s_vital,s_time,s_censor]].dropna(how='any')\n",
    "    if df.columns.isin(['Stage']).any():\n",
    "        df.Stage = df.Stage.replace({'I':1,'II':2,'III':3,'IV':4}).astype('int')\n",
    "    if df.columns.isin(['Grade']).any():\n",
    "        df.Grade = df.Grade.astype('int')\n",
    "    if df.loc[:,s_vital].dtype=='O':\n",
    "        df_dummy = pd.get_dummies(df.loc[:,[s_vital]],drop_first=True)\n",
    "        df.drop(s_vital,axis=1,inplace=True)\n",
    "        s_vital = df_dummy.columns[0]\n",
    "        df[s_vital] = df_dummy\n",
    "    fig, cph = util.cph_plot(df,s_vital,s_time,s_censor,figsize=(3,1.5))\n",
    "    plt.tight_layout()\n",
    "    fig.savefig(f'figures/CPH_single_{s_vital}_all.png')\n",
    "    plt.close(fig)\n",
    "    #break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "837d3f5f",
   "metadata": {},
   "source": [
    "### Patients in Paper \n",
    "\n",
    "n=434 specimens\n",
    "\n",
    "n=422 patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a874c60c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load full patient data, neoadjuvant\n",
    "df_patient = pd.read_excel(f'data/SupplementalDataset1.xlsx',\n",
    "                           sheet_name='Patients - Tab 1')\n",
    "df_patient.rename({'Patient ID':'Public_Patient_ID'},axis=1,inplace=True)\n",
    "df_patient.loc[df_patient.loc[:,'Lung Met Present']=='YES','Cohort'] = 'Lung'\n",
    "df_patient.loc[df_patient.loc[:,'Liver Met Present']=='YES','Cohort'] = 'Liver'\n",
    "#how many in cohorts\n",
    "for s_site in df_patient.Cohort.dropna().unique():\n",
    "    n_patients = df_patient[df_patient.Cohort==s_site].Public_Patient_ID.nunique()\n",
    "    print(f'{s_site} {n_patients}')\n",
    "print(f'number rows {len(df_patient)}')\n",
    "print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')\n",
    "\n",
    "#define neodjuvant binary\n",
    "df_patient.loc[:,'Neoadjuvant Treatment'].unique()\n",
    "df_patient.loc[df_patient.loc[:,'Neoadjuvant Treatment'] == 'Yes Neoadjuvant','Neoadjuvant'] = 'Yes'\n",
    "df_patient['Neoadjuvant'] = df_patient.Neoadjuvant.fillna('No')\n",
    "\n",
    "#load T cell data\n",
    "df_tcell_tumor = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset6.xlsx',\n",
    "                         sheet_name='Tumor Samples')\n",
    "\n",
    "df_tcell_blood = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset6.xlsx',\n",
    "                         sheet_name='Blood Samples')\n",
    "\n",
    "#merge T cell blood and tumor\n",
    "df_tcell = df_tcell_tumor.merge(df_tcell_blood,on='Patient ID',suffixes=(' tumor',' blood'),how='outer')\n",
    "\n",
    "df_tcell.rename({'Productive Rearrangements (Observed Richness)':'Productive_Rearrangements',\n",
    "                'Templates per ng':'Templates_per_ng','Patient ID':'Public_Patient_ID'},axis=1,inplace=True)\n",
    "print(f'Add TCR patients {len(df_tcell)}')\n",
    "#merge\n",
    "df_patient = df_patient.merge(df_tcell,on='Public_Patient_ID',how='left',suffixes=('','_x'))\n",
    "df_patient['INDEX'] = df_patient.index + 1\n",
    "print(f'number rows {len(df_patient)}')\n",
    "print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')\n",
    "#all 422 have OS and vital status\n",
    "print(f\"Missing OS: {df_patient.loc[:,'Days from Diagnosis to FU'].isna().sum()}\")\n",
    "print(f\"Missing Vital Status: {df_patient.loc[:,'Vital Status at FU'].isna().sum()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51098a99",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load Purist\n",
    "d_gsva = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset1.xlsx', sheet_name=None)\n",
    "df_gsva = d_gsva['Specimen Subtype - Tab 2'].loc[:,['PurIST Score','Patient Specimen ID','PurIST Subtype','Patient ID']]\n",
    "df_gsva.rename({'Patient ID':'Public_Patient_ID'},axis=1,inplace=True)\n",
    "df_gsva['Specimen_ID'] = [item.split('-')[-1] for item in df_gsva.loc[:,'Patient Specimen ID']]\n",
    "df_gsva.drop('Patient Specimen ID',axis=1,inplace=True)\n",
    "df_purist = df_gsva.pivot(index='Public_Patient_ID',columns=['Specimen_ID'],values='PurIST Score')\n",
    "df_purist['PurIST_Primary'] = df_purist.loc[:,'T'].fillna(df_purist.F).fillna(df_purist.T2)\n",
    "df_purist['PurIST_Primary_T2'] = df_purist.T2\n",
    "#odd that there are only 71 mets with purist calls\n",
    "df_purist['PurIST_Met'] = df_purist.loc[:,'M'].fillna(df_purist.M2)\n",
    "print(f'Pts with RNAseq: {df_purist.reset_index().Public_Patient_ID.nunique()}')\n",
    "df_patient = df_patient.merge(df_purist.reset_index().loc[:,['PurIST_Primary','PurIST_Met',\n",
    "                            'PurIST_Primary_T2','Public_Patient_ID']],\n",
    "                            on='Public_Patient_ID',how='left')\n",
    "print(f'number rows {len(df_patient)}')\n",
    "print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "800eebff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# a patient who has a period of resection to FU but not to recurrence had no recurrence and \n",
    "#a patient who had a period of resection to recurrence but no liver or lung met had another site of recurrence.\n",
    "\n",
    "#has a recurrence, not in liver or lung\n",
    "b_recur = ~df_patient.loc[:,'Days from Resection to Recurrence'].isna() & df_patient.Cohort.isna() \n",
    "\n",
    "#had a resection, no recurrence,  not liver or lung\n",
    "b_no_recur = ~df_patient.loc[:,'Days from Resection to FU'].isna() & df_patient.loc[:,'Days from Resection to Recurrence'].isna()  & df_patient.Cohort.isna() \n",
    "\n",
    "#add recurrence other\n",
    "df_patient['Other_Recurrence'] = False\n",
    "df_patient.loc[b_recur,'Other_Recurrence'] = True\n",
    "#add recurrence none\n",
    "df_patient['No_Recurrence'] = False\n",
    "df_patient.loc[b_no_recur,'No_Recurrence'] = True\n",
    "# should be 73 and 103, but I get 73 and 110 ...\n",
    "print(df_patient.Other_Recurrence.sum())\n",
    "print(df_patient.No_Recurrence.sum())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94a33917",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#add mutation data, if there are more than 9 patient with a mutation\n",
    "df_gene = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset2.xlsx',\n",
    "                           sheet_name='Mutation Data')\n",
    "#which patients were sequenced\n",
    "se_patients = pd.Series(df_gene.loc[:,'Patient Specimen ID'].unique())\n",
    "se_mets = se_patients[se_patients.str.contains('-M')]\n",
    "ls_met = [item.split('-')[0] + '-' + item.split('-')[1] for item in se_mets]\n",
    "se_pri = se_patients[~se_patients.str.contains('-M')]\n",
    "ls_pri = [item.split('-')[0] + '-' + item.split('-')[1] for item in se_pri]\n",
    "d_fill = {'_Primary':ls_pri,'_Met':ls_met,'':ls_pri+ls_met}\n",
    "ls_genes = []\n",
    "for s_gene in df_gene.Gene.unique():  \n",
    "    df_gene_alt = df_gene.loc[df_gene.Gene==s_gene,['Patient ID','Patient Specimen ID','Gene']].copy()\n",
    "    if df_gene_alt.loc[:,'Patient ID'].nunique() > 9 or s_gene.find('BRCA') > -1:\n",
    "        print(s_gene)\n",
    "        df_gene_alt['Specimen_ID'] = [item.split('-')[-1] for item in df_gene_alt.loc[:,'Patient Specimen ID']]\n",
    "        df_pivot = df_gene_alt.drop_duplicates().pivot(index='Patient ID',columns=['Specimen_ID'],values='Gene')\n",
    "        try:\n",
    "            df_pivot[f'{s_gene}_Altered_Primary'] = df_pivot.loc[:,'T'].fillna(df_pivot.F).replace({s_gene:True})#.fillna(False)\n",
    "        except:\n",
    "            df_pivot[f'{s_gene}_Altered_Primary'] = df_pivot.loc[:,'T'].replace({s_gene:True})#.fillna(False)\n",
    "        if df_pivot.columns.isin(['M']).any():\n",
    "            ls_col = ['Patient ID',f'{s_gene}_Altered_Primary',f'{s_gene}_Altered_Met']\n",
    "            try:\n",
    "                df_pivot[f'{s_gene}_Altered_Met'] = df_pivot.loc[:,'M'].fillna(df_pivot.M2).replace({s_gene:True})#.fillna(False)\n",
    "            except:\n",
    "                df_pivot[f'{s_gene}_Altered_Met'] = df_pivot.loc[:,'M'].replace({s_gene:True})#.fillna(False)\n",
    "        else:\n",
    "            ls_col = ['Patient ID',f'{s_gene}_Altered_Primary']\n",
    "        df_pivot = df_pivot.reset_index().loc[:,ls_col]\n",
    "        df_pivot.rename({'Patient ID':'Public_Patient_ID'},axis=1,inplace=True)\n",
    "        df_pivot.loc[(df_pivot.iloc[:,1::]).any(axis=1),f'{s_gene}_Altered'] = True\n",
    "        df_patient = df_patient.merge(df_pivot,on='Public_Patient_ID',how='left',suffixes=('','__'))\n",
    "        for s_loc in ['_Primary','_Met','']:\n",
    "            try:\n",
    "                b_na = df_patient.loc[:,f'{s_gene}_Altered{s_loc}'].isna()\n",
    "                b_loc = df_patient.Public_Patient_ID.isin(d_fill[s_loc])\n",
    "                df_patient.loc[(b_na) & (b_loc),f'{s_gene}_Altered{s_loc}'] = False #.fillna(False,inplace=True)\n",
    "                ls_genes.append(f'{s_gene}_Altered{s_loc}')\n",
    "            except:\n",
    "                print(f'no {s_loc}')\n",
    "    #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8aaa7d0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add age, filter patients who died after surgery\n",
    "d_ids = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/OLD Versions/Simplified_Public_IDs_Key.xlsx',sheet_name=None)\n",
    "\n",
    "df_id = d_ids['RnaSeqKey']\n",
    "ls_ids = df_id.loc[:,'Public.Specimen.ID']\n",
    "\n",
    "# add patients w/o RNA seq\n",
    "for s_key in ['TcrTumorKey','TcrBloodKey']:\n",
    "    df_add = d_ids[s_key].loc[~d_ids[s_key].loc[:,'Public.Specimen.ID'].isin(ls_ids)]\n",
    "    df_id = pd.concat([df_id,df_add])\n",
    "\n",
    "#check\n",
    "print(df_id.loc[:,'OPTR.Specimen.ID'].duplicated().any())\n",
    "ls_drop = df_id.loc[df_id.loc[:,'OPTR.Specimen.ID'].str.contains('-T2')].index\n",
    "df_unique = df_id.loc[df_id.loc[:,'OPTR.Specimen.ID'].str.contains('-T')].drop(ls_drop)\n",
    "#check\n",
    "print(df_unique.OPTR.duplicated().any())\n",
    "\n",
    "#add id\n",
    "df_vital['Public_Patient_ID'] = df_vital.OPTR.map(dict(zip(df_unique.OPTR,df_unique.loc[:,'Biolibrary.Subject.ID'])))\n",
    "\n",
    "#omics data plus clinical data\n",
    "df_patient = df_patient.merge(df_vital,on='Public_Patient_ID',how='left',suffixes=('','_x'))\n",
    "print('adding cinical data')\n",
    "print(f'number rows {len(df_patient)}')\n",
    "print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')\n",
    "#df_id.to_csv('Patient_IDs.csv')\n",
    "\n",
    "#add age category\n",
    "df_patient.loc[df_patient.loc[:,'Age at Diagnosis'] > 70,'Age'] = '>70'\n",
    "df_patient.loc[df_patient.loc[:,'Age at Diagnosis'] <= 70,'Age'] = '<=70'\n",
    "\n",
    "# drop less than 30 days #died of surgery\n",
    "ls_drop_surgery = df_patient[(df_patient.loc[:,'Days from Resection to FU'] < 30) & (df_patient.loc[:,'cVitalStatus'] == 'Dead')].index\n",
    "print('patients who died 1 month after surgery')\n",
    "print(df_patient[df_patient.index.isin(ls_drop_surgery)].Public_Patient_ID)\n",
    "#df_patient = df_patient.drop(ls_drop_surgery)\n",
    "df_patient.loc[ls_drop_surgery,'Alive_30_days_post_surgery'] = False\n",
    "df_patient.Alive_30_days_post_surgery.fillna(True,inplace=True)\n",
    "print('identifying died of surgery')\n",
    "print(f'{df_patient.Alive_30_days_post_surgery.value_counts()}')\n",
    "print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')\n",
    "\n",
    "#add lung and liver versus all (not just lung, liver and NA)\n",
    "df_patient['Lung_Cohort'] = False\n",
    "df_patient.loc[df_patient.Cohort=='Lung','Lung_Cohort'] = True\n",
    "\n",
    "df_patient['Liver_Cohort'] = False\n",
    "df_patient.loc[df_patient.Cohort=='Liver','Liver_Cohort'] = True\n",
    "\n",
    "# should be 73 and 103 (got 73 and 103)\n",
    "print(df_patient[~df_patient.Public_Patient_ID.duplicated()].Other_Recurrence.sum())\n",
    "print(df_patient[~df_patient.Public_Patient_ID.duplicated()].No_Recurrence.sum())\n",
    "\n",
    "print(f'number rows {len(df_patient)}')\n",
    "print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0178a2c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#all recurrence in one\n",
    "df_patient['Recurrence_Sites_4'] = df_patient.Cohort\n",
    "df_patient.loc[df_patient.Other_Recurrence,'Recurrence_Sites_4'] = 'Other_site'\n",
    "df_patient.loc[df_patient.No_Recurrence,'Recurrence_Sites_4'] = 'No_Recurrence'\n",
    "df_patient.Recurrence_Sites_4.fillna('No_Resection',inplace=True)\n",
    "\n",
    "for s_site in df_patient.Recurrence_Sites_4.unique():\n",
    "    n_patients = df_patient[df_patient.Recurrence_Sites_4==s_site].Public_Patient_ID.nunique()\n",
    "    print(f'{s_site} {n_patients}')\n",
    "\n",
    "print(f'number rows {len(df_patient)}')\n",
    "print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')\n",
    "\n",
    "#looks good\n",
    "130 + 34 + 103 + 73 + 74 #alive after surgery\n",
    "130 + 35 + 110 + 73 + 74 # all patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a04b929f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#add pORG \n",
    "ls_pORG = ['trim_padj_0.2_pORG_Up_55_Genes','Original_pORG_Up_78_Genes']\n",
    "s_pSUB = 'Kallisto55_pSUB1e-04'\n",
    "df_pORG = pd.read_csv(f'../20230608_GSVA_Scores.csv',index_col=0)\n",
    "df_pORG.reset_index(drop=True,inplace=True)\n",
    "df_pORG['Specimen_ID'] = [item.split('-')[-1] for item in df_pORG.Public_Specimen_ID]\n",
    "df_pORG['Public_Patient_ID'] = [item.split('-')[0] + '-' + item.split('-')[1] for item in df_pORG.Public_Specimen_ID]\n",
    "#drop 'T2'\n",
    "df_t2 = df_pORG[df_pORG.Public_Specimen_ID.str.contains('-T2')]\n",
    "ls_drop = df_pORG[df_pORG.Public_Specimen_ID.str.contains('-T2')].index\n",
    "df_pORG = df_pORG.drop(ls_drop)\n",
    "#add pri and met\n",
    "for s_pORG in ls_pORG:\n",
    "    for s_group in ['GSVA_Primary', 'GSVA_Met']: \n",
    "        df_group = df_pORG.loc[df_pORG.Group==s_group,[s_pORG,'Public_Patient_ID']]\n",
    "        df_group.rename({'trim_padj_0.2_pORG_Up_55_Genes':'pORG_0.2',\n",
    "                        'Original_pORG_Up_78_Genes':'pORG_78'},axis=1,inplace=True)\n",
    "        s_suffix = s_group.replace('GSVA','')\n",
    "        df_patient = df_patient.merge(df_group,on='Public_Patient_ID',suffixes=('',s_suffix),how='left')\n",
    "df_patient.rename({'pORG_0.2':'pORG_0.2_Primary',\n",
    "                      'pORG_78':'pORG_78_Primary'},axis=1,inplace=True)\n",
    "\n",
    "#add 'GSVA_All': must pivot\n",
    "df_group = df_pORG.loc[df_pORG.Group=='GSVA_All',[s_pORG,'Public_Patient_ID','Specimen_ID']]\n",
    "df_group = df_group.pivot(index='Public_Patient_ID',columns='Specimen_ID',values=s_pORG)\n",
    "df_group['pORG_0.2_allPrimary'] = df_group.loc[:,'T'].fillna(df_group.F)\n",
    "df_group['pORG_0.2_allMet'] = df_group.loc[:,'M'].fillna(df_group.M2)\n",
    "df_patient = df_patient.merge(df_group.reset_index().loc[:,['Public_Patient_ID','pORG_0.2_allPrimary','pORG_0.2_allMet']],on='Public_Patient_ID',how='left')\n",
    "print(len(df_patient))\n",
    "#add T2 specimens, primary, all\n",
    "d_t2 = dict(zip(df_t2[df_t2.Group=='GSVA_Primary'].Public_Patient_ID,df_t2.loc[df_t2.Group=='GSVA_Primary',s_pORG]))\n",
    "df_patient['pORG_0.2_Primary_T2'] = df_patient.Public_Patient_ID.map(d_t2)\n",
    "d_t2_all = dict(zip(df_t2[df_t2.Group=='GSVA_All'].Public_Patient_ID,df_t2.loc[df_t2.Group=='GSVA_All',s_pORG]))\n",
    "df_patient['pORG_0.2_All_T2'] = df_patient.Public_Patient_ID.map(d_t2)\n",
    "\n",
    "#label 10 matched samples\n",
    "df_patient.loc[(~df_patient['pORG_0.2_allPrimary'].isna()) | (~df_patient['pORG_0.2_allMet'].isna()),'Matched_Primary_Met_RNA'] = False \n",
    "df_patient.loc[(~df_patient['pORG_0.2_allPrimary'].isna()) & (~df_patient['pORG_0.2_allMet'].isna()),'Matched_Primary_Met_RNA'] = True\n",
    "df_patient.Matched_Primary_Met_RNA.value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3fcea0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add pSUB\n",
    "\n",
    "d_rename = {'Kallisto55_pSUB1e-05':'pSUB1e-05', 'Kallisto55_pSUB1e-04':'pSUB1e-04'}\n",
    "for s_group in ['Primaries','Mets','All']:\n",
    "    df_pSUB = pd.read_csv(f'../annotation/GSVA_{s_group}_Kallisto55_pSUB1e-4and1e-5.tsv',sep='\\t')\n",
    "    ls_drop = df_pSUB[df_pSUB.SampleID.str.contains('-T2')].index\n",
    "    df_pSUB = df_pSUB.drop(ls_drop)\n",
    "    #temp: use OPTR\n",
    "    df_pSUB['Patient Specimen ID'] = df_pSUB.SampleID.map(dict(zip(df_id.loc[:,'OPTR.Specimen.ID'],df_id.loc[:,'Public.Specimen.ID'])))\n",
    "    df_pSUB.drop('SampleID',axis=1,inplace=True)\n",
    "    df_pSUB['Public_Patient_ID'] = [item.split('-')[0] + '-' + item.split('-')[1] for item in df_pSUB.loc[:,'Patient Specimen ID']]\n",
    "    df_pSUB['Specimen_ID'] = [item.split('-')[-1] for item in df_pSUB.loc[:,'Patient Specimen ID']]\n",
    "    #df_pSUB = df_pSUB.loc[:,[s_pSUB,'Public_Patient_ID']]\n",
    "    df_pSUB.rename(d_rename,axis=1,inplace=True)\n",
    "    s_suffix = s_group.replace('Mets','Met').replace('Primaries','Primary')\n",
    "    if s_group == 'All':\n",
    "        df_group = df_pSUB.pivot(index='Public_Patient_ID',columns='Specimen_ID',values=d_rename[s_pSUB])\n",
    "        df_group[f'{d_rename[s_pSUB]}_allPrimary'] = df_group.loc[:,'T'].fillna(df_group.F)\n",
    "        df_group[f'{d_rename[s_pSUB]}_allMet'] = df_group.loc[:,'M'].fillna(df_group.M2)\n",
    "        ls_col = [f'{d_rename[s_pSUB]}_allPrimary',f'{d_rename[s_pSUB]}_allMet','Public_Patient_ID']\n",
    "        df_patient = df_patient.merge(df_group.reset_index().loc[:,ls_col],on='Public_Patient_ID',suffixes=('',f'_{s_suffix}'),how='left')\n",
    "    else:\n",
    "        df_patient = df_patient.merge(df_pSUB.loc[:,[d_rename[s_pSUB],'Public_Patient_ID']],on='Public_Patient_ID',suffixes=('',f'_{s_suffix}'),how='left')\n",
    "    #break\n",
    "df_patient.rename({'pSUB1e-05':'pSUB1e-05_Primary', 'pSUB1e-04':'pSUB1e-04_Primary'},axis=1,inplace=True)\n",
    "\n",
    "print(f'number rows {len(df_patient)}')\n",
    "print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f182f1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_carl = pd.read_csv('../JasonsPaperMetaDataWithPublicIDs.tsv',sep='\\t')\n",
    "print(len(df_carl))\n",
    "ls_col = ['Tumor.Cellularity.by.DNA','cHas.DDR.Alteration','Specimen.Site','Public.Specimen.ID']\n",
    "df_carl = df_carl.loc[:,ls_col]\n",
    "df_carl.columns = [item.replace('.','_').replace('cHas_','') for item in df_carl.columns]\n",
    "ls_drop = df_carl[df_carl['Public_Specimen_ID'].str.contains('T2')].index\n",
    "df_carl = df_carl.drop(ls_drop)\n",
    "df_carl['Public_Patient_ID'] = [item.split('-')[0] + '-' + item.split('-')[1] for item in df_carl.Public_Specimen_ID]\n",
    "df_carl['Specimen_ID'] = [item.split('-')[-1] for item in df_carl.Public_Specimen_ID]\n",
    "print(len(df_carl))\n",
    "df_carl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78aac2ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_optr = pd.DataFrame(index=df_carl.Public_Patient_ID.unique(),dtype='O')\n",
    "df_optr.index.name = 'Public_Patient_ID'\n",
    "ls_col = ['Specimen_Site','DDR_Alteration','Tumor_Cellularity_by_DNA']\n",
    "df_group = df_carl.pivot(index='Public_Patient_ID',columns='Specimen_ID',values=ls_col)\n",
    "for s_col in ls_col:\n",
    "    df_col = df_group.loc[:,s_col]\n",
    "    df_optr[f'{s_col}_Primary'] = df_col.loc[:,'T'].fillna(df_col.F)\n",
    "    df_optr[f'{s_col}_Met'] = df_col.loc[:,'M'].fillna(df_col.M2)\n",
    "\n",
    "df_patient = df_patient.merge(df_optr.reset_index(),on='Public_Patient_ID',how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2086c4ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f'number rows {len(df_patient)}')\n",
    "print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa35768d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Panel, but no mutation data: 4044 4139 4404 4701 4731 4793\n",
    "# We need to add a \"place holder\" gene for each of the samples with no calls.\n",
    "# This is so that these will be included in the statistics and shown on the graphs.\n",
    "OPTRsWithPanelButNoMutations = [4044, 4139, 4404, 4701, 4731, 4793]\n",
    "se_optr = df_id.loc[df_id.OPTR.isin(OPTRsWithPanelButNoMutations),'Biolibrary.Subject.ID']\n",
    "df_patient.loc[df_patient.Public_Patient_ID.isin(se_optr),df_patient.columns.str.contains('Altered_Primary')]=False\n",
    "#Primary: 5 or six already had a False in the data (should have been NA)\n",
    "#df_patient.loc[df_patient.Public_Patient_ID.isin(se_optr),df_patient.columns.str.endswith('Altered')]\n",
    "#Met: 6 of 6 had False in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "341a03fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dove = pd.read_excel('../Adaptive-total samples-metadata-mets.xlsx',index_col=0)\n",
    "print(df_dove.index.duplicated().any())\n",
    "df_dove['Public_Patient_ID'] = df_dove.index.map(dict(zip(df_id.OPTR,df_id.loc[:,'Biolibrary.Subject.ID'])))\n",
    "print(df_dove.Public_Patient_ID.duplicated().any())\n",
    "df_patient['TCR_Met_Site'] = df_patient.Public_Patient_ID.map(dict(zip(df_dove.Public_Patient_ID,df_dove.loc[:,'Tumor tissue site'])))\n",
    "d_replace = {'Supraclavicular lymph node':'Lymph Node','peritoneum':'Peritoneum',\n",
    "             'lung':'Lung','omentum':'Omentum'}\n",
    "df_patient['TCR_Met_Site'] = df_patient.TCR_Met_Site.replace(d_replace)\n",
    "df_patient.groupby('Cohort').TCR_Met_Site.value_counts()#.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb91ae66",
   "metadata": {},
   "outputs": [],
   "source": [
    "s_out = '20230828_Patient_Metadata.csv'\n",
    "df_patient.Stage = df_patient.Stage.replace({'I':1,'II':2,'III':3,'IV':4,'0':0}).astype('Int64')\n",
    "df_patient.Grade = df_patient.Grade.astype('Int64')\n",
    "df_patient['Survival'] = df_patient.loc[:,'Vital Status at FU'].replace({'Alive':0,'Dead':1})\n",
    "ls_drop_columns = ['OPTR','Lung Met Present in Patient blood','Liver Met Present in Patient blood',\n",
    "    'Liver Met Present in Patient tumor','Lung Met Present in Patient tumor',\n",
    "    'cDays from Diagnosis to FU','cDays from Earliest Recur to FU','cDays from Resection to FU',\n",
    "    'cDays from Resection to Recurrence','cLiverMet','cLongTermRecurrer','cLongTermSurvivor',\n",
    "    'cLungMet','cNeoadjuvant Treatment','cRapidRecurrer','cVitalStatus']\n",
    "if not os.path.exists(s_out):\n",
    "    print('saving')\n",
    "    df_patient.loc[:,~df_patient.columns.isin(ls_drop_columns)].to_csv(s_out)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01bfb10f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ls_site = [#['pORG_0.2_Primary', 'pORG_0.2_Met'],\n",
    "           ['pORG_0.2_allPrimary', 'pORG_0.2_allMet'],\n",
    "['pSUB1e-04_allPrimary', 'pSUB1e-04_allMet'],\n",
    " ['PurIST_Primary', 'PurIST_Met'] ]\n",
    "for tu_site in ls_site:\n",
    "    df_test = df_patient.loc[:,['Public_Patient_ID']+tu_site].dropna()\n",
    "    s_value = tu_site[0].replace(f\"_{tu_site[0].split('_')[-1]}\",'')\n",
    "    df_test.set_index('Public_Patient_ID',inplace=True)#.stack()\n",
    "    df_test.columns = [item.split('_')[-1].replace('all','') for item in df_test.columns]\n",
    "    df_long = df_test.stack().reset_index().rename({'level_1':'Specimen_Type',0:s_value},axis=1)\n",
    "    df_long['Met_Site'] = df_long.Public_Patient_ID.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Specimen_Site_Met)))\n",
    "    fig,ax=plt.subplots(figsize=(5,2.5),dpi=200)\n",
    "    sns.pointplot(data=df_long,x='Specimen_Type',y=s_value,hue='Met_Site',\n",
    "                  dodge=0.1,ax=ax,alpha=0.7,linestyles='-.')\n",
    "    ax.legend(bbox_to_anchor=(1,1.1),title='Met_Site')\n",
    "    ax.set_title(s_value)\n",
    "    ax.set_ylabel('')\n",
    "    plt.tight_layout()\n",
    "    # average\n",
    "    fig,ax=plt.subplots(figsize=(2,2),dpi=200)\n",
    "    sns.pointplot(data=df_long,x='Specimen_Type',y=s_value,ax=ax)\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_title(ax.get_ylabel())\n",
    "    ax.set_ylabel('')\n",
    "    plt.tight_layout()\n",
    "    #by patient\n",
    "    fig,ax=plt.subplots(figsize=(3.5,2.5),dpi=200)\n",
    "    sns.pointplot(data=df_long,x='Specimen_Type',y=s_value,hue='Public_Patient_ID',ax=ax)\n",
    "    ax.set_xlabel('')\n",
    "    statistic, pvalue = stats.wilcoxon(x=df_test.Primary, y=df_test.Met)\n",
    "    ax.set_title(f'{ax.get_ylabel()} p={pvalue:.2}')\n",
    "    ax.set_ylabel('')\n",
    "    ax.legend(bbox_to_anchor=(1,1.1),title='Patient',fontsize='small')\n",
    "    plt.tight_layout()\n",
    "    #by cohort\n",
    "    df_long['Cohort'] = df_long.Public_Patient_ID.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Cohort)))\n",
    "    fig,ax=plt.subplots(figsize=(3,2),dpi=200)\n",
    "    sns.pointplot(data=df_long,x='Specimen_Type',y=s_value,hue='Cohort',\n",
    "                  ax=ax,palette=d_colorblind)\n",
    "    ax.legend(bbox_to_anchor=(1,1),title='Cohort')\n",
    "    ax.set_title(s_value)\n",
    "    ax.set_ylabel('')\n",
    "    ax.set_xlabel('')\n",
    "    plt.tight_layout()\n",
    "    #break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5bc696e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_long.Cohort.value_counts()/2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "114431bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_test['Met_Site'] = df_test.index.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Specimen_Site_Met)))\n",
    "# df_test.Met_Site.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6297e72",
   "metadata": {},
   "outputs": [],
   "source": [
    "### overlap in omics data samples\n",
    "# #the samples overlap/don't overlap in assays\n",
    "# # all samples \n",
    "\n",
    "# from matplotlib_venn import venn3, venn3_circles\n",
    "# from matplotlib_venn import venn2, venn2_circles, venn2_unweighted\n",
    "# from pyvenn import venn\n",
    "# d_ids = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}annotation/Simplified_Public_IDs_Key.xlsx',sheet_name=None)\n",
    "\n",
    "# print(d_ids.keys())\n",
    "# d_keys = {#'OPTR.Specimen.ID':['RnaSeqKey','TcrTumorKey','DnaPanelKey'], #'',\n",
    "#           'OPTR':['RnaSeqKey','TcrTumorKey','TcrBloodKey','DnaPanelKey'],\n",
    "#           #'OPTR':['RnaSeqKey','TcrTumorKey','DnaPanelKey'],\n",
    "#        }\n",
    "\n",
    "# for s_col ,ls_keys in d_keys.items():\n",
    "#     d_sets = {}\n",
    "#     for s_key in ls_keys:\n",
    "#             print(s_key)\n",
    "#             print(len(d_ids[s_key]))\n",
    "#             d_sets.update({s_key:d_ids[s_key]})\n",
    "#     #plot\n",
    "#     if len(ls_keys) == 4:\n",
    "#         labels = venn.get_labels([set(item.loc[:,s_col]) for key, item in d_sets.items()])\n",
    "#         fig,ax = venn.venn4(labels, names=[key.split('Key')[0] for key, item in d_sets.items()])#,ax=ax\n",
    "#     elif len(ls_keys) == 3:\n",
    "#         fig,ax = plt.subplots(figsize=(3,3),dpi=300)\n",
    "#         venn3([set(item.loc[:,s_col]) for key, item in d_sets.items()], [key.split('Key')[0] for key, item in d_sets.items()],ax=ax)\n",
    "\n",
    "# # #primaries\n",
    "\n",
    "# # from matplotlib_venn import venn3, venn3_circles\n",
    "# # from matplotlib_venn import venn2, venn2_circles, venn2_unweighted\n",
    "# # from pyvenn import venn\n",
    "# # d_ids = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}annotation/Simplified_Public_IDs_Key.xlsx',sheet_name=None)\n",
    "\n",
    "# # print(d_ids.keys())\n",
    "# # d_keys = {#'OPTR.Specimen.ID':['RnaSeqKey','TcrTumorKey','DnaPanelKey'], #'',\n",
    "# #           #'OPTR':['RnaSeqKey','TcrTumorKey','TcrBloodKey','DnaPanelKey'],\n",
    "# #           'OPTR':['RnaSeqKey','TcrTumorKey','DnaPanelKey'],\n",
    "# #        }\n",
    "\n",
    "# # for s_col ,ls_keys in d_keys.items():\n",
    "# #     d_sets = {}\n",
    "# #     for s_key in ls_keys:\n",
    "# #             print(s_key)\n",
    "# #             print(len(d_ids[s_key]))\n",
    "# #             df_set = d_ids[s_key][~d_ids[s_key].loc[:,'OPTR.Specimen.ID'].str.contains('-M')]\n",
    "# #             d_sets.update({s_key:df_set})\n",
    "# #     #plot\n",
    "# #     if len(ls_keys) == 4:\n",
    "# #         labels = venn.get_labels([set(item.loc[:,s_col]) for key, item in d_sets.items()])\n",
    "# #         fig,ax = venn.venn4(labels, names=[key.split('Key')[0] for key, item in d_sets.items()],ax=ax)\n",
    "# #     elif len(ls_keys) == 3:\n",
    "# #         fig,ax = plt.subplots(figsize=(3,3),dpi=300)\n",
    "# #         venn3([set(item.loc[:,s_col]) for key, item in d_sets.items()], [key.split('Key')[0] for key, item in d_sets.items()],ax=ax)\n",
    "# #'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d67689c4",
   "metadata": {},
   "source": [
    "# Survival Analysis - patients in paper <a name=\"clin\"></a>\n",
    "\n",
    "KM, CPH\n",
    "\n",
    "\n",
    "[contents](#contents)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9eece8d8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "s_out = '20230828_Patient_Metadata.csv'\n",
    "df_patient = pd.read_csv(s_out,index_col=0)\n",
    "#add recurrence censor\n",
    "df_patient.loc[~df_patient.loc[:,'Days from Resection to Recurrence'].isna(),'Recurrence'] = 1\n",
    "df_patient.loc[df_patient.No_Recurrence,'Recurrence'] = 0\n",
    "print(len(df_patient))\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05871bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# matches kevin, 58 liver metastasis 18 lung metastasis 83 no documented recurrence \n",
    "#53 other recurrence site 6 no resection\n",
    "print(df_patient[~df_patient.PurIST_Primary.isna()].Recurrence_Sites_4.value_counts())\n",
    "print(df_patient[~df_patient.PurIST_Primary_T2.isna()].Recurrence_Sites_4.value_counts())\n",
    "58 + 18 + 83 + 53 + 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "accc2282",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df_patient[~df_patient.loc[:,'pORG_0.2_Met'].isna()].Recurrence_Sites_4.value_counts().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0779a81a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CPH single variable\n",
    "\n",
    "tu_time_censor = (('Days from Diagnosis to FU','Survival'),\n",
    "                  ('Days from Resection to Recurrence', 'Recurrence'))\n",
    "for (s_time, s_censor) in tu_time_censor:\n",
    "    print(s_censor)\n",
    "    df_result=pd.DataFrame()\n",
    "    se_recur = {'Other_Recurrence', 'No_Recurrence','Lung_Cohort','Liver_Cohort'} #don't compare to the 74 with no resection\n",
    "    ls_vital = [# 'pORG_0.2_Primary','pORG_0.2_Met','Neoadjuvant','Age','Sex','Stage', 'Grade','LV_Invasion','LN_Pos', #'Age at Diagnosis',\n",
    "          'Other_Recurrence', 'No_Recurrence','Lung_Cohort','Liver_Cohort',\n",
    "               'pSUB1e-04_Primary', 'pSUB1e-04_Met',\n",
    "           'PurIST_Primary', 'PurIST_Met']\n",
    "    ls_vital = ls_vital + ls_genes\n",
    "    for s_vital in ls_vital:\n",
    "        print(s_vital)\n",
    "        df = df_patient.loc[df_patient.Alive_30_days_post_surgery,[s_vital,s_time,s_censor]].copy()#.dropna(how='any')\n",
    "        if len(se_recur.intersection(set([s_vital]))):\n",
    "            df.loc[df_patient.Recurrence_Sites_4=='No_Resection',s_vital] = np.nan\n",
    "        df = df.dropna()    \n",
    "        print(len(df))\n",
    "        #break\n",
    "        if df.columns.isin(['Stage']).any():\n",
    "            df.Stage = df.Stage.replace({'I':1,'II':2,'III':3,'IV':4}).astype('int')\n",
    "        if df.columns.isin(['Grade']).any():\n",
    "            df.Grade = df.Grade.astype('int')\n",
    "        if df.loc[:,s_vital].dtype=='O':\n",
    "            df_dummy = pd.get_dummies(df.loc[:,[s_vital]],drop_first=True)\n",
    "            if df_dummy.shape[1]>0:\n",
    "                df.drop(s_vital,axis=1,inplace=True)\n",
    "                df[s_vital] = df_dummy\n",
    "            else:\n",
    "                continue\n",
    "        try: \n",
    "            fig, cph = util.cph_plot(df,s_vital,s_time,s_censor,figsize=(3.8,1.5))\n",
    "            pvalue = cph.summary.p[s_vital]\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/CPH_single_{s_vital}_{pvalue:.2}_{s_censor}.png')\n",
    "            #if pvalue > 0.09:\n",
    "            plt.close(fig)\n",
    "            df_cph = cph.summary.loc[:,['exp(coef)','p']]\n",
    "            df_cph['n'] = len(df)\n",
    "            df_result=pd.concat([df_result,df_cph])\n",
    "        except:\n",
    "            print('')\n",
    "        #break\n",
    "    #break\n",
    "    #'''\n",
    "    #all the recurrence combos\n",
    "    s_vital = 'Recurrence_Sites_4'\n",
    "    for tu_combo in combinations(df_patient.Recurrence_Sites_4.unique(),r=2):\n",
    "        print(tu_combo)\n",
    "        df = df_patient.loc[df_patient.Alive_30_days_post_surgery,[s_vital,s_time,s_censor]]#.dropna(how='any')\n",
    "        print(len(df))\n",
    "        df = df[df.loc[:,s_vital].isin(tu_combo)].dropna()\n",
    "        #order them\n",
    "        df.Recurrence_Sites_4 = df.Recurrence_Sites_4.astype('category')\n",
    "        df.Recurrence_Sites_4 = df.Recurrence_Sites_4.cat.set_categories(sorted(tu_combo))\n",
    "        df_dummy = pd.get_dummies(df.loc[:,[s_vital]],drop_first=False)\n",
    "        df.drop(s_vital,axis=1,inplace=True)\n",
    "        s_compare = \" to \".join(sorted(tu_combo))\n",
    "        df[s_compare] = df_dummy.iloc[:,0]\n",
    "        print(len(df))\n",
    "        try:\n",
    "            fig, cph = util.cph_plot(df,s_compare,s_time,s_censor,figsize=(4,1.5))\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/CPH_single_{s_vital}_{s_compare}_{s_censor}.png')\n",
    "            plt.close(fig)\n",
    "            df_cph = cph.summary.loc[:,['exp(coef)','p']]\n",
    "            df_cph['n'] = len(df)\n",
    "            df_result=pd.concat([df_result,df_cph])\n",
    "        except:\n",
    "            print('cph error')\n",
    "        #break\n",
    "        #save results\n",
    "    df_result.to_csv(f'results/results_single_CPH_{s_censor}.csv')\n",
    "    #'''\n",
    "    #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41b458e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_result[df_result.p<0.05].round(4)#.to_csv('results_single_CPH_summary.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4e136a9",
   "metadata": {},
   "source": [
    "## multivariable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d33d410f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CPH multivariable\n",
    "\n",
    "for (s_time, s_censor) in tu_time_censor:\n",
    "    print(s_censor)\n",
    "    for b_genes in [False, True,]:\n",
    "        if b_genes:\n",
    "            s_covar = 'gene_alterations'\n",
    "            figsize = (5,3)\n",
    "        else:\n",
    "            s_covar = 'clinical_covariates'\n",
    "            figsize = (4,3)\n",
    "        df_result_multi = pd.DataFrame()\n",
    "        ls_multi = ['Other_Recurrence', 'No_Recurrence','Lung_Cohort',\n",
    "                   'PurIST_Primary','pORG_0.2_Primary',#'pORG_0.2_Met', 'pORG_0.2_All',,'PurIST_Met',\n",
    "                'pSUB1e-04_Primary', 'Liver_Cohort',#'pSUB1e-04_Met',#'pSUB1e-04_All'\n",
    "               ]\n",
    "        ls_cats = ['LV_Invasion','LN_Pos','Neoadjuvant'] #categorical\n",
    "        if b_genes:\n",
    "            ls_cats = ['CDKN2A_Altered_Primary', #'CDKN2A_Altered_Met', 'KDM6A_Altered_Met',\n",
    "               'ARID1B_Altered_Primary', 'KMT2D_Altered_Primary',\n",
    "               'MTAP_Altered_Primary']\n",
    "        for s_multi in ls_multi:\n",
    "            if len(se_recur.intersection(set([s_multi]))) and b_genes:\n",
    "                ls_cats = ['CDKN2A_Altered','ARID1B_Altered','KMT2D_Altered','MTAP_Altered','TP53_Altered']\n",
    "            print(s_multi)\n",
    "            df = df_patient.loc[df_patient.Alive_30_days_post_surgery,[s_multi,s_time,s_censor,'Grade','Stage','Public_Patient_ID']].copy()#.dropna() #\n",
    "\n",
    "            if ls_cats[0].find('Altered') > -1:\n",
    "                df = df_patient.loc[df_patient.Alive_30_days_post_surgery,[s_multi,s_time,s_censor,'Public_Patient_ID']].copy()#.dropna()\n",
    "            if len(se_recur.intersection(set([s_multi]))):\n",
    "                df.loc[df_patient.Recurrence_Sites_4=='No_Resection',s_multi] = np.nan\n",
    "            df = df.dropna()    \n",
    "            if df.loc[:,s_multi].dtype=='O':\n",
    "                df_dummy = pd.get_dummies(df.loc[df_patient.Alive_30_days_post_surgery,[s_multi]],drop_first=True)\n",
    "                if df_dummy.shape[1]>0:\n",
    "                    df.drop(s_multi,axis=1,inplace=True)\n",
    "                    s_multi = df_dummy.columns[0]\n",
    "                    df[s_multi] = df_dummy\n",
    "                else:\n",
    "                    continue\n",
    "            df_dummy = pd.get_dummies(df_patient.loc[df_patient.Alive_30_days_post_surgery,ls_cats+['Public_Patient_ID']].dropna().set_index('Public_Patient_ID'),drop_first=True) #\n",
    "            df = df.merge(df_dummy.reset_index(),on='Public_Patient_ID',how='left').set_index('Public_Patient_ID')\n",
    "            df = df.dropna()\n",
    "            print(len(df))\n",
    "            try:\n",
    "                fig, cph = util.cph_plot(df,s_multi,s_time,s_censor,figsize=figsize)\n",
    "                fig.savefig(f'figures/CPH_multi_{s_multi}_{s_covar}_{s_censor}.png')\n",
    "                plt.close(fig)\n",
    "                df_result_model = cph.summary.loc[:,['exp(coef)','p']].reset_index()\n",
    "                df_result_model['model'] = s_multi\n",
    "                df_result_model['n'] = len(df)\n",
    "                df_result_multi=pd.concat([df_result_multi,df_result_model])\n",
    "            except:\n",
    "                print('cph error')\n",
    "#             break\n",
    "#         break\n",
    "#     break\n",
    "    #all the recurrence combos - multi\n",
    "    #'''\n",
    "        s_vital = 'Recurrence_Sites_4'\n",
    "        figsize=(5,3)\n",
    "        for tu_combo in combinations(df_patient.Recurrence_Sites_4.unique(),r=2):\n",
    "            print(tu_combo)\n",
    "            df = df_patient.loc[df_patient.Alive_30_days_post_surgery,[s_vital,s_time,s_censor,'Stage','Grade']]#.dropna(how='any')\n",
    "            df = df[df.loc[:,s_vital].isin(tu_combo)].dropna()\n",
    "            #order them\n",
    "            df.Recurrence_Sites_4 = df.Recurrence_Sites_4.astype('category')\n",
    "            df.Recurrence_Sites_4 = df.Recurrence_Sites_4.cat.set_categories(sorted(tu_combo))\n",
    "            df_dummy = pd.get_dummies(df.loc[df_patient.Alive_30_days_post_surgery,[s_vital]],drop_first=False)\n",
    "            df.drop(s_vital,axis=1,inplace=True)\n",
    "            s_compare = \" to \".join(sorted(tu_combo))\n",
    "            df[s_compare] = df_dummy.iloc[:,0]\n",
    "            df_dummy = pd.get_dummies(df_patient.loc[:,ls_cats].dropna(),drop_first=True) \n",
    "            df = pd.concat([df,df_dummy],axis=1)\n",
    "            df = df.dropna()\n",
    "            print(len(df))\n",
    "            try:\n",
    "                fig, cph = util.cph_plot(df,s_compare,s_time,s_censor,figsize=figsize)\n",
    "                plt.tight_layout()\n",
    "                fig.savefig(f'figures/CPH_mutli_{s_vital}_{s_compare}_{s_covar}_{s_censor}.png')\n",
    "                plt.close(fig)\n",
    "                df_result_model = cph.summary.loc[:,['exp(coef)','p']].reset_index()\n",
    "                df_result_model['model'] = s_compare\n",
    "                df_result_model['n'] = len(df)\n",
    "                df_result_multi=pd.concat([df_result_multi,df_result_model])\n",
    "            except:\n",
    "                print('cph error')\n",
    "        df_result_multi.reset_index(drop=True).to_csv(f'results/results_multi_CPH_{s_covar}_{s_censor}.csv')\n",
    "        #'''\n",
    "#         break\n",
    "#     break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8c480d1",
   "metadata": {},
   "source": [
    "# Section 6 <a name=\"geneexp\"></a>\n",
    "\n",
    "gene expresison correlation\n",
    "\n",
    "The 7 genes used for the IRDS signature were: STAT1, IFI44, IFIT3, OAS1,208\n",
    "IFIT1, G1P2, and MX1\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5e75232",
   "metadata": {},
   "outputs": [],
   "source": [
    "s_out = '20230828_Patient_Metadata.csv'\n",
    "df_patient= pd.read_csv(s_out,index_col=0)\n",
    "\n",
    "df_vst = pd.read_csv('data/VST_Genes_Link.csv',index_col=0)\n",
    "df_rna = df_vst.T.copy()\n",
    "df_rna.index = [item[0:-2] for item in df_rna.index]\n",
    "print(len(df_rna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afa3f19e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ls_add = ['pORG_0.2_Primary','pSUB1e-04_Primary','PurIST_Primary']\n",
    "for s_add in ls_add:\n",
    "    d_map = dict(zip(df_patient.loc[:,'Public_Patient_ID'],df_patient.loc[:,s_add]))\n",
    "    df_rna[s_add] = df_rna.index.map(d_map)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db8c2fde",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "d_rename={'txi_pORG_Up_42_Genes':'pORG_42',\n",
    "          'trim_padj_0.2_pORG_Up_55_Genes':'pORG_55',\n",
    "          'pORG.14':'pORG_14', 'pORG.15':'pORG_15',\n",
    "          'pORG_0.2_Primary':'pORG_0.2',\n",
    "          'pSUB1e-04_Primary':'pSUB1e-04',\n",
    "          'MRC1':'MRC1 (CD206)',\n",
    "          'OLR1':'OLR1 (LOX1)',\n",
    "          'MS4A1':'MS4A1 (CD20)'}\n",
    "from scipy.stats import pearsonr\n",
    "dim = (4,3)\n",
    "    \n",
    "for s_add in ls_add:\n",
    "    ls_marker = ['CD3E','CD4', 'CD8A', 'MS4A1',#'LAMP3','ITGAX','CD209',\n",
    "                 'CD68','OLR1','MRC1','MX1','STAT1',\n",
    "                 s_add] #\n",
    "    df_all = df_rna.loc[:,ls_marker].corr().rename(d_rename,axis=1).rename(d_rename,axis=0)\n",
    "    print(len(df_rna))\n",
    "    g = sns.clustermap(df_all)\n",
    "    plt.close()\n",
    "    categories_order = df_all.iloc[g.dendrogram_col.reordered_ind,:].index.tolist()\n",
    "    df_all = df_all.loc[categories_order,categories_order]\n",
    "    rho = df_rna.loc[:,ls_marker].corr() #df_all.corr()\n",
    "    pval = df_rna.loc[:,ls_marker].corr(method=lambda x, y: pearsonr(x, y)[1]) - np.eye(*rho.shape)\n",
    "    p_vals = pval.applymap(lambda x: ''.join(['*' for t in [0.001,0.005,0.05] if x<=t]))\n",
    "    p_vals = p_vals.rename(d_rename,axis=1).rename(d_rename,axis=0)\n",
    "    p_vals = p_vals.loc[categories_order,categories_order]\n",
    "    fig, ax = plt.subplots(figsize=dim,dpi=300)\n",
    "    sns.heatmap(df_all, vmin=-1, vmax=1, annot=p_vals, fmt = '', cmap='RdBu_r',ax=ax,\n",
    "               cbar_kws={'shrink':0.85,'label':s_add})\n",
    "    #temp\n",
    "    plt.title(f'{s_add} n={len(df_rna)}')\n",
    "    fig, ax = plt.subplots(figsize=dim,dpi=300)\n",
    "    matrix = np.triu(np.ones_like(rho))\n",
    "    np.fill_diagonal(matrix, val=0)\n",
    "    np.fill_diagonal(p_vals.values,'')\n",
    "    sns.heatmap(df_all, vmin=-1, vmax=1, annot=p_vals, fmt = '', cmap='RdBu_r',\n",
    "                ax=ax,mask=matrix,cbar_kws={'shrink':0.85,'label':'Pearson Correlation'},\n",
    "               ) #'anchor':(-1.4,0.0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d14d4247",
   "metadata": {},
   "source": [
    "## gene expresison versus liver lung"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73ad5cdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "Colorblind[8]\n",
    "pal_porg = ('#56B4E9','#E69F00')\n",
    "pal_liv = ('#0072B2','#D55E00')\n",
    "sns.set_palette(pal_liv)\n",
    "#sns.palplot(Colorblind[8])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebe37598",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# add cohort\n",
    "df_rna['Public_Patient_ID'] = df_rna.index\n",
    "df_rna['Cohort'] = df_rna.index.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Cohort)))\n",
    "#liver vs lung (t cell)\n",
    "for s_samples in ['T cell samples','All Samples']:\n",
    "    ls_marker = ['CD3E','CD4', 'CD8A', 'MS4A1','CD68','OLR1','MRC1','MX1','STAT1']\n",
    "    if s_samples == 'T cell samples':\n",
    "        df_plot = df_rna.loc[df_rna.Public_Patient_ID.isin(df_tcell.Public_Patient_ID),ls_marker].unstack().reset_index()\n",
    "    else:\n",
    "        df_plot = df_rna.loc[:,ls_marker].unstack().reset_index()\n",
    "    df_plot['Public_Patient_ID'] = df_plot.level_1#[item[0:-2] for item in df_plot.level_1]\n",
    "    df_plot['Cohort'] = df_plot.Public_Patient_ID.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Cohort)))\n",
    "    df_plot.rename({0:'Expression','level_0':'Gene'},axis=1,inplace=True)\n",
    "    fig,ax = plt.subplots(dpi=200,figsize=(5,3))\n",
    "    sns.stripplot(data=df_plot,x='Gene',y='Expression',hue='Cohort',dodge=True,ax=ax,s=2)\n",
    "    sns.boxplot(data=df_plot,x='Gene',y='Expression',hue='Cohort',ax=ax,showmeans=True,medianprops={'visible': False},\n",
    "                           whiskerprops={'visible': False},meanline=True,showcaps=False,\n",
    "                           meanprops={'color': 'k', 'ls': '-', 'lw': 2},showfliers=False,showbox=False)\n",
    "    h, l = ax.get_legend_handles_labels()\n",
    "    ax.legend(h[0:2],l[0:2],loc='lower left')\n",
    "    pairs = [((item,'Lung'),(item,'Liver')) for item in ls_marker]\n",
    "    annot = Annotator(ax, pairs, data=df_plot,x='Gene',y='Expression',hue='Cohort',\n",
    "                      order=ls_marker,hue_order=('Lung','Liver'))\n",
    "    annot.configure(test='t-test_ind')\n",
    "    annot.apply_and_annotate()\n",
    "    ax.set_title(f'Liver vs Lung: {s_samples}')\n",
    "    plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ce56879",
   "metadata": {},
   "source": [
    "## Section 6 <a name=\"tcell\"></a>\n",
    "\n",
    "re-analyze t cell data\n",
    "\n",
    "289 blood samples with matching 175 primary\n",
    "tumors (141 overlapping with the RNA-seq dataset) and 43 metastatic tumors (33 overlapping with\n",
    "the RNA-seq dataset). (218)\n",
    "\n",
    "**missing 2 tumor (have 216), all 289 blood there**\n",
    "\n",
    "290 unique patients (one primary tumor w/o blood: ST-00018360)\n",
    "\n",
    "215 blood and tumor are matched\n",
    "\n",
    "TOTAL = 174 primary, 42 met (216 total)\n",
    "\n",
    "Of the 290 patients, 284 are analyzed (some dropped surgery)\n",
    "\n",
    "we analyzed blood samples from 77 patients in the liver cohort and 16\n",
    "patients in the lung cohort, of which 60 and 16 were matched with tumor samples from the same\n",
    "patient, respectively\n",
    "\n",
    "TOTAL = 94 blood\n",
    "TOTAL = 76 tumor\n",
    "\n",
    "tumor distinct clones\n",
    "used data from 214 matched pairs of tumor and blood samples\n",
    "\n",
    "**213 are there, missing 1**\n",
    "\n",
    "(TCR tumor: 59 and 16)\n",
    "\n",
    "#### Exclude those who died of surgery\n",
    "\n",
    "yes      ST-00018963\n",
    "\n",
    "yes      ST-00020077\n",
    "\n",
    "yes     ST-00016968\n",
    "\n",
    "yes     ST-00006625\n",
    "\n",
    "yes     ST-00007146\n",
    "\n",
    "yes    ST-00018260\n",
    "\n",
    "#### Actually analyzed \n",
    "Liver    76\n",
    "Lung     16\n",
    "\n",
    "| Cohort  | Tumor Type  |  number pts.  |\n",
    "|---------|-------------|---------------|\n",
    "| Liver   | Met         | 17            |\n",
    "|         | Primary     | 42            |\n",
    "| Lung    | Met         | 3             |\n",
    "|         | Primary     | 13            |\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "339b5717",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_merge.loc[df_merge.Alive_30_days_post_surgery,['Shannon_Entropy_Tumor','Alive_30_days_post_surgery']].dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1357f459",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load full patient data\n",
    "s_out = '20230828_old_Patient_Metadata.csv'\n",
    "df_patient= pd.read_csv(s_out,index_col=0)\n",
    "# old patient data\n",
    "df_patient_old = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset1.xlsx',\n",
    "                           sheet_name='Patients - Tab 1')\n",
    "#load T cell data - tumor\n",
    "df_tcell = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset6.xlsx',\n",
    "                         sheet_name='Tumor Samples')\n",
    "print(f'T cell tumor {len(df_tcell)}')\n",
    "#blood\n",
    "df_tcell_blood = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset6.xlsx',\n",
    "                         sheet_name='Blood Samples')\n",
    "#df_tcell_blood['Tumor Type'] = 'Blood'\n",
    "print(f'T cell blood {len(df_tcell_blood)}')\n",
    "\n",
    "#both\n",
    "df_tcell = df_tcell.merge(df_tcell_blood,on='Patient ID',suffixes=(' tumor',' blood'),how='outer')\n",
    "df_tcell.rename({'Productive Rearrangements (Observed Richness)':'Productive_Rearrangements',\n",
    "                'Templates per ng':'Templates_per_ng','Patient ID':'Public_Patient_ID'},axis=1,inplace=True)\n",
    "print(f'T cell both {len(df_tcell)}')\n",
    "df_merge = df_patient[df_patient.Public_Patient_ID.isin(df_tcell.Public_Patient_ID)]\n",
    "\n",
    "# missing_ids: some were dropped surgery due to death from surgery\n",
    "missing_ids = df_tcell[~df_tcell.Public_Patient_ID.isin(df_merge.Public_Patient_ID)].Public_Patient_ID\n",
    "df_not_matched = df_patient_old[df_patient_old.loc[:,'Patient ID'].isin(missing_ids)]\n",
    "print(f'dropped {len(df_not_matched)} patients; days surv post surgery: {df_not_matched.loc[:,\"Days from Resection to FU\"].max()}')\n",
    "\n",
    "print(f'Unique patients t cell {(df_merge.Public_Patient_ID.nunique())}')\n",
    "print(df_merge.Cohort.value_counts())\n",
    "print(df_merge.groupby(['Cohort','Tumor Type']).count().Public_Patient_ID)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "229714e0",
   "metadata": {},
   "source": [
    "## add calculated Entropy, simpsons "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c41df173",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load data\n",
    "\n",
    "d_load = {#'TCR_Tumor_Distinct_Clones_no_rare.csv':['Public_Patient_ID',\n",
    "          #      'Number Tumor Distinct Clones', 'Fraction Tumor Distinct Clones','Prod. Freq. Tumor Distinct Clones'],\n",
    "         # 'TCR_Tumor_Distinct_Rearrangements_no_rare.csv':['Public_Patient_ID',\n",
    "       # 'Number Tumor Distinct Rearrangements','Fraction Tumor Distinct Rearrangements','Prod. Freq. Tumor Distinct Rearrangements'],\n",
    "          'TCR_Tumor_Distinct_Clones.csv':['Public_Patient_ID','Fraction_Tumor_Distinct_Clones'],#'Number Tumor Distinct Clones (rare)',\n",
    "          'TCR_Simpsons_Evenness_templates.csv':['Public_Patient_ID','Simpsons_Evenness','Simpsons_Evenness_no_CMV'],\n",
    "          'TCR_Simpsons_D_templates.csv':['Public_Patient_ID','Simpsons_D'],\n",
    "          'TCR_Shannon_entropy_productive_frequency.csv':['Public_Patient_ID','Shannon_Entropy','Normalized_Shannon_Entropy'],\n",
    "         }\n",
    "for s_en, ls_markers in d_load.items():\n",
    "    print(s_en)\n",
    "    df_en = pd.read_csv(f'data/{s_en}')\n",
    "    df_en.rename({'Unnamed: 0':'Patient_Specimen_ID','Shannon Entropy':'Shannon_Entropy',#'Percent Tumor Distinct Rearrangements in Tumor':'Percent Tumor Distinct Clones (JL)',   \n",
    "             'Normalized Shannon Entropy':'Normalized_Shannon_Entropy'},axis=1,inplace=True)\n",
    "    if s_en == 'TCR_Tumor_Distinct_Clones.csv':\n",
    "        df_en.rename({'Unnamed: 0':'Patient_Specimen_ID','Fraction Tumor Distinct Clones in Tumor':'Fraction_Tumor_Distinct_Clones'},axis=1,inplace=True)\n",
    "    df_en['dash_end'] = [item.split('-')[-1] for item in df_en.Patient_Specimen_ID]\n",
    "    df_en['Public_Patient_ID'] = [item.split('-')[0] + '-' + item.split('-')[1] for item in df_en.Patient_Specimen_ID]\n",
    "    #patient rows with columns for blood and tumor\n",
    "    df_en_pat = pd.DataFrame(index=df_en.Public_Patient_ID.unique())\n",
    "    df_tum = df_en[df_en.dash_end!='B'].loc[:,ls_markers].set_index('Public_Patient_ID')\n",
    "    print(df_tum.index.duplicated().any())\n",
    "    df_en_pat = df_en_pat.merge(df_tum,left_index=True,right_index=True,how='left')\n",
    "    if s_en.find('Tumor_Distinct') == -1:\n",
    "        df_bld = df_en[df_en.dash_end=='B'].loc[:,ls_markers].set_index('Public_Patient_ID')\n",
    "        print(df_bld.index.duplicated().any())\n",
    "        df_en_pat = df_en_pat.merge(df_bld,left_index=True,right_index=True,how='left',suffixes=('_Tumor','_Blood'))\n",
    "    df_merge = df_merge.merge(df_en_pat.reset_index().rename({'index':'Public_Patient_ID'},axis=1),on='Public_Patient_ID',how='left')\n",
    "\n",
    "    \n",
    "df_merge['Simpsons_Diversity_Tumor'] = 1/df_merge.Simpsons_D_Tumor#1-df_merge.Simpsons_D_Tumor#\n",
    "df_merge['Simpsons_Diversity_Blood'] = 1/df_merge.Simpsons_D_Blood#1-df_merge.Simpsons_D_Blood#\n",
    "df_merge['Clonality_Tumor'] = 1 - df_merge.Normalized_Shannon_Entropy_Tumor\n",
    "df_merge['Clonality_Blood'] = 1 - df_merge.Normalized_Shannon_Entropy_Blood\n",
    "\n",
    "#met or primary\n",
    "df_merge['Percent Tumor Distinct Clones'] = 100 - df_merge.loc[:,'Percentage Tumor-Distinct Clones in Paired Tumor Sample']\n",
    "df_merge.rename({'Patient Specimen ID':'Patient_Specimen_ID','Tumor Type':'Tumor_Type'},axis=1,inplace=True)\n",
    "print((df_merge.Public_Patient_ID.nunique()))\n",
    "\n",
    "# df_merge.loc[df_merge.loc[:,'PurIST_Met'] > 0.5,'PurIST_Subtype'] = 'basal-like'\n",
    "# df_merge.loc[df_merge.loc[:,'PurIST_Met'] <= 0.5,'PurIST_Subtype'] = 'classical'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84e310bc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#add pORG quartiles, plot violins \n",
    "%matplotlib inline\n",
    "sns.set_palette('tab10')\n",
    "importlib.reload(util)\n",
    "alpha = 0.05\n",
    "b_correct= True#False #True #\n",
    "ls_foci = ['Shannon_Entropy_Tumor','Templates_per_ng', 'Shannon_Entropy_Blood',\n",
    "           'Productive_Rearrangements',\n",
    "        'Simpsons_Diversity_Tumor','Simpsons_Diversity_Blood',\n",
    "           'Clonality_Tumor', 'Clonality_Blood',\n",
    "        \"Simpson's Evenness tumor\", \"Simpson's Evenness blood\",\n",
    "        'Fraction_Tumor_Distinct_Clones','Percent Tumor Distinct Clones',\n",
    "        ]\n",
    "d_colorblind = {'Liver':'#d55e00','Lung':'#0072b2',\n",
    "               'high':'#e69f00','low': '#56b4e9',\n",
    "               'basal-like':'#000000','classical':'#cc79a7',\n",
    "               'high pSUB': '#f0e442','low pSUB':'#d55e00'}\n",
    "#non- parametric\n",
    "se_non_para = pd.Series(['Simpsons_Diversity_Blood', \n",
    "    'Templates_per_ng','Productive_Rearrangements',\n",
    "     \"Simpson's Evenness tumor\",\n",
    "    \"Simpson's Evenness blood\",\n",
    "   'Number Tumor Distinct Clones'])\n",
    "d_order =  {'Cohort':['Liver','Lung'],#'PurIST_Subtype':['basal-like','classical'],\n",
    "    'quartiles':['high','low']}\n",
    "for s_porg in ['pORG_0.2_All', 'pORG_0.2_Primary', 'pORG_0.2_Met']:#, 'pORG_78_Primary',\n",
    "    if s_porg.find('Primar') > -1:\n",
    "        print('Primaries')\n",
    "        df_pri = df_merge.loc[(df_merge.Tumor_Type=='Primary')].copy()\n",
    "    elif s_porg.find('All') > -1: \n",
    "        print('Primaries and Mets')\n",
    "        df_whole = pd.DataFrame(columns=df_merge.columns)\n",
    "        for s_all in ['pORG_0.2_allPrimary', 'pORG_0.2_allMet']:\n",
    "            df_half = df_merge[(df_merge.loc[:,s_all].notna())]\n",
    "            df_half = df_half[~df_half.Public_Patient_ID.isin(df_whole.Public_Patient_ID)].copy()\n",
    "            df_half.rename({s_all:'pORG_0.2_All'},axis=1,inplace=True)\n",
    "            print(len(df_half))\n",
    "            df_whole = pd.concat([df_whole,df_half],axis=0,ignore_index=True)\n",
    "        df_pri = df_merge.merge(df_whole.loc[df_whole.loc[:,s_foci].notna(),['Public_Patient_ID','pORG_0.2_All']],on='Public_Patient_ID',how='left')\n",
    "        df_pri.loc[df_pri.Public_Patient_ID.duplicated(),'Cohort'] = np.nan\n",
    "    elif s_porg.find('Met') > -1:\n",
    "        df_pri = df_merge.loc[df_merge.Tumor_Type=='Met'].copy()\n",
    "    df_pri = util.add_quartiles(df_pri,s_porg)#.drop(219) #p. rearrangements outlier \n",
    "    for s_foci in ls_foci:\n",
    "        print(s_foci)\n",
    "        if se_non_para.isin([s_foci]).any():\n",
    "            s_stats = 'non-parametric'\n",
    "        else:\n",
    "            s_stats = 'mean' \n",
    "        df_both,d_pval,order = util.violin_stats2(df_pri,d_order,s_foci,s_stats)\n",
    "        #util.qq_plot_hist(df_pri,s_cat,s_foci)  #anova eval\n",
    "        fig,pvalues,corrected = util.plot_violins2(df_both,d_pval,d_order,s_stats,s_foci,order,d_colorblind,s_porg,b_correct=b_correct)#True#False\n",
    "        fig.savefig(f'figures/violinplot_both_{s_foci}_{list(d_order.keys())[-1]}_{s_porg}_{s_stats}.png')\n",
    "        plt.close(fig)\n",
    "        # SPEARMAN tumor distinct clones\n",
    "        #fig2, pvalues2 = util.plot_pearson(df_pri,'Fraction_Tumor_Distinct_Clones',s_foci,s_stats= 'non-parametric',ls_plots=['Primaries'])\n",
    "        '''\n",
    "        if s_porg.find('All') > -1:\n",
    "            fig2, pvalues2 = util.plot_pearson(df_pri,s_porg,s_foci,s_stats,ls_plots=['Primaries','Mets','Both'])\n",
    "            fig2.savefig(f'figures/scatterplot_pearson_all_{s_foci}_{list(d_order.keys())[-1]}_{s_porg}_{s_stats}.png')\n",
    "        elif s_porg.find('Primar') > -1:\n",
    "            fig2, pvalues2 = util.plot_pearson(df_pri,s_porg,s_foci,s_stats,ls_plots=['Primaries'])\n",
    "            fig2.savefig(f'figures/scatterplot_pearson_primary_{s_foci}_{list(d_order.keys())[-1]}_{s_porg}_{s_stats}.png')\n",
    "        elif s_porg.find('Met') > -1:\n",
    "            fig2, pvalues2 = util.plot_pearson(df_pri,s_porg,s_foci,s_stats,ls_plots=['Mets'])\n",
    "            fig2.savefig(f'figures/scatterplot_pearson_met_{s_foci}_{list(d_order.keys())[-1]}_{s_porg}_{s_stats}.png')\n",
    "        plt.close(fig2)\n",
    "#         break\n",
    "#         #'''\n",
    "    break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f134f27b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_half #153, 141 shannon entropy tumor\n",
    "#len(df_half.loc[:,s_foci].dropna()) #41, 34 have entropy\n",
    "#(141 + 34)/2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65f9b3a2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# #color by met\n",
    "# #add pORG quartiles, plot violins \n",
    "# %matplotlib inline\n",
    "# sns.set_palette('tab10')\n",
    "# importlib.reload(util)\n",
    "\n",
    "# s_porg = 'pORG_0.2_Met'\n",
    "# for s_comp in ['Cohort','quartiles']:#, 'pORG_78_Primary',\n",
    "#     df_pri = df_merge.loc[df_merge.Tumor_Type=='Met'].copy()\n",
    "#     df_pri = util.add_quartiles(df_pri,s_porg)#.drop(219) #p. rearrangements outlier \n",
    "#     for s_foci in ls_foci:\n",
    "#         print(s_foci)\n",
    "#         fig = util.plot_violins3(df_pri,s_stats,s_foci,s_comp,s_porg,hue='TCR_Met_Site',figsize=(5,3))#True#False\n",
    "#         fig.savefig(f'figures/violinplot_met_colors_{s_foci}_{s_porg}_{s_comp}.png')\n",
    "#         break\n",
    "#     break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21428c13",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #group by met\n",
    "# d_replace = {'Gastric serosal':'Other', 'Lower quadrant mass':'Other','Mesocolon':'Other',\n",
    "#              'Tissue near left hepatic artery':'Other',\n",
    "#             'Pelvic nodule':'Other', 'Gallbladder':'Other', 'Omentum':'Other',\n",
    "#              'Retroperitoneum':'Other'}\n",
    "# df_pri['TCR_Met_Site_S'] = df_pri.TCR_Met_Site.replace(d_replace)\n",
    "# for s_foci in ls_foci:\n",
    "#     fig = util.plot_violins3(df_pri,s_stats,s_foci,s_comp='TCR_Met_Site_S',s_porg=s_porg,hue='TCR_Met_Site',figsize=(8,3))#True#False\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "607c8a63",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # check the quartiles code\n",
    "# #s_porg ='pORG_0.2_Primary'#'pORG_78_Primary'#'pORG_0.2_Met'\n",
    "# x = df_pri.loc[:,s_porg].dropna()\n",
    "# b_cut = df_pri.loc[:,[s_porg,'Public_Patient_ID']].dropna().Public_Patient_ID\n",
    "# b_cut = df_pri.loc[:,s_porg].dropna().index\n",
    "# print(len(x))\n",
    "# d_cut = {'quartiles':(4,['low','med-low','med-high','high']),\n",
    "#          'tertiles' : (3,['low','med','high']),\n",
    "#         'medians' : (2,['low','high'])}\n",
    "# for s_col, tu_cut in d_cut.items():\n",
    "#     i_cut = tu_cut[0]\n",
    "#     labels = tu_cut[1]\n",
    "#     q = pd.qcut(x, q=i_cut,labels=labels) \n",
    "#     break\n",
    "# df_pri['test'] = q\n",
    "# fig,ax=plt.subplots()\n",
    "# sns.stripplot(data=df_pri,x='test',y=s_porg,ax=ax)\n",
    "# ax.axhline(df_pri.loc[df_pri.test=='low',s_porg].max())\n",
    "# ax.axhline(df_pri.loc[df_pri.test=='high',s_porg].min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cbc0b4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_pri.loc[~df_pri.Cohort.isna(),['Public_Patient_ID','Cohort','TCR_Met_Site']].sort_values(by=['Cohort','TCR_Met_Site']).to_csv('test.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a9fdfde",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "## TCR survival\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8a0128c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#KM and CPH\n",
    "pal_porg_r = ('#E69F00','#56B4E9')\n",
    "sns.set_palette(pal_porg_r)\n",
    "s_time = 'Days from Diagnosis to FU'\n",
    "s_censor = 'Survival'\n",
    "alpha = 0.05\n",
    "savedir = 'figures'\n",
    "i_cut=4\n",
    "labels= ['low','med-low','med-high','high'] # ['low','highX','high']#\n",
    "#labels = ['low','highX','high']\n",
    "for s_title_str in  ['All','Primary','Met',]:\n",
    "    df_result = pd.DataFrame()\n",
    "    if s_title_str == 'All':\n",
    "        df_km_samples = df_merge.loc[df_merge.Alive_30_days_post_surgery]\n",
    "    elif s_title_str == 'Met':\n",
    "        i_cut=2\n",
    "        labels=['low','high']\n",
    "        df_km_samples = df_merge.loc[(df_merge.loc[:,'Tumor_Type']==s_title_str) & (df_merge.Alive_30_days_post_surgery),[s_time,s_censor,'Public_Patient_ID']+ls_foci]\n",
    "    else:\n",
    "        df_km_samples = df_merge.loc[(df_merge.loc[:,'Tumor_Type']==s_title_str) & (df_merge.Alive_30_days_post_surgery),[s_time,s_censor,'Public_Patient_ID']+ls_foci]\n",
    "    for s_col in ls_foci[2::]:\n",
    "        print(s_col)\n",
    "        df_km, pvalue = util.quartile_km(df_km_samples,s_col,s_title_str,savedir,alpha,i_cut,\n",
    "                                            labels,s_time=s_time,s_censor='Survival')\n",
    "        print(pvalue)\n",
    "        try: #cox\n",
    "            df = df_km_samples.loc[:,[s_col,s_time,s_censor]].dropna()\n",
    "            fig2, cph = util.cph_plot(df,s_col,s_time,s_censor,figsize=(4,1.5))\n",
    "            plt.tight_layout()\n",
    "            fig2.savefig(f'figures/CPH_TCR_{s_col}_{s_title_str}.png')\n",
    "            df_result_model = cph.summary.loc[:,['exp(coef)','p']].reset_index()\n",
    "            df_result=pd.concat([df_result,df_result_model])\n",
    "            plt.close(fig2)\n",
    "        except:\n",
    "            print('')\n",
    "        #print(s_title_str)\n",
    "        break\n",
    "    df_result.sort_values(by='p').set_index('covariate').to_csv(f'results_CPH_TCR_{s_title_str}.csv')\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e0376ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "77 + 16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "512a24a1",
   "metadata": {},
   "source": [
    "### TCR heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83fc85de",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#heatmap\n",
    "if s_porg == 'pORG_0.2_All':\n",
    "    import matplotlib as mpl\n",
    "    df_pri['Survival interval'] = pd.qcut(df_merge.loc[:,s_time],6)\n",
    "    df_pri.rename({'medians':f'{s_porg} (median)'},axis=1,inplace=True)\n",
    "    ls_annot = ['Tumor_Type', 'Cohort',f'{s_porg} (median)' ,'Survival interval']\n",
    "    cmap=plt.cm.get_cmap('Blues', 6)\n",
    "    ls_color=[mpl.cm.Set1.colors,('#D55E00', '#0072B2'),('#E69F00', '#56B4E9'),\n",
    "             [cmap(item) for item in np.arange(0,1,.17)]]\n",
    "    df_annot = pd.DataFrame()\n",
    "    dd_color = {}\n",
    "    for idx, s_annot in enumerate(ls_annot):\n",
    "        color_palette = ls_color[idx]\n",
    "        d_color = dict(zip(sorted(df_pri.loc[:,s_annot].dropna().unique()),color_palette[0:len(df_pri.loc[:,s_annot].dropna().unique())]))\n",
    "        d_color.update({'NA':'lightgray'})\n",
    "        network_colors = df_pri.loc[:,s_annot].astype('object').fillna('NA').map(d_color)\n",
    "        df_annot[s_annot] = pd.DataFrame(network_colors)\n",
    "        dd_color.update({s_annot:d_color})\n",
    "    from matplotlib.pyplot import gcf\n",
    "    g = sns.clustermap(df_pri.loc[:,ls_foci].dropna(),z_score=1,dendrogram_ratio=0.1,cmap='RdBu_r',vmin=-5,vmax=5,\n",
    "          cbar_pos=(0.01, 0.94, 0.04, 0.08),figsize=(8, 10),row_colors=df_annot,cbar_kws={'label': 'z-score'})\n",
    "    for idx, (s_annot, d_color) in enumerate(dd_color.items()):\n",
    "        g.ax_col_dendrogram.bar(0, 0, color='w',label=' ', linewidth=0)\n",
    "        for label,color in d_color.items():\n",
    "            g.ax_col_dendrogram.bar(0, 0, color=color,label=label, linewidth=0)\n",
    "    g.ax_heatmap.set_yticks([])\n",
    "    l1 = g.ax_col_dendrogram.legend(loc=\"right\", ncol=1,bbox_to_anchor=(0, 0.7),bbox_transform=gcf().transFigure)\n",
    "    g.ax_heatmap.set_title(s_porg)\n",
    "    g.savefig(f'figures/heatmap_TCR_{s_porg}.png')\n",
    "    #plt.close(g.fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c476cb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#high low pORG quartile versus gene expresison\n",
    "sns.set_palette(pal_porg)\n",
    "if s_porg == 'pORG_0.2_Primary':\n",
    "    #df_rna['Public_Patient_ID'] = [item.split('-T')[0] for item in df_rna.index]\n",
    "    #liver vs lung (run t cell section to load t cell data!)\n",
    "    ls_marker = ['CD3E','CD4', 'CD8A', 'MS4A1','CD68','OLR1','MRC1','MX1','STAT1',]\n",
    "    df_plot = df_rna.loc[df_rna.index.isin(df_merge.Public_Patient_ID),ls_marker].unstack().reset_index()\n",
    "    df_plot['Public_Patient_ID'] = df_plot.level_1 #[item[0:-2] for item in df_plot.level_1]\n",
    "    df_plot['pORG'] = df_plot.Public_Patient_ID.map(dict(zip(df_pri.Public_Patient_ID,df_pri.quartiles)))\n",
    "    df_plot.rename({0:'Expression','level_0':'Gene'},axis=1,inplace=True)\n",
    "    fig,ax = plt.subplots(dpi=200,figsize=(5,3))\n",
    "    sns.stripplot(data=df_plot,x='Gene',y='Expression',hue='pORG',dodge=True,ax=ax,s=2)\n",
    "    sns.boxplot(data=df_plot,x='Gene',y='Expression',hue='pORG',ax=ax,showmeans=True,medianprops={'visible': False},\n",
    "                           whiskerprops={'visible': False},meanline=True,showcaps=False,\n",
    "                           meanprops={'color': 'k', 'ls': '-', 'lw': 2},showfliers=False,showbox=False)\n",
    "    h, l = ax.get_legend_handles_labels()\n",
    "    ax.legend(h[0:2],l[0:2],loc='lower left')\n",
    "    pairs = [((item,'low'),(item,'high')) for item in ls_marker]\n",
    "    annot = Annotator(ax, pairs, data=df_plot,x='Gene',y='Expression',hue='pORG',\n",
    "                      order=ls_marker,hue_order=('low','high'))\n",
    "    annot.configure(test='t-test_ind', text_format='simple',fontsize=7,comparisons_correction='fdr_bh') #\n",
    "    annot.apply_test()\n",
    "    d_pval = dict([(res.data.group1[0],res.data.pvalue) for res in annot.annotations])\n",
    "    pvalues = [d_pval[item] for item in ls_marker]\n",
    "    reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='fdr_bh')\n",
    "    formatted_pvalues = [f'p={pvalue:.2}' for pvalue in list(corrected)]\n",
    "    annot.set_custom_annotations(formatted_pvalues)\n",
    "    annot.annotate()\n",
    "    ax.set_title(f'{s_porg} Quartiles') \n",
    "    plt.tight_layout()\n",
    "    fig.savefig(f'figures/gene_expression_binary_{s_porg}_Quartiles.png')\n",
    "    plt.close(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8b802b9",
   "metadata": {},
   "source": [
    "## Section 7 <a name=\"split\"></a>\n",
    "\n",
    "\n",
    "GSEA plots. I think we should say we are showing any pathway with FDRq < 0.15 and NES>1.5 for all for datasets, so if a pathway doesn’t meet these cutoffs, they are not shown as a bar in the two graphs: liver/lung and high/low pORG, and high/low PurIST and high/low pSUB. \n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "056afc85",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#generate more GSEA\n",
    "ls_columns = ['NAME', 'SIZE', 'ES', 'NES', 'NOM.p.val', 'FDR.q.val', 'FWER.p.val',#'RANK.AT.MAX'\n",
    "       ]\n",
    "ls_gsea = ['Top4th_PurIST.Score_vs_Bottom4th_PurIST.Score_h.xlsx',\n",
    "     'basal-like_vs_classical_h.xlsx','LiverCohort_vs_LungNotLiverCohort_h.xlsx',\n",
    " 'Top4th_pSUB.1eNeg4_vs_Bottom4th_pSUB.1eNeg4_h.xlsx',\n",
    " 'Top4th_pSUB.1eNeg5_vs_Bottom4th_pSUB.1eNeg5_h.xlsx',\n",
    " 'Top4th_pORG.20_vs_Bottom4th_pORG.20_h.xlsx'\n",
    "    ]\n",
    "d_en = {}\n",
    "for s_gsea in ls_gsea:\n",
    "    d_load =  pd.read_excel(f'data/{s_gsea}',sheet_name=None)#,index_col=0\n",
    "    df_up = d_load['GSEA_UP'].loc[:,ls_columns] #df[df.ES>0]\n",
    "    df_down = d_load['GSEA_DN'].loc[:,ls_columns]#df[df.ES<0]\n",
    "    d_en.update({f\"{s_gsea.split('_h.xlsx')[0]}_UP\":df_up})\n",
    "    d_en.update({f\"{s_gsea.split('_h.xlsx')[0]}_DN\":df_down})\n",
    "    #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a7f0a1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_labels = {'TopVsBot4th_PurIST':'Top vs Bottom Quartile by PurIST',\n",
    "            'TopVsBot4th_pSUB':'Top vs Bottom Quartile by pSUB',\n",
    "            'basal-like_vs_classical':'Basal-Like vs. Classical',\n",
    "            'Top4th_PurIST.Score_vs_Bottom4th_PurIST.Score':'Top vs Bottom Quartile by PurIST',\n",
    "           'TopVsBot4th_pORG':'Top vs Bottom Quartile by pORG 78 gene',\n",
    "            'LungVsLiver':'Liver Cohort vs. Lung Cohort',\n",
    "            'LiverCohort_vs_LungNotLiverCohort':'Liver Cohort vs. Lung Cohort',\n",
    "           'Top4th_pORG.14_vs_Bottom4th_pORG.14':'Top vs Bottom Quartile by pORG 0.14',\n",
    "           'Top4th_pORG.20_vs_Bottom4th_pORG.20':'Top vs Bottom Quartile by pORG 0.2',\n",
    "           'Top4th_pSUB.1eNeg5_vs_Bottom4th_pSUB.1eNeg5':'Top vs Bottom Quartile by pSUB 1e-5',\n",
    "           'Top4th_pSUB.1eNeg4_vs_Bottom4th_pSUB.1eNeg4':'Top vs Bottom Quartile by pSUB 1e-4'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d85cf3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#bubble plots\n",
    "# organo\n",
    "sorter_combined =  ['HALLMARK_MYOGENESIS' ,'HALLMARK_CHOLESTEROL_HOMEOSTASIS',\n",
    "  'HALLMARK_ANDROGEN_RESPONSE',\n",
    "  'HALLMARK_OXIDATIVE_PHOSPHORYLATION',\n",
    "  'HALLMARK_DNA_REPAIR',\n",
    "  'HALLMARK_INTERFERON_ALPHA_RESPONSE',\n",
    "  'HALLMARK_E2F_TARGETS',\n",
    "  'HALLMARK_MYC_TARGETS_V1',\n",
    "  'HALLMARK_G2M_CHECKPOINT',\n",
    "  'HALLMARK_MITOTIC_SPINDLE',\n",
    "  'HALLMARK_GLYCOLYSIS',\n",
    "  'HALLMARK_MTORC1_SIGNALING',\n",
    "  'HALLMARK_PROTEIN_SECRETION',\n",
    "                  ]\n",
    "sns.set_palette(('#E69F00','#56B4E9','#D55E00','#0072B2',))\n",
    "ls_plot_items = ['Top4th_pORG.20_vs_Bottom4th_pORG.20',\n",
    "    'LiverCohort_vs_LungNotLiverCohort']\n",
    "df_plot_long = pd.DataFrame()\n",
    "for s_comp in ls_plot_items:\n",
    "    for s_direction in ['UP','DN']:\n",
    "        s_compare =f'{s_comp}_{s_direction}'\n",
    "        print(s_compare)\n",
    "        #find the genes up or down\n",
    "        df_plot_o = d_en[s_compare]\n",
    "        df_plot_o['comparison'] = s_compare\n",
    "        df_plot_long = pd.concat([df_plot_long,df_plot_o])\n",
    "#plot\n",
    "df_plot = df_plot_long[df_plot_long.NAME.isin(sorter_combined)]\n",
    "d_rename = {'Top4th_pORG.20_vs_Bottom4th_pORG.20_UP':'pORG_UP',\n",
    "       'Top4th_pORG.20_vs_Bottom4th_pORG.20_DN':'pORG_DOWN',\n",
    "       'LiverCohort_vs_LungNotLiverCohort_UP':'Liver_UP',\n",
    "       'LiverCohort_vs_LungNotLiverCohort_DN':'Liver_DN'}\n",
    "df_plot['comparison'] = df_plot.comparison.replace(d_rename)\n",
    "fig, ax = plt.subplots(dpi=300,figsize=(8,4))\n",
    "sns.scatterplot(data=df_plot, y='NAME',x='NES',hue='comparison',size='FDR.q.val',ax=ax,\n",
    "            )\n",
    "ax.set_title('Organotropism')\n",
    "ax.legend(bbox_to_anchor=(1,1))\n",
    "ax.set_ylabel('')\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53e7045c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "609d1540",
   "metadata": {},
   "outputs": [],
   "source": [
    "#subtype\n",
    "sorter_combined = ['HALLMARK_XENOBIOTIC_METABOLISM',\n",
    "  'HALLMARK_PEROXISOME',\n",
    "  'HALLMARK_FATTY_ACID_METABOLISM',\n",
    "  'HALLMARK_BILE_ACID_METABOLISM',\n",
    "  'HALLMARK_PANCREAS_BETA_CELLS','HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION',\n",
    "  'HALLMARK_APICAL_JUNCTION',\n",
    "  'HALLMARK_HYPOXIA',\n",
    "  'HALLMARK_GLYCOLYSIS']\n",
    "\n",
    "sns.set_palette(('#000000','#CC79A7','#F0E442','#009E73',))\n",
    "ls_plot_items = [ 'Top4th_PurIST.Score_vs_Bottom4th_PurIST.Score',\n",
    "    'Top4th_pSUB.1eNeg4_vs_Bottom4th_pSUB.1eNeg4',\n",
    "    ]\n",
    "df_plot_long = pd.DataFrame()\n",
    "for s_comp in ls_plot_items:\n",
    "    for s_direction in ['UP','DN']:\n",
    "        s_compare =f'{s_comp}_{s_direction}'\n",
    "        print(s_compare)\n",
    "        #find the genes up or down\n",
    "        df_plot_o = d_en[s_compare]\n",
    "        df_plot_o['comparison'] = s_compare\n",
    "        df_plot_long = pd.concat([df_plot_long,df_plot_o])\n",
    "#plot\n",
    "df_plot = df_plot_long[df_plot_long.NAME.isin(sorter_combined)]\n",
    "d_rename = {'Top4th_pSUB.1eNeg4_vs_Bottom4th_pSUB.1eNeg4_UP':'pSUB_UP',\n",
    "       'Top4th_pSUB.1eNeg4_vs_Bottom4th_pSUB.1eNeg4_DN':'pSUB_DN',\n",
    "       'Top4th_PurIST.Score_vs_Bottom4th_PurIST.Score_UP':'PurIST_UP',\n",
    "       'Top4th_PurIST.Score_vs_Bottom4th_PurIST.Score_DN':'PurIST_DN'}\n",
    "df_plot['comparison'] = df_plot.comparison.replace(d_rename)\n",
    "fig, ax = plt.subplots(dpi=300,figsize=(8,4))\n",
    "sns.scatterplot(data=df_plot, y='NAME',x='NES',hue='comparison',size='FDR.q.val',ax=ax,\n",
    "            alpha=0.8\n",
    "               )\n",
    "ax.set_title('Subtype')\n",
    "ax.legend(bbox_to_anchor=(1,1))\n",
    "ax.set_ylabel('')\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9873a6e9",
   "metadata": {},
   "source": [
    "## figure 2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5558c5a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "216 + 71 - 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "affab052",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df_patient.PurIST_Primary.notna().sum())\n",
    "print(df_patient.loc[:,'pORG_0.2_Primary'].notna().sum())\n",
    "print(df_patient.loc[:,'pSUB1e-04_Primary'].notna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbf0fcbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df_patient.PurIST_Met.notna().sum()) #why only 71 mets have purist?\n",
    "print(df_patient.loc[:, 'pSUB1e-04_Met'].notna().sum())\n",
    "print(df_patient.loc[:, 'pORG_0.2_Met'].notna().sum())\n",
    "216 + 72 - 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63261117",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df_patient.loc[:,['PurIST_Primary','PurIST_Met']].notna().all(axis=1).sum())\n",
    "print(df_patient.loc[:,['pSUB1e-04_Primary','pSUB1e-04_Met']].notna().all(axis=1).sum())\n",
    "print(df_patient.loc[:,['pORG_0.2_Primary','pORG_0.2_Met']].notna().all(axis=1).sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "700e1dfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df_patient.loc[:,'Cohort'].notna().sum())\n",
    "print(df_patient.loc[df_patient.loc[:,'pORG_0.2_Primary'].notna(),'Cohort'].notna().sum())\n",
    "print(df_patient.loc[df_patient.loc[:,'pORG_0.2_Met'].notna(),'Cohort'].notna().sum())\n",
    "print(df_patient.loc[df_patient.loc[:,'PurIST_Met'].notna(),'Cohort'].notna().sum())\n",
    "print(df_patient.loc[df_patient.loc[:,['pORG_0.2_Primary','pORG_0.2_Met']].notna().all(axis=1),'Cohort'].notna().sum())\n",
    "75 + 38 - 7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2a89b66",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#lper patient\n",
    "#Can you redo this removing the non lung and non liver mets in their respective cohorts:\n",
    "importlib.reload(util)\n",
    "import warnings\n",
    "s_out = '20230828_Patient_Metadata.csv'\n",
    "df_patient= pd.read_csv(s_out,index_col=0)\n",
    "s_stats = 'mean'\n",
    "d_order =  {'Cohort':['Liver','Lung'],\n",
    "            'PurIST Subtype':['basal-like','classical']}\n",
    "lls_score = [['pORG_0.2_Primary','pSUB1e-04_Primary','PurIST_Primary'],\n",
    "           ['pORG_0.2_All', 'pSUB1e-04_All','PurIST_All'],\n",
    "          ['pORG_0.2_Met','pSUB1e-04_Met','PurIST_Met'],\n",
    "          ]\n",
    "for ls_score in lls_score:\n",
    "    if ls_score[0].find('All') > -1:\n",
    "        ls_population = ['match','pORG_0.2_Met','pORG_0.2_Primary']\n",
    "    elif ls_score[0].find('Met') > -1:\n",
    "        ls_population = ['pORG_0.2_Met']\n",
    "    else:\n",
    "        ls_population = ['pORG_0.2_Primary']\n",
    "    for s_population in ls_population:\n",
    "        print(f'population {s_population}')\n",
    "        if s_population.find('_Met')>-1:\n",
    "            lb_met = [True,False]\n",
    "        else:\n",
    "            lb_met = [False]\n",
    "        for b_met in lb_met:\n",
    "            for s_score in ls_score:\n",
    "                print(f'score {s_score}')\n",
    "                if s_population == 'match':\n",
    "                    s_population = s_score\n",
    "                    print(f'population {s_population}')\n",
    "                if s_score.find('_All')> -1: #need to calculate and add \"all\" scores\n",
    "                    print('Primaries and Mets')\n",
    "                    df_pri = df_patient.copy()\n",
    "                    df_pri['PurIST_allPrimary'] = df_pri.PurIST_Primary.copy()\n",
    "                    df_pri['PurIST_allMet'] = df_pri.PurIST_Met.copy()\n",
    "                    for s_score_all in ['PurIST','pORG_0.2','pSUB1e-04']:\n",
    "                        df_whole = pd.DataFrame(columns=df_pri.columns)\n",
    "                        for s_all in [f'{s_score_all}_allPrimary',f'{s_score_all}_allMet']:\n",
    "                            #select the mets or primaries (this will prioritze primary over met for matched)\n",
    "                            df_half = df_pri[df_pri.loc[:,s_all.replace(s_score_all,'pORG_0.2')].notna()].copy()\n",
    "                            #only fill in the non duplicate \n",
    "                            df_half = df_half[~df_half.Public_Patient_ID.isin(df_whole.Public_Patient_ID)].copy()\n",
    "                            df_half.rename({s_all:f'{s_score_all}_All'},axis=1,inplace=True)\n",
    "                            #df_whole.index = [str(item) + 'a' for item in df_whole.index]\n",
    "                            with warnings.catch_warnings():\n",
    "                                warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "                                df_whole = pd.concat([df_whole,df_half],axis=0,ignore_index=True)\n",
    "                        #print(len(df_whole))\n",
    "                        df_pri = df_pri.merge(df_whole.loc[:,['Public_Patient_ID',f'{s_score_all}_All']],on='Public_Patient_ID',how='left',suffixes=('','_'))\n",
    "                else:\n",
    "                    df_pri = df_patient.copy()\n",
    "                if s_population.find('_Met')> -1:\n",
    "                    print('mets')\n",
    "                    if b_met: #select mets specific to cohort\n",
    "                        df_pri = df_pri[(df_pri.loc[:,s_population].notna()) & (df_patient.Specimen_Site_Met.isin(['Liver','Lung']))].copy()\n",
    "                    else: #select mets\n",
    "                        df_pri = df_pri[(df_pri.loc[:,s_population].notna())].copy()\n",
    "                elif s_population.find('_Primary')> -1: #select primaries\n",
    "                    print('primaries')\n",
    "                    df_pri = df_pri[df_pri.loc[:,s_population].notna()].copy()\n",
    "                df_pri['PurIST_All'] = df_pri.PurIST_Primary.fillna(df_pri.PurIST_Met)\n",
    "                df_pri.loc[df_pri.loc[:,f'PurIST_{s_score.split(\"_\")[-1]}'] > 0.5,'PurIST Subtype'] = 'basal-like'\n",
    "                df_pri.loc[df_pri.loc[:,f'PurIST_{s_score.split(\"_\")[-1]}'] <= 0.5,'PurIST Subtype'] = 'classical'\n",
    "                df_pri.loc[df_pri.Public_Patient_ID.duplicated(),'PurIST Subtype'] = np.nan\n",
    "                df_both,d_pval,order = util.violin_stats2(df_pri,d_order,s_score,s_stats)\n",
    "                fig,pvalues,corrected = util.plot_violins2(df_both,d_pval,d_order,s_stats,s_score,order,\n",
    "                                                           d_colorblind,s_population,b_correct=True,figsize=(4.7,2.8))\n",
    "                plt.ylim(-0.7, 1.13)\n",
    "                if s_score.find('PurIST') > -1:\n",
    "                    plt.ylim(-0.1, 1.2)\n",
    "                fig.savefig(f'figures/violinplot_fig2_{s_score}_{s_population.split(\"_\")[-1]}_mets_only{b_met}.png')\n",
    "            #break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "207770d8",
   "metadata": {},
   "source": [
    "## per specimen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7cef8c9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#per specimen\n",
    "#Can you redo this removing the non lung and non liver mets in their respective cohorts:\n",
    "importlib.reload(util)\n",
    "import warnings\n",
    "s_out = '20230828_Patient_Metadata.csv'\n",
    "df_patient= pd.read_csv(s_out,index_col=0)\n",
    "df_patient['PurIST_allPrimary'] = df_patient.PurIST_Primary.copy()\n",
    "df_patient['PurIST_allMet'] = df_patient.PurIST_Met.copy()\n",
    "s_stats = 'mean'\n",
    "d_order =  {'Cohort':['Liver','Lung'],\n",
    "            'PurIST Subtype':['basal-like','classical']}\n",
    "lls_score = [['pORG_0.2_All', 'pSUB1e-04_All','PurIST_All']]\n",
    "for ls_score in lls_score:\n",
    "    ls_population = ['match']\n",
    "    for s_population in ls_population:\n",
    "        print(f'population {s_population}')\n",
    "        if s_population.find('_Met')>-1:\n",
    "            lb_met = [True,False]\n",
    "        else:\n",
    "            lb_met = [False]\n",
    "        for b_met in lb_met:\n",
    "            for s_score in ls_score:\n",
    "                print(f'score {s_score}')\n",
    "                if s_population == 'match':\n",
    "                    s_population = s_score\n",
    "                print(f'population {s_population}')\n",
    "\n",
    "                df_pri = pd.concat([df_pri,df_met],axis=0,ignore_index=True)\n",
    "                print(df_pri.Public_Specimen_ID.duplicated().any())\n",
    "                print(f'specimen data {len(df_pri)}')\n",
    "                met_index = df_patient.loc[:,['pORG_0.2_allMet','pSUB1e-04_allMet']].dropna(how='any').index #72\n",
    "                df_met = df_patient.loc[met_index,['Public_Patient_ID','pORG_0.2_allMet','pSUB1e-04_allMet','PurIST_allMet']] #72\n",
    "                df_met['Public_Specimen_ID'] = [item + f\"-Met\" for item in df_met.Public_Patient_ID]\n",
    "                print(len(df_met))         \n",
    "                pri_index = df_patient.loc[:,['pORG_0.2_allPrimary','pSUB1e-04_allPrimary']].dropna(how='any').index #72\n",
    "                df_primary = df_patient.loc[pri_index,['Public_Patient_ID','pORG_0.2_allPrimary','pSUB1e-04_allPrimary','PurIST_allPrimary']] #216\n",
    "                df_primary['Public_Specimen_ID'] = [item + f\"-Primary\" for item in df_primary.Public_Patient_ID]\n",
    "                print(len(df_primary))   \n",
    "                df_pri = pd.concat([df_met,df_primary],ignore_index=True)\n",
    "                df_pri['pORG_0.2_All'] = df_pri.loc[:,'pORG_0.2_allPrimary'].fillna(df_pri.loc[:,'pORG_0.2_allMet'])\n",
    "                df_pri['pSUB1e-04_All'] = df_pri.loc[:,'pSUB1e-04_allPrimary'].fillna(df_pri.loc[:,'pSUB1e-04_allMet'])\n",
    "                df_pri['PurIST_All'] = df_pri.PurIST_allPrimary.fillna(df_pri.PurIST_allMet)\n",
    "                df_pri.loc[df_pri.loc[:,f'PurIST_{s_score.split(\"_\")[-1]}'] > 0.5,'PurIST Subtype'] = 'basal-like'\n",
    "                df_pri.loc[df_pri.loc[:,f'PurIST_{s_score.split(\"_\")[-1]}'] <= 0.5,'PurIST Subtype'] = 'classical'\n",
    "                df_pri['Cohort'] = df_pri.Public_Patient_ID.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Cohort)))\n",
    "                df_both,d_pval,order = util.violin_stats2(df_pri,d_order,s_score,s_stats)\n",
    "                fig,pvalues,corrected = util.plot_violins2(df_both,d_pval,d_order,s_stats,s_score,order,\n",
    "                                                           d_colorblind,s_population,b_correct=True,figsize=(4.7,2.8))\n",
    "                plt.ylim(-0.7, 1.13)\n",
    "                if s_score.find('PurIST') > -1:\n",
    "                    plt.ylim(-0.1, 1.2)\n",
    "                fig.savefig(f'figures/violinplot_fig2_spec_{s_score}_{s_population.split(\"_\")[-1]}_mets_only{b_met}.png')\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8af1883",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8209dd2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "               \n",
    "#                 df_pri = df_patient.copy()\n",
    "\n",
    "#                 df_met = df_pri[df_pri.loc[:,['pORG_0.2_Primary','pORG_0.2_Met']].notna().all(axis=1)].copy()\n",
    "#                 df_met.loc[:,['pORG_0.2_allPrimary','pSUB1e-04_allPrimary','PurIST_allPrimary']] = np.nan\n",
    "#                 break\n",
    "#                 df_pri.loc[:,['pORG_0.2_allMet','pSUB1e-04_allMet','PurIST_allMet']] = np.nan\n",
    "#                 #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec49979c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "012dba64",
   "metadata": {},
   "outputs": [],
   "source": [
    "#old bar plots of GSEA\n",
    "# #subtype\n",
    "# sorter_combined = ['HALLMARK_PEROXISOME','HALLMARK_XENOBIOTIC_METABOLISM', #0.05\n",
    "#    'HALLMARK_PANCREAS_BETA_CELLS', 'HALLMARK_BILE_ACID_METABOLISM',\n",
    "#     'HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION','HALLMARK_GLYCOLYSIS',\n",
    "#                   ]\n",
    "# sorter_combined = ['HALLMARK_XENOBIOTIC_METABOLISM',\n",
    "#   'HALLMARK_PEROXISOME',\n",
    "#   'HALLMARK_FATTY_ACID_METABOLISM',\n",
    "#   'HALLMARK_BILE_ACID_METABOLISM',\n",
    "#   'HALLMARK_PANCREAS_BETA_CELLS','HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION',\n",
    "#   'HALLMARK_APICAL_JUNCTION',\n",
    "#   'HALLMARK_HYPOXIA',\n",
    "#   'HALLMARK_GLYCOLYSIS']\n",
    "\n",
    "# #sorter_all\n",
    "# b_manual = True #True #False\n",
    "# sorter_all = {}\n",
    "# #Rosie ask: rank by NES in pORG\n",
    "# rescale = lambda y: (y - np.min(y)) / (np.max(y) - np.min(y))\n",
    "# d_p = {}\n",
    "# es_add = set()#set(['HALLMARK_MYOGENESIS'])\n",
    "# s_pval = 'FDR.q.val'#'NES'#'NOM.p.val'#\n",
    "# alpha = 0.15#0.2#\n",
    "# df_all = pd.DataFrame()\n",
    "# es_marker = set()\n",
    "# ls_plot_items = [#'basal-like_vs_classical',\n",
    "#     #'Top4th_pSUB.1eNeg5_vs_Bottom4th_pSUB.1eNeg5',\n",
    "#     'Top4th_pSUB.1eNeg4_vs_Bottom4th_pSUB.1eNeg4',\n",
    "#     'Top4th_PurIST.Score_vs_Bottom4th_PurIST.Score'\n",
    "#                 ]\n",
    "\n",
    "# for s_comp in ls_plot_items:\n",
    "#     df_plot_long = pd.DataFrame()\n",
    "#     df_plot_long2 = pd.DataFrame()\n",
    "#     for s_direction in ['UP','DN']:\n",
    "#         s_compare =f'{s_comp}_{s_direction}'\n",
    "#         print(s_compare)\n",
    "#         #find the genes up or down\n",
    "#         df_plot_o = d_en[s_compare][(d_en[s_compare].loc[:,s_pval]<alpha)].sort_values(by='NES',ascending=False)\n",
    "#         if s_direction == 'DN':\n",
    "#             print('down')\n",
    "#             #print(len(df_plot))\n",
    "#             sorter = df_plot_o.sort_values(by='NES',ascending=False).NAME.tolist()\n",
    "#         else:\n",
    "#             print('up')\n",
    "#             #print(len(df_plot))\n",
    "#             sorter = df_plot_o.sort_values(by='NES').NAME.tolist()\n",
    "#         #override\n",
    "#         if b_manual:\n",
    "#             df_plot = d_en[s_compare].loc[d_en[s_compare].NAME.isin(sorter_combined),ls_columns]\n",
    "#         else:\n",
    "#             df_plot = df_plot_o\n",
    "#         print(len(df_plot))\n",
    "#         df_plot_long = pd.concat([df_plot_long,df_plot])\n",
    "#         #df_plot_long2 = pd.concat([df_plot_long2,d_en[s_compare]])\n",
    "#         es_marker = es_marker.union(set(df_plot.NAME)).union(es_add)\n",
    "#         sorter_all.update({s_compare:sorter})\n",
    "#     #'''\n",
    "#     df_plot_long = df_plot_long.sort_values(by='NES')\n",
    "#     df_plot_long['comparison'] = s_comp\n",
    "#     #df_plot_long2 = df_plot_long2.sort_values(by='NES',ascending=False)\n",
    "#     #print(df_plot_long2.head())\n",
    "#     df_all = pd.concat([df_all,df_plot_long])\n",
    "#     d_p.update({s_comp:df_plot_long.NAME.unique()})\n",
    "#     #break\n",
    "# #'''\n",
    "# df_all.NAME = df_all.NAME.astype('category')\n",
    "# df_all.NAME = df_all.NAME.cat.set_categories(sorter_combined)\n",
    "# df_plot_bar = df_all[df_all.NAME.isin(sorter_combined)].sort_values(by=['comparison','NAME'])\n",
    "# df_plot_bar['FDR_color'] = rescale(df_plot_bar.loc[:,'FDR.q.val'])\n",
    "# #erase low FDR.q\n",
    "# #df_plot_bar.loc[(abs(df_plot_bar.loc[:,'NES']) < 1.5) | (df_plot_bar.loc[:,'FDR.q.val'] > 0.15),'NES'] = 0\n",
    "# fig, ax = plt.subplots(dpi=200)\n",
    "# my_cmap = plt.get_cmap(\"Reds_r\")\n",
    "# height=0.4\n",
    "# for idx, s_comp in enumerate(ls_plot_items):\n",
    "#     df_comp = df_plot_bar[df_plot_bar.comparison==s_comp]\n",
    "#     df_comp.set_index('NAME',inplace=True)\n",
    "#     if idx == 0:\n",
    "#         indices = np.arange(len(df_comp.index))\n",
    "#         ax.barh(y=indices+height/2, width=df_comp.loc[sorter_combined,'NES'],height=height,\n",
    "#             color=[my_cmap(item) for item in df_comp.FDR_color],label=s_comp,\n",
    "#            )\n",
    "#     else: \n",
    "#         ax.barh(y=indices-height/2, width=df_comp.loc[sorter_combined,'NES'],height=height,hatch='//',\n",
    "#             color=[my_cmap(item) for item in df_comp.FDR_color],label=s_comp,\n",
    "#            )\n",
    "#     ax.set_yticks(range(len(df_comp.index)))\n",
    "#     ax.set_yticklabels(df_comp.index)\n",
    "# norm = mpl.colors.Normalize(vmin=df_plot_bar.loc[:,'FDR.q.val'].min(),vmax=df_plot_bar.loc[:,'FDR.q.val'].max())\n",
    "# #norm = mpl.colors.Normalize(vmin=0,vmax=0.2)\n",
    "# mappable = mpl.cm.ScalarMappable(norm=norm, cmap=my_cmap)\n",
    "# fig.colorbar(mappable=mappable,ax=ax,label='FDR.q.val')\n",
    "# ax.set_title('Subtype')\n",
    "# handles = [mpl.patches.Patch(facecolor='lightgray', edgecolor='black',\n",
    "#                          label=d_labels[ls_plot_items[0]]),\n",
    "#                   mpl.patches.Patch(facecolor='lightgray', edgecolor='black',\n",
    "#                          label=d_labels[ls_plot_items[1]],#label='Top vs Bottom Quartile by PurIST',\n",
    "#                                     hatch='//')]\n",
    "# ax.legend(handles=handles,bbox_to_anchor = (1.25,1),markerscale=1,title='Comparison')\n",
    "# ax.set_xlabel('NES')\n",
    "# #'''\n",
    "\n",
    "# # organo\n",
    "# sorter_combined =  ['HALLMARK_MYOGENESIS' ,'HALLMARK_CHOLESTEROL_HOMEOSTASIS',\n",
    "#   'HALLMARK_ANDROGEN_RESPONSE',\n",
    "#   'HALLMARK_OXIDATIVE_PHOSPHORYLATION',\n",
    "#   'HALLMARK_DNA_REPAIR',\n",
    "#   'HALLMARK_INTERFERON_ALPHA_RESPONSE',\n",
    "#   'HALLMARK_E2F_TARGETS',\n",
    "#   'HALLMARK_MYC_TARGETS_V1',\n",
    "#   'HALLMARK_G2M_CHECKPOINT',\n",
    "#   'HALLMARK_MITOTIC_SPINDLE',\n",
    "#   'HALLMARK_GLYCOLYSIS',\n",
    "#   'HALLMARK_MTORC1_SIGNALING',\n",
    "#   'HALLMARK_PROTEIN_SECRETION',\n",
    "#                   ]\n",
    "# b_manual = True #True #False\n",
    "# sorter_all = {}\n",
    "# #Rosie ask: rank by NES in pORG\n",
    "# rescale = lambda y: (y - np.min(y)) / (np.max(y) - np.min(y))\n",
    "# d_p = {}\n",
    "# es_add = set()#set(['HALLMARK_MYOGENESIS'])\n",
    "# s_pval = 'FDR.q.val'#'NES'#'NOM.p.val'#\n",
    "# alpha = 0.05#0.2#\n",
    "# df_all = pd.DataFrame()\n",
    "# es_marker = set()\n",
    "# ls_plot_items = ['Top4th_pORG.20_vs_Bottom4th_pORG.20',\n",
    "#     'LiverCohort_vs_LungNotLiverCohort']\n",
    "\n",
    "# for s_comp in ls_plot_items:\n",
    "#     df_plot_long = pd.DataFrame()\n",
    "#     df_plot_long2 = pd.DataFrame()\n",
    "#     for s_direction in ['UP','DN']:\n",
    "#         s_compare =f'{s_comp}_{s_direction}'\n",
    "#         print(s_compare)\n",
    "#         #find the genes up or down\n",
    "#         df_plot_o = d_en[s_compare][(d_en[s_compare].loc[:,s_pval]<alpha)].sort_values(by='NES',ascending=False)\n",
    "#         if s_direction == 'DN':\n",
    "#             print('down')\n",
    "#             #print(len(df_plot))\n",
    "#             sorter = df_plot_o.sort_values(by='NES',ascending=False).NAME.tolist()\n",
    "#         else:\n",
    "#             print('up')\n",
    "#             #print(len(df_plot))\n",
    "#             sorter = df_plot_o.sort_values(by='NES').NAME.tolist()\n",
    "#         #override\n",
    "#         if b_manual:\n",
    "#             df_plot = d_en[s_compare].loc[d_en[s_compare].NAME.isin(sorter_combined),ls_columns]\n",
    "#         else:\n",
    "#             df_plot = df_plot_o\n",
    "#         print(len(df_plot))\n",
    "#         df_plot_long = pd.concat([df_plot_long,df_plot])\n",
    "#         #df_plot_long2 = pd.concat([df_plot_long2,d_en[s_compare]])\n",
    "#         es_marker = es_marker.union(set(df_plot.NAME)).union(es_add)\n",
    "#         sorter_all.update({s_compare:sorter})\n",
    "#     #'''\n",
    "#     df_plot_long = df_plot_long.sort_values(by='NES')\n",
    "#     df_plot_long['comparison'] = s_comp\n",
    "#     #df_plot_long2 = df_plot_long2.sort_values(by='NES',ascending=False)\n",
    "#     #print(df_plot_long2.head())\n",
    "#     df_all = pd.concat([df_all,df_plot_long])\n",
    "#     d_p.update({s_comp:df_plot_long.NAME.unique()})\n",
    "#     #break\n",
    "# #'''\n",
    "# df_all.NAME = df_all.NAME.astype('category')\n",
    "# df_all.NAME = df_all.NAME.cat.set_categories(sorter_combined)\n",
    "# df_plot_bar = df_all[df_all.NAME.isin(sorter_combined)].sort_values(by=['comparison','NAME'])\n",
    "# df_plot_bar['FDR_color'] = rescale(df_plot_bar.loc[:,'FDR.q.val'])\n",
    "# #erase low FDR.q\n",
    "# #df_plot_bar.loc[(abs(df_plot_bar.loc[:,'NES']) < 1.5) | (df_plot_bar.loc[:,'FDR.q.val'] > 0.15),'NES'] = 0\n",
    "# fig, ax = plt.subplots(dpi=200)\n",
    "# my_cmap = plt.get_cmap(\"Reds_r\")\n",
    "# height=0.4\n",
    "# for idx, s_comp in enumerate(ls_plot_items):\n",
    "#     df_comp = df_plot_bar[df_plot_bar.comparison==s_comp]\n",
    "#     df_comp.set_index('NAME',inplace=True)\n",
    "#     if idx == 0:\n",
    "#         indices = np.arange(len(df_comp.index))\n",
    "#         ax.barh(y=indices+height/2, width=df_comp.loc[sorter_combined,'NES'],height=height,\n",
    "#             color=[my_cmap(item) for item in df_comp.FDR_color],label=s_comp,\n",
    "#            )\n",
    "#     else: \n",
    "#         ax.barh(y=indices-height/2, width=df_comp.loc[sorter_combined,'NES'],height=height,hatch='//',\n",
    "#             color=[my_cmap(item) for item in df_comp.FDR_color],label=s_comp,\n",
    "#            )\n",
    "#     ax.set_yticks(range(len(df_comp.index)))\n",
    "#     ax.set_yticklabels(df_comp.index)\n",
    "# norm = mpl.colors.Normalize(vmin=df_plot_bar.loc[:,'FDR.q.val'].min(),vmax=df_plot_bar.loc[:,'FDR.q.val'].max())\n",
    "# #norm = mpl.colors.Normalize(vmin=0,vmax=0.2)\n",
    "# mappable = mpl.cm.ScalarMappable(norm=norm, cmap=my_cmap)\n",
    "# fig.colorbar(mappable=mappable,ax=ax,label='FDR.q.val')\n",
    "# ax.set_title('Organotropism')\n",
    "# handles = [mpl.patches.Patch(facecolor='lightgray', edgecolor='black',\n",
    "#                          label=d_labels[ls_plot_items[0]]),\n",
    "#                   mpl.patches.Patch(facecolor='lightgray', edgecolor='black',\n",
    "#                          label=d_labels[ls_plot_items[1]],#label='Top vs Bottom Quartile by PurIST',\n",
    "#                                     hatch='//')]\n",
    "# ax.legend(handles=handles,bbox_to_anchor = (1.25,1),markerscale=1,title='Comparison')\n",
    "# ax.set_xlabel('NES')\n",
    "# #'''\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
