{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78a56982",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import shutil\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "from skimage import io\n",
    "import tifffile\n",
    "from scipy.ndimage import median_filter\n",
    "from skimage.util import img_as_ubyte,  img_as_float\n",
    "import skimage\n",
    "from skimage.feature import blob_dog, blob_log, blob_doh\n",
    "from math import sqrt\n",
    "import scipy\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "\n",
    "import lifelines\n",
    "from lifelines import KaplanMeierFitter, CoxPHFitter\n",
    "from lifelines.statistics import multivariate_logrank_test\n",
    "\n",
    "from statannotations.Annotator import Annotator\n",
    "from itertools import combinations\n",
    "import statsmodels\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning) \n",
    "\n",
    "# Set Paths\n",
    "codedir = os.getcwd()\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fb9d6a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#images: dowload from synapse.org syn51068458\n",
    "# (free account required)\n",
    "rootdir = '/home/groups/BCC_Chin_Lab/ChinData/Cyclic_Workflow/cmIF_2021-05-03_PDAC'\n",
    "regdir = f'{rootdir}/RegisteredImages'\n",
    "segdir = f'{codedir}/Segmentation'\n",
    "segdiro = f'{rootdir}/Segmentation'\n",
    "\n",
    "# clone mplex_image at https://gitlab.com/engje/mplex_image\n",
    "os.chdir('/home/groups/graylab_share/data/engje/Data/')\n",
    "from mplex_image import preprocess, mpimage, cmif\n",
    "os.chdir(codedir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ddd69b4",
   "metadata": {},
   "source": [
    "# Table of contents <a name=\"contents\"></a>\n",
    "0. [functions](#func)\n",
    "1. [skimage blobs](#sk)\n",
    "2. [Foci analysis](#focifoci)\n",
    "3. [mIHC analysis](#mihc)\n",
    "4. [CPH modeling](#clin)\n",
    "5. [NEW CPH modeling](#newsurv)\n",
    "5. [gene expression analysis](#geneexp)\n",
    "6. [TCR analysis](#tcell)\n",
    "7. [High v. low pORG](#split)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f1ef3f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy\n",
    "from skimage.filters import unsharp_mask\n",
    "import skimage\n",
    "from skimage.restoration import (denoise_tv_chambolle, denoise_bilateral,\n",
    "                                 denoise_wavelet, estimate_sigma)\n",
    "from skimage import color, morphology\n",
    "from skimage.transform import rescale"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "604aaac5",
   "metadata": {},
   "source": [
    "## functions <a name=\"func\"></a> \n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a57878d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_blobs2(image_gray,min_sigma,max_sigma,threshold,exclude_border):\n",
    "\n",
    "    blobs_dog = blob_dog(image_gray,  min_sigma=min_sigma, max_sigma=max_sigma, threshold=threshold,exclude_border=exclude_border)\n",
    "    blobs_dog[:, 2] = blobs_dog[:, 2] * sqrt(2)\n",
    "\n",
    "    blobs_list = [image_gray,  blobs_dog] #blobs_doh ,\n",
    "    colors = ['red','red', ]\n",
    "    titles = ['Original','Difference of Gaussian',#\n",
    "              ]\n",
    "    sequence = zip(blobs_list, colors, titles)\n",
    "\n",
    "    fig, axes = plt.subplots(1, 2, figsize=(6, 3), sharex=True, sharey=True)\n",
    "    ax = axes.ravel()\n",
    "\n",
    "    for idx, (blobs, color, title) in enumerate(sequence):\n",
    "        if idx == 1:\n",
    "            ax[idx].set_title(f'{title}\\nmin={min_sigma} max={max_sigma} thresh={threshold}')\n",
    "        else:\n",
    "            ax[idx].set_title(f'{title}')\n",
    "        ax[idx].imshow(image)\n",
    "        if not title == 'Original':\n",
    "            for blob in blobs:\n",
    "                y, x, r = blob\n",
    "                c = plt.Circle((x, y), r, color=color, linewidth=2, fill=False)\n",
    "                ax[idx].add_patch(c)\n",
    "        #ax[idx].set_axis_off()\n",
    "\n",
    "    plt.tight_layout()\n",
    "    plt.close(fig)\n",
    "    return(blobs_dog,fig)\n",
    "\n",
    "def km_plot(df,s_col,s_time,s_censor):\n",
    "    results = multivariate_logrank_test(event_durations=df.loc[:,s_time],\n",
    "                                    groups=df.loc[:,s_col], event_observed=df.loc[:,s_censor])        \n",
    "    kmf = KaplanMeierFitter()\n",
    "    fig, ax = plt.subplots(figsize=(4,4),dpi=300)\n",
    "    ls_order = sorted(df.loc[:,s_col].dropna().unique())\n",
    "    for s_group in ls_order:\n",
    "        print(s_group)\n",
    "        df_abun = df[df.loc[:,s_col]==s_group]\n",
    "        durations = df_abun.loc[:,s_time]\n",
    "        event_observed = df_abun.loc[:,s_censor]\n",
    "        kmf.fit(durations,event_observed,label=s_group)\n",
    "        kmf.plot(ax=ax,ci_show=True,show_censors=True)\n",
    "    ax.set_title(f'{s_col}\\np={results.summary.p[0]:.2} n={[df.loc[:,s_col].value_counts()[item] for item in ls_order]}')\n",
    "    ax.set_ylim(-0.05,1.05)\n",
    "    return(fig,ls_order)\n",
    "\n",
    "def cph_plot(df,s_multi,s_time,s_censor,figsize=(3,3)):\n",
    "    cph = CoxPHFitter()  #penalizer=0.1\n",
    "    if df.columns.isin(['Stage']).any():\n",
    "        df.Stage = df.Stage.replace({'I':1,'II':2,'III':3,'IV':4})\n",
    "    cph.fit(df.dropna(), s_time, event_col=s_censor) \n",
    "    fig, ax = plt.subplots(figsize=figsize,dpi=300)\n",
    "    cph.plot(ax=ax)\n",
    "    pvalue = cph.summary.p[s_multi]\n",
    "    ax.set_title(f'{s_multi}\\np={pvalue:.2} n={(len(df.dropna()))}')\n",
    "    plt.tight_layout()\n",
    "    return(fig,cph)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ba7147b",
   "metadata": {},
   "source": [
    "##  Section 1: skimage blob detection\n",
    "\n",
    "Foci are detected here from images plus segmentation masks\n",
    "\n",
    "**You can skip and use pre-computed foci counts**\n",
    "\n",
    " <a name=\"sk\"></a> \n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f737cb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#intensity data: dowload from synapse.org syn51068458\n",
    "df_mi = pd.read_csv(f'{codedir}/data/20220720_U54-TMA_FilteredMeanIntensity_Link.csv',index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0f1a0ba",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#skimage blob detection\n",
    "#requires images: dowload from synapse.org syn51068458\n",
    "threshold=0.002\n",
    "d_thresh = {'pRPA':1100, 'gH2AX':1100, 'RAD51':1300}\n",
    "os.chdir(regdir)\n",
    "ls_slide = sorted(set(df_mi.slide_scene)) \n",
    "ls_marker = ['pRPA','gH2AX','RAD51',] \n",
    "df_result_all = pd.DataFrame()\n",
    "for s_slide in ls_slide:\n",
    "    print(s_slide)\n",
    "    os.chdir(f'{segdiro}/U54-TMA-9_CellposeSegmentation') #change this to path for your downloaded segmentation masks\n",
    "    s_seg = pd.Series(sorted(os.listdir()),dtype='object')[pd.Series(index=sorted(os.listdir()),dtype='object').index.str.contains(f'{s_slide}_nuc30_NucleiSegmentationBasins')].iloc[0]\n",
    "    label_image = io.imread(s_seg)\n",
    "    os.chdir(f'{regdir}/{s_slide}') # change this to path to your downloaded registered tifs\n",
    "    df_img=mpimage.parse_org()\n",
    "    for idxs, s_marker in enumerate(ls_marker):\n",
    "        intensity_image = io.imread(df_img[df_img.marker==s_marker].index[0])\n",
    "        props = skimage.measure.regionprops_table(label_image, intensity_image=intensity_image, properties=('label','bbox','mean_intensity')) # 'image','intensity_image',\n",
    "        df_props = pd.DataFrame(props,dtype='float').set_index('label')\n",
    "        df_props.columns = [item.replace('-','') for item in df_props.columns]\n",
    "        ls_index = df_props[df_props.mean_intensity>d_thresh[s_marker]].index\n",
    "        print(f'{s_marker} {len(ls_index)}')\n",
    "        for i_cell in ls_index:\n",
    "            se_cell = df_props.loc[i_cell].dropna().astype('int')\n",
    "            image = intensity_image[se_cell.bbox0:se_cell.bbox2,se_cell.bbox1:se_cell.bbox3]\n",
    "            if s_marker == 'pRPA':\n",
    "                blobs, fig = get_blobs2(image,min_sigma=0.1,max_sigma=2,threshold=threshold,exclude_border=1)\n",
    "            elif s_marker == 'gH2AX':\n",
    "                blobs, fig = get_blobs2(image,min_sigma=1,max_sigma=2,threshold=threshold,exclude_border=1)\n",
    "            else:\n",
    "                blobs, fig = get_blobs2(image,min_sigma=0.1,max_sigma=2,threshold=threshold,exclude_border=1)\n",
    "            df_props.loc[i_cell,'blobs'] = len(blobs)\n",
    "            if len(blobs) > 1:\n",
    "                fig.suptitle(s_marker)\n",
    "                plt.tight_layout()\n",
    "                fig.savefig(f'{codedir}/blobs/{s_marker}/{s_slide}_{s_marker}_{i_cell}.png',dpi=200)\n",
    "                plt.close(fig)\n",
    "        df_props.index = [s_slide + '_cell' + str(item) for item in df_props.index]\n",
    "        if len(ls_index)==0:\n",
    "                df_props['blobs'] = np.NaN\n",
    "        if idxs == 0:\n",
    "            df_result = df_props.rename({'blobs':f'{s_marker}_foci'},axis=1)\n",
    "        else:\n",
    "            df_result.loc[df_props.index,f'{s_marker}_foci'] = df_props.blobs\n",
    "    df_result_all = df_result_all.append(df_result)\n",
    "\n",
    "#uncomment to save (this data has been pre-computed and saved in the repo)\n",
    "#df_result_all.loc[:,df_result_all.columns.str.contains('foci')].dropna(how='all').to_csv(f'{codedir}/foci_U54-TMA-9_{\".\".join(ls_marker)}_{threshold}.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b487760",
   "metadata": {},
   "source": [
    "### Section 2: FOCI Analysis <a name=\"focifoci\"></a>\n",
    "\n",
    "Load saved foci for plotting and downstream analysis\n",
    "\n",
    "**You don't need to run section 1 to run this**\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1de50de",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_surv = pd.read_csv(f'{codedir}/data/u54_tma_sampleannot_Link.csv')\n",
    "len(df_surv)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1d510f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pORG = pd.read_csv(f'{codedir}/data/GSVA_Scores_Link.csv')#=,index_col='Public_Specimen_ID'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cf32cc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_primary = df_surv[((df_surv.Tissue=='PDAC') | (df_surv.Tissue=='Intestinal'))].copy()\n",
    "print(len(df_primary))\n",
    "df_primary = df_primary.merge(df_pORG,on='Public_Specimen_ID',how='left')\n",
    "df_primary.set_index(\"Public_Specimen_ID\", inplace = True)\n",
    "len(df_primary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d508cacc",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mapper = df_primary[~df_primary.index.duplicated()]\n",
    "ls_add = ['txi_pORG_Up_42_Genes','Original_pORG_Up_78_Genes', \n",
    "       'Original_pSUB_Up_100_Genes', 'txi_pSUB_Up_100_Genes']\n",
    "for s_add in ls_add:\n",
    "    d_map = dict(zip(df_mapper.index,df_mapper.loc[:,s_add]))\n",
    "    print(len(d_map))\n",
    "    df_surv[s_add] = df_surv.Public_Specimen_ID.map(d_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f21d2717",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('..')\n",
    "import util\n",
    "os.chdir(codedir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e3ec3a5",
   "metadata": {},
   "source": [
    "### TMA survival"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bb0c145",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59b765f6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "##\n",
    "\n",
    "alpha = 0.05\n",
    "s_propo = ''\n",
    "savedir = f'{codedir}'\n",
    "s_time=  'Survival_time'#\n",
    "s_censor='Survival'#\n",
    "s_subtype = ''\n",
    "s_cell = ''\n",
    "s_type_title = ''\n",
    " \n",
    "for s_col in ls_add[1:4]:\n",
    "    print(s_col)\n",
    "    for cutp in [0.33,0.5,0.66]: #np.round(np.arange(0.25,1,0.25),3):#\n",
    "            print(cutp)\n",
    "            df_km, pvalue = util.single_km(df_surv[~df_surv.Public_Patient_ID.duplicated()],s_cell,s_subtype,s_type_title,s_col,savedir,alpha,cutp, #\n",
    "                                       s_time,s_censor,s_propo)\n",
    "            print(pvalue)\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e4366b4",
   "metadata": {},
   "source": [
    "### define high/low pORG\n",
    "\n",
    "based on survival differences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7916167",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pri = df_surv[~(df_surv.Public_Patient_ID.duplicated(keep='first')) & ((df_surv.Tissue=='PDAC'))]\n",
    "i_pORG_ori = np.quantile(df_pri.Original_pORG_Up_78_Genes.dropna(),0.66)\n",
    "print(i_pORG_ori)\n",
    "i_pORG_txi = np.quantile(df_pri.txi_pORG_Up_42_Genes.dropna(),0.66)\n",
    "print(i_pORG_txi)\n",
    "df_surv.loc[df_surv['Original_pORG_Up_78_Genes'] >= i_pORG_ori,'pORG_binary_orig'] = 'high'\n",
    "df_surv.loc[df_surv['Original_pORG_Up_78_Genes'] < i_pORG_ori,'pORG_binary_orig'] = 'low'\n",
    "\n",
    "df_surv.loc[df_surv['txi_pORG_Up_42_Genes'] >= i_pORG_txi,'pORG_binary_txi'] = 'high'\n",
    "df_surv.loc[df_surv['txi_pORG_Up_42_Genes'] < i_pORG_txi,'pORG_binary_txi'] = 'low'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc795272",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_surv.to_csv(f'{codedir}/data/u54_tma_sampleannot_Link_new.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "192637e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df_surv.pORG_binary_orig.value_counts())\n",
    "print(df_surv.pORG_binary_txi.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "899c9d63",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load combined\n",
    "s_sample = '20220721_U54-TMA'#'20220711_U54-TMA' #'20220409_JP-TMAs_IMC-TMAs'\n",
    "s_names = 'Combined' #'unnamed' #\n",
    "s_type = 'PDAC'\n",
    "df_lei = pd.read_csv(f'{codedir}/data/{s_sample}_{s_names}Celltypes_{s_type}_Link.csv',index_col=0)\n",
    "\n",
    "d_patient = dict(zip(df_surv.coor_mplexable,df_surv.Public_Patient_ID))\n",
    "\n",
    "df_lei['Tissue'] = df_lei.Patient.map(dict(zip(df_surv.Public_Patient_ID,df_surv.Tissue)))\n",
    "df_epi = pd.read_csv(f'{codedir}/data/results_20220721_U54-TMA_CellTypeCounts_byPatient_byleidencelltype5_PDAC_Link.csv',index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f19c9989",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ls_marker = ['gH2AX','pRPA','RAD51']\n",
    "threshold=0.002\n",
    "df_foci2 = pd.read_csv(f'{codedir}/data/foci_U54-TMA-9_{\".\".join(ls_marker)}_{threshold}.csv',index_col=0)\n",
    "df_foci2.index = [item.replace('.0','') for item in df_foci2.index]\n",
    "df_foci2 = df_foci2.fillna(0)\n",
    "df_foci2['scene'] = [item.split('_cell')[0] for item in df_foci2.index]\n",
    "df_foci2['Patient'] = df_foci2.scene.map(d_patient)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68276502",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot Ki67 versus pRPA foci\n",
    "s_define = 'pORG_binary_txi'#'pORG_binary_orig'#'pORG_binary'\n",
    "df_lei['Cohort'] = df_lei.Patient.map(dict(zip(df_surv.Public_Patient_ID,df_surv.Cohort)))\n",
    "df_lei['pORG_binary'] = df_lei.Patient.map(dict(zip(df_surv.Public_Patient_ID,df_surv.loc[:,s_define])))\n",
    "df_lei['Ki67pos'] = np.nan\n",
    "df_lei.loc[df_lei.Ki67>3*256,'Ki67pos'] = 'Ki67+'\n",
    "df_lei.Ki67pos.fillna('Ki67-',inplace=True)\n",
    "ls_foci =['pRPA_foci','gH2AX_foci','RAD51_foci']\n",
    "df_lei_foci = df_lei.merge(df_foci2.loc[:,ls_foci],left_index=True,right_index=True,how='left')\n",
    "df_lei_foci.loc[:,ls_foci] = df_lei_foci.loc[:,ls_foci].fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1274b7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#use df foci sum 2 (lower threshold)\n",
    "ls_index = df_lei[df_lei.leidencelltype5=='epithelial'].index\n",
    "df_foci_sum2 = df_foci2.loc[df_foci2.index.isin(ls_index)].groupby('Patient').sum()\n",
    "for s_marker in ls_marker:\n",
    "    df_foci_sum2[f'log_{s_marker}_foci'] = np.log(df_foci_sum2.loc[:,f'{s_marker}_foci'] + 1)\n",
    "    df_foci_sum2[f'mean_{s_marker}_foci'] = (df_foci_sum2.loc[df_foci_sum2.index.isin(df_epi.index),f'{s_marker}_foci']/df_epi.epithelial).fillna(0)\n",
    "\n",
    "    \n",
    "df_foci_sum2['Public_Patient_ID'] = df_foci_sum2.index\n",
    "df_surv = df_surv.merge(df_foci_sum2,on='Public_Patient_ID',how='left',suffixes=('_1',''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05d2fab1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#compare Ki67+ versus negative, double violinplots\n",
    "#compare all the combinations, double violinplots\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from statannotations.Annotator import Annotator\n",
    "from itertools import combinations\n",
    "import statsmodels\n",
    "from statsmodels.stats.multicomp import pairwise_tukeyhsd\n",
    "\n",
    "s_compare = 'Ki67'#'all' #\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning) \n",
    "for s_cat in ['pORG_binary','Cohort']:\n",
    "    ls_order = ['Ki67-','Ki67+']\n",
    "    for s_foci in ls_foci:\n",
    "        if s_compare == 'all':\n",
    "            figsize=(3.5,3)\n",
    "        else:\n",
    "            figsize=(2.8,2.8)\n",
    "        fig,ax=plt.subplots(dpi=300,figsize=figsize)\n",
    "        df_both = pd.DataFrame()\n",
    "        order = []\n",
    "        d_pval = {}\n",
    "        for idx, s_cohort in enumerate(df_lei_foci.loc[:,s_cat].dropna().unique()):\n",
    "            print(s_cohort)\n",
    "            df_plot_foci =df_lei_foci[(df_lei_foci.leidencelltype5=='epithelial')\n",
    "                                      & (df_lei_foci.Tissue=='PDAC') & (df_lei_foci.loc[:,s_cat]==s_cohort)]\n",
    "            statistic, pvalue = stats.ttest_ind(df_plot_foci.loc[df_plot_foci.Ki67pos=='Ki67+',s_foci],\n",
    "                                                df_plot_foci.loc[df_plot_foci.Ki67pos=='Ki67-',s_foci],\n",
    "                                               alternative='greater')\n",
    "            for s_order in ls_order:\n",
    "                order.append((s_cohort,s_order))\n",
    "            df_both = pd.concat([df_both,df_plot_foci])\n",
    "            d_pval.update({s_cohort:pvalue})\n",
    "        sns.violinplot(data=df_both,hue='Ki67pos',y=s_foci,x=s_cat,ax=ax,alpha=0.5,linewidth=0.5)\n",
    "        sns.stripplot(data=df_both,hue='Ki67pos',y=s_foci,x=s_cat,s=1,dodge=True,ax=ax,palette='dark',jitter=0.2)\n",
    "        #annotate\n",
    "        if s_compare == 'all':\n",
    "            pairs = list(combinations(order, r=2))\n",
    "            annotator = Annotator(ax, pairs=pairs, data=df_both, y=s_foci,x=s_cat,hue='Ki67pos')\n",
    "            annotator.configure(test=\"t-test_ind\",line_width=1)#,alternative='greater\n",
    "            pvalues = annotator.apply_test().annotations #annotator.apply_and_annotate() # \n",
    "            pvalues = [item.data.pvalue for item in pvalues]     \n",
    "            reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='bonferroni')\n",
    "            formatted_pvalues = [f'p={pvalue:.2}' for pvalue in list(corrected)]\n",
    "            annotator.set_custom_annotations(formatted_pvalues)\n",
    "            annotator.annotate()\n",
    "        else:\n",
    "            pairs = [(order[0],order[1]),(order[2],order[3])]\n",
    "            pvalues = [d_pval[pairs[0][0][0]],d_pval[pairs[1][0][0]]]\n",
    "            reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='bonferroni')\n",
    "            formatted_pvalues = [f'p={pvalue:.2}' for pvalue in list(corrected)]\n",
    "            annotator = Annotator(ax, pairs=pairs, data=df_both, y=s_foci,x=s_cat,hue='Ki67pos')\n",
    "            annotator.set_custom_annotations(formatted_pvalues)\n",
    "            annotator.annotate()\n",
    "        \n",
    "        ax.set_title(f\"{s_foci.replace('_',' ')} vs. {s_cat.split('_')[0]}\", fontsize='x-large')\n",
    "        ax.set_xlabel(s_cat)\n",
    "        ax.set_ylabel(f\"No. {s_foci.replace('_',' ')}\")\n",
    "        h, l = ax.get_legend_handles_labels()\n",
    "        labels =  [f'_{item}' if ind < 2 else item for ind,item in enumerate(l)]\n",
    "        ax.legend(h,labels,title='',fontsize='small',markerscale=.5,bbox_to_anchor=(1.01,0.9))\n",
    "        plt.tight_layout()\n",
    "        fig.savefig(f'{codedir}/violinplot_both_{s_foci}_{s_cohort}_{s_compare}.png')\n",
    "        #df_plot_foci.loc[:,ls_col].to_csv(f'results_foci_Ki67_{s_cohort}.csv')\n",
    "        #break\n",
    "    #break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be8bfceb",
   "metadata": {},
   "source": [
    "### Section 3: mIHC Analysis <a name=\"mihc\"></a>\n",
    "\n",
    "Stacked barplot\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "693d274e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#survival\n",
    "\n",
    "df_surv = pd.read_csv(f'{codedir}/data/u54_tma_sampleannot_Link_new.csv',index_col=0)\n",
    "df_surv['Sample_ID'] = [item.split('-T')[0][-6::] for item in df_surv.loc[:,'Public_Patient_ID'].fillna('none')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebe97b0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_surv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7ce65c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dowload mIHC data from https://www.synapse.org/#!Synapse:syn51078766\n",
    "# combine ROIs into large dataframe and save\n",
    "if not os.path.exists(f'data/20221123_mIHC_LiverLung_Celltypes.csv'):\n",
    "    ls_col = ['Sample_ID','class','Location_Center_X', 'Location_Center_Y', \n",
    "           'GRZB_func', 'KI67_func', 'PD1_func', 'PDL1_func', 'CD163_func',\n",
    "           'CCR2_func', 'HLAII_func', 'EOMES_func','Area']\n",
    "    df_ll =pd.DataFrame()\n",
    "    for s_file in sorted(os.listdir('mIHC_Data')):\n",
    "        df = pd.read_csv(f'mIHC_Data/{s_file}',index_col=0)\n",
    "        s_sample = s_file.split('LiverLungBCC')[1].split('.')[0]\n",
    "        df['Sample_ID'] = s_sample\n",
    "        df.index = [f'{s_sample.split(\"Nuclei_\")[1].replace(\"ROI\",\"_scene\")}_cell{item}' for item in df.index]\n",
    "        df_ll = pd.concat([df_ll,df.loc[:,ls_col]])\n",
    "        break\n",
    "    df_ll['Organ'] = df_ll.Sample_ID.map(dict(zip(df_ll_annot.index,df_ll_annot.loc[:,'Met Site'])))\n",
    "    df_ll['Location'] = df_ll.Sample_ID.map(dict(zip(df_ll_annot.index,df_ll_annot.loc[:,'Location'])))\n",
    "    df_ll['Desc'] = df_ll.Sample_ID.map(dict(zip(df_ll_annot.index,df_ll_annot.loc[:,'Desc'])))\n",
    "    df_ll.to_csv(f'data/20221123_mIHC_LiverLung_Celltypes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a6cbcba",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load data\n",
    "df_ll = pd.read_csv(f'data/20221123_mIHC_LiverLung_Celltypes.csv',index_col=0,low_memory=False)\n",
    "print(len(df_ll))\n",
    "\n",
    "df_ll_ann = pd.read_csv('data/LiverLung_annotations.csv',index_col=0)\n",
    "df_ll_roi = pd.read_csv('data/annotated_LiverLung_perROI.csv',index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db35cd8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add annotation\n",
    "df_ll['Sample_ID_short'] = [item.split('Nuclei_')[1].split('ROI')[0] for item in df_ll.Sample_ID]\n",
    "df_ll['Sample_ID_int'] = [int(item) for item in df_ll.Sample_ID_short]\n",
    "df_ll['ROI'] = [int(item.split('ROI')[1]) for item in df_ll.Sample_ID]\n",
    "df_ll['Cohort'] = df_ll.Sample_ID_int.map(dict(zip(df_ll_ann.index, df_ll_ann.Cohort)))\n",
    "df_ll['Patient'] = df_ll.Sample_ID_int.map(dict(zip(df_ll_ann.index, df_ll_ann.Patient)))\n",
    "df_ll['Sample_ROI'] = df_ll.Sample_ID_int.astype('str') + '_' + df_ll.ROI.astype('str')\n",
    "df_ll_roi['Sample_ROI'] = df_ll_roi.index.astype('str') + '_' + df_ll_roi.ROI.astype('str')\n",
    "df_ll['Location'] = df_ll.Sample_ROI.map(dict(zip(df_ll_roi.Sample_ROI,df_ll_roi.loc[:,'Location'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9a04a6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ll['classII'] = df_ll.loc[:,'class'].replace({'T-regulatory CD4 cells':'T cells',\n",
    "                                                 'CD4 T helper cells':'T cells','CD8 T cells':'T cells'}) #\n",
    "df_ll.rename({'class':'classI'},axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ab5c426",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_color = dict(zip(['ST-00016289', 'ST-00017078', 'ST-00017310', 'ST-00017381',\n",
    "       'ST-00018269', 'ST-00018955', 'ST-00019367', 'ST-00019368',\n",
    "       'ST-00020181'],sns.color_palette('Purples',9)))\n",
    "d_color.update(dict(zip(['ST-00015839', 'ST-00017440', 'ST-00017804'],sns.color_palette('Blues',3))))\n",
    "\n",
    "ls_mihc = [ 'PanCK+','aSMA+','CD4 T helper cells','CD8 T cells', 'T-regulatory CD4 cells','B cells', 'Granulocytes', 'Monocyte',\n",
    " 'Macrophage', 'Mature DC','Immature DC',]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a65fb8b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#by location\n",
    "s_define = 'pORG_Score' #'txi_pORG_Up_42_Genes'#''#\n",
    "s_group = 'Cohort'\n",
    "alpha = 0.05\n",
    "s_column ='classI'\n",
    "s_patient = 'Public_Patient_ID'\n",
    "\n",
    "for s_loc in ['all','T','B','D']: \n",
    "    if s_loc == 'all':\n",
    "        df_loc = df_ll\n",
    "    else:\n",
    "        df_loc = df_ll[df_ll.Location==s_loc]\n",
    "    print(len(df_loc.Sample_ID_short.unique()))\n",
    "    df_group = (df_loc.groupby(['Patient',s_column]).count().Sample_ID/(df_loc.groupby(['Patient']).count().Sample_ID)).unstack()\n",
    "    df_group[s_patient] = df_group.index\n",
    "    df_group['Cohort'] = df_group.index.map(dict(zip(df_ll_roi.Patient,df_ll_roi.loc[:,'Cohort'])))\n",
    "    df_group = df_group.merge(df_surv.loc[:,['Survival','Survival_time',s_define,s_patient]],on=s_patient)\n",
    "    df_group = df_group[~df_group.loc[:,s_patient].duplicated()]\n",
    "    df_group = df_group[df_group.Cohort!='Liver met'].fillna(0)\n",
    "    ls_order = sorted(df_group.loc[:,s_group].unique())\n",
    "    for s_marker in ls_mihc:\n",
    "        try:\n",
    "            s_high = df_group.loc[:,s_group].unique()[0]\n",
    "            s_low = df_group.loc[:,s_group].unique()[1]\n",
    "        except:\n",
    "            continue\n",
    "        n_high = sum(df_group.loc[:,s_group]==s_high)\n",
    "        n_low = sum(df_group.loc[:,s_group]==s_low)\n",
    "        statistic,pvalue = stats.ttest_ind(df_group.loc[df_group.loc[:,s_group]==s_high,s_marker],\n",
    "                                           df_group.loc[df_group.loc[:,s_group]==s_low,s_marker])\n",
    "        if pvalue <= alpha:\n",
    "            df_group_roi = (df_loc.groupby(['Sample_ROI',s_column]).count().Sample_ID/(df_loc.groupby(['Sample_ROI']).count().Sample_ID)).unstack()\n",
    "            df_group_roi['Cohort'] = df_group_roi.index.map(dict(zip(df_ll_roi.Sample_ROI,df_ll_roi.loc[:,'Cohort'])))\n",
    "            df_group_roi['Patient'] = df_group_roi.index.map(dict(zip(df_ll_roi.Sample_ROI,df_ll_roi.loc[:,'Patient'])))\n",
    "            df_group_roi = df_group_roi[df_group_roi.Cohort.isin(['liver_cohort','lung_cohort'])]\n",
    "            fig, ax = plt.subplots(figsize=(3,3),dpi=300)\n",
    "            sns.boxplot(data=df_group_roi,x=s_group,y=s_marker,showfliers=False,ax=ax,order=[str(item) for item in ls_order],palette=['mediumpurple','deepskyblue'])\n",
    "            sns.stripplot(data=df_group_roi,x=s_group,y=s_marker,ax=ax,hue='Patient',s=3,palette=d_color)\n",
    "            ax.set_ylim(ax.get_ylim()[0],ax.get_ylim()[1])\n",
    "            ax.set_title(f'{s_group} versus\\n {s_marker}\\n p={pvalue:.4f} (n={n_low}, {n_high})')\n",
    "            ax.set_ylabel(f'Fraction {s_marker} in {s_loc}')\n",
    "            ax.set_ylabel(f'{s_marker}')\n",
    "            ax.get_legend().remove()\n",
    "            plt.tight_layout()\n",
    "            #fig.savefig(f'{s_date}/boxplot_mIHC_{s_marker}_versus_{s_group}_in_{s_loc}.png')\n",
    "        #break\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e71ca46a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "print(df_group.groupby('Cohort').mean().loc[:,s_define])\n",
    "#print(df_group.groupby('Cohort').sem().loc[:,s_define])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fedca39",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_ll.classI.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f01e0a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax=plt.subplots(dpi=300)\n",
    "df_group.groupby('Cohort').mean().loc[:,ls_mihc].plot(kind='bar',width=.9,stacked=True,ax=ax,colormap='Paired')\n",
    "ax.legend(bbox_to_anchor=(1,.9))\n",
    "ax.set_ylabel('Fraction in Tissue')\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d999eea3",
   "metadata": {},
   "source": [
    "# Section 4 <a name=\"clin\"></a>\n",
    "\n",
    "pORG mutltivarable survival analysis with clinicopathological variables\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f7649de",
   "metadata": {},
   "source": [
    "### clinicopathological variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63f160b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load patient vital status\n",
    "#also had stage/ grade/ lymph nodes etc.\n",
    "df_vital = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/FMP_Patients_Nov17_2021.xlsx',sheet_name='Sheet1')\n",
    "\n",
    "#Stage 1 pancreatic cancer means the cancer is not more than 4cm in size and it hasn't spread outside the pancreas. It is split into 1A and 1B.\n",
    "# 1A In TNM staging, this is the same as T1, N0, M0 1B In TNM staging, this is the same as T2, N0, M0\n",
    "\n",
    "# Stage 2A means the cancer is bigger than 4cm but is still within the pancreas. It has not spread to the lymph nodes Open a glossary item or other areas of the body.\n",
    "#2A in TNM staging, this is the same as T3, N0, M0. 2B In TNM staging, this is the same as T1, 2 or 3, N1, M0.\n",
    "\n",
    "#3 Stage 3 can mean that the cancer is any size within the pancreas and has spread to 4 or more nearby lymph nodes Open a glossary item.\n",
    "#In TNM staging, this is the same as T1, 2 or 3, N2, M0.\n",
    "#Or stage 3 can mean the cancer has started to grow outside the pancreas into the major blood vessels nearby. It may or may not have spread into the lymph nodes. It hasn't spread to any other areas of the body.\n",
    "#In TNM staging, this is the same as T4, Any N, M0.\n",
    "\n",
    "#4 Your doctor might call this advanced (metastatic) cancer.\n",
    "#In TNM staging, this is the same as Any T, Any N, M1.\n",
    "\n",
    "#collapse stage\n",
    "d_stage = {'2B - IIB':'II', 'p2A':'II', 'p2B':'II', '4 - IV':'IV', '2A - IIA':'II', 'p3':'III', '1B - IB':'I',\n",
    "    'c1B':'I', 'c2B':'I', 'nan':np.NaN, 'p4':'IV','2B - T1-3, N1, M0':'II','p0':'0','p1B':'I','c1':'I',\n",
    "    'c4':'IV','c3':'III','p1A':'I','pNA':np.NaN,'3 - T4, Any N, M0':'III','c2A':'II','p4B':'IV','c4B':'IV',\n",
    "     '2B - T1, N1, M0 / T2, N1, M0 / T3, N1, M0':'II','pUNK':np.NaN,'p2':'II','p3B':'III','c2':'II','p3A':'III',\n",
    "    '3 - III':'III','99 - Unknown':np.NaN,'1A - IA':'I','c3A':'III','c4A':'IV','p1':'I','c1A':'I','p4A':'IV',\n",
    "    '88 - Not applicable to 7th Edition staging':np.NaN,\n",
    "           '88 - No classification is recommended in 6th Edition':np.NaN,\n",
    "    '2A - T3, N0, M0':'II','4 - Any T, Any N, M1':'IV'}\n",
    "df_vital['Stage'] = df_vital.loc[:,'Stage Grouping _ Dominant'].replace(d_stage) \n",
    "\n",
    "#collapse grade\n",
    "\n",
    "d_grade = {'Grade II  Moderately Diff / Mod Well Diff':'2',\n",
    "       'Grade I   Well Differentiated/Differentiated':'1',\n",
    "       'Cell type not determined; not stated;N/A;Unk; high grade dysplas':np.NaN,\n",
    "       'Grade III Poorly Differentiated':'3', 'nan':np.NaN,'Grade IV Undifferentiated, Anaplastic':'4',\n",
    "         'B-CELL    LYMPHOMA OR LEUKEMIA ONLY':np.NaN}\n",
    "df_vital['Grade'] = df_vital.loc[:,'Grade_Differentiation'].replace(d_grade) \n",
    "\n",
    "#collapse LV invasion\n",
    "\n",
    "\n",
    "\n",
    "d_replace = {'nan':np.NaN, 'LYMPHOVASCULAR INVASION STATED AS NOT PRESENT':'NO',\n",
    "       'LYMPHOVASCULAR INVASION PRESENT/IDENTIFIED':'YES',\n",
    "       'Unknown/Indeterminate':np.NaN, 'NOT APPLICABLE':np.NaN,\n",
    "       'Lymph-vascular Invasion Present/Identified':'YES',\n",
    "       'LYMPHATIC AND SMALL VESSEL INVASION ONLY (L)':'YES',\n",
    "       'BOTH LYMPHATIC AND SMALL VESSEL AND VENOUS (LARGE VESSEL) INVASION':'YES'}\n",
    "df_vital['LV_Invasion'] = df_vital.loc[:,'Lymph_vascular Invasion'].replace(d_replace)\n",
    "\n",
    "# LN positivity\n",
    "df_vital['LN_Pos'] = df_vital.loc[:,'Regional Lymph Nodes Positive'] >= 1\n",
    "df_vital.loc[df_vital.loc[:,'Regional Lymph Nodes Positive'].isna(),'LN_Pos'] = np.NaN\n",
    "df_vital['LN_Pos']  = df_vital.LN_Pos.replace({True:'YES',False:'NO'})\n",
    "\n",
    "df_vital['Survival'] = df_vital.cVitalStatus.replace({'Alive':0,'Dead':1})\n",
    "print(df_vital.Grade.unique())\n",
    "print(df_vital.Stage.unique())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6e9bceb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #save\n",
    "# #'Stage Grouping _ Dominant'\n",
    "# pd.Series(d_stage, name='Stage').to_csv('Stage_Mapping.csv')\n",
    "# # 'Grade_Differentiation'\n",
    "# pd.Series(d_grade, name='Grade').to_csv('Grade_Mapping.csv')\n",
    "# #'Lymph_vascular Invasion'\n",
    "# pd.Series(d_replace, name='LV_Invasion').to_csv('LV_Invasion_Mapping.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a37f9f5a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# %matplotlib inline\n",
    "# # check prognostic value of clinicopathologiocal variables\n",
    "\n",
    "# ls_vital = ['Stage', 'Grade','LV_Invasion','LN_Pos'] \n",
    "# s_time = 'cDays from Diagnosis to FU'\n",
    "# s_censor = 'Survival'\n",
    "# for s_vital in ls_vital:\n",
    "#     print(s_vital)\n",
    "#     df = df_vital.loc[df_vital.loc[:,'Primary Site _ Major Groups For Staging']=='Pancreas',[s_vital,s_time,s_censor]].dropna(how='any')\n",
    "#     fig, __ = km_plot(df,s_vital,s_time,s_censor)\n",
    "#     fig.savefig(f'figures/KM_clinicopath_{s_vital}.png')\n",
    "#     #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fb38c71",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CPH\n",
    "s_time = 'cDays from Diagnosis to FU'\n",
    "df_vital['Survival_time'] = df_vital.loc[:,s_time]\n",
    "s_censor = 'Survival'\n",
    "ls_vital = ['LV_Invasion','Stage', 'Grade','LN_Pos'] #'Age at Diagnosis','Sex',\n",
    "for s_vital in ls_vital:\n",
    "    print(s_vital)\n",
    "    df = df_vital.loc[df_vital.loc[:,'Primary Site _ Major Groups For Staging']=='Pancreas',[s_vital,s_time,s_censor]].dropna(how='any')\n",
    "\n",
    "    if df.columns.isin(['Stage']).any():\n",
    "        df.Stage = df.Stage.replace({'I':1,'II':2,'III':3,'IV':4}).astype('int')\n",
    "    if df.columns.isin(['Grade']).any():\n",
    "        df.Grade = df.Grade.astype('int')\n",
    "    if df.loc[:,s_vital].dtype=='O':\n",
    "        df_dummy = pd.get_dummies(df.loc[:,[s_vital]],drop_first=True)\n",
    "        df.drop(s_vital,axis=1,inplace=True)\n",
    "        s_vital = df_dummy.columns[0]\n",
    "        df[s_vital] = df_dummy\n",
    "    fig, cph = cph_plot(df,s_vital,s_time,s_censor,figsize=(3,1.5))\n",
    "    plt.tight_layout()\n",
    "    fig.savefig(f'figures/CPH_single_{s_vital}_all.png')\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "837d3f5f",
   "metadata": {},
   "source": [
    "### omics data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a874c60c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load full patient data, neoadjuvant\n",
    "df_patient = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset1_age.xlsx',\n",
    "                           sheet_name='Patients - Tab 1')\n",
    "df_patient.rename({'Patient ID':'Public_Patient_ID'},axis=1,inplace=True)\n",
    "df_patient.loc[df_patient.loc[:,'Lung Met Present']=='YES','Cohort'] = 'Lung'\n",
    "df_patient.loc[df_patient.loc[:,'Liver Met Present']=='YES','Cohort'] = 'Liver'\n",
    "\n",
    "#df_patient['Public.Specimen.ID'] = [item + '-T' for item in df_patient.loc[:,'Patient ID']]\n",
    "\n",
    "#neo\n",
    "df_patient.loc[:,'Neoadjuvant Treatment'].unique()\n",
    "df_patient.loc[df_patient.loc[:,'Neoadjuvant Treatment'] == 'Yes Neoadjuvant','Neoadjuvant'] = 'Yes'\n",
    "df_patient['Neoadjuvant'] = df_patient.Neoadjuvant.fillna('No')\n",
    "\n",
    "\n",
    "#load T cell data\n",
    "df_tcell = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset6.xlsx',\n",
    "                         sheet_name='Tumor Samples')\n",
    "\n",
    "df_tcell_blood = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset6.xlsx',\n",
    "                         sheet_name='Blood Samples')\n",
    "\n",
    "#both\n",
    "df_tcell = df_tcell.merge(df_tcell_blood,on='Patient ID',suffixes=(' tumor',' blood'),how='outer')\n",
    "\n",
    "df_tcell.rename({'Productive Rearrangements (Observed Richness)':'Productive_Rearrangements',\n",
    "                'Templates per ng':'Templates_per_ng','Patient ID':'Public_Patient_ID'},axis=1,inplace=True)\n",
    "\n",
    "#merge\n",
    "df_patient = df_patient.merge(df_tcell,on='Public_Patient_ID',how='outer')\n",
    "\n",
    "#load pSUB\n",
    "df_gsva = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset1.xlsx', sheet_name='GSVA Scores - Tab 3')\n",
    "df_gsva.rename({'pSUB\\nPrimaries':'pSUB_Primary','pORG\\nPrimaries':'pORG_Primary',\n",
    "                'Patient ID':'Public_Patient_ID'},axis=1,inplace=True)\n",
    "\n",
    "d_gsva = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset1.xlsx', sheet_name=None)\n",
    "\n",
    "df_gsva = df_gsva.merge(d_gsva['Specimen Subtype - Tab 2'].loc[:,['PurIST Score','Patient Specimen ID','PurIST Subtype']],on='Patient Specimen ID')\n",
    "\n",
    "df_patient = df_patient.merge(df_gsva.loc[:,~df_gsva.columns.str.contains('\\n')],on='Public_Patient_ID',how='outer')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b969285",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_gene = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset2.xlsx',\n",
    "                           sheet_name='Mutation Data')\n",
    "\n",
    "ls_genes = []\n",
    "for s_gene in df_gene.Gene.unique():  \n",
    "    ls_patient = df_gene.loc[df_gene.Gene==s_gene,'Patient Specimen ID']\n",
    "    if len(ls_patient) > 10:\n",
    "        print(s_gene)\n",
    "        df_patient.loc[df_patient.loc[:,'Patient Specimen ID'].isin(ls_patient),f'{s_gene}_altered'] = True\n",
    "        df_patient.loc[:,f'{s_gene}_altered'].fillna(False,inplace=True)\n",
    "        ls_genes.append(f'{s_gene}_altered')\n",
    "    #break\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e2e06f9",
   "metadata": {},
   "source": [
    "### overlap in omics data samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d558ed7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#the samples overlap/don't overlap in assays\n",
    "# all samples \n",
    "\n",
    "from matplotlib_venn import venn3, venn3_circles\n",
    "from matplotlib_venn import venn2, venn2_circles, venn2_unweighted\n",
    "from pyvenn import venn\n",
    "d_ids = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}annotation/Simplified_Public_IDs_Key.xlsx',sheet_name=None)\n",
    "\n",
    "print(d_ids.keys())\n",
    "d_keys = {#'OPTR.Specimen.ID':['RnaSeqKey','TcrTumorKey','DnaPanelKey'], #'',\n",
    "          'OPTR':['RnaSeqKey','TcrTumorKey','TcrBloodKey','DnaPanelKey'],\n",
    "          #'OPTR':['RnaSeqKey','TcrTumorKey','DnaPanelKey'],\n",
    "       }\n",
    "\n",
    "for s_col ,ls_keys in d_keys.items():\n",
    "    d_sets = {}\n",
    "    for s_key in ls_keys:\n",
    "            print(s_key)\n",
    "            print(len(d_ids[s_key]))\n",
    "            d_sets.update({s_key:d_ids[s_key]})\n",
    "    #plot\n",
    "    if len(ls_keys) == 4:\n",
    "        labels = venn.get_labels([set(item.loc[:,s_col]) for key, item in d_sets.items()])\n",
    "        fig,ax = venn.venn4(labels, names=[key.split('Key')[0] for key, item in d_sets.items()])#,ax=ax\n",
    "    elif len(ls_keys) == 3:\n",
    "        fig,ax = plt.subplots(figsize=(3,3),dpi=300)\n",
    "        venn3([set(item.loc[:,s_col]) for key, item in d_sets.items()], [key.split('Key')[0] for key, item in d_sets.items()],ax=ax)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5537241e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# #primaries\n",
    "\n",
    "# from matplotlib_venn import venn3, venn3_circles\n",
    "# from matplotlib_venn import venn2, venn2_circles, venn2_unweighted\n",
    "# from pyvenn import venn\n",
    "# d_ids = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}annotation/Simplified_Public_IDs_Key.xlsx',sheet_name=None)\n",
    "\n",
    "# print(d_ids.keys())\n",
    "# d_keys = {#'OPTR.Specimen.ID':['RnaSeqKey','TcrTumorKey','DnaPanelKey'], #'',\n",
    "#           #'OPTR':['RnaSeqKey','TcrTumorKey','TcrBloodKey','DnaPanelKey'],\n",
    "#           'OPTR':['RnaSeqKey','TcrTumorKey','DnaPanelKey'],\n",
    "#        }\n",
    "\n",
    "# for s_col ,ls_keys in d_keys.items():\n",
    "#     d_sets = {}\n",
    "#     for s_key in ls_keys:\n",
    "#             print(s_key)\n",
    "#             print(len(d_ids[s_key]))\n",
    "#             df_set = d_ids[s_key][~d_ids[s_key].loc[:,'OPTR.Specimen.ID'].str.contains('-M')]\n",
    "#             d_sets.update({s_key:df_set})\n",
    "#     #plot\n",
    "#     if len(ls_keys) == 4:\n",
    "#         labels = venn.get_labels([set(item.loc[:,s_col]) for key, item in d_sets.items()])\n",
    "#         fig,ax = venn.venn4(labels, names=[key.split('Key')[0] for key, item in d_sets.items()],ax=ax)\n",
    "#     elif len(ls_keys) == 3:\n",
    "#         fig,ax = plt.subplots(figsize=(3,3),dpi=300)\n",
    "#         venn3([set(item.loc[:,s_col]) for key, item in d_sets.items()], [key.split('Key')[0] for key, item in d_sets.items()],ax=ax)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8aaa7d0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_id = d_ids['RnaSeqKey']\n",
    "ls_ids = df_id.loc[:,'Public.Specimen.ID']\n",
    "\n",
    "# add patients w/o RNA seq\n",
    "for s_key in ['TcrTumorKey','TcrBloodKey']:\n",
    "    df_add = d_ids[s_key].loc[~d_ids[s_key].loc[:,'Public.Specimen.ID'].isin(ls_ids)]\n",
    "    df_id = pd.concat([df_id,df_add])\n",
    "\n",
    "#check\n",
    "print(df_id.loc[:,'OPTR.Specimen.ID'].duplicated().any())\n",
    "\n",
    "ls_drop = df_id.loc[df_id.loc[:,'OPTR.Specimen.ID'].str.contains('-T2')].index\n",
    "df_unique = df_id.loc[df_id.loc[:,'OPTR.Specimen.ID'].str.contains('-T')].drop(ls_drop)\n",
    "\n",
    "#check\n",
    "print(df_unique.OPTR.duplicated().any())\n",
    "\n",
    "#add id\n",
    "df_vital['Public_Patient_ID'] = df_vital.OPTR.map(dict(zip(df_unique.OPTR,df_unique.loc[:,'Biolibrary.Subject.ID'])))\n",
    "\n",
    "#omics data plus clinical data\n",
    "df_patient = df_patient.merge(df_vital,on='Public_Patient_ID',how='left')\n",
    "\n",
    "#df_id.to_csv('Patient_IDs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2747b837",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add age category\n",
    "df_patient.loc[df_patient.loc[:,'Age at Diagnosis'] > 70,'Age'] = '>70'\n",
    "df_patient.loc[df_patient.loc[:,'Age at Diagnosis'] <= 70,'Age'] = '<=70'\n",
    "df_patient = df_patient.drop(ls_drop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "745c0aa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop less than 30 days\n",
    "# died of surgery\n",
    "ls_drop_surgery = df_patient[(df_patient.loc[:,'Days from Resection to FU'] < 30) & (df_patient.loc[:,'cVitalStatus'] == 'Dead')].index\n",
    "df_patient = df_patient.drop(ls_drop_surgery)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d67689c4",
   "metadata": {},
   "source": [
    "## Survival Analysis - patients in paper\n",
    "\n",
    "KM, CPH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0779a81a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CPH\n",
    "df_result=pd.DataFrame()\n",
    "ls_vital = ['Neoadjuvant','Age','Sex','Stage', 'Grade','LV_Invasion','LN_Pos', #'Age at Diagnosis',\n",
    "       'PurIST Score']\n",
    "#ls_vital = ls_genes\n",
    "for s_vital in ls_vital:\n",
    "    print(s_vital)\n",
    "    df = df_patient.loc[:,[s_vital,s_time,s_censor]].dropna(how='any')\n",
    "    if df.columns.isin(['Stage']).any():\n",
    "        df.Stage = df.Stage.replace({'I':1,'II':2,'III':3,'IV':4}).astype('int')\n",
    "    if df.columns.isin(['Grade']).any():\n",
    "        df.Grade = df.Grade.astype('int')\n",
    "    if df.loc[:,s_vital].dtype=='O':\n",
    "        df_dummy = pd.get_dummies(df.loc[:,[s_vital]],drop_first=True)\n",
    "        df.drop(s_vital,axis=1,inplace=True)\n",
    "        s_vital = df_dummy.columns[0]\n",
    "        df[s_vital] = df_dummy\n",
    "    if s_vital.find('_Up_')>-1:\n",
    "        d_rename = {s_vital:s_vital.split('_Up')[0]}\n",
    "        df=df.rename(d_rename,axis=1)\n",
    "        s_vital = s_vital.split('_Up')[0]\n",
    "    fig, cph = cph_plot(df,s_vital,s_time,s_censor,figsize=(3,1.5))\n",
    "    plt.tight_layout()\n",
    "    fig.savefig(f'figures/CPH_single_{s_vital}_subset.png')\n",
    "    df_result=pd.concat([df_result,cph.summary.loc[:,['exp(coef)','p']]])\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd19134a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "ls_drop = df_patient[~df_patient.loc[:,'Patient Specimen ID'].str.contains('-T2').fillna(False)].index\n",
    "# check prognostic value of clinicopathologiocal variables\n",
    "import lifelines\n",
    "from lifelines import KaplanMeierFitter, CoxPHFitter\n",
    "from lifelines.statistics import multivariate_logrank_test\n",
    "ls_vital = ['Neoadjuvant','Age','Sex','Stage', 'Grade','LV_Invasion','LN_Pos'] \n",
    "#ls_vital  = ls_genes\n",
    "s_time = 'cDays from Diagnosis to FU'\n",
    "s_censor = 'Survival'\n",
    "for s_vital in ls_vital:\n",
    "    print(s_vital)\n",
    "    df = df_patient.loc[df_patient.Stage!='0',[s_vital,s_time,s_censor]].dropna(how='any')\n",
    "    df.loc[:,s_vital] = df.loc[:,s_vital].replace({True:'altered',False:'WT'})\n",
    "    fig, __ = km_plot(df,s_vital,s_time,s_censor)\n",
    "    fig.savefig(f'figures/KM_clinicopath_{s_vital}_subset.png')\n",
    "    break\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b01d2749",
   "metadata": {},
   "source": [
    "## gene signature scores\n",
    "\n",
    "pORG, pSUB GSVA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16561bb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pORG = pd.read_csv(f'{codedir}/data/GSVA_Scores_Link.csv')#=,index_col='Public_Specimen_ID'\n",
    "df_pORG['Patient Specimen ID'] = df_pORG.Public_Specimen_ID#[item.split('-')[0] + '-' + item.split('-')[1] for item in df_pORG.Public_Specimen_ID]\n",
    "df_patient = df_patient.merge(df_pORG,on='Patient Specimen ID',suffixes=('',\"_x\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "682b0ded",
   "metadata": {},
   "outputs": [],
   "source": [
    "ls_drop_T2 = df_patient[df_patient.Public_Specimen_ID.str.contains('-T2')].index\n",
    "df_patient.drop(ls_drop_T2,axis=0,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "577f5ba9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CPH\n",
    "#df_result=pd.DataFrame()\n",
    "ls_vital = [ #'Age at Diagnosis',\n",
    "      'Original_pORG_Up_78_Genes', 'txi_pORG_Up_42_Genes',\n",
    "       'Original_pSUB_Up_100_Genes', 'txi_pSUB_Up_100_Genes', ]\n",
    "for s_vital in ls_vital:\n",
    "    print(s_vital)\n",
    "    df = df_patient.loc[:,[s_vital,s_time,s_censor]].dropna(how='any')\n",
    "    if df.columns.isin(['Stage']).any():\n",
    "        df.Stage = df.Stage.replace({'I':1,'II':2,'III':3,'IV':4}).astype('int')\n",
    "    if df.columns.isin(['Grade']).any():\n",
    "        df.Grade = df.Grade.astype('int')\n",
    "    if df.loc[:,s_vital].dtype=='O':\n",
    "        df_dummy = pd.get_dummies(df.loc[:,[s_vital]],drop_first=True)\n",
    "        df.drop(s_vital,axis=1,inplace=True)\n",
    "        s_vital = df_dummy.columns[0]\n",
    "        df[s_vital] = df_dummy\n",
    "    if s_vital.find('_Up_')>-1:\n",
    "        d_rename = {s_vital:s_vital.split('_Up')[0]}\n",
    "        df=df.rename(d_rename,axis=1)\n",
    "        s_vital = s_vital.split('_Up')[0]\n",
    "    fig, cph = cph_plot(df,s_vital,s_time,s_censor,figsize=(3,1.5))\n",
    "    plt.tight_layout()\n",
    "    fig.savefig(f'figures/CPH_single_{s_vital}_subset.png')\n",
    "    df_result=pd.concat([df_result,cph.summary.loc[:,['exp(coef)','p']]])\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d33d410f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CPH\n",
    "df_result_multi = pd.DataFrame()\n",
    "ls_drop = df_patient[~df_patient.loc[:,'Patient Specimen ID'].str.contains('-T2').fillna(False) & ~df_patient.pSUB_Primary.isna()].index\n",
    "ls_multi = ['txi_pORG_Up_42_Genes','pORG_Primary','pSUB_Primary','PurIST Score','txi_pSUB_Up_100_Genes',]\n",
    "ls_cats = ['TP53_altered','CDKN2A_altered','LV_Invasion','LN_Pos','Age']\n",
    "for s_multi in ls_multi:\n",
    "    #why does PurIST have extra patients?\n",
    "    df = df_patient.loc[ls_drop,[s_multi,s_time,s_censor,'Grade','Stage']] ##'Neoadjuvant','Age at Diagnosis'\n",
    "    df_dummy = pd.get_dummies(df_patient.loc[:,ls_cats],drop_first=True) #'TP53_altered','CDKN2A_altered',,'Sex'\n",
    "    df = pd.concat([df,df_dummy],axis=1)\n",
    "    fig, cph = cph_plot(df,s_multi,s_time,s_censor)\n",
    "    fig.savefig(f'figures/CPH_{s_multi}.png')\n",
    "    df_result_multi=pd.concat([df_result_multi,cph.summary.loc[:,['exp(coef)','p']]])\n",
    "    #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e937f8ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save out the metadata\n",
    "ls_drop = ['Liver Met Present in Patient tumor',\n",
    "       'Lung Met Present in Patient tumor','Lung Met Present in Patient blood',\n",
    "       'Liver Met Present in Patient blood','Unnamed: 0',#'OPTR',\n",
    "          ]\n",
    "if not os.path.exists('Patient_metadata_clean.csv'):\n",
    "    print('saving')\n",
    "    df_patient.drop(ls_drop,axis=1).to_csv('Patient_metadata_clean.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1df09965",
   "metadata": {},
   "source": [
    "# Section 5 <a name=\"newsurv\"></a>\n",
    "\n",
    "new patient survival\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8aedb9ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_id = pd.read_csv('Patient_IDs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e6601f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # load new survival and de-identify (done)\n",
    "# df_new_surv = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}annotation/cancer_participant_overview_optr_2023-05-04_16-55-18.xlsx')#,dtype=object\n",
    "# df_new_surv.loc[:,'Participant_ID'] = df_new_surv.loc[:,'Participant ID'].astype('str')\n",
    "# df_new_surv['Vital_Status'] = df_new_surv.loc[:,'Vital Status at Follow Up'].replace({'Alive':0,'Dead':1})\n",
    "# #load new pORG\n",
    "# df_pORG = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}data/PartialReport_pORG_pSUB_Update.xlsx',sheet_name='GSVA_Scores')\n",
    "# df_pORG['Participant_ID'] = [item.split('-')[0] for item in df_pORG.SpecimenID]\n",
    "# df_pORG['Type_TM'] = [item.split('-')[1] for item in df_pORG.SpecimenID]\n",
    "# ls_drop = df_pORG[df_pORG.Type_TM=='T2'].index\n",
    "# df_pORG.drop(ls_drop,inplace=True)\n",
    "# #merge scores\n",
    "# df_new = df_pORG.merge(df_new_surv,on='Participant_ID',how='left',suffixes=('','_x'))\n",
    "# d_zip = dict(zip(df_id.loc[:,'OPTR.Specimen.ID'],df_id.loc[:,'Public.Specimen.ID']))\n",
    "# df_new['Public_Specimen_ID'] = df_new.SpecimenID.map(d_zip)\n",
    "# # save\n",
    "# ls_drop_col = ['SpecimenID','Participant_ID'] #[]#\n",
    "# df_new.drop(ls_drop_col,axis=1).to_csv('Updated_survival_scores_Link.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69100155",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_new = pd.read_csv('Updated_survival_scores_OPTR.csv',index_col=0) #Updated_survival_scores_Link\n",
    "df_patient= pd.read_csv('Patient_metadata_clean.csv',index_col=0)\n",
    "df_patient['SpecimenID'] = df_patient.Public_Specimen_ID.map(dict(zip(df_id.loc[:,'Public.Specimen.ID'],df_id.loc[:,'OPTR.Specimen.ID'])))\n",
    "\n",
    "#merge\n",
    "df_new_pt = df_patient.merge(df_new,on='SpecimenID',suffixes=('','_x'))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c615e0b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#list od patients still alive\n",
    "df_alive = df_patient[(df_patient.Survival==0)& (df_patient.Cohort=='Lung')] #  & (df_patient.Cohort=='Lung').Public_Specimen_ID\n",
    "ls_index = df_new[df_new.Public_Specimen_ID.isin(df_alive.Public_Specimen_ID)].loc[:,'Date of Last Follow Up'].sort_values()[-3::]#.index\n",
    "([print(item) for item in df_new.loc[ls_index.index,'SpecimenID']])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44600c3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CPH\n",
    "s_censor = 'Vital_Status'\n",
    "s_time = 'Overall Survival (Days)'\n",
    "df_result_multi = pd.DataFrame()\n",
    "ls_drop = df_new_pt[~df_new_pt.loc[:,'SpecimenID'].str.contains('-T2').fillna(False) & ~df_new_pt.txi_pORG_Up_42_Genes.isna()].index\n",
    "ls_multi = ['txi_pORG_Up_42_Genes','pORG_Primary','pSUB_Primary','PurIST Score','txi_pSUB_Up_100_Genes',]\n",
    "ls_cats = ['TP53_altered','CDKN2A_altered','LV_Invasion','LN_Pos','Age']\n",
    "for s_multi in ls_multi:\n",
    "    #why does PurIST have extra patients?\n",
    "    df = df_new_pt.loc[ls_drop,[s_multi,s_time,s_censor,'Grade','Stage']] ##'Neoadjuvant','Age at Diagnosis'\n",
    "    df_dummy = pd.get_dummies(df_patient.loc[:,ls_cats],drop_first=True) #'TP53_altered','CDKN2A_altered',,'Sex'\n",
    "    df = pd.concat([df,df_dummy],axis=1)\n",
    "    fig, cph = cph_plot(df,s_multi,s_time,s_censor)\n",
    "    fig.savefig(f'figures/CPH_{s_multi}.png')\n",
    "    df_result_multi=pd.concat([df_result_multi,cph.summary.loc[:,['exp(coef)','p']]])\n",
    "    #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "420b8c75",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_new_surv.loc[:,'Participant ID'].isin([4329]).any()\n",
    "df_new_surv.loc[:,'Participant ID'].isin([4104]).any()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8c480d1",
   "metadata": {},
   "source": [
    "# Section 6 <a name=\"geneexp\"></a>\n",
    "\n",
    "gene expresison correlation\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "761916e0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5e75232",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_vst = pd.read_csv('data/VST_Genes_Link.csv',index_col=0)\n",
    "df_rna = df_vst.T.copy()\n",
    "df_rna.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afa3f19e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pORG = pd.read_csv(f'{codedir}/data/GSVA_Scores_Link.csv')#=,index_col='Public_Specimen_ID'\n",
    "df_pORG['Patient Specimen ID'] = df_pORG.Public_Specimen_ID\n",
    "df_pORG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c6cc732",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ls_add = ['Original_pORG_Up_78_Genes',\n",
    " 'txi_pORG_Up_42_Genes',\n",
    " 'Original_pSUB_Up_100_Genes',\n",
    " 'txi_pSUB_Up_100_Genes']\n",
    "\n",
    "for s_add in ls_add:\n",
    "    d_map = dict(zip(df_pORG.loc[:,'Patient Specimen ID'],df_pORG.loc[:,s_add]))\n",
    "    df_rna[s_add] = df_rna.index.map(d_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06bcc103",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_rna.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db8c2fde",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "d_rename={#'pORG_Up_78':'pORG',\n",
    "          'MS4A1':'MS4A1 (CD20)'}\n",
    "from scipy.stats import pearsonr\n",
    "dim = (4,3)\n",
    "    \n",
    "for s_add in ls_add:\n",
    "    ls_marker = ['CD3E','CD4', 'CD8A', 'MS4A1',s_add]\n",
    "    df_all = df_rna.loc[:,ls_marker].corr().rename(d_rename,axis=1).rename(d_rename,axis=0)\n",
    "    print(len(df_rna))\n",
    "    g = sns.clustermap(df_all)\n",
    "    plt.close()\n",
    "    categories_order = df_all.iloc[g.dendrogram_col.reordered_ind,:].index.tolist()\n",
    "    df_all = df_all.loc[categories_order,categories_order]\n",
    "    rho = df_rna.loc[:,ls_marker].corr() #df_all.corr()\n",
    "    pval = df_rna.loc[:,ls_marker].corr(method=lambda x, y: pearsonr(x, y)[1]) - np.eye(*rho.shape)\n",
    "    p_vals = pval.applymap(lambda x: ''.join(['*' for t in [0.001,0.005,0.05] if x<=t]))\n",
    "    p_vals = p_vals.rename(d_rename,axis=1).rename(d_rename,axis=0)\n",
    "    p_vals = p_vals.loc[categories_order,categories_order]\n",
    "    fig, ax = plt.subplots(figsize=dim,dpi=300)\n",
    "    sns.heatmap(df_all, vmin=-1, vmax=1, annot=p_vals, fmt = '', cmap='RdBu_r',ax=ax,\n",
    "               cbar_kws={'shrink':0.85,'label':s_add})\n",
    "\n",
    "    #ax.set_title(f'Cell Type versus Scores', fontdict={'fontsize':16}, pad=12);\n",
    "    #fig.savefig(f'{codedir}/{s_date}/heatmap_Celltype_vs_pORG_purIST.png', dpi=300, bbox_inches='tight')\n",
    "    '''\n",
    "    fig, ax = plt.subplots(figsize=dim,dpi=300)\n",
    "    matrix = np.triu(np.ones_like(rho))\n",
    "    np.fill_diagonal(matrix, val=0)\n",
    "    np.fill_diagonal(p_vals.values,'')\n",
    "    sns.heatmap(df_all, vmin=-1, vmax=1, annot=p_vals, fmt = '', cmap='RdBu_r',\n",
    "                ax=ax,mask=matrix,cbar_kws={'shrink':0.85,'label':'Pearson Correlation'}) #'anchor':(-1.4,0.0)\n",
    "    #'''\n",
    "    #pval\n",
    "    #break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ce56879",
   "metadata": {},
   "source": [
    "## Section 6 <a name=\"tcell\"></a>\n",
    "\n",
    "re-analyze t cell data\n",
    "\n",
    "289 blood samples with matching 175 primary\n",
    "tumors (141 overlapping with the RNA-seq dataset) and 43 metastatic tumors (33 overlapping with\n",
    "the RNA-seq dataset). (218)\n",
    "\n",
    "**missing 2 tumor, all 289 blood there**\n",
    "\n",
    "TOTAL = 174\n",
    "\n",
    "(total pORG = 171)\n",
    "\n",
    "\n",
    "we analyzed blood samples from 77 patients in the liver cohort and 16\n",
    "patients in the lung cohort, of which 60 and 16 were matched with tumor samples from the same\n",
    "patient, respectively\n",
    "\n",
    "TOTAL = 94 blood\n",
    "TOTAL = 76 tumor\n",
    "\n",
    "tumor distinct clones\n",
    "used data from 214 matched pairs of tumor and blood samples\n",
    "\n",
    "**213 are there, missing 1**\n",
    "\n",
    "(TCR tumor: 59 and 16)\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1357f459",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load full patient data, neoadjuvant\n",
    "from scipy.stats import f_oneway\n",
    "df_patient = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset1_age.xlsx',\n",
    "                           sheet_name='Patients - Tab 1')\n",
    "df_patient.rename({'Patient ID':'Public_Patient_ID'},axis=1,inplace=True)\n",
    "df_patient.loc[df_patient.loc[:,'Lung Met Present']=='YES','Cohort'] = 'Lung'\n",
    "df_patient.loc[df_patient.loc[:,'Liver Met Present']=='YES','Cohort'] = 'Liver'\n",
    "\n",
    "#df_patient['Public.Specimen.ID'] = [item + '-T' for item in df_patient.loc[:,'Patient ID']]\n",
    "\n",
    "#neo\n",
    "df_patient.loc[:,'Neoadjuvant Treatment'].unique()\n",
    "df_patient.loc[df_patient.loc[:,'Neoadjuvant Treatment'] == 'Yes Neoadjuvant','Neoadjuvant'] = 'Yes'\n",
    "df_patient['Neoadjuvant'] = df_patient.Neoadjuvant.fillna('No')\n",
    "\n",
    "\n",
    "#load T cell data\n",
    "df_tcell = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset6.xlsx',\n",
    "                         sheet_name='Tumor Samples')\n",
    "print(len(df_tcell))\n",
    "\n",
    "df_tcell_blood = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset6.xlsx',\n",
    "                         sheet_name='Blood Samples')\n",
    "print(len(df_tcell_blood))\n",
    "#both\n",
    "df_tcell = df_tcell.merge(df_tcell_blood,on='Patient ID',suffixes=(' tumor',' blood'),how='outer')\n",
    "\n",
    "df_tcell.rename({'Productive Rearrangements (Observed Richness)':'Productive_Rearrangements',\n",
    "                'Templates per ng':'Templates_per_ng','Patient ID':'Public_Patient_ID'},axis=1,inplace=True)\n",
    "\n",
    "\n",
    "#load pSUB, pORG\n",
    "df_gsva = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset1.xlsx', sheet_name='GSVA Scores - Tab 3')\n",
    "df_gsva.rename({'pSUB\\nPrimaries':'pSUB_Primary','pORG\\nPrimaries':'pORG_Primary','pSUB\\nAll':'pSUB_All',\n",
    "                'pORG\\nAll':'pORG_All',\n",
    "                'Patient Specimen ID':'Patient_Specimen_ID'},axis=1,inplace=True)\n",
    "\n",
    "# # #FNA #changes to 34 high\n",
    "# ls_fna = df_gsva.Patient_Specimen_ID[df_gsva.Patient_Specimen_ID.str.contains('-F')]\n",
    "# ls_fna_pt = [item.split('-F')[0] for item in ls_fna]\n",
    "# ls_fna_index = df_tcell.loc[(df_tcell.Public_Patient_ID.isin(ls_fna_pt)) & (df_tcell.loc[:,'Tumor Type'].isna()),:].index\n",
    "# df_tcell.loc[ls_fna_index,'Tumor Type'] = '-F'\n",
    "\n",
    "#met or primary\n",
    "df_tcell['Type_TM'] = df_tcell.loc[:,'Tumor Type'].replace({'Primary':'-T', 'Met':'-M'})\n",
    "df_tcell['Patient_Specimen_ID'] = df_tcell.Public_Patient_ID + df_tcell.Type_TM\n",
    "\n",
    "df_tcell['Cohort'] = df_tcell.Public_Patient_ID.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Cohort)))\n",
    "for s_cat in ['pORG_Primary','pSUB_Primary','pSUB_All','pORG_All']:\n",
    "    df_tcell[s_cat] = df_tcell.Patient_Specimen_ID.map(dict(zip(df_gsva.Patient_Specimen_ID,df_gsva.loc[:,s_cat])))\n",
    "\n",
    "# load purIST\n",
    "df_gsva = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset1.xlsx', sheet_name='Specimen Subtype - Tab 2')\n",
    "df_gsva.rename({'Patient Specimen ID':'Patient_Specimen_ID'},axis=1,inplace=True)\n",
    "for s_cat in ['PurIST Subtype']:\n",
    "    df_tcell[s_cat] = df_tcell.Patient_Specimen_ID.map(dict(zip(df_gsva.Patient_Specimen_ID,df_gsva.loc[:,s_cat])))\n",
    "    \n",
    "df_merge = df_tcell.copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84e310bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add pORG quartiles\n",
    "s_porg = 'pORG_Primary'# 'pORG_All'#\n",
    "x = df_merge.loc[:,s_porg]\n",
    "\n",
    "d_cut = {'quartiles':(4,['low','med-low','med-high','high']),\n",
    "        'medians' : (2,['low','high'])}\n",
    " \n",
    "for s_col, tu_cut in d_cut.items():\n",
    "    i_cut = tu_cut[0]\n",
    "    labels = tu_cut[1]\n",
    "    q = pd.qcut(x, q=i_cut,labels=labels) \n",
    "    if s_col == 'quartiles':\n",
    "        df_merge[s_col] = q.replace({'med-low':np.NaN,'med-high':np.NaN})\n",
    "    else:\n",
    "        df_merge[s_col] = q\n",
    "    print(df_merge[s_col].value_counts())\n",
    "    \n",
    "df_merge['Tumor Distinct Clones 100-x'] = 100 - df_merge.loc[:,'Percentage Tumor-Distinct Clones in Paired Tumor Sample']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32df8716",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# the greater the absolute value of the slope, the greater the diversity.\n",
    "# redefine liver patient?\n",
    "\n",
    "#b_primary = False #  True #\n",
    "if s_porg == 'pORG_Primary': \n",
    "    df_pri = df_merge[df_merge.loc[:,'Tumor Type']=='Primary']  # \n",
    "else:\n",
    "    df_pri = df_merge\n",
    "\n",
    "\n",
    "# if b_primary:\n",
    "#     df_pri.loc[df_pri.loc[:,'Tumor Type'] != 'Primary',['quartiles','median']] = np.nan\n",
    "#     print(df_pri.Cohort.isna().sum())\n",
    "\n",
    "\n",
    "ls_foci = ['Templates_per_ng',\n",
    "           \"Simpson's Evenness tumor\",\n",
    "           \"Simpson's Evenness blood\",\n",
    "          'Tumor Distinct Clones 100-x',\n",
    "          ]\n",
    "\n",
    "d_order =  {\n",
    "    'PurIST Subtype':['basal-like','classical'],\n",
    "    'Cohort':['Liver','Lung'],\n",
    "    'quartiles':['high','low'],\n",
    "    #'medians':['high','low']\n",
    "     }#tertiles_h # quartiles\n",
    "\n",
    "for s_foci in ls_foci:\n",
    "    print(s_foci)\n",
    "    figsize=(3.5,2.8)\n",
    "    fig,ax=plt.subplots(dpi=300,figsize=figsize)\n",
    "    order = []\n",
    "    ls_ticks = []\n",
    "    d_pval = {}\n",
    "    df_both = pd.DataFrame()\n",
    "    for idx, s_cat in enumerate(d_order.keys()):\n",
    "        #print(s_cat)\n",
    "        ls_order = d_order[s_cat]\n",
    "        s_bad = ls_order[0]\n",
    "        s_good = ls_order[1]\n",
    "        d_replace = {s_bad:'bad',s_good:'good'}\n",
    "        a = df_pri.loc[df_pri.loc[:,s_cat]==ls_order[0],s_foci].dropna()\n",
    "        b = df_pri.loc[df_pri.loc[:,s_cat]==ls_order[1],s_foci].dropna()\n",
    "        ##statistic, pvalue = stats.ttest_ind(a,b,alternative=s_alt)\n",
    "        #print(len(a) + len(b))\n",
    "        statistic, pvalue = f_oneway(b,a)\n",
    "        df_pri['hue'] = df_pri.loc[:,s_cat].replace(d_replace)\n",
    "        df_pri['x'] = s_cat\n",
    "        df_both=pd.concat([df_both,df_pri.loc[:,['x','hue',s_foci]]])\n",
    "        for s_test in ls_order:\n",
    "            order.append((s_cat,d_replace[s_test]))\n",
    "            ls_ticks.append(s_test)\n",
    "        d_pval.update({s_cat:pvalue})\n",
    "    sns.violinplot(data=df_both,y=s_foci,x='x',hue='hue',ax=ax,alpha=0.7,linewidth=1,cut=0,inner='quartile',hue_order=['bad','good'])#hue='Ki67pos'\n",
    "    sns.stripplot(data=df_both,y=s_foci,x='x',hue='hue',s=2,dodge=True,ax=ax,palette='dark',jitter=0.2,alpha=0.3,hue_order=['bad','good']) #hue='Ki67pos'\n",
    "    #annotate\n",
    "    if len(order) == 6:\n",
    "        pairs = [(order[0],order[1]),(order[2],order[3]),(order[4],order[5])]\n",
    "        pvalues = [d_pval[list(d_order.keys())[0]],d_pval[list(d_order.keys())[1]],d_pval[list(d_order.keys())[2]]]\n",
    "    else:\n",
    "        pairs = [(order[0],order[1]),(order[2],order[3]),(order[4],order[5]),(order[6],order[7])]\n",
    "        pvalues = [d_pval[list(d_order.keys())[0]],d_pval[list(d_order.keys())[1]],d_pval[list(d_order.keys())[2]],d_pval[list(d_order.keys())[3]]]\n",
    "    reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='fdr_bh')\n",
    "    formatted_pvalues = [f'p={pvalue:.2}' for pvalue in list(corrected)]#corrected #pvalues\n",
    "    annotator = Annotator(ax, pairs=pairs, data=df_both,y=s_foci,x='x',hue='hue',verbose=False)\n",
    "    annotator.set_custom_annotations(formatted_pvalues)\n",
    "    #annotator.configure(line_height=0.9,line_offset=10)\n",
    "    annotator.annotate()\n",
    "    ax.legend().remove()\n",
    "    if len(order) == 6:\n",
    "        ax.set_xticks([-0.2,0.2, 0.8,1.2,1.8,2.2])\n",
    "    else:\n",
    "        ax.set_xticks([-0.2,0.2, 0.8,1.2,1.8,2.2,2.8,3.2])\n",
    "    ax.set_xticklabels(ls_ticks,rotation=45)\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_title(f\"{s_foci}\", fontsize='x-large') #\n",
    "    plt.tight_layout()\n",
    "    fig.savefig(f'figures/violinplot_both_{s_foci}_{list(d_order.keys())[-1]}_all.png')\n",
    "    #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9f7874b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_both.dropna().groupby(['x','hue']).count()\n",
    "#35, 35 low pORG primary!!\n",
    "#16 lung, 59 liver \n",
    "# 45 basal like"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ee21288",
   "metadata": {},
   "source": [
    "### additional TCR data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a936b543",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_patient = pd.read_csv('data/Patient_metadata_clean.csv',index_col=0)\n",
    "\n",
    "\n",
    "df_tum = pd.read_excel('Diversity Metrics/Diversity Metrics - All Tumors 210809.xlsx',sheet_name='diversity')\n",
    "#add cohort\n",
    "df_tum['Cohort'] = (df_tum.Liver.astype('str') + df_tum.Lung.astype('str')).replace({'00':np.nan,'10':'Liver','11':'Liver','01':'Lung'})\n",
    "print(df_tum['Cohort'].value_counts())\n",
    "#add pORG quartiles\n",
    "s_porg = 'pORG-78UP-All'#'pORG-78UP-Primaries Halves' #\n",
    "x = df_tum.loc[:,s_porg]\n",
    "d_cut = {'quartiles':(4,['low','med-low','med-high','high']),\n",
    "        'median' : (2,['low','high']), \n",
    "         'tertiles' : (3,['low','med','high']),\n",
    "        }\n",
    " \n",
    "for s_col, tu_cut in d_cut.items():\n",
    "    i_cut = tu_cut[0]\n",
    "    labels = tu_cut[1]\n",
    "\n",
    "    q = pd.qcut(x, q=i_cut,labels=labels) \n",
    "    if s_col == 'quartiles':\n",
    "        df_tum[s_col] = q.replace({'med-low':np.NaN,'med-high':np.NaN})\n",
    "    elif s_col == 'tertiles':\n",
    "        df_tum[s_col] = q.replace({'med':np.NaN})\n",
    "    else:\n",
    "        df_tum[s_col] = q\n",
    "    print(df_tum[s_col].value_counts())\n",
    "\n",
    "## blood\n",
    "df_bld = pd.read_excel('Diversity Metrics/Diversity Metrics - All Blood 210804.xlsx',sheet_name='Sheet1')\n",
    "df_bld['OPTR ID'] = df_bld.loc[:,'sample'].astype('Int64',errors='ignore')\n",
    "df_bld = df_bld[~df_bld['OPTR ID'].isna()].merge(df_tum.loc[:,['Cohort','OPTR ID',s_porg,'Subtype']],on='OPTR ID')\n",
    "#add cohort\n",
    "#df_bld['Cohort'] = (df_bld.Liver.astype('str') + df_bld.Lung.astype('str')).replace({'00':np.nan,'10':'Liver','11':'Liver','01':'Lung'})\n",
    "\n",
    "#add pORG quartiles\n",
    "x = df_bld.loc[:,s_porg]\n",
    "\n",
    "d_cut = {'quartiles':(4,['low','med-low','med-high','high']),\n",
    "        'median' : (2,['low','high']),\n",
    "        'tertiles' : (3,['low','med','high']),\n",
    "        }\n",
    " \n",
    "for s_col, tu_cut in d_cut.items():\n",
    "    i_cut = tu_cut[0]\n",
    "    labels = tu_cut[1]\n",
    "    q = pd.qcut(x, q=i_cut,labels=labels) \n",
    "    if s_col == 'quartiles':\n",
    "        df_bld[s_col] = q.replace({'med-low':np.NaN,'med-high':np.NaN})\n",
    "    elif s_col == 'tertiles':\n",
    "        df_bld[s_col] = q.replace({'med':np.NaN})\n",
    "    else:\n",
    "        df_bld[s_col] = q\n",
    "    print(df_bld[s_col].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac436e7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_order =  {\n",
    "    'Subtype':['basal-like','classical'],\n",
    "    'Cohort':['Liver','Lung'],\n",
    "    #'quartiles':['high','low'],\n",
    "    'tertiles':['high','low']\n",
    "     }#tertiles_h # quartiles\n",
    "d_type = {\n",
    "      'Tumor' :[\"Simpson's Evenness\", 'Pielou Evenness',\n",
    "       'Clone Distribution Slope', \n",
    "       'Total Productive Templates', 'Templates per ng',\n",
    "       'Productive Rearrangements (Observed Richness)', 'Prod Rearr per ng',\n",
    "       'iChao1', 'Efron Thisted Estimator', 'Richness Index',\n",
    "        \"Simpson's D\", 'Simpson Clonality',\n",
    "       'Max Prod Freq',  'Z-score Simpson Clonality',\n",
    "       \"Z-score Simpson's D\", '% Tumor', 'Estimator Multiplier'],\n",
    "        'Blood':[\n",
    "            \"Simpson's Evenness\", 'Pielou Evenness',\n",
    "       'Clone Distribution Slope', 'Total Productive Templates',\n",
    "       'Productive Rearrangements (Observed Richness)', 'iChao1',\n",
    "       'Efron Thisted Estimator', \n",
    "       'Max Prod Freq', 'Simpson Clonality', \"Simpson's D\",\n",
    "        \"Simpson's D Z-score\", 'Clone Distribution Slope Z-score',\n",
    "       \"Simpson's Evenness Z-score\", \"Simpson Clonality Z-score\",\n",
    "       'Estimator Multiplier', 'Estimator Extrapolation Value']\n",
    "         }\n",
    "\n",
    "b_primary = True #True\n",
    "\n",
    "for s_type, ls_foci in d_type.items():\n",
    "    if s_type == 'Tumor':\n",
    "        df_pri = df_tum  #\n",
    "        if b_primary:\n",
    "            df_pri = df_pri.loc[df_pri.loc[:,'Tumor Site']=='Pancreas']\n",
    "    elif s_type == 'Blood':\n",
    "        df_pri = df_bld  # df_bld #\n",
    "        if b_primary:\n",
    "            df_pri = df_pri.loc[df_pri.loc[:,'Blood Collected During Metastatic Disease'] == 'NO']\n",
    "    for s_foci in ls_foci:\n",
    "        print(s_foci)\n",
    "        figsize=(3.5,2.8)\n",
    "        fig,ax=plt.subplots(dpi=300,figsize=figsize)\n",
    "        order = []\n",
    "        ls_ticks = []\n",
    "        d_pval = {}\n",
    "        df_both = pd.DataFrame()\n",
    "        for idx, s_cat in enumerate(d_order.keys()):\n",
    "            #print(s_cat)\n",
    "            ls_order = d_order[s_cat]\n",
    "            s_bad = ls_order[0]\n",
    "            s_good = ls_order[1]\n",
    "            d_replace = {s_bad:'bad',s_good:'good'}\n",
    "            a = df_pri.loc[df_pri.loc[:,s_cat]==ls_order[0],s_foci].dropna()\n",
    "            b = df_pri.loc[df_pri.loc[:,s_cat]==ls_order[1],s_foci].dropna()\n",
    "            ##statistic, pvalue = stats.ttest_ind(a,b,alternative=s_alt)\n",
    "            #print(len(a) + len(b))\n",
    "            statistic, pvalue = f_oneway(b,a)\n",
    "            df_pri['hue'] = df_pri.loc[:,s_cat].replace(d_replace)\n",
    "            df_pri['x'] = s_cat\n",
    "            df_both=pd.concat([df_both,df_pri.loc[:,['x','hue',s_foci]]])\n",
    "            for s_test in ls_order:\n",
    "                order.append((s_cat,d_replace[s_test]))\n",
    "                ls_ticks.append(s_test)\n",
    "            d_pval.update({s_cat:pvalue})\n",
    "        sns.violinplot(data=df_both,y=s_foci,x='x',hue='hue',ax=ax,alpha=0.7,linewidth=1,cut=0,inner='quartile',hue_order=['bad','good'])#hue='Ki67pos'\n",
    "        sns.stripplot(data=df_both,y=s_foci,x='x',hue='hue',s=2,dodge=True,ax=ax,palette='dark',jitter=0.2,alpha=0.3,hue_order=['bad','good']) #hue='Ki67pos'\n",
    "        #annotate\n",
    "        if len(order) == 6:\n",
    "            pairs = [(order[0],order[1]),(order[2],order[3]),(order[4],order[5])]\n",
    "            pvalues = [d_pval[list(d_order.keys())[0]],d_pval[list(d_order.keys())[1]],d_pval[list(d_order.keys())[2]]]\n",
    "        else:\n",
    "            pairs = [(order[0],order[1]),(order[2],order[3]),(order[4],order[5]),(order[6],order[7])]\n",
    "            pvalues = [d_pval[list(d_order.keys())[0]],d_pval[list(d_order.keys())[1]],d_pval[list(d_order.keys())[2]],d_pval[list(d_order.keys())[3]]]\n",
    "        reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='fdr_bh')\n",
    "        formatted_pvalues = [f'p={pvalue:.2}' for pvalue in list(pvalues)]#corrected #pvalues\n",
    "        annotator = Annotator(ax, pairs=pairs, data=df_both,y=s_foci,x='x',hue='hue',verbose=False)\n",
    "        annotator.set_custom_annotations(formatted_pvalues)\n",
    "        #annotator.configure(line_height=0.9,line_offset=10)\n",
    "        annotator.annotate()\n",
    "        ax.legend().remove()\n",
    "        if len(order) == 6:\n",
    "            ax.set_xticks([-0.2,0.2, 0.8,1.2,1.8,2.2])\n",
    "        else:\n",
    "            ax.set_xticks([-0.2,0.2, 0.8,1.2,1.8,2.2,2.8,3.2])\n",
    "        ax.set_xticklabels(ls_ticks,rotation=45)\n",
    "        ax.set_xlabel('')\n",
    "        ax.set_title(f\"{s_foci} {s_type}\", fontsize='x-large') #\n",
    "        plt.tight_layout()\n",
    "        if np.array([item < 0.05 for item in pvalues]).any():\n",
    "            fig.savefig(f'figures/violinplot_original_data_{s_type}_{s_foci}_{list(d_order.keys())[-1]}_all.png')\n",
    "        else:\n",
    "            plt.close()\n",
    "        #break\n",
    "    break\n",
    "          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd610dbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "alpha = 0.05\n",
    "# pearsons\n",
    "d_type = {\n",
    "#       'Tumor' :[\"Simpson's Evenness\", 'Pielou Evenness',\n",
    "#        'Clone Distribution Slope', \n",
    "#        'Total Productive Templates', 'Templates per ng',\n",
    "#        'Productive Rearrangements (Observed Richness)', 'Prod Rearr per ng',\n",
    "#        'iChao1', 'Efron Thisted Estimator', 'Richness Index',\n",
    "#         \"Simpson's D\", 'Simpson Clonality',\n",
    "#        'Max Prod Freq',  'Z-score Simpson Clonality',\n",
    "#        \"Z-score Simpson's D\", '% Tumor', 'Estimator Multiplier'],\n",
    "        'Blood':[\n",
    "            \"Simpson's Evenness\", 'Pielou Evenness',\n",
    "       'Clone Distribution Slope', 'Total Productive Templates',\n",
    "       'Productive Rearrangements (Observed Richness)', 'iChao1',\n",
    "       'Efron Thisted Estimator', \n",
    "       'Max Prod Freq', 'Simpson Clonality', \"Simpson's D\",\n",
    "        \"Simpson's D Z-score\", 'Clone Distribution Slope Z-score',\n",
    "       \"Simpson's Evenness Z-score\", \"Simpson Clonality Z-score\",\n",
    "       'Estimator Multiplier', 'Estimator Extrapolation Value']\n",
    "         }\n",
    "\n",
    "b_primary = True #False # \n",
    "x_value = 'pORG-78UP-All'\n",
    "for s_type, ls_foci in d_type.items():\n",
    "    if s_type == 'Tumor':\n",
    "        df_pri = df_tum  #\n",
    "        if b_primary:\n",
    "            s_type = s_type + \" Primaries\" #'Mets'\n",
    "            df_pri = df_pri.loc[df_pri.loc[:,'Tumor Site']=='Pancreas']\n",
    "    elif s_type == 'Blood':\n",
    "        df_pri = df_bld  # df_bld #\n",
    "        if b_primary:\n",
    "            s_type = s_type + \" Primaries\"  #' Mets'#\n",
    "            df_pri = df_pri.loc[df_pri.loc[:,'Blood Collected During Metastatic Disease'] == 'NO']# \n",
    "    for y_value in ls_foci:\n",
    "        fig,ax=plt.subplots(figsize=(3,2.8),dpi=200)\n",
    "        df_pearson = df_pri.loc[:,[x_value,y_value]].dropna()\n",
    "        sns.regplot(data=df_pearson, x=x_value, y=y_value,ax=ax,line_kws={'color':'k'})\n",
    "        stat, pvalue = stats.pearsonr(x=df_pearson.loc[:,x_value], y=df_pearson.loc[:,y_value])\n",
    "        ax.set_title(f'{y_value} {s_type}\\np={pvalue:.3} n={len(df_pearson)}')\n",
    "        if pvalue > alpha:\n",
    "            plt.close(fig)\n",
    "        #break\n",
    "    #break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8b802b9",
   "metadata": {},
   "source": [
    "## Section 7 <a name=\"split\"></a>\n",
    "reproduce figures\n",
    "\n",
    "evaluate high/low\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "056afc85",
   "metadata": {},
   "outputs": [],
   "source": [
    "#1/3 splits\n",
    "df_patient = pd.read_csv('data/Patient_metadata_clean.csv',index_col=0)\n",
    "#df_patient.loc[:,'OPTR ID'] = df_pri.OPTR.astype('Int64')\n",
    "#df_patient = df_patient[~df_patient.loc[:,'OPTR ID'].isna()].merge(df_tum.loc[:,['OPTR ID','Cohort']],on='OPTR ID',suffixes=('_x',''))\n",
    "#primaries\n",
    "df_pri = df_patient[(~df_patient.loc[:,'Patient Specimen ID'].str.contains('-T2')) ].copy() #(~df_patient.loc[:,'txi_pORG_Up_42_Genes'].isna())\n",
    "#& (df_pri.loc[:,'Primary Site _ Major Groups For Staging'])\n",
    "##df_pri.to_csv('Patient_List.csv')\n",
    "df_pri = df_patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a01526c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CPH\n",
    "ls_multi = [#'txi_pORG_Up_42_Genes',\n",
    "            'Original_pORG_Up_78_Genes', \n",
    "       #'Original_pSUB_Up_100_Genes', \n",
    "    #'txi_pSUB_Up_100_Genes',\n",
    "    'PurIST Score']\n",
    "d_col = {'quartiles':(4,['low','med-low','med-high','high']), #\n",
    "         #'median':(2,['low','high']),\n",
    "         #'tertiles':(3,['low','med','high']),\n",
    "        #'tertiles_h':([0, .33, 1.],['low','high']),\n",
    "        # 'tertiles_l':([0, .66, 1.],['low','high'])\n",
    "        }\n",
    "b_plot = True\n",
    "b_two = True\n",
    "for s_col, tu_cut in d_col.items():\n",
    "    i_cut = tu_cut[0]\n",
    "    labels = tu_cut[1]\n",
    "    for s_multi in ls_multi:\n",
    "        x = df_pri.loc[:,s_multi]\n",
    "        q = pd.qcut(x, q=i_cut,labels=labels) \n",
    "        if b_two:\n",
    "            if s_col == 'quartiles':\n",
    "                df_pri[s_col] = q.replace({'med-low':np.NaN,'med-high':np.NaN})\n",
    "            elif s_col == 'tertiles':\n",
    "                df_pri[s_col] = q.replace({'med':np.NaN})\n",
    "                #print('')\n",
    "            else:\n",
    "                df_pri[s_col] = q \n",
    "        else:\n",
    "            df_pri[s_col] = q \n",
    "        if b_plot:\n",
    "            df = df_pri.loc[:,[s_col,s_time,s_censor]].dropna()\n",
    "            fig, ls_order = km_plot(df,s_col,s_time,s_censor)\n",
    "            fig.suptitle(s_multi,y=.93)\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/KM_{s_col}_{s_multi}.png')\n",
    "#         break\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be3c3058",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ls_multi = ['quartiles', 'median', 'tertiles', 'tertiles_h', 'tertiles_l']\n",
    "# s_time = 'cDays from Diagnosis to FU'\n",
    "# s_censor = 'Survival'\n",
    "# for s_multi in ls_multi:\n",
    "#     #why does PurIST have extra patients?\n",
    "#     df = df_pri.loc[:,[s_time,s_censor,'Grade','Stage']]\n",
    "#     df_dummy = pd.get_dummies(df_pri.loc[:,['LV_Invasion','LN_Pos','Neoadjuvant',s_multi]],drop_first=True)\n",
    "#     df = pd.concat([df,df_dummy],axis=1)\n",
    "#     fig = cph_plot(df,f'{s_multi}_high',s_time,s_censor)\n",
    "#     fig.savefig(f'figures/CPH_{s_multi}.png')\n",
    "#     #break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6f92657",
   "metadata": {},
   "source": [
    "## figure 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "619b26ab",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# the greater the absolute value of the slope, the greater the diversity.\n",
    "# redefine liver patient?\n",
    "from scipy.stats import f_oneway\n",
    "df_pri['Tumor Distinct Clones'] = 100 - df_pri.loc[:,'Percentage Tumor-Distinct Clones in Paired Tumor Sample']\n",
    "df_pri[\"Simpson's Diversity tumor\"] = 1 - df_pri.loc[:,\"Simpson's Evenness tumor\"]\n",
    "df_pri[\"Simpson's Diversity blood\"] = 1 - df_pri.loc[:,\"Simpson's Evenness blood\"]\n",
    "\n",
    "ls_foci = [#'pORG_Primary', sanity\n",
    "#     ('Templates_per_ng','less'),\n",
    "#     ('Productive_Rearrangements','less'),\n",
    "#     ('Clone Distribution Slope blood','less'),\n",
    "#     ('Pielou Evenness blood','less'),\n",
    "#     (\"Simpson's Evenness blood\",'less'),\n",
    "#     ('Clone Distribution Slope tumor','less'),\n",
    "#     ('Pielou Evenness tumor','less'),\n",
    "#     (\"Simpson's Evenness tumor\",'greater'),\n",
    "#     (\"Simpson's Diversity tumor\",'greater'),\n",
    "#     (\"Simpson's Diversity blood\",'greater'),\n",
    "    ('Tumor Distinct Clones','less'),\n",
    "    ]\n",
    "d_order =  {\n",
    "    'PurIST Subtype':['basal-like','classical'],\n",
    "    'Cohort':['Liver','Lung'],\n",
    "    'quartiles':['high','low']}#tertiles_h # quartiles\n",
    "for s_foci_alt in ls_foci:\n",
    "    s_foci = s_foci_alt[0]\n",
    "    print(s_foci)\n",
    "    s_alt = s_foci_alt[1]\n",
    "    figsize=(3.5,2.8)\n",
    "    fig,ax=plt.subplots(dpi=300,figsize=figsize)\n",
    "    order = []\n",
    "    ls_ticks = []\n",
    "    d_pval = {}\n",
    "    df_both = pd.DataFrame()\n",
    "    for idx, s_cat in enumerate(d_order.keys()):\n",
    "        print(s_cat)\n",
    "        ls_order = d_order[s_cat]\n",
    "        s_bad = ls_order[0]\n",
    "        s_good = ls_order[1]\n",
    "        d_replace = {s_bad:'bad',s_good:'good'}\n",
    "        a = df_pri.loc[df_pri.loc[:,s_cat]==ls_order[0],s_foci].dropna()\n",
    "        b = df_pri.loc[df_pri.loc[:,s_cat]==ls_order[1],s_foci].dropna()\n",
    "        statistic, pvalue = stats.ttest_ind(a,b,alternative=s_alt)\n",
    "        print(len(a) + len(b))\n",
    "        \n",
    "        statistic, pvalue = f_oneway(df_pri.loc[df_pri.loc[:,s_cat]==ls_order[0],s_foci].dropna(),\n",
    "                                    df_pri.loc[df_pri.loc[:,s_cat]==ls_order[1],s_foci].dropna(),#alternative='greater'\n",
    "                                             )\n",
    "        df_pri['hue'] = df_pri.loc[:,s_cat].replace(d_replace)\n",
    "        df_pri['x'] = s_cat\n",
    "        df_both=pd.concat([df_both,df_pri.loc[:,['x','hue',s_foci]]])\n",
    "        for s_test in ls_order:\n",
    "            order.append((s_cat,d_replace[s_test]))\n",
    "            ls_ticks.append(s_test)\n",
    "        d_pval.update({s_cat:pvalue})\n",
    "    sns.violinplot(data=df_both,y=s_foci,x='x',hue='hue',ax=ax,alpha=0.5,linewidth=1,cut=0,inner='quartile')#hue='Ki67pos'\n",
    "    sns.stripplot(data=df_both,y=s_foci,x='x',hue='hue',s=2,dodge=True,ax=ax,palette='dark',jitter=0.2) #hue='Ki67pos'\n",
    "    #annotate\n",
    "    pairs = [(order[0],order[1]),(order[2],order[3]),(order[4],order[5])]\n",
    "    pvalues = [d_pval[list(d_order.keys())[0]],d_pval[list(d_order.keys())[1]],d_pval[list(d_order.keys())[2]]]\n",
    "    reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='fdr_bh')\n",
    "    formatted_pvalues = [f'p={pvalue:.3}' for pvalue in list(pvalues)]#corrected #pvalues\n",
    "    annotator = Annotator(ax, pairs=pairs, data=df_both,y=s_foci,x='x',hue='hue')\n",
    "    annotator.set_custom_annotations(formatted_pvalues)\n",
    "    annotator.annotate()\n",
    "    ax.legend().remove()\n",
    "    ax.set_xticks([-0.2,0.2, 0.8,1.2,1.8,2.2])\n",
    "    ax.set_xticklabels(ls_ticks,rotation=45)\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_title(f\"{s_foci}\", fontsize='x-large') #\n",
    "    plt.tight_layout()\n",
    "    fig.savefig(f'figures/violinplot_both_{s_foci}_{list(d_order.keys())[-1]}_primary.png')\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e718c0f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_both.groupby('x').count()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9873a6e9",
   "metadata": {},
   "source": [
    "## figure 2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7952a3dc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#need to get mets \n",
    "\n",
    "ls_foci = ['Original_pORG_Up_78_Genes', 'txi_pORG_Up_42_Genes',\n",
    "       'Original_pSUB_Up_100_Genes', 'txi_pSUB_Up_100_Genes',]\n",
    "d_order =  {'Cohort':['Liver','Lung'],\n",
    "            'PurIST Subtype':['basal-like','classical']}\n",
    "for s_foci in ls_foci:\n",
    "    figsize=(4,2.8)\n",
    "    fig,ax=plt.subplots(dpi=300,figsize=figsize)\n",
    "    order = []\n",
    "    ls_ticks = []\n",
    "    d_pval = {}\n",
    "    df_both = pd.DataFrame()\n",
    "    for idx, s_cat in enumerate(d_order.keys()):\n",
    "        print(s_cat)\n",
    "        ls_order = d_order[s_cat]\n",
    "        s_bad = ls_order[0]\n",
    "        s_good = ls_order[1]\n",
    "        d_replace = {s_bad:'bad',s_good:'good'}\n",
    "        a = df_pri.loc[df_pri.loc[:,s_cat]==ls_order[0],s_foci]\n",
    "        b =df_pri.loc[df_pri.loc[:,s_cat]==ls_order[1],s_foci]\n",
    "        statistic, pvalue = stats.ttest_ind(a,b, )#    alternative='greater' )\n",
    "        print(len(a) + len(b))\n",
    "        df_pri['hue'] = df_pri.loc[:,s_cat].replace(d_replace)\n",
    "        df_pri['x'] = s_cat\n",
    "        df_both=pd.concat([df_both,df_pri.loc[:,['x','hue',s_foci]]])\n",
    "        for s_test in ls_order:\n",
    "            order.append((s_cat,d_replace[s_test]))\n",
    "            ls_ticks.append(s_test)\n",
    "        d_pval.update({s_cat:pvalue})\n",
    "    sns.violinplot(data=df_both,y=s_foci,x='x',hue='hue',ax=ax,alpha=0.5,linewidth=1,inner='quartile')#hue='Ki67pos'\n",
    "    sns.stripplot(data=df_both,y=s_foci,x='x',hue='hue',s=2,dodge=True,ax=ax,palette='dark',jitter=0.2) #hue='Ki67pos'\n",
    "    #annotate\n",
    "    pairs = [(order[0],order[1]),(order[2],order[3])]\n",
    "    pvalues = [d_pval[list(d_order.keys())[0]],d_pval[list(d_order.keys())[1]]]\n",
    "    reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='fdr_bh')\n",
    "    formatted_pvalues = [f'p={pvalue:.2}' for pvalue in list(corrected)]\n",
    "    annotator = Annotator(ax, pairs=pairs, data=df_both,y=s_foci,x='x',hue='hue')\n",
    "    annotator.set_custom_annotations(formatted_pvalues)\n",
    "    annotator.annotate()\n",
    "    ax.legend().remove()\n",
    "    ax.set_xticks([-0.2,0.2, 0.8,1.2])\n",
    "    ax.set_xticklabels(ls_ticks)\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_title(f\"{s_foci.split('_')[0]} GSVA Scores\", fontsize='x-large') #{s_foci.replace('_',' ')} vs. \n",
    "    plt.tight_layout()\n",
    "    fig.savefig(f'figures/violinplot_both_{s_foci}_primary.png')\n",
    "    #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cea2e481",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CPH\n",
    "\n",
    "for s_vital_ in ls_foci:\n",
    "    s_multi = s_vital_[0]\n",
    "    df = df_pri.loc[:,[s_multi,s_time,s_censor,'Grade','Stage']]\n",
    "    df_dummy = pd.get_dummies(df_pri.loc[:,['LV_Invasion','LN_Pos','Neoadjuvant']],drop_first=True)\n",
    "    df = pd.concat([df,df_dummy],axis=1)\n",
    "    fig, cph = cph_plot(df,s_multi,s_time,s_censor,figsize=(5,3.5))\n",
    "    plt.tight_layout()\n",
    "    fig.savefig(f'figures/CPH_{s_multi}.png')\n",
    "    #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b56dfe7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "d_col = {#'quartiles':(4,['low','med-low','med-high','high']), #\n",
    "         #'median':(2,['low','high']),\n",
    "         'tertiles':(3,['low','med','high']),\n",
    "         #'#tertiles_h':([0, .33, 1.],['low','high']),\n",
    "         #'tertiles_l':([0, .66, 1.],['low','high'])\n",
    "        }\n",
    "\n",
    "for s_col, tu_cut in d_col.items():\n",
    "    i_cut = tu_cut[0]\n",
    "    labels = tu_cut[1]\n",
    "    break\n",
    "\n",
    "for s_vital_ in ls_foci:\n",
    "    s_vital = s_vital_[0]\n",
    "    print(s_vital)\n",
    "    df = df_pri.copy()\n",
    "    x = df.loc[:,s_vital]\n",
    "    q = pd.qcut(x, q=i_cut,labels=labels) \n",
    "    if s_col == 'quartiles':\n",
    "        df[s_vital] = q.replace({'med-low':np.NaN,'med-high':np.NaN})\n",
    "    elif s_col == 'tertiles':\n",
    "        df[s_vital] = q.replace({'med':np.NaN})\n",
    "    else:\n",
    "        df[s_vital] = q \n",
    "    df = df.loc[:,[s_vital,s_time,s_censor]].dropna(how='any')\n",
    "    fig, __ = km_plot(df,s_vital,s_time,s_censor)\n",
    "    fig.savefig(f'figures/KM_clinicopath_{s_vital}_subset.png')\n",
    "    #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e36e6430",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3.9.5",
   "language": "python",
   "name": "base"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
