{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78a56982",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import shutil\n",
    "import importlib\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "from skimage import io\n",
    "import tifffile\n",
    "from scipy.ndimage import median_filter\n",
    "from skimage.util import img_as_ubyte,  img_as_float\n",
    "import skimage\n",
    "from skimage.feature import blob_dog, blob_log, blob_doh\n",
    "from math import sqrt\n",
    "import scipy\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from skimage.filters import unsharp_mask\n",
    "from skimage.restoration import (denoise_tv_chambolle, denoise_bilateral,\n",
    "                                 denoise_wavelet, estimate_sigma)\n",
    "from skimage import color, morphology\n",
    "from skimage.transform import rescale\n",
    "import matplotlib as mpl\n",
    "import lifelines\n",
    "from lifelines import KaplanMeierFitter, CoxPHFitter\n",
    "from lifelines.statistics import multivariate_logrank_test\n",
    "\n",
    "from statannotations.Annotator import Annotator\n",
    "from itertools import combinations\n",
    "import statsmodels\n",
    "import warnings\n",
    "import matplotlib as mpl\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning) \n",
    "from bokeh.palettes import Colorblind\n",
    "d_colorblind = {'Liver':'#d55e00','Lung':'#0072b2',\n",
    "               'high':'#e69f00','low': '#56b4e9',\n",
    "               'basal-like':'#000000','classical':'#cc79a7',\n",
    "               'high pSUB': '#f0e442','low pSUB':'#009E73'}\n",
    "\n",
    "pal_porg_r = ('#56B4E9','#E69F00',)\n",
    "pal_liv_r = ('#0072B2','#D55E00',)\n",
    "# Set Paths\n",
    "codedir = os.getcwd()\n",
    "import util\n",
    "linebreak = '\\n'\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fb9d6a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#images: dowload from synapse.org syn51068458\n",
    "# (free account required)\n",
    "rootdir = codedir#'/home/groups/BCC_Chin_Lab/ChinData/Cyclic_Workflow/cmIF_2021-05-03_PDAC'\n",
    "regdir = f'{rootdir}/RegisteredImages'\n",
    "segdir = f'{codedir}/Segmentation'\n",
    "segdiro = f'{rootdir}/Segmentation'\n",
    "\n",
    "# clone mplex_image at https://gitlab.com/engje/mplex_image\n",
    "os.chdir('../..')\n",
    "from mplex_image import preprocess, mpimage #, cmif\n",
    "os.chdir(codedir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ddd69b4",
   "metadata": {},
   "source": [
    "# Table of contents <a name=\"contents\"></a>\n",
    "0. [functions](#func)\n",
    "1. [skimage blobs](#sk)\n",
    "2. [Foci analysis](#focifoci)\n",
    "3. [mIHC analysis](#mihc)\n",
    "4. [Patient metadata](#meta)  [Primary vs met](#primet)\n",
    "\n",
    "   [Gene alteration vs GSVA](#genes) [DDR vs TMB](#DDR) \n",
    "4. [Survival Analysis](#km) [Cohort vs clinical covariates](#livlungclin) [CPH modeling](#clin)  [CPH forest plots](#cphplot)\n",
    "5. [gene expression analysis](#geneexp) correlation heatmap, quadrant plot, gene expression versus cohort\n",
    "6. [TCR analysis](#tcell) [TCR survival](#tcells) [TCR Met site](#TCRmet)\n",
    "7. [GSVA violins](#violin2)  [GSEA bar plots](#bars2)\n",
    "8. [OLD](#split) (not used)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "604aaac5",
   "metadata": {},
   "source": [
    "## functions <a name=\"func\"></a> \n",
    "\n",
    "import from util.py\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79d07d01-797f-4454-b6b5-f9e6924f6278",
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "def annotated_stripplot_color(plotting,ls_groups,color,b_correct=False,figsize=(2.4,3)):\n",
    "    '''\n",
    "    no hue, show pvalues\n",
    "    default no correction\n",
    "    '''\n",
    "    if figsize == 0:\n",
    "        fig, ax = plt.subplots(figsize=(1.3*len(ls_groups),3.2),dpi=300)\n",
    "    else:\n",
    "        fig, ax = plt.subplots(figsize=figsize,dpi=300)\n",
    "    \n",
    "    sns.swarmplot(**plotting,hue=color,s=3,ax=ax,alpha=0.8,label='__')\n",
    "    sns.boxplot(**plotting,ax=ax,showmeans=True,medianprops={'visible': False},\n",
    "                                   whiskerprops={'visible': False},meanline=True,showcaps=False,\n",
    "                           meanprops={'color': 'k', 'ls': '-', 'lw': 2},showfliers=False,showbox=False)\n",
    "    pairs = [item for item in itertools.combinations(ls_groups,2)]\n",
    "    #dummy\n",
    "    f,a = plt.subplots()\n",
    "    annotd = Annotator(a,pairs,**plotting,verbose=False)\n",
    "    annotd.configure(test='t-test_ind',text_format=\"star\")\n",
    "    annotd.apply_test()\n",
    "    a, test_results = annotd.apply_test().annotate()\n",
    "    plt.close(f)\n",
    "    #extract annotations\n",
    "    pvalues = []\n",
    "    pairs2 = []\n",
    "    for res in test_results:\n",
    "        pairs2.append((res.data.group1,res.data.group2))\n",
    "        pvalues.append(res.data.pvalue)\n",
    "    #correct\n",
    "    reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='fdr_bh')\n",
    "    formatted_pvalues = [f'P={pvalue:.2}' for pvalue in list(pvalues)]\n",
    "    if b_correct:\n",
    "        formatted_pvalues = [f'FDR={pvalue:.2}' for pvalue in list(corrected)]\n",
    "    # create real annotator\n",
    "    annot = Annotator(ax,pairs2,**plotting,verbose=False) #annot = Annotator(ax=ax, pairs=pairs2,data=df_numbers,x=x,y=y,order=order)\n",
    "    annot.configure(fontsize='small')\n",
    "    annot.set_custom_annotations(formatted_pvalues)\n",
    "    annot.annotate()\n",
    "    return(fig,ax,test_results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ba7147b",
   "metadata": {},
   "source": [
    "##  Section 1: skimage blob detection\n",
    "\n",
    "Foci are detected here from images plus segmentation masks\n",
    "\n",
    "**You can skip and use pre-computed foci counts**\n",
    "\n",
    " <a name=\"sk\"></a> \n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f737cb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#intensity data: dowload from synapse.org syn51068458\n",
    "'''\n",
    "df_mi = pd.read_csv(f'{codedir}/data/20220720_U54-TMA_FilteredMeanIntensity_Link.csv',index_col=0)\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0f1a0ba",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#skimage blob detection (pre-calculated)\n",
    "#requires images: dowload from synapse.org syn51068458\n",
    "'''\n",
    "threshold=0.002\n",
    "d_thresh = {'pRPA':1100, 'gH2AX':1100, 'RAD51':1300}\n",
    "os.chdir(regdir)\n",
    "ls_slide = sorted(set(df_mi.slide_scene)) \n",
    "ls_marker = ['pRPA','gH2AX','RAD51',] \n",
    "df_result_all = pd.DataFrame()\n",
    "for s_slide in ls_slide:\n",
    "    print(s_slide)\n",
    "    os.chdir(f'{segdiro}/U54-TMA-9_CellposeSegmentation') #change this to path for your downloaded segmentation masks\n",
    "    s_seg = pd.Series(sorted(os.listdir()),dtype='object')[pd.Series(index=sorted(os.listdir()),dtype='object').index.str.contains(f'{s_slide}_nuc30_NucleiSegmentationBasins')].iloc[0]\n",
    "    label_image = io.imread(s_seg)\n",
    "    os.chdir(f'{regdir}/{s_slide}') # change this to path to your downloaded registered tifs\n",
    "    df_img=mpimage.parse_org()\n",
    "    for idxs, s_marker in enumerate(ls_marker):\n",
    "        intensity_image = io.imread(df_img[df_img.marker==s_marker].index[0])\n",
    "        props = skimage.measure.regionprops_table(label_image, intensity_image=intensity_image, properties=('label','bbox','mean_intensity')) # 'image','intensity_image',\n",
    "        df_props = pd.DataFrame(props,dtype='float').set_index('label')\n",
    "        df_props.columns = [item.replace('-','') for item in df_props.columns]\n",
    "        ls_index = df_props[df_props.mean_intensity>d_thresh[s_marker]].index\n",
    "        print(f'{s_marker} {len(ls_index)}')\n",
    "        for i_cell in ls_index:\n",
    "            se_cell = df_props.loc[i_cell].dropna().astype('int')\n",
    "            image = intensity_image[se_cell.bbox0:se_cell.bbox2,se_cell.bbox1:se_cell.bbox3]\n",
    "            if s_marker == 'pRPA':\n",
    "                blobs, fig = util.get_blobs2(image,min_sigma=0.1,max_sigma=2,threshold=threshold,exclude_border=1)\n",
    "            elif s_marker == 'gH2AX':\n",
    "                blobs, fig = util.get_blobs2(image,min_sigma=1,max_sigma=2,threshold=threshold,exclude_border=1)\n",
    "            else:\n",
    "                blobs, fig = util.get_blobs2(image,min_sigma=0.1,max_sigma=2,threshold=threshold,exclude_border=1)\n",
    "            df_props.loc[i_cell,'blobs'] = len(blobs)\n",
    "            if len(blobs) > 1:\n",
    "                fig.suptitle(s_marker)\n",
    "                plt.tight_layout()\n",
    "                fig.savefig(f'{codedir}/blobs/{s_marker}/{s_slide}_{s_marker}_{i_cell}.png',dpi=200)\n",
    "                plt.close(fig)\n",
    "        df_props.index = [s_slide + '_cell' + str(item) for item in df_props.index]\n",
    "        if len(ls_index)==0:\n",
    "                df_props['blobs'] = np.NaN\n",
    "        if idxs == 0:\n",
    "            df_result = df_props.rename({'blobs':f'{s_marker}_foci'},axis=1)\n",
    "        else:\n",
    "            df_result.loc[df_props.index,f'{s_marker}_foci'] = df_props.blobs\n",
    "    df_result_all = df_result_all.append(df_result)\n",
    "\n",
    "#uncomment to save (this data has been pre-computed and saved in the repo)\n",
    "#df_result_all.loc[:,df_result_all.columns.str.contains('foci')].dropna(how='all').to_csv(f'{codedir}/foci_U54-TMA-9_{\".\".join(ls_marker)}_{threshold}.csv')\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b487760",
   "metadata": {},
   "source": [
    "### Section 2: FOCI Analysis <a name=\"focifoci\"></a>\n",
    "\n",
    "Load saved foci for plotting and downstream analysis\n",
    "\n",
    "liver/lung\n",
    "\n",
    "**You don't need to run section 1 to run this**\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1de50de",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_surv = pd.read_csv(f'{codedir}/data/u54_tma_sampleannot_Link.csv')\n",
    "len(df_surv)\n",
    "\n",
    "# df_pORG = pd.read_csv(f'data/20230608_GSVA_Scores.csv')\n",
    "# #df_pORG\n",
    "# df_primary = df_surv[((df_surv.Tissue=='PDAC') | (df_surv.Tissue=='Intestinal'))].copy()\n",
    "# print(len(df_primary))\n",
    "# df_primary = df_primary.merge(df_pORG[df_pORG.Group=='GSVA_All'],on='Public_Specimen_ID',how='left')\n",
    "# df_primary.set_index(\"Public_Specimen_ID\", inplace = True)\n",
    "# len(df_primary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d508cacc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_mapper = df_primary[~df_primary.index.duplicated()]\n",
    "# ls_add = [#'txi_pORG_Up_42_Genes',\n",
    "#           'trim_padj_0.2_pORG_Up_55_Genes',\n",
    "#          #'trim_padj_0.2_pSUB_Up_100_Genes',\n",
    "#         #'txi_pSUB_Up_100_Genes'\n",
    "#   ]\n",
    "# for s_add in ls_add:\n",
    "#     d_map = dict(zip(df_mapper.index,df_mapper.loc[:,s_add]))\n",
    "#     print(len(d_map))\n",
    "#     df_surv[s_add] = df_surv.Public_Specimen_ID.map(d_map)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e3ec3a5",
   "metadata": {},
   "source": [
    "### TMA survival"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59b765f6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# ##\n",
    "# %matplotlib inline\n",
    "# alpha = 0.06\n",
    "# s_propo = ''\n",
    "# savedir = f'{codedir}'\n",
    "# s_time=  'Survival_time'#\n",
    "# s_censor='Survival'#\n",
    "# s_subtype = ''\n",
    "# s_cell = ''\n",
    "# s_type_title = ''\n",
    " \n",
    "# for s_col in ls_add:\n",
    "#     print(s_col)\n",
    "#     for cutp in [0.33,0.5,0.66]: #np.round(np.arange(0.25,1,0.25),3):#\n",
    "#             print(cutp)\n",
    "#             df_km, pvalue = util.single_km(df_surv[~df_surv.Public_Patient_ID.duplicated()],s_cell,s_subtype,s_type_title,s_col,savedir,alpha,cutp, #\n",
    "#                                        s_time,s_censor,s_propo)\n",
    "#             print(pvalue)\n",
    "#     #break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e4366b4",
   "metadata": {},
   "source": [
    "### define high/low pORG\n",
    "\n",
    "based on survival differences on TMA samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76214823-7b9a-4974-ac2f-6c239fc1996f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_surv.loc[:,[s_select_porg,'coor_mplexable']].dropna().coor_mplexable.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7916167",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_pri = df_surv[~(df_surv.Public_Patient_ID.duplicated(keep='first')) & (df_surv.Tissue.isin(['PDAC','Intestinal']))]\n",
    "# s_select_porg = 'trim_padj_0.2_pORG_Up_55_Genes'#'txi_pORG_Up_42_Genes'\n",
    "# i_pORG_txi = 0.0249 #np.quantile(df_pri.loc[:,s_select_porg].dropna(),0.66)#0.66 #0.5\n",
    "# print(i_pORG_txi)\n",
    "# # df_surv.loc[df_surv['Original_pORG_Up_78_Genes'] >= i_pORG_ori,'pORG_binary_orig'] = 'high'\n",
    "# # df_surv.loc[df_surv['Original_pORG_Up_78_Genes'] < i_pORG_ori,'pORG_binary_orig'] = 'low'\n",
    "\n",
    "# df_surv.loc[df_surv[s_select_porg] > i_pORG_txi,'pORG_binary'] = 'High'\n",
    "# df_surv.loc[df_surv[s_select_porg] <= i_pORG_txi,'pORG_binary'] = 'Low'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bff736f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save the 0608\n",
    "#df_surv.drop(['Unnamed: 0','pORG_binary','pORG_Score'],axis=1).to_csv(f'{codedir}/data/u54_tma_sampleannot_Link_new.csv')\n",
    "#df_surv.Tissue\n",
    "df_surv.Cohort.replace({'liver_cohort':'Liver','lung_cohort':'Lung'},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "192637e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(df_surv.pORG_binary_orig.value_counts())\n",
    "print(df_surv[(df_surv.Tissue.isin(['PDAC','Intestinal'])) & (~df_surv.Public_Specimen_ID.duplicated())].pORG_binary.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "899c9d63",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load combined\n",
    "s_sample = '20220721_U54-TMA'#'20220711_U54-TMA' #'20220409_JP-TMAs_IMC-TMAs'\n",
    "s_names = 'Combined' #'unnamed' #\n",
    "s_type = 'PDAC'\n",
    "df_lei = pd.read_csv(f'{codedir}/data/{s_sample}_{s_names}Celltypes_{s_type}_Link.csv',index_col=0)\n",
    "\n",
    "d_patient = dict(zip(df_surv.coor_mplexable,df_surv.Public_Patient_ID))\n",
    "\n",
    "df_lei['Tissue'] = df_lei.Patient.map(dict(zip(df_surv.Public_Patient_ID,df_surv.Tissue)))\n",
    "# number of epithelial\n",
    "df_epi = pd.read_csv(f'{codedir}/data/results_20220721_U54-TMA_CellTypeCounts_byPatient_byleidencelltype5_PDAC_Link.csv',index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f19c9989",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ls_marker = ['gH2AX','pRPA','RAD51']\n",
    "threshold=0.002\n",
    "df_foci2 = pd.read_csv(f'{codedir}/data/foci_U54-TMA-9_{\".\".join(ls_marker)}_{threshold}.csv',index_col=0)\n",
    "df_foci2.index = [item.replace('.0','') for item in df_foci2.index]\n",
    "df_foci2 = df_foci2.fillna(0)\n",
    "df_foci2['scene'] = [item.split('_cell')[0] for item in df_foci2.index]\n",
    "df_foci2['Patient'] = df_foci2.scene.map(d_patient)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68276502",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot Ki67 versus pRPA foci\n",
    "s_define = 'pORG_binary'#'pORG_binary_txi'#'pORG_binary_orig'#\n",
    "df_lei['Cohort'] = df_lei.Patient.map(dict(zip(df_surv.Public_Patient_ID,df_surv.Cohort)))\n",
    "df_lei['pORG_binary'] = df_lei.Patient.map(dict(zip(df_surv.Public_Patient_ID,df_surv.loc[:,s_define])))\n",
    "df_lei['Ki67pos'] = np.nan\n",
    "df_lei.loc[df_lei.Ki67>3*256,'Ki67pos'] = 'Ki67+'\n",
    "df_lei.Ki67pos.fillna('Ki67-',inplace=True)\n",
    "ls_foci =['pRPA_foci','gH2AX_foci','RAD51_foci']\n",
    "df_lei_foci = df_lei.merge(df_foci2.loc[:,ls_foci],left_index=True,right_index=True,how='left')\n",
    "df_lei_foci.loc[:,ls_foci] = df_lei_foci.loc[:,ls_foci].fillna(0)\n",
    "for s_col in ls_foci:\n",
    "    print(s_col)\n",
    "    df_lei_foci[f'{s_col}_pos'] = df_lei_foci.loc[:,s_col] > 0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ca7f5fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#use df foci sum 2 (lower threshold)\n",
    "\n",
    "ls_index = df_lei[(df_lei.leidencelltype5=='epithelial')].index\n",
    "print(len(ls_index))\n",
    "df_foci_sum2 = df_lei_foci.loc[df_lei_foci.index.isin(ls_index)].groupby('Patient').sum()\n",
    "df_foci_sum2.drop(df_foci_sum2.columns[df_foci_sum2.dtypes=='object'],axis=1,inplace=True)\n",
    "for s_marker in ls_marker:\n",
    "    print(s_marker)\n",
    "    df_foci_sum2[f'mean_{s_marker}_foci'] = (df_foci_sum2.loc[df_foci_sum2.index.isin(df_epi.index),f'{s_marker}_foci']/df_epi.epithelial).fillna(0)\n",
    "    df_foci_sum2[f'log_mean_{s_marker}_foci'] = np.log(df_foci_sum2.loc[:,f'mean_{s_marker}_foci'] + 1)\n",
    "    df_foci_sum2[f'mean_{s_marker}_foci_pos'] = (df_foci_sum2.loc[df_foci_sum2.index.isin(df_epi.index),f'{s_marker}_foci_pos']/df_epi.epithelial).fillna(0)\n",
    "\n",
    "df_foci_sum2['Public_Patient_ID'] = df_foci_sum2.index\n",
    "df_surv = df_surv.merge(df_foci_sum2,on='Public_Patient_ID',how='left',suffixes=('_1',''))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bcf3a0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#use df foci sum 2 (Ki67 lower threshold)\n",
    "for s_ki67 in df_lei.Ki67pos.dropna().unique():\n",
    "    ls_index = df_lei[(df_lei.Ki67pos==s_ki67) & (df_lei.leidencelltype5=='epithelial')].index\n",
    "    df_ki67 = df_lei.loc[ls_index].groupby('Patient').count()\n",
    "    df_foci_sum2 = df_lei_foci.loc[df_lei_foci.index.isin(ls_index)].groupby('Patient').sum()\n",
    "    df_foci_sum2.drop(df_foci_sum2.columns[df_foci_sum2.dtypes=='object'],axis=1,inplace=True)\n",
    "    for s_marker in ls_marker:\n",
    "        df_foci_sum2[f'mean_{s_marker}_foci_in_{s_ki67}'] = (df_foci_sum2.loc[df_foci_sum2.index.isin(df_epi.index),f'{s_marker}_foci']/df_ki67.Ki67).fillna(0)\n",
    "        df_foci_sum2[f'log_mean_{s_marker}_foci_in_{s_ki67}'] = np.log(df_foci_sum2.loc[:,f'mean_{s_marker}_foci_in_{s_ki67}'] + 1)\n",
    "        df_foci_sum2[f'mean_{s_marker}_foci_pos_in_{s_ki67}'] = (df_foci_sum2.loc[df_foci_sum2.index.isin(df_epi.index),f'{s_marker}_foci_pos']/df_ki67.Ki67).fillna(0)\n",
    "    \n",
    "df_foci_sum2['Public_Patient_ID'] = df_foci_sum2.index\n",
    "df_surv = df_surv.merge(df_foci_sum2,on='Public_Patient_ID',how='left',suffixes=('_1',''))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93bd50b1-a867-431a-96f9-5ca93b515366",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_surv.pORG_binary.replace('low','Low',inplace=True)\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01f67704-02db-4430-9160-7424a7b5fd62",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_id = pd.read_csv('../annotation/Patient_IDs.csv')\n",
    "# d_id = dict(zip(df_id.loc[:,'Biolibrary.Subject.ID'],df_id.loc[:,'OPTR']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d32c6ef",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "importlib.reload(util)\n",
    "pal_porg = ('#E69F00','#56B4E9',)\n",
    "pal_liv = ('#0072B2','#D55E00')\n",
    "sns.set_palette(pal_porg)\n",
    "d_groups = {'pORG_binary':['High','Low'],'Cohort':['Liver','Lung']}\n",
    "d_result_fig = {}\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    for s_group in ['pORG_binary','Cohort']:\n",
    "        if s_group == 'pORG_binary':\n",
    "            figsize=(2.1,3)\n",
    "        else:\n",
    "            figsize=(2.1,2.1)\n",
    "        s_type = ''\n",
    "        s_cell = 'epithelial'\n",
    "        ls_tissue = ['Intestinal', #'Liver_Met','Lung_Met',\n",
    "                     'PDAC',]\n",
    "        s_propo='in'\n",
    "        df_plot = df_surv.loc[(df_surv.Tissue.isin(ls_tissue))& (~df_surv.coor_mplexable.duplicated())]#& (~df_surv.Public_Patient_ID.duplicated())\n",
    "        ls_foci = ['mean_pRPA_foci',#'log_mean_pRPA_foci','mean_pRPA_foci_pos',\n",
    "                         #'log_mean_pRPA_foci_in_Ki67+',  'mean_pRPA_foci_pos_in_Ki67+'\n",
    "                   'mean_pRPA_foci_in_Ki67+',\n",
    "                      ]\n",
    "        linebreak = '\\n'\n",
    "        for s_marker in ls_foci:\n",
    "            #util.qq_plot_hist(df_plot,s_group,s_marker)\n",
    "            df_mean = df_plot.loc[:,[s_marker,'Public_Patient_ID']].groupby('Public_Patient_ID').mean()\n",
    "            print(df_plot.loc[:,[s_marker,'Public_Patient_ID','coor_mplexable',s_group]].dropna().coor_mplexable.nunique())\n",
    "            df_mean[s_group] = df_mean.index.map(dict(zip(df_plot.Public_Patient_ID,df_plot.loc[:,s_group])))\n",
    "            fig, __,__,__ = util.categorical_correlation_boxplot(df_mean,s_group,s_marker,s_type,s_cell,\n",
    "                                                 alpha=1.05,s_propo=s_propo,b_ttest=True,figsize=(2.3,2.8)) #\n",
    "            if s_marker == 'mean_pRPA_foci_in_Ki67+':\n",
    "                fig.get_axes()[0].set_ylabel(f\"{s_marker.replace('_',' ').replace('mean','Mean').replace('in',f'per{linebreak}')} {s_cell} cell\")\n",
    "            else:\n",
    "                fig.get_axes()[0].set_ylabel(f\"{s_marker.replace('_',' ').replace('mean','Mean')}\\n{s_propo.replace('in','per')} {s_cell} cell\")\n",
    "            fig.get_axes()[0].set_xlabel(f\"{s_group.split('_')[0]}\")\n",
    "            fig.get_axes()[0].set_title(fig.get_axes()[0].title.get_text().split('\\n ')[-1],fontsize='medium')\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/boxplotFigure3_d_{s_marker}_{s_group}.pdf')#break\n",
    "            plt.close(fig)\n",
    "            ls_groups=d_groups[s_group]\n",
    "            plotting = {'data':df_mean,'x':s_group,'y':s_marker,'order':ls_groups}\n",
    "            fig,ax,test_results = util.annotated_stripplot(plotting,ls_groups,figsize=figsize,b_correct=False)\n",
    "            ax.set_xlabel(f'{s_group.split(\"_\")[0]}')\n",
    "            if s_marker == 'mean_pRPA_foci_in_Ki67+':\n",
    "                ax.set_ylabel(f\"{s_marker.replace('_',' ').replace('mean','Mean').replace('in',f'per{linebreak}')} {s_cell} cell\")\n",
    "            else:\n",
    "                ax.set_ylabel(f\"{s_marker.replace('_',' ').replace('mean','Mean')}\\n{s_propo.replace('in','per')} {s_cell} cell\")\n",
    "            plt.tight_layout()\n",
    "            \n",
    "            if s_group == 'Cohort':\n",
    "                s_save = f'ED_Fig3_6_{s_marker}_{s_group}'\n",
    "                d_result_fig.update({s_save:df_mean})\n",
    "            else:\n",
    "                s_save = f'Figure3_d_{s_marker}_{s_group}'\n",
    "                #d_result_fig.update({s_save:df_mean})\n",
    "            #fig.savefig(f'figures/{s_save}.pdf')#break\n",
    "            # df_mean['OPTR'] = df_mean.index.map(d_id).astype('Int64')\n",
    "            # df_mean.to_csv(f'results_foci_{s_group}_{s_marker}.csv')\n",
    "        #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "005d4efd-60fa-4018-b403-3beb828d0248",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for key, df in d_result_fig.items():\n",
    "#     if key == 'ED_Fig3_6_mean_pRPA_foci_Cohort':\n",
    "#         df.to_csv(f'Source_Extended_Data_Figure6_A_pt1.csv')\n",
    "#     else:\n",
    "#         df.to_csv(f'Source_Extended_Data_Figure6_A_pt2.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04a4d594-53d0-4925-a25d-68b354a34c22",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #d_result_fig\n",
    "# #excel func\n",
    "# def open_write_excel(d_result_fig,filename='Source_Data_Figure3.xlsx'):\n",
    "#     if not os.path.exists(filename):\n",
    "#         with pd.ExcelWriter(filename,mode='w',engine=\"openpyxl\") as writer:\n",
    "#             for sheet_name, df in d_result_fig.items():\n",
    "#                 print(f'saving {sheet_name[0:30]}')\n",
    "#                 df.to_excel(writer, sheet_name=sheet_name[0:30])\n",
    "#     else:\n",
    "#         with pd.ExcelWriter(filename,mode='a',if_sheet_exists='replace',engine=\"openpyxl\") as writer:\n",
    "#             for sheet_name, df in d_result_fig.items():\n",
    "#                 print(f'saving {sheet_name[0:30]}')\n",
    "#                 df.to_excel(writer, sheet_name=sheet_name[0:30])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36d4944c-9071-4293-abc6-7217c0215143",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# #df_plot.loc[:,[s_marker,'Public_Patient_ID']].groupby('Public_Patient_ID').mean()\n",
    "# df_mean = df_plot.loc[:,['pRPA_foci','Public_Patient_ID']].groupby('Public_Patient_ID').sum()\n",
    "# df_mean['OPTR'] = df_mean.index.map(d_id).astype('Int64')\n",
    "# df_mean.to_csv(f'results_foci_total_pRPA_foci.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81c68e2d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# #linear correlation\n",
    "# sns.set_palette(pal_liv)\n",
    "# s_score = 'pORG_Score'\n",
    "# for s_foci in ls_foci:\n",
    "#     df_plot = df_plot[(~df_plot.Public_Patient_ID.duplicated()) & (df_surv.Tissue.isin(ls_tissue))] # primary tumors\n",
    "#     b_score = df_plot.loc[:,s_score].notna()\n",
    "#     r, pvalue = stats.pearsonr(x=df_plot.loc[b_score,s_foci].fillna(0),y=df_plot.loc[b_score,s_score])\n",
    "#     #if s_foci.find('Ki67') > -1:\n",
    "#     #r, pvalue =stats.spearmanr(df_plot.loc[b_score,s_foci].fillna(0),df_plot.loc[b_score,s_score])\n",
    "#     print(pvalue)\n",
    "#     if pvalue < 1.1:\n",
    "#         fig, ax = plt.subplots(figsize=(3,2),dpi=200)\n",
    "#         sns.regplot(x=df_plot.loc[b_score,s_foci],y=df_plot.loc[b_score,s_score],\n",
    "#                     ax=ax,scatter_kws={'s':3})\n",
    "#         ax.set_title(f'{s_foci} p = {pvalue:.3}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "331246e7-b312-401c-97f0-a2ccae6c3a74",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_lei_foci.pORG_binary.replace({'low':'Low'},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05d2fab1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# #compare Ki67+ versus negative, double violinplots\n",
    "# #compare all the combinations, double violinplots\n",
    "# %matplotlib inline\n",
    "# import seaborn as sns\n",
    "# from scipy import stats\n",
    "# from statannotations.Annotator import Annotator\n",
    "# from itertools import combinations\n",
    "# import statsmodels\n",
    "# from statsmodels.stats.multicomp import pairwise_tukeyhsd\n",
    "# ls_foci =['pRPA_foci','gH2AX_foci','RAD51_foci']\n",
    "# s_compare = 'Ki67'#'all' #\n",
    "# import warnings\n",
    "# with warnings.catch_warnings():\n",
    "#     warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "#     for s_cat in ['pORG_binary','Cohort']:\n",
    "#         ls_order = ['Ki67-','Ki67+']\n",
    "#         for s_foci in [ls_foci[0]]:\n",
    "#             if s_compare == 'all':\n",
    "#                 figsize=(3.5,3)\n",
    "#             else:\n",
    "#                 figsize=(2.8,2.8)\n",
    "#             fig,ax=plt.subplots(dpi=300,figsize=figsize)\n",
    "#             df_both = pd.DataFrame()\n",
    "#             order = []\n",
    "#             d_pval = {}\n",
    "#             for idx, s_cohort in enumerate(df_lei_foci.loc[:,s_cat].dropna().unique()):\n",
    "#                 print(s_cohort)\n",
    "#                 df_plot_foci =df_lei_foci[(df_lei_foci.leidencelltype5=='epithelial')\n",
    "#                                           & (df_lei_foci.Tissue=='PDAC') & (df_lei_foci.loc[:,s_cat]==s_cohort)]\n",
    "#                 statistic, pvalue = stats.ttest_ind(df_plot_foci.loc[df_plot_foci.Ki67pos=='Ki67+',s_foci],\n",
    "#                                                     df_plot_foci.loc[df_plot_foci.Ki67pos=='Ki67-',s_foci],\n",
    "#                                                    alternative='greater')\n",
    "#                 for s_order in ls_order:\n",
    "#                     order.append((s_cohort,s_order))\n",
    "#                 df_both = pd.concat([df_both,df_plot_foci])\n",
    "#                 d_pval.update({s_cohort:pvalue})\n",
    "#             sns.violinplot(data=df_both,hue='Ki67pos',y=s_foci,x=s_cat,ax=ax,alpha=0.5,linewidth=0.5)\n",
    "#             sns.stripplot(data=df_both,hue='Ki67pos',y=s_foci,x=s_cat,s=1,dodge=True,ax=ax,palette='dark',jitter=0.2)\n",
    "#             #annotate\n",
    "#             if s_compare == 'all':\n",
    "#                 pairs = list(combinations(order, r=2))\n",
    "#                 annotator = Annotator(ax, pairs=pairs, data=df_both, y=s_foci,x=s_cat,hue='Ki67pos')\n",
    "#                 annotator.configure(test=\"t-test_ind\",line_width=1)#,alternative='greater\n",
    "#                 pvalues = annotator.apply_test().annotations #annotator.apply_and_annotate() # \n",
    "#                 pvalues = [item.data.pvalue for item in pvalues]     \n",
    "#                 reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='bonferroni')\n",
    "#                 formatted_pvalues = [f'p={pvalue:.2}' for pvalue in list(corrected)]\n",
    "#                 annotator.set_custom_annotations(formatted_pvalues)\n",
    "#                 annotator.annotate()\n",
    "#             else:\n",
    "#                 pairs = [(order[0],order[1]),(order[2],order[3])]\n",
    "#                 pvalues = [d_pval[pairs[0][0][0]],d_pval[pairs[1][0][0]]]\n",
    "#                 reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='bonferroni')\n",
    "#                 formatted_pvalues = [f'p={pvalue:.2}' for pvalue in list(corrected)]\n",
    "#                 annotator = Annotator(ax, pairs=pairs, data=df_both, y=s_foci,x=s_cat,hue='Ki67pos')\n",
    "#                 annotator.set_custom_annotations(formatted_pvalues)\n",
    "#                 annotator.annotate()\n",
    "            \n",
    "#             ax.set_title(f\"{s_foci.replace('_',' ')} vs. {s_cat.split('_')[0]}\", fontsize='x-large')\n",
    "#             ax.set_xlabel(s_cat)\n",
    "#             ax.set_ylabel(f\"No. {s_foci.replace('_',' ')}\")\n",
    "#             h, l = ax.get_legend_handles_labels()\n",
    "#             labels =  [f'_{item}' if ind < 2 else item for ind,item in enumerate(l)]\n",
    "#             ax.legend(h,labels,title='',fontsize='small',markerscale=.5,bbox_to_anchor=(1.01,0.9))\n",
    "#             plt.tight_layout()\n",
    "#             #fig.savefig(f'{codedir}/violinplot_both_{s_foci}_{s_cohort}_{s_compare}.png')\n",
    "#         if s_cat == 'pORG_binary':\n",
    "#             df_both.to_csv(f'results_foci_Ki67.csv')\n",
    "#             #break\n",
    "#         #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f068f888",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# ## make the bargraphs\n",
    "# # percent pRPA epithelial cells\n",
    "# ls_foci =['pRPA_foci','gH2AX_foci','RAD51_foci']\n",
    "# d_cats = {'pORG_binary':['high','low'],'Cohort':['liver_cohort','lung_cohort']}\n",
    "# for s_cat, ls_order in d_cats.items():\n",
    "#     for s_foci in [ls_foci[0]]:\n",
    "#         figsize=(2.8,2.5)\n",
    "#         fig,ax=plt.subplots(dpi=300,figsize=figsize)\n",
    "#         df_both = pd.DataFrame()\n",
    "#         d_pval = {}\n",
    "#         for idx, s_cohort in enumerate(df_lei_foci.loc[:,s_cat].dropna().unique()):\n",
    "#             print(s_cohort)\n",
    "#             df_plot_foci =df_lei_foci[(df_lei_foci.leidencelltype5=='epithelial')\n",
    "#                                       & (df_lei_foci.Tissue=='PDAC') & (df_lei_foci.loc[:,s_cat]==s_cohort)]\n",
    "#             df_both = pd.concat([df_both,df_plot_foci])\n",
    "#         df_obs = df_both[df_both.loc[:,s_foci] > 0].loc[:,s_cat]\n",
    "#         df_exp = df_both.loc[:,s_cat]\n",
    "#         #chi\n",
    "#         f_obs = df_obs.value_counts().loc[ls_order]\n",
    "#         f_exp = f_obs.sum() * df_exp.value_counts(normalize=True).loc[ls_order]\n",
    "#         statistic, pvalue =  stats.chisquare(f_obs, f_exp)\n",
    "#         df_obs.value_counts(normalize=True).loc[ls_order].plot(kind='bar',ax=ax,color=['mediumpurple','deepskyblue'])\n",
    "#         ax.set_title(f'{s_foci} vs. {s_cat}\\nChi-squared p={pvalue:.3}')\n",
    "#         ax.set_ylabel(f'Fraction of {s_foci.split(\"_\")[0]}+ Epithelial')\n",
    "#         ax.set_xticklabels([item.replace('_cohort','') for item in ls_order])\n",
    "#         plt.tight_layout()\n",
    "#         xtickslocs = ax.get_xticks()\n",
    "#         for idx, s_order in enumerate(ls_order):\n",
    "#             s = f'{df_obs.value_counts().loc[s_order]}\\nof\\n{df_exp.value_counts().loc[s_order]}'\n",
    "#             plt.text(-0.2+idx,0.3,s)\n",
    "#         break\n",
    "#     #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05a48341-7bbf-4b28-8429-9559974f258a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## make the boxplots\n",
    "# percent pRPA epithelial cells\n",
    "pal_porg = ('#E69F00','#56B4E9',)\n",
    "pal_liv = ('#0072B2','#D55E00')\n",
    "sns.set_palette(pal_porg)\n",
    "d_cats = {'pORG_binary':['high','low'],'Cohort':['liver_cohort','lung_cohort']}\n",
    "df_out = pd.DataFrame()\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    for s_cat, ls_order in d_cats.items():\n",
    "        for s_foci in [ls_foci[0]]:\n",
    "            figsize=(2.1,2.2)\n",
    "            #fig,ax=plt.subplots(dpi=300,figsize=figsize)\n",
    "            df_both = pd.DataFrame()\n",
    "            d_pval = {}\n",
    "            for idx, s_cohort in enumerate(df_lei_foci.loc[:,s_cat].dropna().unique()):\n",
    "                print(s_cohort)\n",
    "                df_plot_foci =df_lei_foci[(df_lei_foci.leidencelltype5=='epithelial') #& (df_lei_foci.loc[:,s_foci] > 0)\n",
    "                                          & (df_lei_foci.Tissue.isin(ls_tissue)) & (df_lei_foci.loc[:,s_cat]==s_cohort)]\n",
    "                print(df_plot_foci.Patient.nunique())\n",
    "                df_both = pd.concat([df_both,df_plot_foci])\n",
    "            df_mean = pd.DataFrame(df_both.loc[:,[f'{s_foci}_pos','Patient']].groupby('Patient').mean())\n",
    "            df_mean[s_cat] = df_mean.index.map(dict(zip(df_both.Patient,df_both.loc[:,s_cat])))\n",
    "            fig, __, __, __ = util.categorical_correlation_boxplot(df_mean,s_cat,f'{s_foci}_pos',s_type='',s_cell=f\"epithelial\",\n",
    "                                             alpha=1.05,s_propo='in',b_ttest=True,figsize=(2,2))\n",
    "            #fig.get_axes()[0].set_ylabel(f\"{s_marker.replace('_',' ').replace('mean ','')}\\nper {s_cell} cell\")\n",
    "            fig.get_axes()[0].set_xlabel(f\"{s_cat.split('_')[0]}\")\n",
    "            fig.get_axes()[0].set_title(fig.get_axes()[0].title.get_text().replace('_',' '),fontsize='small')#.split('\\n ')[-1]\n",
    "            fig.savefig(f'figures/Supp13_{s_foci}_{s_cat}.pdf')\n",
    "            plt.close(fig)\n",
    "            y=f'{s_foci}_pos'\n",
    "            ls_groups=d_groups[s_cat]\n",
    "            df_mean['value'] = y\n",
    "            plotting = {'data':df_mean,'x':s_cat,'y':y,'order':ls_groups}\n",
    "            fig,ax,test_results = util.annotated_stripplot(plotting,ls_groups,figsize=figsize,b_correct=False)\n",
    "            ax.set_xlabel(f'{s_cat.split(\"_\")[0]}')\n",
    "            ax.set_ylabel(f'Fraction {y.replace(\"_foci_pos\",\" foci+\")}{linebreak}in epithelial cells')\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/Ext6_{s_foci}_{s_cat}.pdf')\n",
    "            df_out = pd.concat([df_out,df_mean])\n",
    "            #break\n",
    "        #break\n",
    "print(len(df_out))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc8f635c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# ## make the bargraphs\n",
    "# # percent pRPA epithelial cells\n",
    "# # percent of those that are Ki67+\n",
    "# d_cats = {'pORG_binary':['high','low'],'Cohort':['liver_cohort','lung_cohort']}\n",
    "# for s_cat, ls_order in d_cats.items():\n",
    "#     for s_foci in [ls_foci[0]]:\n",
    "#         figsize=(2.8,2.5)\n",
    "#         fig,ax=plt.subplots(dpi=300,figsize=figsize)\n",
    "#         df_both = pd.DataFrame()\n",
    "#         d_pval = {}\n",
    "#         for idx, s_cohort in enumerate(df_lei_foci.loc[:,s_cat].dropna().unique()):\n",
    "#             print(s_cohort)\n",
    "#             df_plot_foci =df_lei_foci[(df_lei_foci.leidencelltype5=='epithelial') & (df_lei_foci.loc[:,s_foci] > 0)\n",
    "#                                       & (df_lei_foci.Tissue=='PDAC') & (df_lei_foci.loc[:,s_cat]==s_cohort)]\n",
    "#             df_both = pd.concat([df_both,df_plot_foci])\n",
    "#         df_obs = df_both[df_both.loc[:,'Ki67pos']=='Ki67+'].loc[:,s_cat]\n",
    "#         df_exp = df_both.loc[:,s_cat]\n",
    "#         #chi\n",
    "#         f_obs = df_obs.value_counts().loc[ls_order]\n",
    "#         f_exp = f_obs.sum() * df_exp.value_counts(normalize=True).loc[ls_order]\n",
    "#         statistic, pvalue =  stats.chisquare(f_obs, f_exp)\n",
    "#         df_plot = df_obs.value_counts()/df_exp.value_counts()\n",
    "#         df_plot.loc[ls_order].plot(kind='bar',ax=ax,color=['mediumpurple','deepskyblue'])\n",
    "#         ax.set_title(f'{s_foci} vs. {s_cat}\\nChi-squared p={pvalue:.3}')\n",
    "#         ax.set_ylabel(f'Fration of {s_foci.split(\"_\")[0]}+ Epithelial\\nthat are Proliferating')\n",
    "#         ax.set_xticklabels([item.replace('_cohort','') for item in ls_order])\n",
    "#         plt.tight_layout()\n",
    "#         xtickslocs = ax.get_xticks()\n",
    "#         for idx, s_order in enumerate(ls_order):\n",
    "#             s = f'{df_obs.value_counts().loc[s_order]}\\nof\\n{df_exp.value_counts().loc[s_order]}'\n",
    "#             plt.text(-0.2+idx,0.15,s)\n",
    "#         #break\n",
    "#     #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe760931",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## make the boxplots\n",
    "# percent pRPA epithelial cells\n",
    "# percent of those that are Ki67+\n",
    "pal_porg = ('#E69F00','#56B4E9',)\n",
    "pal_liv = ('#0072B2','#D55E00')\n",
    "sns.set_palette(pal_porg)\n",
    "d_cats = {'pORG_binary':['high','low'],'Cohort':['liver_cohort','lung_cohort']}\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    for s_cat, ls_order in d_cats.items():\n",
    "        for s_foci in [ls_foci[0]]:\n",
    "            figsize=(2.1,2.2)\n",
    "            #fig,ax=plt.subplots(dpi=300,figsize=figsize)\n",
    "            df_both = pd.DataFrame()\n",
    "            d_pval = {}\n",
    "            for idx, s_cohort in enumerate(df_lei_foci.loc[:,s_cat].dropna().unique()):\n",
    "                print(s_cohort)\n",
    "                df_plot_foci =df_lei_foci[(df_lei_foci.leidencelltype5=='epithelial') & (df_lei_foci.loc[:,s_foci] > 0)\n",
    "                                          & (df_lei_foci.Tissue.isin(ls_tissue)) & (df_lei_foci.loc[:,s_cat]==s_cohort)]\n",
    "                df_both = pd.concat([df_both,df_plot_foci])\n",
    "            df_sum = df_both.loc[:,[s_cat,'Patient','Ki67pos']].groupby('Patient').value_counts().unstack().reset_index().fillna(0)\n",
    "            df_sum['Fraction Ki67+'] = df_sum.loc[:,'Ki67+']/(df_sum.loc[:,['Ki67+','Ki67-']].sum(axis=1))\n",
    "            fig, __, __, __ = util.categorical_correlation_boxplot(df_sum,s_cat,'Fraction Ki67+',s_type='',s_cell=f\"{s_foci.split('_')[0]}+\",\n",
    "                                             alpha=1.05,s_propo='in',b_ttest=True,figsize=(2,2))\n",
    "            #fig.get_axes()[0].set_ylabel(f\"{s_marker.replace('_',' ').replace('mean ','')}\\nper {s_cell} cell\")\n",
    "            fig.get_axes()[0].set_xlabel(f\"{s_cat.split('_')[0]}\")\n",
    "            fig.get_axes()[0].set_title(fig.get_axes()[0].title.get_text().replace('_',' '),fontsize='small')#.split('\\n ')[-1]\n",
    "            fig.savefig(f'figures/Supp13_{s_foci}_{s_cat}_ki67.pdf')#break\n",
    "            plt.close(fig)\n",
    "            y='Fraction Ki67+'\n",
    "            df_mean['value'] = y\n",
    "            ls_groups=d_groups[s_cat]\n",
    "            plotting = {'data':df_sum,'x':s_cat,'y':y,'order':ls_groups}\n",
    "            fig,ax,test_results = util.annotated_stripplot(plotting,ls_groups,figsize=figsize,b_correct=False)\n",
    "            ax.set_xlabel(f'{s_cat.split(\"_\")[0]}')\n",
    "            ax.set_ylabel(f'{y.replace(\"_\",\" \")} in{linebreak}{s_foci.split(\"_\")[0]}+ epithelial cells')\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/Ext6_frac_prolif_in_{s_foci}+_{s_cat}.pdf')\n",
    "            df_out = pd.concat([df_out,df_mean])\n",
    "        #break\n",
    "print(len(df_out))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "327ae575-f305-4d55-84bc-d2efb8aab7e7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## make the boxplots\n",
    "# percent pRPA epithelial cells\n",
    "# percent of those that are Ki67+\n",
    "# percent of prolif with pRPA+\n",
    "pal_porg = ('#E69F00','#56B4E9',)\n",
    "pal_liv = ('#0072B2','#D55E00')\n",
    "sns.set_palette(pal_porg)\n",
    "d_cats = {'pORG_binary':['high','low'],'Cohort':['liver_cohort','lung_cohort']}\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    for s_cat, ls_order in d_cats.items():\n",
    "        for s_foci in [ls_foci[0]]:\n",
    "            figsize=(2.1,2.2)\n",
    "            #fig,ax=plt.subplots(dpi=300,figsize=figsize)\n",
    "            df_both = pd.DataFrame()\n",
    "            d_pval = {}\n",
    "            for idx, s_cohort in enumerate(df_lei_foci.loc[:,s_cat].dropna().unique()):\n",
    "                print(s_cohort)\n",
    "                df_plot_foci =df_lei_foci[(df_lei_foci.leidencelltype5=='epithelial') & (df_lei_foci.loc[:,'Ki67pos']=='Ki67+')\n",
    "                                          & (df_lei_foci.Tissue.isin(ls_tissue)) & (df_lei_foci.loc[:,s_cat]==s_cohort)]\n",
    "                df_both = pd.concat([df_both,df_plot_foci])\n",
    "            df_sum = df_both.loc[:,[s_cat,'Patient',f'{s_foci}_pos']].groupby('Patient').value_counts().unstack().reset_index().fillna(0)\n",
    "            df_sum.columns = df_sum.columns.astype('str')\n",
    "            df_sum[f'Fraction {s_foci}+'] = df_sum.loc[:,'True']/(df_sum.loc[:,['True','False']].sum(axis=1))\n",
    "            fig, __, __, __ = util.categorical_correlation_boxplot(df_sum,s_cat,f'Fraction {s_foci}+',s_type='',s_cell=f\"Ki67+\",\n",
    "                                             alpha=1.05,s_propo='in',b_ttest=True,figsize=(2,2))\n",
    "            #fig.get_axes()[0].set_ylabel(f\"{s_marker.replace('_',' ').replace('mean ','')} per\\n{s_cell} cell\")\n",
    "            fig.get_axes()[0].set_xlabel(f\"{s_group.split('_')[0]}\")\n",
    "            fig.get_axes()[0].set_title(fig.get_axes()[0].title.get_text().replace('_',' '),fontsize='small')#\n",
    "            fig.savefig(f'figures/Supp13_{s_foci}_{s_cat}_percent.pdf')#break\n",
    "            plt.close(fig)\n",
    "            ls_groups=d_groups[s_cat]\n",
    "            y=f'Fraction {s_foci}+'\n",
    "            df_mean['value'] = y\n",
    "            plotting = {'data':df_sum,'x':s_cat,'y':y,'order':ls_groups}\n",
    "            fig,ax,test_results = util.annotated_stripplot(plotting,ls_groups,figsize=figsize,b_correct=False)\n",
    "            ax.set_xlabel(f'{s_cat.split(\"_\")[0]}')\n",
    "            ax.set_ylabel(f'{y.split(\"_\")[0]}+ in{linebreak}Ki67+ epithelial cells')\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/Ext6_frac_{s_foci}+-in_prolif__{s_cat}.pdf')\n",
    "            df_out = pd.concat([df_out,df_mean])\n",
    "print(len(df_out))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deadb14d-f26b-40dd-8309-c7b68dd390b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_out.to_csv(f'Source_Extended_Data_Figure6_B-D.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be8bfceb",
   "metadata": {},
   "source": [
    "### Section 3: mIHC Analysis <a name=\"mihc\"></a>\n",
    "\n",
    "Stacked barplot\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6331dd2f-2899-464c-a3c6-15df58c03e5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_patient_all = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "                          index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2226fa59",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load sam's data\n",
    "df_patient = pd.read_csv(f'{codedir}/data/BCCLiverLung_V4_sampleavg_Density.csv')\n",
    "\n",
    "# look as carl's youden cutoff for pORG\n",
    "cutoff = 0.0249\n",
    "df_patient.loc[df_patient.loc[:,'pORG_0.2_Primary'] > cutoff,f'pORG>{cutoff}'] = 'high' # 'pORG_0.2_Primary','pORG_0.2_Met','pORG_0.2_allPrimary','pORG_0.2_allMet',\n",
    "df_patient.loc[df_patient.loc[:,'pORG_0.2_Primary'] <= cutoff,f'pORG>{cutoff}'] = 'low'\n",
    "df_patient.pORG.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e799a921-990b-4be3-a0f7-9de0afad0063",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_patient = df_patient.merge(df_patient_all.loc[:,['Days from Diagnosis to FU','Vital Status at FU','Public_Patient_ID']],on='Public_Patient_ID',how='left')\n",
    "ls_lung_short = ['ST-00007206',\n",
    " 'ST-00011168',\n",
    " 'ST-00017440',\n",
    " 'ST-00018232',\n",
    " 'ST-00020081',\n",
    " 'ST-00020149',\n",
    " 'ST-00020207',\n",
    " 'ST-00021880']\n",
    "#df_patient = df_patient[~df_patient.Public_Patient_ID.isin(ls_lung_short)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0616ff57-1916-4061-b5ae-5a652e6dda53",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_patient.loc[:,['pORG>0.0249','pORG_0.2_Primary']].groupby('pORG>0.0249').std()\n",
    "from scipy.stats import sem\n",
    "for s_col in ['pORG>0.0249','Cohort']:\n",
    "    for s_col_col in df_patient.loc[:,s_col].dropna().unique():\n",
    "        print(s_col_col)\n",
    "        print(sem(df_patient.loc[df_patient.loc[:,s_col]==s_col_col,'pORG_0.2_Primary']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "693d274e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#survival\n",
    "df_surv = pd.read_csv(f'{codedir}/data/u54_tma_sampleannot_Link.csv',index_col=0)\n",
    "df_surv['Sample_ID'] = [item.split('-T')[0][-6::] for item in df_surv.loc[:,'Public_Patient_ID'].fillna('none')]\n",
    "s_out = 'annotation/20231205_Patient_Metadata.csv'#'annotation/20230921_Patient_Metadata.csv'\n",
    "df_surv = pd.read_csv(s_out,index_col=0)\n",
    "df_surv = df_surv[df_surv.Public_Patient_ID.isin(df_patient.Public_Patient_ID)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15566846-bf18-41ef-95ce-8d6e307e6f38",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_density = pd.read_csv(f'{codedir}/pORGCohort_DBT_ROI_Density.csv')\n",
    "for s_col in ['pORG>0.0249','Cohort']:\n",
    "    df_density[s_col] = df_density.Sample_ID.map(dict(zip(df_patient.Sample_ID,df_patient.loc[:,s_col])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b450489-16aa-481e-835e-83afba966d36",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sns.palplot(color_palette)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "238dfffa-1550-47cd-a7be-3087c1715703",
   "metadata": {},
   "outputs": [],
   "source": [
    "color_palette = [(1.0, 0.5019607843137255, 0.054901960784313725), #orange\n",
    " (0.0, 0.4196078431372549, 0.6431372549019608), #blue\n",
    " (0.7843137254901961, 0.3215686274509804, 0.0),#orange\n",
    "  (0.34901960784313724, 0.34901960784313724, 0.34901960784313724),#gray\n",
    "  (0.6352941176470588, 0.7843137254901961, 0.9254901960784314),#blue\n",
    " (0.6705882352941176, 0.6705882352941176, 0.6705882352941176), #gray\n",
    "                 (1.0, 0.7372549019607844, 0.4745098039215686),#orange\n",
    " (0.37254901960784315, 0.6196078431372549, 0.8196078431372549),#blue(0.9372549019607843, 0.9529411764705882, 1.0),#blue\n",
    " (0.5372549019607843, 0.5372549019607843, 0.5372549019607843),#gray\n",
    " (0.8117647058823529, 0.8117647058823529, 0.8117647058823529)]#gray\n",
    "sns.palplot(color_palette)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73e18666-0196-461f-85e2-8bf07b09dc21",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sns.palplot(qual_palettes.Blues_4.mpl_colors)\n",
    "#qual_palettes.Blues_4.mpl_colors\n",
    "d_fig4 = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b1955bd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# stacked bar\n",
    "import palettable.colorbrewer.sequential as qual_palettes\n",
    "import palettable.tableau as tbl_palettes\n",
    "#color_palette =  tbl_palettes.ColorBlind_10.mpl_colors  #qual_palettes.Bold_9.mpl_colors#qual_palettes.Safe_10.mpl_colors \n",
    "ls_mihc = ['CD4 T helper cells', 'CD8 T cells', 'Macrophage', 'Granulocytes',\n",
    "       'Immature DC', 'Monocyte', 'B cells', 'T-regulatory CD4 cells',\n",
    "       'Mature DC' #'Immune Other', 'PanCK+',\n",
    "       #'aSMA+', 'Other Cells', \n",
    "          ]\n",
    "ls_mihc = ['Granulocytes','CD4 T helper cells', 'Macrophage', 'CD8 T cells', \n",
    "       'B cells','Monocyte','Immature DC',   'T-regulatory CD4 cells',\n",
    "       'Mature DC' #'Immune Other', 'PanCK+',\n",
    "       #'aSMA+', 'Other Cells', \n",
    "          ]\n",
    "d_color = dict(zip(ls_mihc,color_palette))\n",
    "for s_group in ['Cohort','pORG>0.0249']:\n",
    "    fig,ax=plt.subplots(figsize=(5,4.5),dpi=300)\n",
    "    df_mean = df_density.loc[:,ls_mihc+[s_group]].groupby(s_group).mean()\n",
    "    df_mean.plot(kind='bar',width=.9,stacked=True,ax=ax,color=d_color)\n",
    "    ax.legend(bbox_to_anchor=(1,.9))\n",
    "    ax.set_ylabel('Cell density (cells/mm$^2$)')\n",
    "    plt.tight_layout()\n",
    "    fig.savefig(f'figures/mIHC_{s_group}.pdf')\n",
    "    d_fig4.update({f'Figure4_C_mIHC_{s_group}':df_mean})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3441ff71-1975-4b65-a01d-e1800d987f20",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_density.loc[:,s_group].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3a24a06-024d-4100-815c-050a963f8918",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_density.loc[:,'Cohort'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08281495-af3a-4562-97b5-45abd298cd82",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# liver to lung\n",
    "sorter = [8049,18268,1127,232215,15475,8122,6951,19234, 7080,164354,1749,248041]\n",
    "df_patient.Sample_ID.astype('category')\n",
    "df_patient.Sample_ID = df_patient.Sample_ID.astype('category')\n",
    "df_patient.Sample_ID = df_patient.Sample_ID.cat.set_categories(sorter)\n",
    "df_patient = df_patient.sort_values(by='Sample_ID')\n",
    "ls_pat = ['ST-00016289',\n",
    " 'ST-00017310',\n",
    " 'ST-00018955',\n",
    " 'ST-00018269',\n",
    " 'ST-00017078',\n",
    " 'ST-00020181',\n",
    " 'ST-00019367',\n",
    " 'ST-00017381',\n",
    " 'ST-00019368',\n",
    " 'ST-00015839',\n",
    " 'ST-00017804',\n",
    " 'ST-00017440',]\n",
    "df_patient['Sample ID'] = ls_pat\n",
    "df_patient.loc[:,['Sample_ID','Sample ID','Cohort']]\n",
    "d_color_mihc = dict(zip(ls_pat,qual_palettes.Oranges_9.mpl_colors+qual_palettes.Blues_3.mpl_colors))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d673954-e443-48df-b844-8e2f1b4a3527",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sns.palplot(qual_palettes.Oranges_9.mpl_colors)\n",
    "# sns.palplot(qual_palettes.Blues_3.mpl_colors)\n",
    "mihc_pal = qual_palettes.Oranges_9.mpl_colors+qual_palettes.Blues_3.mpl_colors\n",
    "sns.palplot(mihc_pal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9a4418e-76da-4309-a403-4738c2304fc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_density.loc[:,ls_mihc].median().sort_values(ascending=False).index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89df688d-5681-471b-a886-2f23ee71fae0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# roi\n",
    "df_density['Sample ID'] = df_density.Sample_ID.map(dict(zip(df_patient['Sample_ID'],df_patient['Sample ID'])))\n",
    "mpl.rcParams.update({'font.size':6})\n",
    "from statannotations.Annotator import Annotator\n",
    "importlib.reload(util)\n",
    "d_hue = {'Cohort':['Lung','Liver',],'pORG>0.0249':['low','high',]}\n",
    "x = \"cell_type\"\n",
    "y = \"density\"\n",
    "d_order = {'pORG>0.0249':[#'CD8 T cells',\n",
    "         'CD4 T helper cells','Granulocytes', 'Macrophage',\n",
    "         'Immature DC', 'B cells','T-regulatory CD4 cells', #'Monocyte',\n",
    "          'Mature DC'],\n",
    "          'Cohort':['CD4 T helper cells',\n",
    "         'Immature DC', 'B cells','T-regulatory CD4 cells']}\n",
    "ls_mihc = ['Granulocytes','CD4 T helper cells', 'Macrophage', 'CD8 T cells', \n",
    "       'B cells','Monocyte','Immature DC',   'T-regulatory CD4 cells',\n",
    "       'Mature DC' #'Immune Other', 'PanCK+',\n",
    "       #'aSMA+', 'Other Cells', \n",
    "          ]\n",
    "pal_porg_r = ('#56B4E9','#E69F00',)\n",
    "pal_liv_r = ('#0072B2','#D55E00',)\n",
    "%matplotlib inline\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    for hue in ['Cohort','pORG>0.0249']:\n",
    "        order = ls_mihc#d_order[hue]\n",
    "        sns.set_palette(['white','lightgray'])\n",
    "        for s_region in ['All']:#,'Tumor','Border','Distal'\n",
    "            if s_region == 'All':\n",
    "                df_mean = df_density.loc[:,ls_mihc+['Sample ID']]#.groupby(['Sample_ID']).mean()\n",
    "            else:\n",
    "                continue\n",
    "                #df_mean = df_density.loc[df_density.Location==s_region[0],ls_mihc+['Sample_ID']].groupby(['Sample_ID']).mean()\n",
    "            df_mean = df_mean.reset_index().set_index('Sample ID')\n",
    "            df_long = df_mean.stack().reset_index().rename({'level_1':'cell_type',0:'density'},axis=1)\n",
    "            df_long[hue] = df_long.loc[:,'Sample ID'].map(dict(zip(df_patient.loc[:,'Sample ID'],df_patient.loc[:,hue])))#.map(d_color_mihc)#\n",
    "            #plot\n",
    "            figsize = (5,4.5)\n",
    "            hue_order = d_hue[hue]\n",
    "            fig, ax = util.annotated_stripplot_hue(df_long,x,y,hue,order,hue_order,figsize,b_correct=True,s=1,b_box=True)\n",
    "            ls_result = []\n",
    "            for sex, color in zip(ls_pat, mihc_pal):\n",
    "                df_per_sex = df_long[df_long['Sample ID'] == sex]\n",
    "                sns.stripplot(data=df_per_sex, x=x, y=y, hue=hue,\n",
    "                              order=order, hue_order=hue_order,\n",
    "                              dodge=True, palette=[color],\n",
    "                              marker='o', size=1.5, linewidth=0,ax=ax)\n",
    "                ls_result.append([color] * 2)\n",
    "            ax.set_xticklabels(ax.get_xticklabels(),rotation=90,fontsize=10)\n",
    "            ax.set_yticklabels(ax.get_yticklabels(),fontsize=8)\n",
    "            ax.set_ylabel('Cell density (cells/mm$^2$)',fontsize=10)\n",
    "            ax.set_xlabel('')\n",
    "            s_region = f\"{hue_order[0]}: N={sum(df_density.loc[:,hue]==hue_order[0])} ROIs vs. {hue_order[1]}: N={sum(df_density.loc[:,hue]==hue_order[1])} ROIs\"\n",
    "            ax.set_title(f'{s_region}',fontsize=11)\n",
    "            handleso, labels = ax.get_legend_handles_labels()\n",
    "            handles = [handleso[2], handleso[3], handleso[0]] + handleso[4::2]\n",
    "            labels = hue_order + ['Sample ID'] + ls_pat\n",
    "            sns.set_palette(['white','lightgray','white']+mihc_pal)\n",
    "            ax.legend(handles=handles,labels=labels,bbox_to_anchor=(1,1),loc='upper left',markerscale=0.5,\n",
    "                      fontsize=8,title_fontsize=8,title=hue,frameon=False)\n",
    "            ax.spines[['right','top']].set_visible(False)\n",
    "            plt.tight_layout()\n",
    "            plt.savefig(f'figures/mIHC_per_roi_by_{hue}_density_in_{s_region}.pdf', dpi=300, bbox_inches='tight')\n",
    "            d_fig4.update({f'Figure4_D_mIHC_{hue}':df_long})\n",
    "#break\n",
    "mpl.rcParams.update({'font.size': 10})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f411d905-2758-4265-8aa5-af44d4245196",
   "metadata": {},
   "outputs": [],
   "source": [
    "# d_fig4.keys()\n",
    "# util.open_write_excel(d_fig4,filename='Source_Data_Figure4.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "273d75da-c442-4502-9c7e-92e28965d269",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_density.groupby(hue).mean(numeric_only=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27469278-0514-45dd-807f-84b7152b75af",
   "metadata": {},
   "outputs": [],
   "source": [
    "#high low pORG\n",
    "pvals = np.array([0.32718848,4.37E-06,0.000201767,0.002671442,0.005402357,\n",
    "0.277169408,0.038489462,0.022766419,6.05E-07])\n",
    "reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvals,method='fdr_bh')\n",
    "[print(f'{item:.3}')  for item in corrected]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "988d6ccd-e733-47f1-aa4d-d98c992a7fe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "pvals = np.array([0.156317606,8.49E-14,2.37E-09,0.000461868,\n",
    "0.402545821,0.039025283,0.374874377,0.983725258,0.000113532,])\n",
    "reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvals,method='fdr_bh')\n",
    "[print(f'{item:.3}')  for item in corrected]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2177cef1-7aad-4a2e-9b69-33a94bdcaff0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#patient level\n",
    "importlib.reload(util)\n",
    "d_hue = {'Cohort':['Lung','Liver',],'pORG>0.0249':['low','high',]}\n",
    "x = \"cell_type\"\n",
    "y = \"density\"\n",
    "d_order = {'Cohort':['CD4 T helper cells', #'CD8 T cells', 'Granulocytes', 'Macrophage',\n",
    "       'Immature DC', 'B cells',  'T-regulatory CD4 cells'],#'Monocyte',       'Mature DC'],\n",
    "           'pORG>0.0249':['CD4 T helper cells',  'Granulocytes', 'Macrophage',#'CD8 T cells',\n",
    "       'Immature DC', 'B cells',  'T-regulatory CD4 cells','Mature DC']}#'Monocyte',       ],\n",
    "pal_porg_r = ('#56B4E9','#E69F00',)\n",
    "pal_liv_r = ('#0072B2','#D55E00',)\n",
    "df_out = pd.DataFrame()\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    for hue in ['Cohort','pORG>0.0249']:\n",
    "        order = d_order[hue]\n",
    "        if hue == 'Cohort':\n",
    "            figsize = (5,4)\n",
    "            sns.set_palette(pal_liv_r)\n",
    "        else:\n",
    "            figsize = (7,4)\n",
    "            sns.set_palette(pal_porg_r)\n",
    "        for s_region in ['All','Border','Tumor','Distal']:#\n",
    "            if s_region == 'All':\n",
    "                df_mean = df_density.loc[:,ls_mihc+['Sample_ID']].groupby(['Sample_ID']).mean()\n",
    "            else:\n",
    "                df_mean = df_density.loc[df_density.Location==s_region[0],ls_mihc+['Sample_ID']].groupby(['Sample_ID']).mean()\n",
    "            df_mean = df_mean.reset_index().set_index('Sample_ID')\n",
    "            df_long = df_mean.stack().reset_index().rename({'level_1':'cell_type',0:'density'},axis=1)\n",
    "            df_long[hue] = df_long.Sample_ID.map(dict(zip(df_patient.Sample_ID,df_patient.loc[:,hue])))\n",
    "            #plot\n",
    "            hue_order = d_hue[hue]\n",
    "            df_out = pd.concat([df_out,df_long])\n",
    "            fig, ax = util.annotated_stripplot_hue(df_long,x,y,hue,order,hue_order,figsize,b_correct=True,loc='upper right',s=5)\n",
    "            ax.set_xticklabels(ax.get_xticklabels(),rotation=45)\n",
    "            ax.set_ylabel('Cell density (cells/mm$^2$)')\n",
    "            ax.set_xlabel('')\n",
    "            ax.set_title(f'{s_region} ROIs: Mean per Patient')\n",
    "            plt.tight_layout()\n",
    "            plt.savefig(f'figures/mIHC_per_patient_by_{hue}_density_in_{s_region}.pdf', dpi=300, bbox_inches='tight')\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc9e5b09-a151-4d28-aeba-c9f376a5ebaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #df_out.to_csv(f'Source_Extended_Data_Figure6_I.csv')\n",
    "# s_find = 'Source_Extended_Data_Figure6'\n",
    "# d_fig = excel_dict_from_csvs(s_find,s_prefix='ED_Fig6')\n",
    "# util.open_write_excel(d_fig,filename='Source_Extended_Data_Figure6.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbc40703",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# # boxplot starthere\n",
    "\n",
    "# df = df_patient.loc[:,ls_mihc].unstack().reset_index().rename({'level_0':'cell type',0:'Density'},axis=1)\n",
    "# df['pORG'] = df.level_1.map(dict(zip(df_patient.index,df_patient.pORG)))\n",
    "# df['Cohort'] = df.level_1.map(dict(zip(df_patient.index,df_patient.Cohort)))\n",
    "# for s_group in ['Cohort','pORG']:\n",
    "#     if s_group == 'Cohort':\n",
    "#         pal_liv = ('#D55E00','#0072B2',)\n",
    "#         sns.set_palette(pal_liv)\n",
    "#     else:\n",
    "#         pal_porg = ('#E69F00','#56B4E9',)\n",
    "#         sns.set_palette(pal_porg)\n",
    "#     fig,ax=plt.subplots(figsize=(8,4),dpi=200)\n",
    "#     #sns.violinplot(data=df,x='cell type',y='Density',hue=s_group,cut=0,ax=ax)\n",
    "#     sns.stripplot(data=df,x='cell type',y='Density',hue=s_group,dodge=True,ax=ax)\n",
    "#     sns.boxplot(data=df,x='cell type',y='Density',hue=s_group,ax=ax,showmeans=True,\n",
    "#                 medianprops={'visible': False},whiskerprops={'visible': False},\n",
    "#                 meanline=True,showcaps=False,\n",
    "#                 meanprops={'color': 'k', 'ls': '-', 'lw': 2},\n",
    "#                 showfliers=False,showbox=False)#\n",
    "#     # Rotating X-axis labels\n",
    "#     ax.set_xticklabels(ax.get_xticklabels(), rotation = 50)\n",
    "#     ax.set_xlabel('')\n",
    "#     h, l = ax.get_legend_handles_labels()\n",
    "#     ax.legend(h[0:2],l[0:2],loc='upper right',title=s_group)\n",
    "#     ls_group = sorted(df.loc[:,s_group].unique())\n",
    "#     pairs = [((item,ls_group[1]),(item,ls_group[0])) for item in ls_mihc]\n",
    "#     annot = Annotator(ax, pairs, data=df,x='cell type',y='Density',hue=s_group,\n",
    "#                       order=ls_mihc,hue_order=ls_group)\n",
    "#     annot.configure(test='t-test_ind', text_format='star',fontsize=7) #,comparisons_correction='fdr_bh'\n",
    "#     res = annot.apply_test()\n",
    "#     d_pval = dict([(res.data.group1[0],res.data.pvalue) for res in annot.annotations])\n",
    "#     pvalues = [d_pval[item] for item in ls_mihc]\n",
    "#     reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='fdr_bh')\n",
    "#     formatted_pvalues = [f'p={pvalue:.2}' for pvalue in list(pvalues)]\n",
    "#     annot.set_custom_annotations(formatted_pvalues)\n",
    "#     annot.annotate()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7ce65c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # dowload mIHC data from https://www.synapse.org/#!Synapse:syn51078766\n",
    "# # combine ROIs into large dataframe and save\n",
    "# if not os.path.exists(f'data/20221123_mIHC_LiverLung_Celltypes.csv'):\n",
    "#     ls_col = ['Sample_ID','class','Location_Center_X', 'Location_Center_Y', \n",
    "#            'GRZB_func', 'KI67_func', 'PD1_func', 'PDL1_func', 'CD163_func',\n",
    "#            'CCR2_func', 'HLAII_func', 'EOMES_func','Area']\n",
    "#     df_ll =pd.DataFrame()\n",
    "#     for s_file in sorted(os.listdir('mIHC_Data')):\n",
    "#         df = pd.read_csv(f'mIHC_Data/{s_file}',index_col=0)\n",
    "#         s_sample = s_file.split('LiverLungBCC')[1].split('.')[0]\n",
    "#         df['Sample_ID'] = s_sample\n",
    "#         df.index = [f'{s_sample.split(\"Nuclei_\")[1].replace(\"ROI\",\"_scene\")}_cell{item}' for item in df.index]\n",
    "#         df_ll = pd.concat([df_ll,df.loc[:,ls_col]])\n",
    "#         break\n",
    "#     df_ll['Organ'] = df_ll.Sample_ID.map(dict(zip(df_ll_annot.index,df_ll_annot.loc[:,'Met Site'])))\n",
    "#     df_ll['Location'] = df_ll.Sample_ID.map(dict(zip(df_ll_annot.index,df_ll_annot.loc[:,'Location'])))\n",
    "#     df_ll['Desc'] = df_ll.Sample_ID.map(dict(zip(df_ll_annot.index,df_ll_annot.loc[:,'Desc'])))\n",
    "#     df_ll.to_csv(f'data/20221123_mIHC_LiverLung_Celltypes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a6cbcba",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#load data\n",
    "df_ll = pd.read_csv(f'data/20221123_mIHC_LiverLung_Celltypes.csv',index_col=0,low_memory=False)\n",
    "print(len(df_ll))\n",
    "\n",
    "df_ll_ann = pd.read_csv('data/LiverLung_annotations.csv',index_col=0)\n",
    "df_ll_roi = pd.read_csv('data/annotated_LiverLung_perROI.csv',index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db35cd8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add annotation\n",
    "df_ll['Sample_ID_short'] = [item.split('Nuclei_')[1].split('ROI')[0] for item in df_ll.Sample_ID]\n",
    "df_ll['Sample_ID_int'] = [int(item) for item in df_ll.Sample_ID_short]\n",
    "df_ll['ROI'] = [int(item.split('ROI')[1]) for item in df_ll.Sample_ID]\n",
    "df_ll['Cohort'] = df_ll.Sample_ID_int.map(dict(zip(df_ll_ann.index, df_ll_ann.Cohort)))\n",
    "df_ll['Patient'] = df_ll.Sample_ID_int.map(dict(zip(df_ll_ann.index, df_ll_ann.Patient)))\n",
    "df_ll['Sample_ROI'] = df_ll.Sample_ID_int.astype('str') + '_' + df_ll.ROI.astype('str')\n",
    "df_ll_roi['Sample_ROI'] = df_ll_roi.index.astype('str') + '_' + df_ll_roi.ROI.astype('str')\n",
    "df_ll['Location'] = df_ll.Sample_ROI.map(dict(zip(df_ll_roi.Sample_ROI,df_ll_roi.loc[:,'Location'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9a04a6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ll['classII'] = df_ll.loc[:,'class'].replace({'T-regulatory CD4 cells':'T cells',\n",
    "                                                 'CD4 T helper cells':'T cells','CD8 T cells':'T cells'}) #\n",
    "df_ll.rename({'class':'classI'},axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ab5c426",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_color = dict(zip(['ST-00016289', 'ST-00017078', 'ST-00017310', 'ST-00017381',\n",
    "       'ST-00018269', 'ST-00018955', 'ST-00019367', 'ST-00019368',\n",
    "       'ST-00020181'],sns.color_palette('Purples',9)))\n",
    "d_color.update(dict(zip(['ST-00015839', 'ST-00017440', 'ST-00017804'],sns.color_palette('Blues',3))))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f45031c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# # boxplot figure friendly\n",
    "# # look as carl's youden cutoff for pORG\n",
    "# s_column = 'classI'\n",
    "# if s_column == 'classI':\n",
    "#     cutoff = 0.0249\n",
    "#     df_group.loc[df_group.loc[:,'pORG_Primary'] > cutoff,f'pORG'] = 'high' # 'pORG_0.2_Primary','pORG_0.2_Met','pORG_0.2_allPrimary','pORG_0.2_allMet',\n",
    "#     df_group.loc[df_group.loc[:,'pORG_Primary'] <= cutoff,f'pORG'] = 'low'\n",
    "#     print(df_group.pORG.value_counts())\n",
    "#     ls_mihc = [ #'PanCK+','aSMA+',\n",
    "#                'CD4 T helper cells','CD8 T cells','B cells', 'T-regulatory CD4 cells',\n",
    "#         'Granulocytes', 'Monocyte',\n",
    "#      'Macrophage', 'Mature DC','Immature DC']\n",
    "#     df = df_group.loc[:,ls_mihc].unstack().reset_index().rename({'level_0':'cell type',0:'Fraction'},axis=1)\n",
    "#     df['pORG'] = df.level_1.map(dict(zip(df_group.index,df_group.pORG)))\n",
    "#     df['Cohort'] = df.level_1.map(dict(zip(df_group.index,df_group.Cohort)))\n",
    "#     for s_group in ['Cohort','pORG']:\n",
    "#         if s_group == 'Cohort':\n",
    "#             pal_liv = ('#0072B2','#D55E00')\n",
    "#             sns.set_palette(pal_liv)\n",
    "#         else:\n",
    "#             pal_porg = ('#56B4E9','#E69F00',)\n",
    "#             sns.set_palette(pal_porg)\n",
    "#         fig,ax=plt.subplots(figsize=(8,4),dpi=200)\n",
    "#         #sns.violinplot(data=df,x='cell type',y='Density',hue=s_group,cut=0,ax=ax)\n",
    "#         sns.stripplot(data=df,x='cell type',y='Fraction',hue=s_group,dodge=True,ax=ax)\n",
    "#         sns.boxplot(data=df,x='cell type',y='Fraction',hue=s_group,ax=ax,showmeans=True,\n",
    "#                     medianprops={'visible': False},whiskerprops={'visible': False},\n",
    "#                     meanline=True,showcaps=False,\n",
    "#                     meanprops={'color': 'k', 'ls': '-', 'lw': 2},\n",
    "#                     showfliers=False,showbox=False)#\n",
    "#         # Rotating X-axis labels\n",
    "#         ax.set_xticklabels(ax.get_xticklabels(), rotation = 50)\n",
    "#         ax.set_xlabel('')\n",
    "#         h, l = ax.get_legend_handles_labels()\n",
    "#         ax.legend(h[0:2],l[0:2],loc='upper right',title=s_group)\n",
    "#         ls_group = sorted(df.loc[:,s_group].unique())\n",
    "#         pairs = [((item,ls_group[1]),(item,ls_group[0])) for item in ls_mihc]\n",
    "#         annot = Annotator(ax, pairs, data=df,x='cell type',y='Fraction',hue=s_group,\n",
    "#                           order=ls_mihc,hue_order=ls_group)\n",
    "#         annot.configure(test='t-test_ind', text_format='star',fontsize=7) #,comparisons_correction='fdr_bh'\n",
    "#         res = annot.apply_test()\n",
    "#         d_pval = dict([(res.data.group1[0],res.data.pvalue) for res in annot.annotations])\n",
    "#         pvalues = [d_pval[item] for item in ls_mihc]\n",
    "#         reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='fdr_bh')\n",
    "#         formatted_pvalues = [f'p={pvalue:.2}' for pvalue in list(pvalues)]\n",
    "#         annot.set_custom_annotations(formatted_pvalues)\n",
    "#         annot.annotate()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a65fb8b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#by location\n",
    "%matplotlib inline\n",
    "s_define = 'pORG_Primary'#'trim_padj_0.2_pORG_Up_55_Genes'#'txi_pORG_Up_42_Genes'#''#\n",
    "s_group = 'Cohort'\n",
    "alpha = 0.05\n",
    "s_column = 'classI'#'classII'#\n",
    "if s_column == 'classI':\n",
    "    ls_mihc = [ 'T-regulatory CD4 cells', 'CD4 T helper cells','PanCK+','aSMA+','CD8 T cells','B cells', 'Granulocytes', 'Monocyte',\n",
    " 'Macrophage', 'Mature DC','Immature DC','Immune Other','Other Cells' ]\n",
    "elif s_column == 'classII':\n",
    "    ls_mihc = ['T cells', #'B cells', 'Granulocytes', 'Monocyte', 'Macrophage',\n",
    "       #'Mature DC', 'Immature DC', 'Immune Other', 'PanCK+', 'aSMA+','Other Cells'\n",
    "       ]\n",
    "s_patient = 'Public_Patient_ID'\n",
    "\n",
    "for s_loc in ['all','T','B','D']: \n",
    "    if s_loc == 'all':\n",
    "        df_loc = df_ll\n",
    "    else:\n",
    "        df_loc = df_ll[df_ll.Location==s_loc]\n",
    "    print(len(df_loc.Sample_ID_short.unique()))\n",
    "    df_group = (df_loc.groupby(['Patient',s_column]).count().Sample_ID/(df_loc.groupby(['Patient']).count().Sample_ID)).unstack()\n",
    "    df_group[s_patient] = df_group.index\n",
    "    df_group['Cohort'] = df_group.index.map(dict(zip(df_ll_roi.Patient,df_ll_roi.loc[:,'Cohort'])))\n",
    "    df_group = df_group.merge(df_surv.loc[:,['Survival','Days from Diagnosis to FU',s_define,s_patient]],on=s_patient)\n",
    "    df_group = df_group[~df_group.loc[:,s_patient].duplicated()]\n",
    "    df_group = df_group[df_group.Cohort!='Liver met'].fillna(0)\n",
    "    ls_order = sorted(df_group.loc[:,s_group].unique())\n",
    "    df_group.iloc[:,0:-4].to_csv(f'../data/mIHC_cell_freq_{s_loc}.csv')\n",
    "    for s_marker in ls_mihc:\n",
    "        try:\n",
    "            s_high = df_group.loc[:,s_group].unique()[0]\n",
    "            s_low = df_group.loc[:,s_group].unique()[1]\n",
    "        except:\n",
    "            continue\n",
    "        n_high = sum(df_group.loc[:,s_group]==s_high)\n",
    "        n_low = sum(df_group.loc[:,s_group]==s_low)\n",
    "        statistic,pvalue = stats.ttest_ind(df_group.loc[df_group.loc[:,s_group]==s_high,s_marker],\n",
    "                                           df_group.loc[df_group.loc[:,s_group]==s_low,s_marker])\n",
    "        if pvalue <= alpha:\n",
    "            df_group_roi = (df_loc.groupby(['Sample_ROI',s_column]).count().Sample_ID/(df_loc.groupby(['Sample_ROI']).count().Sample_ID)).unstack()\n",
    "            df_group_roi['Cohort'] = df_group_roi.index.map(dict(zip(df_ll_roi.Sample_ROI,df_ll_roi.loc[:,'Cohort'])))\n",
    "            df_group_roi['Patient'] = df_group_roi.index.map(dict(zip(df_ll_roi.Sample_ROI,df_ll_roi.loc[:,'Patient'])))\n",
    "            df_group_roi = df_group_roi[df_group_roi.Cohort.isin(['liver_cohort','lung_cohort'])]\n",
    "            fig, ax = plt.subplots(figsize=(3,3),dpi=300)\n",
    "            sns.boxplot(data=df_group_roi,x=s_group,y=s_marker,showfliers=False,ax=ax,order=[str(item) for item in ls_order],palette=['mediumpurple','deepskyblue'])\n",
    "            sns.stripplot(data=df_group_roi,x=s_group,y=s_marker,ax=ax,hue='Patient',s=3,palette=d_color)\n",
    "            ax.set_ylim(ax.get_ylim()[0],ax.get_ylim()[1])\n",
    "            ax.set_title(f'{s_group} versus\\n {s_marker}\\n p={pvalue:.4f} (n={n_low}, {n_high})')\n",
    "            ax.set_ylabel(f'{s_marker} in {s_loc}')\n",
    "            #ax.set_ylabel(f'{s_marker}')\n",
    "            ax.get_legend().remove()\n",
    "            plt.tight_layout()\n",
    "            break\n",
    "            #fig.savefig(f'{s_date}/boxplot_mIHC_{s_marker}_versus_{s_group}_in_{s_loc}.png')\n",
    "        break\n",
    "    #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e71ca46a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "print(df_group.groupby('Cohort').mean(numeric_only=True).loc[:,s_define])\n",
    "print(df_group.groupby('Cohort').sem(numeric_only=True).loc[:,s_define])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44bc87c1-c25b-43bb-ba63-dd9f06f883c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_palette('tab20')\n",
    "weight_counts = df_group.mean(numeric_only=True).loc[ls_mihc[::-1]]\n",
    "fig, ax = plt.subplots(figsize=(3,4))\n",
    "bottom=0\n",
    "for boolean, weight_count in weight_counts.items():\n",
    "    p = ax.bar(1, weight_count, width=.9, label=boolean, bottom=bottom)\n",
    "    bottom += weight_count\n",
    "ax.legend(bbox_to_anchor=(1,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f01e0a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fig,ax=plt.subplots(dpi=300)\n",
    "# df_group.groupby('Cohort').mean().loc[:,ls_mihc].plot(kind='bar',width=.9,stacked=True,ax=ax,colormap='Paired')\n",
    "# ax.legend(bbox_to_anchor=(1,.9))\n",
    "# ax.set_ylabel('Fraction in Tissue')\n",
    "# plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d999eea3",
   "metadata": {},
   "source": [
    "# Section 4 <a name=\"meta\"></a>\n",
    "\n",
    "old code to create patient metadata tidy data\n",
    "\n",
    "saved as supplemental dataset 1\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f7649de",
   "metadata": {},
   "source": [
    "### Patients in LabKey"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63f160b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # load patient vital status\n",
    "# #also had stage/ grade/ lymph nodes etc.\n",
    "# df_vital = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/FMP_Patients_Nov17_2021.xlsx',sheet_name='Sheet1')\n",
    "\n",
    "# #Stage 1 pancreatic cancer means the cancer is not more than 4cm in size and it hasn't spread outside the pancreas. It is split into 1A and 1B.\n",
    "# # 1A In TNM staging, this is the same as T1, N0, M0 1B In TNM staging, this is the same as T2, N0, M0\n",
    "\n",
    "# # Stage 2A means the cancer is bigger than 4cm but is still within the pancreas. It has not spread to the lymph nodes Open a glossary item or other areas of the body.\n",
    "# #2A in TNM staging, this is the same as T3, N0, M0. 2B In TNM staging, this is the same as T1, 2 or 3, N1, M0.\n",
    "\n",
    "# #3 Stage 3 can mean that the cancer is any size within the pancreas and has spread to 4 or more nearby lymph nodes Open a glossary item.\n",
    "# #In TNM staging, this is the same as T1, 2 or 3, N2, M0.\n",
    "# #Or stage 3 can mean the cancer has started to grow outside the pancreas into the major blood vessels nearby. It may or may not have spread into the lymph nodes. It hasn't spread to any other areas of the body.\n",
    "# #In TNM staging, this is the same as T4, Any N, M0.\n",
    "\n",
    "# #4 Your doctor might call this advanced (metastatic) cancer.\n",
    "# #In TNM staging, this is the same as Any T, Any N, M1.\n",
    "\n",
    "# #collapse stage\n",
    "# d_stage = {'2B - IIB':'II', 'p2A':'II', 'p2B':'II', '4 - IV':'IV', '2A - IIA':'II', 'p3':'III', '1B - IB':'I',\n",
    "#     'c1B':'I', 'c2B':'I', 'nan':pd.NA, 'p4':'IV','2B - T1-3, N1, M0':'II','p0':'0','p1B':'I','c1':'I',\n",
    "#     'c4':'IV','c3':'III','p1A':'I','pNA':pd.NA,'3 - T4, Any N, M0':'III','c2A':'II','p4B':'IV','c4B':'IV',\n",
    "#      '2B - T1, N1, M0 / T2, N1, M0 / T3, N1, M0':'II','pUNK':pd.NA,'p2':'II','p3B':'III','c2':'II','p3A':'III',\n",
    "#     '3 - III':'III','99 - Unknown':pd.NA,'1A - IA':'I','c3A':'III','c4A':'IV','p1':'I','c1A':'I','p4A':'IV',\n",
    "#     '88 - Not applicable to 7th Edition staging':pd.NA,\n",
    "#            '88 - No classification is recommended in 6th Edition':pd.NA,\n",
    "#     '2A - T3, N0, M0':'II','4 - Any T, Any N, M1':'IV'}\n",
    "# df_vital['Stage'] = pd.NA\n",
    "# df_vital['Stage'] = df_vital.loc[:,'Stage Grouping _ Dominant'].replace(d_stage) \n",
    "\n",
    "# #collapse grade\n",
    "\n",
    "# d_grade = {'Grade II  Moderately Diff / Mod Well Diff':'2',\n",
    "#        'Grade I   Well Differentiated/Differentiated':'1',\n",
    "#        'Cell type not determined; not stated;N/A;Unk; high grade dysplas':pd.NA,\n",
    "#        'Grade III Poorly Differentiated':'3', 'nan':pd.NA,'Grade IV Undifferentiated, Anaplastic':'4',\n",
    "#          'B-CELL    LYMPHOMA OR LEUKEMIA ONLY':pd.NA}\n",
    "# df_vital['Grade'] = pd.NA\n",
    "# df_vital['Grade'] = df_vital.loc[:,'Grade_Differentiation'].replace(d_grade)\n",
    "\n",
    "# #collapse LV invasion\n",
    "# d_replace = {'nan':pd.NA,np.nan:pd.NA, 'LYMPHOVASCULAR INVASION STATED AS NOT PRESENT':'NO',\n",
    "#        'LYMPHOVASCULAR INVASION PRESENT/IDENTIFIED':'YES',\n",
    "#        'Unknown/Indeterminate':pd.NA, 'NOT APPLICABLE':pd.NA,\n",
    "#        'Lymph-vascular Invasion Present/Identified':'YES',\n",
    "#        'LYMPHATIC AND SMALL VESSEL INVASION ONLY (L)':'YES',\n",
    "#        'BOTH LYMPHATIC AND SMALL VESSEL AND VENOUS (LARGE VESSEL) INVASION':'YES'}\n",
    "# df_vital['LV_Invasion'] = pd.NA\n",
    "# df_vital['LV_Invasion'] = df_vital.loc[:,'Lymph_vascular Invasion'].replace(d_replace)\n",
    "\n",
    "# # LN positivity\n",
    "# df_vital['LN_Pos'] = pd.NA\n",
    "# df_vital.loc[df_vital.loc[:,'Regional Lymph Nodes Positive'] >= 1,'LN_Pos'] = True\n",
    "# df_vital.loc[df_vital.loc[:,'Regional Lymph Nodes Positive'] < 1,'LN_Pos'] = False\n",
    "# df_vital.loc[df_vital.loc[:,'Regional Lymph Nodes Positive'].isna(),'LN_Pos'] = pd.NA\n",
    "# df_vital['LN_Pos']  = df_vital.LN_Pos.replace({True:'YES',False:'NO'})\n",
    "\n",
    "# df_vital['Survival'] = df_vital.cVitalStatus.replace({'Alive':0,'Dead':1})\n",
    "# print(df_vital.Grade.unique())\n",
    "# print(df_vital.Stage.unique())\n",
    "# print(df_vital.LN_Pos.unique())\n",
    "# print(df_vital.LV_Invasion.unique())\n",
    "# print(len(df_vital))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fb38c71",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# # kaplan meier\n",
    "# # %matplotlib inline\n",
    "# # # check prognostic value of clinicopathologiocal variables\n",
    "\n",
    "# # ls_vital = ['Stage', 'Grade','LV_Invasion','LN_Pos'] \n",
    "# # s_time = 'cDays from Diagnosis to FU'\n",
    "# # s_censor = 'Survival'\n",
    "# # for s_vital in ls_vital:\n",
    "# #     print(s_vital)\n",
    "# #     df = df_vital.loc[df_vital.loc[:,'Primary Site _ Major Groups For Staging']=='Pancreas',[s_vital,s_time,s_censor]].dropna(how='any')\n",
    "# #     fig, __ = util.km_plot(df,s_vital,s_time,s_censor)\n",
    "# #     fig.savefig(f'figures/KM_clinicopath_{s_vital}.png')\n",
    "# #     #break\n",
    "\n",
    "# #CPH\n",
    "# s_time = 'cDays from Diagnosis to FU'\n",
    "# #df_vital['Survival_time'] = df_vital.loc[:,s_time]\n",
    "# s_censor = 'Survival'\n",
    "# ls_vital = ['LV_Invasion','Stage', 'Grade','LN_Pos'] #,'Age'\n",
    "# for s_vital in ls_vital:\n",
    "#     print(s_vital)\n",
    "#     df = df_vital.loc[df_vital.loc[:,'Primary Site _ Major Groups For Staging']=='Pancreas',[s_vital,s_time,s_censor]].dropna(how='any')\n",
    "#     if df.columns.isin(['Stage']).any():\n",
    "#         df.Stage = df.Stage.replace({'I':1,'II':2,'III':3,'IV':4}).astype('int')\n",
    "#     if df.columns.isin(['Grade']).any():\n",
    "#         df.Grade = df.Grade.astype('int')\n",
    "#     if df.loc[:,s_vital].dtype=='O':\n",
    "#         df_dummy = pd.get_dummies(df.loc[:,[s_vital]],drop_first=True)\n",
    "#         df.drop(s_vital,axis=1,inplace=True)\n",
    "#         s_vital = df_dummy.columns[0]\n",
    "#         df[s_vital] = df_dummy\n",
    "#     fig, cph = util.cph_plot(df,s_vital,s_time,s_censor,figsize=(3,1.5))\n",
    "#     plt.tight_layout()\n",
    "#     fig.savefig(f'figures/CPH_single_{s_vital}_all.png')\n",
    "#     plt.close(fig)\n",
    "#     break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "837d3f5f",
   "metadata": {},
   "source": [
    "### Patients in Paper \n",
    "\n",
    "n=434 specimens\n",
    "\n",
    "n=422 patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7137cf94-b18c-4f36-8fc9-a5cf55df2d20",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temus300\n",
    "# 1 was a data duplicate from tempus49M91\n",
    "# 3 were duplicates sent in.\n",
    "# 7 were intestinal ampulla of vater\n",
    "# 184 - 44\n",
    "#386 + 13 + 23"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a874c60c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load full patient data, neoadjuvant\n",
    "df_patient = pd.read_excel(f'data/SupplementalDataset1.xlsx',\n",
    "                           sheet_name='Patients - Tab 1')\n",
    "df_patient.rename({'Patient ID':'Public_Patient_ID'},axis=1,inplace=True)\n",
    "df_patient['Cohort'] = pd.NA\n",
    "df_patient.loc[df_patient.loc[:,'Lung Met Present']=='YES','Cohort'] = 'Lung'\n",
    "df_patient.loc[df_patient.loc[:,'Liver Met Present']=='YES','Cohort'] = 'Liver'\n",
    "#how many in cohorts\n",
    "for s_site in df_patient.Cohort.dropna().unique():\n",
    "    n_patients = df_patient[df_patient.Cohort==s_site].Public_Patient_ID.nunique()\n",
    "    print(f'{s_site} {n_patients}')\n",
    "print(f'number rows {len(df_patient)}')\n",
    "print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')\n",
    "\n",
    "#define neodjuvant binary\n",
    "df_patient.loc[:,'Neoadjuvant Treatment'].unique()\n",
    "df_patient.loc[df_patient.loc[:,'Neoadjuvant Treatment'] == 'Yes Neoadjuvant','Neoadjuvant'] = 'Yes'\n",
    "df_patient['Neoadjuvant'] = df_patient.Neoadjuvant.fillna('No').replace('nan','No')\n",
    "\n",
    "#load T cell data\n",
    "df_tcell_tumor = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset6.xlsx',\n",
    "                         sheet_name='Tumor Samples')\n",
    "\n",
    "df_tcell_blood = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset6.xlsx',\n",
    "                         sheet_name='Blood Samples')\n",
    "\n",
    "#merge T cell blood and tumor\n",
    "df_tcell = df_tcell_tumor.merge(df_tcell_blood,on='Patient ID',suffixes=(' tumor',' blood'),how='outer')\n",
    "\n",
    "df_tcell.rename({'Productive Rearrangements (Observed Richness)':'Productive_Rearrangements',\n",
    "                'Templates per ng':'Templates_per_ng','Patient ID':'Public_Patient_ID'},axis=1,inplace=True)\n",
    "print(f'Add TCR patients {len(df_tcell)}')\n",
    "# #merge - no, add TCR later\n",
    "# df_patient = df_patient.merge(df_tcell,on='Public_Patient_ID',how='left',suffixes=('','_x'))\n",
    "# df_patient['INDEX'] = df_patient.index + 1\n",
    "print(f'number rows {len(df_patient)}')\n",
    "print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')\n",
    "#all 422 have OS and vital status\n",
    "print(f\"Missing OS: {df_patient.loc[:,'Days from Diagnosis to FU'].isna().sum()}\")\n",
    "print(f\"Missing Vital Status: {df_patient.loc[:,'Vital Status at FU'].isna().sum()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b957ffd-b13d-4bb6-87e6-59fa289ae338",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #61?\n",
    "# set([item.split('-')[2] for item in df_gsva.loc[:,'Patient Specimen ID']])\n",
    "# df_purist.loc['ST-00014524']\n",
    "# df_gsva[df_gsva.Public_Patient_ID=='ST-00014524']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51098a99",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load Purist\n",
    "d_gsva = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset1.xlsx', sheet_name=None)\n",
    "df_gsva = d_gsva['Specimen Subtype - Tab 2'].loc[:,['PurIST Score','Patient Specimen ID','PurIST Subtype','Patient ID']]\n",
    "df_gsva.rename({'Patient ID':'Public_Patient_ID'},axis=1,inplace=True)\n",
    "df_gsva['Specimen_ID'] = [item.split('-')[-1] for item in df_gsva.loc[:,'Patient Specimen ID']]\n",
    "df_gsva.drop('Patient Specimen ID',axis=1,inplace=True)\n",
    "df_purist = df_gsva.pivot(index='Public_Patient_ID',columns=['Specimen_ID'],values='PurIST Score')\n",
    "df_purist['PurIST_Primary'] = df_purist.loc[:,'T'].fillna(df_purist.F).fillna(df_purist.T2)\n",
    "df_purist['PurIST_Primary_T2'] = df_purist.T2\n",
    "#odd that there are only 71 mets with purist calls\n",
    "df_purist['PurIST_Met'] = df_purist.loc[:,'M'].fillna(df_purist.M2)\n",
    "print(f'Pts with RNAseq: {df_purist.reset_index().Public_Patient_ID.nunique()}')\n",
    "df_patient = df_patient.merge(df_purist.reset_index().loc[:,['PurIST_Primary','PurIST_Met',\n",
    "                            'PurIST_Primary_T2','Public_Patient_ID']],\n",
    "                            on='Public_Patient_ID',how='left')\n",
    "print(f'number rows {len(df_patient)}')\n",
    "print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e79f54ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene2 = pd.read_csv('data/HighMedImpactVariantDataForStatsFromMH.tsv',sep='\\t')\n",
    "df_gene2.columns = [item.replace('.',' ') for item in df_gene2.columns]\n",
    "\n",
    "#use tempus\n",
    "df_gene = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset2.xlsx',\n",
    "                            sheet_name='Mutation Data')\n",
    "\n",
    "def catch(func, handle=lambda e : e, *args, **kwargs):\n",
    "    try:\n",
    "        return func(*args, **kwargs)\n",
    "    except Exception as e:\n",
    "        return np.nan\n",
    "df_gene['Alteration_Function'] = [catch(lambda : item.split(' - ')[1]) for item in df_gene.loc[:,'Alteration Type']]\n",
    "df_gene['Gene_Function'] = df_gene.Gene + '_' + df_gene.Alteration_Function.fillna('__')\n",
    "df_gene['Gene_Function'] = [item.replace('___','') for item in df_gene['Gene_Function']]\n",
    "df_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57761991-7726-4842-a308-8449d075efd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_gene.Alteration_Function.value_counts()/len(df_gene)\n",
    "# df_gene.Alteration_Function.isna().sum()/len(df_gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94a33917",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#add mutation data, if there are more than 9 patient with a mutation\n",
    "# df_gene = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset2.xlsx',\n",
    "#                            sheet_name='Mutation Data')\n",
    "#which patients were sequencedL\n",
    "se_patients = pd.Series(df_gene.loc[:,'Patient Specimen ID'].unique())\n",
    "print(f\"number of genomics specimens: {len(se_patients)}\")\n",
    "se_mets = se_patients[se_patients.str.contains('-M')]\n",
    "ls_met = [item.split('-')[0] + '-' + item.split('-')[1] for item in se_mets]\n",
    "se_pri = se_patients[~se_patients.str.contains('-M')]\n",
    "ls_pri = [item.split('-')[0] + '-' + item.split('-')[1] for item in se_pri]\n",
    "print(f'no. primary specimens {len(ls_pri)}')\n",
    "print(f'no. met specimens {len(ls_met)}')\n",
    "print(f'no. unique pts {len(set(ls_pri+ls_met))}')\n",
    "print(f'no. matched pts {len(set(ls_pri).intersection(set(ls_met)))}')\n",
    "print(f'no. duplicated pri specimens {pd.Series(ls_pri).duplicated().sum()}')\n",
    "print(f'no. duplicated met specimens {pd.Series(ls_met).duplicated().sum()}')\n",
    "\n",
    "d_fill = {'_Primary':ls_pri,'_Met':ls_met,'':ls_pri+ls_met}\n",
    "ls_genes = []\n",
    "for s_gene in df_gene.Gene.unique():  \n",
    "    df_gene_alt = df_gene.loc[df_gene.Gene==s_gene,['Patient ID','Patient Specimen ID','Gene']].copy()\n",
    "    if df_gene_alt.loc[:,'Patient ID'].nunique() > 9 or s_gene.find('BRCA') > -1or s_gene.find('PALB') > -1:\n",
    "        print(s_gene)\n",
    "        df_gene_alt['Specimen_ID'] = [item.split('-')[-1] for item in df_gene_alt.loc[:,'Patient Specimen ID']]\n",
    "        df_pivot = df_gene_alt.drop_duplicates().pivot(index='Patient ID',columns=['Specimen_ID'],values='Gene')\n",
    "        try:\n",
    "            df_pivot[f'{s_gene}_Altered_Primary'] = df_pivot.loc[:,'T'].fillna(df_pivot.F).replace({s_gene:True})#.fillna(False)\n",
    "        except:\n",
    "            df_pivot[f'{s_gene}_Altered_Primary'] = df_pivot.loc[:,'T'].replace({s_gene:True})#.fillna(False)\n",
    "        if df_pivot.columns.isin(['M']).any():\n",
    "            ls_col = [f'{s_gene}_Altered_Primary',f'{s_gene}_Altered_Met']\n",
    "            try:\n",
    "                df_pivot[f'{s_gene}_Altered_Met'] = df_pivot.loc[:,'M'].fillna(df_pivot.M2).replace({s_gene:True})#.fillna(False)\n",
    "            except:\n",
    "                df_pivot[f'{s_gene}_Altered_Met'] = df_pivot.loc[:,'M'].replace({s_gene:True})#.fillna(False)\n",
    "        else:\n",
    "            ls_col = [f'{s_gene}_Altered_Primary']\n",
    "        df_pivot = df_pivot.loc[:,ls_col]\n",
    "        df_pivot[f'{s_gene}_Altered'] = pd.NA \n",
    "        df_pivot = df_pivot.reset_index().rename({'Patient ID':'Public_Patient_ID'},axis=1)#,inplace=True\n",
    "        df_pivot.loc[(df_pivot.iloc[:,1::]).any(axis=1),f'{s_gene}_Altered'] = True\n",
    "        df_patient = df_patient.merge(df_pivot,on='Public_Patient_ID',how='left',suffixes=('','__'))\n",
    "        # add the false values (differnet from na)\n",
    "        for s_loc in ['_Primary','_Met','']:\n",
    "            try:\n",
    "                b_na = df_patient.loc[:,f'{s_gene}_Altered{s_loc}'].isna()\n",
    "                b_loc = df_patient.Public_Patient_ID.isin(d_fill[s_loc])\n",
    "                df_patient.loc[(b_na) & (b_loc),f'{s_gene}_Altered{s_loc}'] = False \n",
    "                #print(df_patient.loc[:,f'{s_gene}_Altered{s_loc}'].value_counts())\n",
    "                ls_genes.append(f'{s_gene}_Altered{s_loc}')\n",
    "            except:\n",
    "                print(f'no {s_loc}')\n",
    "    #break\n",
    "#gain versus loss\n",
    "#'''\n",
    "ls_arid_pts =['ST-00007313','ST-00007175','ST-00015114','ST-00020482',\n",
    " 'ST-00019171','ST-00019171','ST-00007120','ST-00007120'] #LOF\n",
    "print(df_patient.ARID1A_Altered.value_counts())\n",
    "df_patient['ARID1aLOF_Altered'] = pd.NA\n",
    "df_patient.loc[df_patient.ARID1A_Altered.notna(),'ARID1aLOF_Altered'] = False\n",
    "df_patient.loc[(df_patient.Public_Patient_ID.isin(ls_arid_pts)) & ((df_patient.ARID1A_Altered)),'ARID1aLOF_Altered'] = True\n",
    "df_patient.loc[df_patient.Public_Patient_ID.isin(ls_arid_pts),'ARID1A_Altered'] = False\n",
    "ls_genes.append('ARID1aLOF_Altered')\n",
    "print(df_patient.ARID1A_Altered.value_counts())\n",
    "print(df_patient.ARID1aLOF_Altered.value_counts())\n",
    "\n",
    "ls_more_than_10 = sorted(set([item.split('_')[0] for item in ls_genes]))\n",
    "se_test = df_patient.loc[:,[f'{item}_Altered' for item in ls_more_than_10]].sum()\n",
    "print(f\"genes with 10 or more: {len(ls_more_than_10) - len(se_test[se_test < 10])}\")\n",
    "#'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3b4989e-05a9-44d1-b35e-bc87e2026228",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(df_patient.RNF43_Altered.notna().sum())\n",
    "print(df_patient.PALB2_Altered_Primary.sum())\n",
    "# print(df_patient.RNF43_Altered_Primary.notna().sum())\n",
    "# print(df_patient.RNF43_Altered_Met.notna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd00e409-3118-4573-bb9c-e36bd2aeb255",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_patient[df_patient.PALB2_Altered_Met==1].Public_Patient_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8aaa7d0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #add age, filter patients who died after surgery\n",
    "# d_ids = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/OLD Versions/Simplified_Public_IDs_Key.xlsx',sheet_name=None)\n",
    "\n",
    "# df_id = d_ids['RnaSeqKey']\n",
    "# ls_ids = df_id.loc[:,'Public.Specimen.ID']\n",
    "\n",
    "# # add patients w/o RNA seq\n",
    "# for s_key in ['TcrTumorKey','TcrBloodKey','SuppTable2Key','DnaPanelKey']:\n",
    "#     df_add = d_ids[s_key].loc[~d_ids[s_key].loc[:,'Public.Specimen.ID'].isin(ls_ids)]\n",
    "#     df_id = pd.concat([df_id,df_add])\n",
    "\n",
    "# #check\n",
    "# print(df_id.loc[:,'OPTR.Specimen.ID'].duplicated().any())\n",
    "# ls_drop = df_id.loc[df_id.loc[:,'OPTR.Specimen.ID'].str.contains('-T2')].index\n",
    "# df_unique = df_id.loc[df_id.loc[:,'OPTR.Specimen.ID'].str.contains('-T')].drop(ls_drop)\n",
    "# #check\n",
    "# print(df_unique.OPTR.duplicated().any())\n",
    "\n",
    "# #add id\n",
    "# df_vital['Public_Patient_ID'] = df_vital.OPTR.map(dict(zip(df_unique.OPTR,df_unique.loc[:,'Biolibrary.Subject.ID'])))\n",
    "\n",
    "# #omics data plus clinical data\n",
    "# df_patient = df_patient.merge(df_vital,on='Public_Patient_ID',how='left',suffixes=('','_x'))\n",
    "# print('adding cinical data')\n",
    "# print(f'number rows {len(df_patient)}')\n",
    "# print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')\n",
    "# #df_id.to_csv('Patient_IDs.csv')\n",
    "\n",
    "# #add age category\n",
    "# df_patient.loc[df_patient.loc[:,'Age at Diagnosis'] > 70,'Age'] = '>70'\n",
    "# df_patient.loc[df_patient.loc[:,'Age at Diagnosis'] <= 70,'Age'] = '<=70'\n",
    "\n",
    "# # drop less than 30 days #died of surgery\n",
    "# ls_drop_surgery = df_patient[(df_patient.loc[:,'Days from Resection to FU'] < 30) & (df_patient.loc[:,'cVitalStatus'] == 'Dead')].index\n",
    "# print('patients who died 1 month after surgery')\n",
    "# print(df_patient[df_patient.index.isin(ls_drop_surgery)].Public_Patient_ID)\n",
    "# df_patient['Alive_30_days_post_surgery'] = True\n",
    "# df_patient.loc[ls_drop_surgery,'Alive_30_days_post_surgery'] = False\n",
    "# #df_patient.Alive_30_days_post_surgery.fillna(True,inplace=True)\n",
    "# print('identifying died of surgery')\n",
    "# print(f'{df_patient.Alive_30_days_post_surgery.value_counts()}')\n",
    "# print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')\n",
    "\n",
    "# #add lung and liver versus all (not just lung, liver and NA)\n",
    "# df_patient['Lung_Cohort'] = False\n",
    "# df_patient.loc[df_patient.Cohort=='Lung','Lung_Cohort'] = True\n",
    "\n",
    "# df_patient['Liver_Cohort'] = False\n",
    "# df_patient.loc[df_patient.Cohort=='Liver','Liver_Cohort'] = True\n",
    "\n",
    "\n",
    "# print(f'number rows {len(df_patient)}')\n",
    "# print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e576756-1c60-4e4c-8fba-96c11b18c0b6",
   "metadata": {},
   "source": [
    "## Find survival diff in GOF, LOF, altered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85744607-19e1-458f-9180-8a073f3e37f0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# sns.set_palette('tab10')\n",
    "# s_time = 'Days from Diagnosis to FU'\n",
    "# s_censor = 'Survival'\n",
    "# df_patient['Survival'] = df_patient.loc[:,'Vital Status at FU'].replace({'Alive':0,'Dead':1})\n",
    "# for s_gene in sorted(set([item.split('_')[0] for item in ls_genes])):\n",
    "\n",
    "#     print(s_gene)\n",
    "#     df_altered = df_gene[(df_gene.loc[:,'Gene_Function'].str.contains(f'{s_gene}_')) | (df_gene.loc[:,'Gene_Function'] == (f'{s_gene}'))].copy()\n",
    "#     df_altered.rename({'Patient ID':'Public_Patient_ID'},axis=1,inplace=True)\n",
    "#     df_merge = df_patient.merge(df_altered,on='Public_Patient_ID',how='left')\n",
    "#     #df_merge.loc[(df_merge.loc[:,'Gene_Function'].isna()) & (df_merge.loc[:,'MTAP_Altered'].notna()),'Gene_Function'] = 'WT'\n",
    "#     df_km = df_merge.loc[df_merge.Alive_30_days_post_surgery,['Public_Patient_ID',s_time,s_censor,'Gene_Function']].dropna()\n",
    "#     try:\n",
    "#         fig,ax,ls_order = util.km_plot(df_km,'Gene_Function',s_time,s_censor)\n",
    "#         ax.set_ylabel('Fraction Alive')\n",
    "#         ax.set_xlabel('Overall Survival (Days)')\n",
    "#         fig.savefig(f'figures/KM_GOF_LOF_{s_gene}.png')\n",
    "#         plt.close(fig)\n",
    "#     except:\n",
    "#         if s_gene == 'ARID1A':\n",
    "#             ls_arid_pts = df_km.loc[df_km.Gene_Function=='ARID1A_LOF','Public_Patient_ID']\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d91322e2-5fab-405c-bb60-51fa38369ce8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# %matplotlib inline\n",
    "# for s_gene in ['ARID1A_Altered','NOTCH1_Altered','ARID1aLOF_Altered']:\n",
    "#     df_km = df_patient.loc[df_patient.Alive_30_days_post_surgery,['Public_Patient_ID',s_time,s_censor,s_gene]].dropna()\n",
    "#     df_km.loc[:,s_gene].replace({True:'Yes',False:'No'},inplace=True)\n",
    "#     fig,ax,ls_order = util.km_plot(df_km,s_gene,s_time,s_censor)\n",
    "#     ax.set_ylabel('Fraction Alive')\n",
    "#     ax.set_xlabel('Overall Survival (Days)')\n",
    "#     fig.savefig(f'figures/KM_GOF_LOF_{s_gene}.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "993bf5ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # a patient who has a period of resection to FU but not to recurrence had no recurrence and \n",
    "# #a patient who had a period of resection to recurrence but no liver or lung met had another site of recurrence.\n",
    "\n",
    "# #has a recurrence, not in liver or lung\n",
    "# b_recur = df_patient.loc[:,'Days from Resection to Recurrence'].notna() #& df_patient.Cohort.isna() \n",
    "\n",
    "# #had a resection, no recurrence,  not liver or lung\n",
    "# b_no_recur = df_patient.loc[:,'Days from Resection to FU'].notna() & df_patient.loc[:,'Days from Resection to Recurrence'].isna()  #& df_patient.Cohort.isna() \n",
    "\n",
    "# #add recurrence other\n",
    "# df_patient['Recurrence'] = False\n",
    "# df_patient.loc[b_recur,'Recurrence'] = True\n",
    "# #add recurrence none\n",
    "# df_patient['No_Recurrence'] = False\n",
    "# df_patient.loc[b_no_recur,'No_Recurrence'] = True\n",
    "# # should be 73 and 103, but I get 184 with a recurrence and 113 without a recurrence\n",
    "# print(df_patient.Recurrence.sum())\n",
    "# print(df_patient.No_Recurrence.sum())\n",
    "\n",
    "# print((df_patient.Recurrence & df_patient.No_Recurrence).sum())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1711790",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #what is up with these patients?\n",
    "# df_patient.loc[(df_patient.Cohort.notna()) & (df_patient.No_Recurrence)].loc[:,['Public_Patient_ID','Cohort','Days from Resection to FU','Days from Resection to Recurrence']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0178a2c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #all recurrence in one\n",
    "# df_patient.Cohort.replace('nan',np.nan,inplace=True)\n",
    "# df_patient['Recurrence_Sites_4'] = np.nan\n",
    "# df_patient['Recurrence_Sites_4'] = df_patient['Recurrence_Sites_4'].astype('object')\n",
    "# df_patient.loc[df_patient.Recurrence,'Recurrence_Sites_4'] = 'Other_site'\n",
    "# df_patient.loc[df_patient.No_Recurrence,'Recurrence_Sites_4'] = 'No_Doc_Recur'\n",
    "# df_patient.loc[(df_patient.Cohort.notna()) & (df_patient.Recurrence),'Recurrence_Sites_4'] = df_patient.Cohort\n",
    "# print(df_patient.Recurrence_Sites_4.isna().sum())\n",
    "# df_patient.Recurrence_Sites_4.fillna('No_Resection',inplace=True)\n",
    "# #lung ST-00020218\n",
    "# df_patient.loc[df_patient.Public_Patient_ID=='ST-00020218','Recurrence_Sites_4'] = 'Lung'\n",
    "# #liver ST-00024980\n",
    "# df_patient.loc[df_patient.Public_Patient_ID=='ST-00024980','Recurrence_Sites_4'] = 'Liver'\n",
    "\n",
    "# for s_site in df_patient.Recurrence_Sites_4.unique():\n",
    "#     n_patients = df_patient[df_patient.Recurrence_Sites_4==s_site].Public_Patient_ID.nunique()\n",
    "#     print(f'{s_site} {n_patients}')\n",
    "# print(df_patient.Cohort.dropna().value_counts())\n",
    "# print(f'number rows {len(df_patient)}')\n",
    "# print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')\n",
    "\n",
    "# #looks good\n",
    "# print(82 + 125 + 113 + 29 + 73) # all patients\n",
    "# print(82 + 29 + 113 + 73)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69a7df09",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #number of pateints alive 30 days after surgery and include in survival analysis\n",
    "# for s_site in df_patient.Recurrence_Sites_4.unique():\n",
    "#     n_patients = df_patient[(df_patient.Recurrence_Sites_4==s_site) &(df_patient.Alive_30_days_post_surgery)].Public_Patient_ID.nunique()\n",
    "#     print(f'{s_site} {n_patients}')\n",
    "# print(df_patient.loc[df_patient.Alive_30_days_post_surgery,'Cohort'].dropna().value_counts())\n",
    "\n",
    "# # 82 + 29 + 73 + 105\n",
    "# # 82 + 29 + 73 + 105 + 125\n",
    "# # 82 + 29 + 113 + 73\n",
    "# #add column for was resected\n",
    "# df_patient['Resected'] = False\n",
    "# df_patient.loc[df_patient.loc[:,'Days from Resection to FU'].notna(),'Resected'] = True\n",
    "# print(df_patient.Resected.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99f07e3a-db24-4cd0-a11a-29690336369f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ls_file = ['data/GSVA_All_Kallisto55_pORG_Up_55_pSUB_Up_51.tsv',\n",
    "#            'data/GSVA_Met_Kallisto55_pORG_Up_55_pSUB_Up_51.tsv',\n",
    "#            'data/GSVA_Primary_Kallisto55_pORG_Up_55_pSUB_Up_51.tsv']\n",
    "# d_rename = {'pORG_Up_55':'pORG', 'pSUB_Up_51':'pSUB'}\n",
    "# d_public_id = dict(zip(d_ids['RnaSeqKey'].loc[:,'OPTR.Specimen.ID'],d_ids['RnaSeqKey'].loc[:,'Public.Specimen.ID']))\n",
    "\n",
    "# for s_file in ls_file:\n",
    "#     df_pORG = pd.read_csv(s_file,sep='\\t',index_col=0)\n",
    "#     s_type = s_file.split('_')[1]\n",
    "\n",
    "#     if s_type == 'All':\n",
    "#         #add 'GSVA_All': must pivot\n",
    "#         # add public IDs\n",
    "#         df_pORG['Public_Specimen_ID'] = df_pORG.index.map(d_public_id)\n",
    "#         df_pORG['Specimen_ID'] = [item.split('-')[-1] for item in df_pORG.Public_Specimen_ID]\n",
    "#         df_pORG['Public_Patient_ID'] = [item.split('-')[0] + '-' + item.split('-')[1] for item in df_pORG.Public_Specimen_ID]\n",
    "#         for key, item in d_rename.items():\n",
    "#             df_group = df_pORG.pivot(index='Public_Patient_ID',columns='Specimen_ID',values=key)\n",
    "#             df_group[f'{item}_allPrimary'] = df_group.loc[:,'T'].fillna(df_group.F)\n",
    "#             df_group[f'{item}_allMet'] = df_group.loc[:,'M'].fillna(df_group.M2)\n",
    "#             df_patient = df_patient.merge(df_group.reset_index().loc[:,['Public_Patient_ID',f'{item}_allPrimary',f'{item}_allMet']],on='Public_Patient_ID',how='left',suffixes=('','_x'))\n",
    "#             df_patient[f'{item}_{s_type}_T2'] = df_patient.Public_Patient_ID.map(df_group.loc[:,'T2'])\n",
    "#             print(len(df_patient))\n",
    "\n",
    "#     else:\n",
    "#         df_pORG.columns = [f'{d_rename[item]}_{s_type}' for item in df_pORG.columns]\n",
    "#         # add public IDs\n",
    "#         df_pORG['Public_Specimen_ID'] = df_pORG.index.map(d_public_id)\n",
    "#         df_pORG['Public_Patient_ID'] = [item.split('-')[0] + '-' + item.split('-')[1] for item in df_pORG.Public_Specimen_ID]\n",
    "#         #drop 'T2'\n",
    "#         df_t2 = df_pORG[df_pORG.Public_Specimen_ID.str.contains('-T2')]\n",
    "#         ls_drop = df_pORG[df_pORG.Public_Specimen_ID.str.contains('-T2')].index\n",
    "#         #merge\n",
    "#         df_patient = df_patient.merge(df_pORG.drop(ls_drop),on='Public_Patient_ID',suffixes=('','_x'),how='left')\n",
    "#         if s_type == 'Primary':\n",
    "#             for key, item in d_rename.items():\n",
    "#                 se_t2 = df_pORG.loc[ls_drop,['Public_Patient_ID',f'{item}_{s_type}']].set_index('Public_Patient_ID')\n",
    "#                 df_patient[f'{item}_{s_type}_T2'] = df_patient.Public_Patient_ID.map(se_t2.to_dict()[f'{item}_{s_type}'])\n",
    "#             print(len(df_patient))\n",
    "#     #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f182f1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_carl = pd.read_csv('../annotation/JasonsPaperMetaDataWithPublicIDs.tsv',sep='\\t')\n",
    "# print(len(df_carl))\n",
    "# ls_col = ['Tumor.Cellularity.by.DNA','cHas.DDR.Alteration','Specimen.Site','Public.Specimen.ID']\n",
    "# df_carl = df_carl.loc[:,ls_col]\n",
    "# df_carl.columns = [item.replace('.','_').replace('cHas_','') for item in df_carl.columns]\n",
    "# ls_drop = df_carl[df_carl['Public_Specimen_ID'].str.contains('T2')].index\n",
    "# df_carl = df_carl.drop(ls_drop)\n",
    "# df_carl['Public_Patient_ID'] = [item.split('-')[0] + '-' + item.split('-')[1] for item in df_carl.Public_Specimen_ID]\n",
    "# df_carl['Specimen_ID'] = [item.split('-')[-1] for item in df_carl.Public_Specimen_ID]\n",
    "# print(len(df_carl))\n",
    "# df_carl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78aac2ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_optr = pd.DataFrame(index=df_carl.Public_Patient_ID.unique(),dtype='O')\n",
    "# df_optr.index.name = 'Public_Patient_ID'\n",
    "# ls_col = ['Specimen_Site','DDR_Alteration','Tumor_Cellularity_by_DNA']\n",
    "# df_group = df_carl.pivot(index='Public_Patient_ID',columns='Specimen_ID',values=ls_col)\n",
    "# for s_col in ls_col:\n",
    "#     df_col = df_group.loc[:,s_col]\n",
    "#     if s_col == 'Specimen_Site':\n",
    "#         df_optr[f'RNA_DNA_{s_col}_Primary'] = df_col.loc[:,'T'].fillna(df_col.F)\n",
    "#         df_optr[f'RNA_DNA_{s_col}_Met'] = df_col.loc[:,'M'].fillna(df_col.M2)\n",
    "#     else:\n",
    "#         df_optr[f'{s_col}_Primary'] = df_col.loc[:,'T'].fillna(df_col.F)\n",
    "#         df_optr[f'{s_col}_Met'] = df_col.loc[:,'M'].fillna(df_col.M2)\n",
    "\n",
    "# df_patient = df_patient.merge(df_optr.reset_index(),on='Public_Patient_ID',how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2086c4ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(f'number rows {len(df_patient)}')\n",
    "# print(f'number unique pts. {df_patient.Public_Patient_ID.nunique()}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce67b469-7ee3-4c43-b25f-d15e3a901161",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_gene[df_gene.loc[:,'Patient ID'].isin(se_optr)]\n",
    "# df_gene.loc[:,'Patient Specimen ID'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa35768d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Panel, but no mutation data: 4044 4139 4404 4701 4731 4793\n",
    "# # We need to add a \"place holder\" gene for each of the samples with no calls.\n",
    "# # This is so that these will be included in the statistics and shown on the graphs.\n",
    "# OPTRsWithPanelButNoMutations = [4044, 4139, 4404, 4701, 4731, 4793]\n",
    "# se_optr = df_id.loc[df_id.OPTR.isin(OPTRsWithPanelButNoMutations),'Biolibrary.Subject.ID']\n",
    "# df_patient.loc[df_patient.Public_Patient_ID.isin(se_optr),df_patient.columns.str.contains('Altered_Primary')]=False\n",
    "# #Primary: 5 or six already had a False in the data (should have been NA)\n",
    "# #df_patient.loc[df_patient.Public_Patient_ID.isin(se_optr),df_patient.columns.str.endswith('Altered')]\n",
    "# #Met: 6 of 6 had False in data\n",
    "# #now we have 202 pri - but one has no RNA\n",
    "# print(df_patient.TP53_Altered_Primary.notna().sum())\n",
    "# len(set(df_patient.loc[df_patient.TP53_Altered_Primary.notna(),'Public_Patient_ID']).intersection(\n",
    "#     df_patient.loc[df_patient.pORG_Primary.notna(),'Public_Patient_ID']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "341a03fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_dove = pd.read_excel('../annotation/Adaptive-total samples-metadata-mets.xlsx',index_col=0)\n",
    "# print(df_dove.index.duplicated().any())\n",
    "# df_dove['Public_Patient_ID'] = df_dove.index.map(dict(zip(df_id.OPTR,df_id.loc[:,'Biolibrary.Subject.ID'])))\n",
    "# print(df_dove.Public_Patient_ID.duplicated().any())\n",
    "# df_patient['TCR_Met_Site'] = df_patient.Public_Patient_ID.map(dict(zip(df_dove.Public_Patient_ID,df_dove.loc[:,'Tumor tissue site'])))\n",
    "# d_replace = {'Supraclavicular lymph node':'Lymph Node','peritoneum':'Peritoneum',\n",
    "#              'lung':'Lung','omentum':'Omentum'}\n",
    "# df_patient['TCR_Met_Site'] = df_patient.TCR_Met_Site.replace(d_replace)\n",
    "# df_patient.groupby('Cohort').TCR_Met_Site.value_counts()#.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3362e11-f0eb-49a9-8a5d-95c148c26f2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #histology: add Desmoplasia +/- !!\n",
    "\n",
    "# d_ids = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/OLD Versions/Simplified_Public_IDs_Key.xlsx',sheet_name=None)\n",
    "\n",
    "# df_id = d_ids['RnaSeqKey']\n",
    "# ls_ids = df_id.loc[:,'Public.Specimen.ID']\n",
    "\n",
    "# # add patients w/o RNA seq\n",
    "# for s_key in ['TcrTumorKey','TcrBloodKey','SuppTable2Key','DnaPanelKey']:\n",
    "#     df_add = d_ids[s_key].loc[~d_ids[s_key].loc[:,'Public.Specimen.ID'].isin(ls_ids)]\n",
    "#     df_id = pd.concat([df_id,df_add])\n",
    "# df_ter = pd.read_excel('annotation/Histology Analyses - Terry and Brian.xlsx',sheet_name='final for TM batch4 list2')\n",
    "# df_ter['OPTR'] = [int(item.split('-')[0]) for item in df_ter.loc[:,'OPTR Specimen ID']]\n",
    "# df_ter.rename({'OPTR Specimen ID':'SpecimenID','PNI (Y/N)':'PNI'},axis=1,inplace=True)\n",
    "# df_ter.loc[:,'Tertiary Lymph Strucures (Peritumoral, intratumoral)'].replace({'Rare peritumoral':'Peritumoral'},inplace=True)\n",
    "# df_ter['TLS'] = df_ter.loc[:,'Tertiary Lymph Strucures (Peritumoral, intratumoral)'].replace({'Peritumoral':'Y', 'Intratumoral and peritumoral':'Y', 'Intratumoral':'Y'})\n",
    "\n",
    "# s_replace = 'Acute, Chronic, and/or plasmacytoid inflammation'\n",
    "# d_replace_imm = {'Acute, chronic':'Y', 'Acute':'Y', 'Chronic, plasmacytoid':'Y',\n",
    "#  'Acute, chronic, plasmacytoid':'Y', 'Chronic':'Y', 'None':'N', ' Acute, chronic':'Y'}\n",
    "\n",
    "# d_replace_acute = {'Acute, chronic':'Y', 'Acute':'Y', 'Chronic, plasmacytoid':'N',\n",
    "#  'Acute, chronic, plasmacytoid':'Y', 'Chronic':'N', 'None':'N', ' Acute, chronic':'Y'}\n",
    "\n",
    "# d_replace_chr = {'Acute, chronic':'Y', 'Acute':'N', 'Chronic, plasmacytoid':'Y',\n",
    "#  'Acute, chronic, plasmacytoid':'Y', 'Chronic':'Y', 'None':'N', ' Acute, chronic':'Y'}\n",
    "\n",
    "# d_replace_pl = {'Acute, chronic':'N', 'Acute':'N', 'Chronic, plasmacytoid':'Y',\n",
    "#  'Acute, chronic, plasmacytoid':'Y', 'Chronic':'N', 'None':'N', ' Acute, chronic':'N'}\n",
    "# d_pni = {'                           ':np.nan,'Y+':'Y',\n",
    "#         'in  slide 2, but not circled ROI':'Y', 'n':'N', 'yes':'Y', 'N/a':np.nan}\n",
    "\n",
    "# df_ter['Inflammation'] = df_ter.loc[:,s_replace].replace(d_replace_imm)\n",
    "# df_ter['Acute Inflammation'] =df_ter.loc[:,s_replace].replace(d_replace_acute)\n",
    "# df_ter['Chronic Inflammation'] =df_ter.loc[:,s_replace].replace(d_replace_chr)\n",
    "# df_ter['Plasmacytoid Inflammation'] =df_ter.loc[:,s_replace].replace(d_replace_pl)\n",
    "# df_ter['Tumor_Type'] = [item.split('-')[-1].replace('F','T') for item in df_ter.SpecimenID]\n",
    "# df_ter['PNI'] =df_ter.loc[:,'PNI'].replace(d_pni)\n",
    "# df_ter['Desmoplasia (Y/N)'].replace({'Y+':'Y'},inplace=True)\n",
    "\n",
    "# # second\n",
    "# df_ter2 = pd.read_excel('annotation/Digital Slide Tempus Final Dx tkm_bs_2021.xlsx',sheet_name='final for TM batch4 list2')\n",
    "# df_ter2['OPTR'] = df_ter2.loc[:,'Subject ID (BC ID#)'].astype(pd.Int64Dtype())\n",
    "# df_ter2.rename({'PN +/-':'PNI'},axis=1,inplace=True)\n",
    "# #two\n",
    "# #Acute or Chronic Inflammation\n",
    "# # Diagnosis TM_Ben\n",
    "# # for s_col in df_ter2.columns:\n",
    "# #     print(s_col)\n",
    "# #     ls_un = df_ter2.loc[:,s_col].unique()\n",
    "# #     if len(ls_un) < 25:\n",
    "# #         print(ls_un)\n",
    "        \n",
    "\n",
    "\n",
    "# d_replace_tls = { 'Negative':'N', 'LN aggregates with plasma cells':'Y',\n",
    "#  'Severe acute inflammation':'N', 'mild chronic with plasma cells':'N',\n",
    "#  'Acute inflammation':'N', 'Acute &  chronic with plasma cells':'N',\n",
    "#  'LN aggregates &  plasma cells (acute in infarcted zone= false pos)':'Y',\n",
    "#  'Areas of elastosis are negative for PDAC (seems FP were targeted for LCM)':np.nan,\n",
    "#  'LN aggregates with numerous plasma cells':'Y', 'N/a':np.nan,\n",
    "#  'mild chronic with plasma cells (at edges of tumor there are LN aggregates, not in tumor)':'Y',\n",
    "#  'Marked LN aggregates with plasma cells throughout tumor':'Y',\n",
    "#  'Amazing sheets of lymphocytes in papillary stroma of IPMN':'Y',\n",
    "#  'Marked acute &  chronic with plasma cells':'N',\n",
    "#  'Negative (inflammation only around PanIN 2)':'N',\n",
    "#  'Negative (acute only in area of tumor infarction ~biopsy site)':'N'}\n",
    "\n",
    "# d_replace_imm = { 'Negative':'N', 'LN aggregates with plasma cells':'Y',\n",
    "#  'Severe acute inflammation':'Y', 'mild chronic with plasma cells':'Y',\n",
    "#  'Acute inflammation':'Y', 'Acute &  chronic with plasma cells':'Y',\n",
    "#  'LN aggregates &  plasma cells (acute in infarcted zone= false pos)':'Y',\n",
    "#  'Areas of elastosis are negative for PDAC (seems FP were targeted for LCM)':np.nan,\n",
    "#  'LN aggregates with numerous plasma cells':'Y', 'N/a':np.nan,\n",
    "#  'mild chronic with plasma cells (at edges of tumor there are LN aggregates, not in tumor)':'Y',\n",
    "#  'Marked LN aggregates with plasma cells throughout tumor':'Y',\n",
    "#  'Amazing sheets of lymphocytes in papillary stroma of IPMN':'Y',\n",
    "#  'Marked acute &  chronic with plasma cells':'Y',\n",
    "#  'Negative (inflammation only around PanIN 2)':'N',\n",
    "#  'Negative (acute only in area of tumor infarction ~biopsy site)':'N'}\n",
    "\n",
    "# d_replace_acute = { 'Negative':'N', 'LN aggregates with plasma cells':'N',\n",
    "#  'Severe acute inflammation':'Y', 'mild chronic with plasma cells':'N',\n",
    "#  'Acute inflammation':'Y', 'Acute &  chronic with plasma cells':'Y',\n",
    "#  'LN aggregates &  plasma cells (acute in infarcted zone= false pos)':'N',\n",
    "#  'Areas of elastosis are negative for PDAC (seems FP were targeted for LCM)':np.nan,\n",
    "#  'LN aggregates with numerous plasma cells':'N', 'N/a':np.nan,\n",
    "#  'mild chronic with plasma cells (at edges of tumor there are LN aggregates, not in tumor)':'N',\n",
    "#  'Marked LN aggregates with plasma cells throughout tumor':'N',\n",
    "#  'Amazing sheets of lymphocytes in papillary stroma of IPMN':'N',\n",
    "#  'Marked acute &  chronic with plasma cells':'Y',\n",
    "#  'Negative (inflammation only around PanIN 2)':'N',\n",
    "#  'Negative (acute only in area of tumor infarction ~biopsy site)':'N'}\n",
    "\n",
    "# d_replace_chr = { 'Negative':'N', 'LN aggregates with plasma cells':'Y',\n",
    "#  'Severe acute inflammation':'N', 'mild chronic with plasma cells':'Y',\n",
    "#  'Acute inflammation':'N', 'Acute &  chronic with plasma cells':'Y',\n",
    "#  'LN aggregates &  plasma cells (acute in infarcted zone= false pos)':'Y',\n",
    "#  'Areas of elastosis are negative for PDAC (seems FP were targeted for LCM)':np.nan,\n",
    "#  'LN aggregates with numerous plasma cells':'Y', 'N/a':np.nan,\n",
    "#  'mild chronic with plasma cells (at edges of tumor there are LN aggregates, not in tumor)':'Y',\n",
    "#  'Marked LN aggregates with plasma cells throughout tumor':'Y',\n",
    "#  'Amazing sheets of lymphocytes in papillary stroma of IPMN':'Y',\n",
    "#  'Marked acute &  chronic with plasma cells':'Y',\n",
    "#  'Negative (inflammation only around PanIN 2)':'N',\n",
    "#  'Negative (acute only in area of tumor infarction ~biopsy site)':'N'}\n",
    "\n",
    "\n",
    "# d_replace_pl = { 'Negative':'N', 'LN aggregates with plasma cells':'Y',\n",
    "#  'Severe acute inflammation':'N', 'mild chronic with plasma cells':'Y',\n",
    "#  'Acute inflammation':'N', 'Acute &  chronic with plasma cells':'Y',\n",
    "#  'LN aggregates &  plasma cells (acute in infarcted zone= false pos)':'Y',\n",
    "#  'Areas of elastosis are negative for PDAC (seems FP were targeted for LCM)':np.nan,\n",
    "#  'LN aggregates with numerous plasma cells':'Y', 'N/a':np.nan,\n",
    "#  'mild chronic with plasma cells (at edges of tumor there are LN aggregates, not in tumor)':'Y',\n",
    "#  'Marked LN aggregates with plasma cells throughout tumor':'Y',\n",
    "#  'Amazing sheets of lymphocytes in papillary stroma of IPMN':'N',\n",
    "#  'Marked acute &  chronic with plasma cells':'Y',\n",
    "#  'Negative (inflammation only around PanIN 2)':'N',\n",
    "#  'Negative (acute only in area of tumor infarction ~biopsy site)':'N'}\n",
    "\n",
    "# d_replace_type = {'PDAC':'T', 'Adenosquamous':'T', 'PDAC + Pan1':'T', 'IPMN 3/PDAC':'T',\n",
    "#        'IPMN 2':np.nan, 'Met PDAC (lung)':'M', 'PDAC (post Tx?) +Pan3':'T', 'PDAC ':'T',\n",
    "#        'Negative (post Tx?)':np.nan, 'Colloid PDAC':'T', 'Met vs primary lung':'M',\n",
    "#        'Met to liver (r/o Angiosarcoma)':'M', 'Met (PDAC) to lymph node':'M',\n",
    "#        'PDAC (post Tx?)':'T', 'ACA r/o colon primary':np.nan, 'Cytology PDAC ':'T',\n",
    "#        'IPMN ':np.nan, 'PDAC + Pan2':'T', 'Ampulllary ACA':'T'}\n",
    "# d_ali = { 'n':'N', 'YES':'Y', 'N/a':np.nan, 'no':'N'}\n",
    "# s_replace = 'Acute or Chronic Inflammation'\n",
    "# df_ter2['TLS'] = df_ter2.loc[:,s_replace].replace(d_replace_tls)\n",
    "# df_ter2['Inflammation'] = df_ter2.loc[:,s_replace].replace(d_replace_imm)\n",
    "# df_ter2['Acute Inflammation'] =df_ter2.loc[:,s_replace].replace(d_replace_acute)\n",
    "# df_ter2['Chronic Inflammation'] =df_ter2.loc[:,s_replace].replace(d_replace_chr)\n",
    "# df_ter2['Plasmacytoid Inflammation'] =df_ter2.loc[:,s_replace].replace(d_replace_pl)\n",
    "# df_ter2['Tumor_Type'] =df_ter2.loc[:,'Diagnosis TM_Ben'].replace(d_replace_type)\n",
    "# df_ter2['PNI'] =df_ter2.loc[:,'PNI'].replace(d_pni)\n",
    "# df_ter2['Desmoplasia (Y/N)'] = df_ter2.loc[:,'Desmoplasia +/-'].replace({np.nan:pd.NA, 'y':'Y',\n",
    "#                                                                          'NO':'N', 'n':'N', \n",
    "#                                                                          'N/a':pd.NA, 'yes':'Y',})\n",
    "# df_ter2['ALI (Y/N)'] =df_ter2.loc[:,'ALI +/-'].replace(d_ali)\n",
    "\n",
    "# # \n",
    "# ls_col = [ 'TLS', 'Inflammation','PNI','ALI (Y/N)','Desmoplasia (Y/N)',\n",
    "#        'Acute Inflammation', 'Chronic Inflammation',\n",
    "#        'Plasmacytoid Inflammation','Tumor_Type','OPTR',]\n",
    "# #'''\n",
    "# df_ter = df_ter.loc[:,ls_col].dropna()\n",
    "# df_ter2 = df_ter2.loc[:,ls_col].dropna()\n",
    "# ls_drop = df_ter2[(df_ter2.OPTR.isin(df_ter.OPTR)) & (df_ter2.Tumor_Type=='T')].index\n",
    "\n",
    "# #concat\n",
    "# df_hist = pd.concat([df_ter,df_ter2.drop(ls_drop)])\n",
    "# df_hist['Public_Patient_ID'] = df_hist.OPTR.map(dict(zip(df_id.OPTR,df_id.loc[:,'Biolibrary.Subject.ID'])))\n",
    "# print(f'missing public ID: {df_hist.Public_Patient_ID.isna().sum()}')\n",
    "# df_hist_tum = df_hist[df_hist.Tumor_Type.str.contains('T').fillna(False)].copy()\n",
    "# print(f'duplicated primary {df_hist_tum.OPTR.duplicated().sum()}')\n",
    "# df_hist_met = df_hist[df_hist.Tumor_Type.str.contains('M').fillna(False)].copy()\n",
    "# print(f'duplicated mets {df_hist_met.OPTR.duplicated().sum()}')\n",
    "# #merge\n",
    "# df_patient = df_patient.merge(df_hist_tum.loc[:,ls_col[0:5]+['Public_Patient_ID']],on='Public_Patient_ID',how='left',suffixes=('','_x'))\n",
    "# df_patient = df_patient.merge(df_hist_met.loc[:,ls_col[0:5]+['Public_Patient_ID']],on='Public_Patient_ID',how='left',suffixes=('','_Met'))\n",
    "# #'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "876b52ca-fef9-46cb-af95-e9eb39734446",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_hist.to_csv('histology_for_Dove.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2d9438b-8994-4c66-b8eb-00b40841bed8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #add recurrence censor\n",
    "# df_patient['Recurrence'] = np.nan\n",
    "# df_patient.loc[~df_patient.loc[:,'Days from Resection to Recurrence'].isna(),'Recurrence'] = 1\n",
    "# df_patient.loc[df_patient.No_Recurrence,'Recurrence'] = 0\n",
    "# print(len(df_patient))\n",
    "# %matplotlib inline\n",
    "# # add purist subtype\n",
    "# # primary after met\n",
    "# df_patient['PurIST_Subtype'] = pd.NA\n",
    "# df_patient.loc[(df_patient.PurIST_Met <= 0.5),'PurIST_Subtype_Met'] = 'classical'\n",
    "# df_patient.loc[(df_patient.PurIST_Primary <= 0.5),'PurIST_Subtype'] = 'classical'\n",
    "# df_patient.loc[(df_patient.PurIST_Met > 0.5),'PurIST_Subtype_Met'] = 'basal-like'\n",
    "# df_patient.loc[(df_patient.PurIST_Primary > 0.5),'PurIST_Subtype'] = 'basal-like'\n",
    "# df_patient['PurIST_Subtype'] = df_patient.PurIST_Subtype.fillna(df_patient.PurIST_Subtype_Met)\n",
    "# df_patient.loc[df_patient.PurIST_Subtype=='classical','Classical_Cohort'] = df_patient.loc[df_patient.PurIST_Subtype=='classical','Cohort']\n",
    "# print(df_patient.Classical_Cohort.value_counts())\n",
    "\n",
    "# #fix no recur and other recur - out of resected patients\n",
    "# df_patient.loc[:,'No_Doc_Recur'] = pd.NA #np.nan #False\n",
    "# df_patient.loc[df_patient.Recurrence_Sites_4!='No_Resection','No_Doc_Recur'] = False#0 #False\n",
    "# df_patient.loc[df_patient.Recurrence_Sites_4=='No_Doc_Recur','No_Doc_Recur'] = True#1 #True\n",
    "# print(df_patient.No_Recurrence.value_counts())\n",
    "\n",
    "# df_patient.loc[:,'Other_Recurrence'] = pd.NA #np.nan #False\n",
    "# df_patient.loc[df_patient.Recurrence_Sites_4!='No_Resection','Other_Recurrence'] = False #0#\n",
    "# df_patient.loc[df_patient.Recurrence_Sites_4=='Other_site','Other_Recurrence'] = True #1#\n",
    "# print(df_patient.Other_Recurrence.value_counts())\n",
    "\n",
    "# d_replace = {'Yes Neoadjuvant':'Yes', 'No Neoadjuvant (but Yes Chemo)':'No',\n",
    "#        'No Chemotherapy':'No', 'No Resection':np.nan, 'No Chemo and No Resection':np.nan}\n",
    "# df_patient['Neoadjuvant'] = df_patient.loc[:,'Neoadjuvant Treatment'].replace(d_replace)\n",
    "# print(df_patient.Neoadjuvant.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0284aa33-63c5-47ad-8c28-da135e259001",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #lung wedge resection, liver met resection versus survival\n",
    "# df_lung = pd.read_excel(f'lung cohort-wedge resections_HPI_complete-v2.xlsx',index_col=0)\n",
    "# se_lung = df_lung[df_lung.loc[:,'lung wedge resection?']=='yes'].index\n",
    "\n",
    "# df_liver =  pd.read_csv(f'Liverlesions_supplemental_table122123.csv',index_col=0)\n",
    "# se_liver = df_liver.loc[df_liver.loc[:,'Liver Met Resection']=='Yes','Public_Patient_ID']\n",
    "\n",
    "# se_lung_pub = se_lung.map(dict(zip(df_id.OPTR,df_id.loc[:,'Biolibrary.Subject.ID'])))\n",
    "# df_patient['Lung_Met_Resection'] = pd.NA\n",
    "# df_patient.loc[df_patient.Cohort=='Lung','Lung_Met_Resection'] = 'No'\n",
    "# df_patient.loc[df_patient.Public_Patient_ID.isin(se_lung_pub),'Lung_Met_Resection'] = 'Yes'\n",
    "\n",
    "# df_patient['Cohort_Met_Resection'] = df_patient.Cohort.copy()\n",
    "# df_patient.loc[df_patient.Public_Patient_ID.isin(se_lung_pub),'Cohort_Met_Resection'] = 'Lung_Met_Res'\n",
    "# df_patient.loc[df_patient.Public_Patient_ID.isin(se_liver),'Cohort_Met_Resection'] = 'Liver_Met_Res'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ab60cec-4ddc-44a9-ad2a-a9fd79beb8d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #add TMB\n",
    "# os.listdir()\n",
    "# df = pd.read_csv('SelectedMetaDataForDnaSamples.tsv',sep='\\t')\n",
    "# df.columns = [item.replace('.','_') for item in df.columns]\n",
    "# df['Public_Patient_ID'] = ['ST-'+item.split('-')[1] for item in df.PublicSampleID]\n",
    "# df['Pri_met'] = [item.split('-')[-1] for item in df.PublicSampleID]\n",
    "# d_tum = {'Primary':['T',  'F'],# \n",
    "#          'PrimaryT2':['T2' ],\n",
    "#          'Met':['M','M2',]}\n",
    "# ls_add = ['PercentTumorPathologistEstimate',\n",
    "#           'Tumor_Mutation_Burden','Microsatellite_Instability']\n",
    "# for s_type, ls_types in d_tum.items():\n",
    "#     print(s_type)\n",
    "#     df_tum = df[df.Pri_met.isin(ls_types)].copy()\n",
    "#     print(df_tum.Public_Patient_ID.duplicated().sum())\n",
    "#     df_tum.rename(dict(zip(ls_add, [item + f'_{s_type}' for item in ls_add])),axis=1,inplace=True)\n",
    "#     df_tum_merge = df_tum.loc[:,['Public_Patient_ID']+[item + f'_{s_type}' for item in ls_add]]\n",
    "#     df_patient = df_patient.merge(df_tum_merge,on='Public_Patient_ID',\n",
    "#                 how='left', suffixes=('',f'_x'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c3482f6-304c-4a94-bd09-7f3674413c55",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# sorted(df_patient.columns.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb91ae66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# s_out = '20230105_Patient_Metadata.csv'#'20230921_Patient_Metadata.csv'\n",
    "\n",
    "# df_patient.Stage = df_patient.Stage.replace({'I':1,'II':2,'III':3,'IV':4,'0':0}).astype('Int64')\n",
    "# df_patient.Grade = df_patient.Grade.astype('Int64')\n",
    "# df_patient['Survival'] = df_patient.loc[:,'Vital Status at FU'].replace({'Alive':0,'Dead':1})\n",
    "# ls_drop_columns = ['Public_Specimen_ID','OPTR',\n",
    "#                    'Lung Met Present in Patient blood','Liver Met Present in Patient blood',\n",
    "#     'Liver Met Present in Patient tumor','Lung Met Present in Patient tumor',\n",
    "#     'cDays from Diagnosis to FU','cDays from Earliest Recur to FU','cDays from Resection to FU',\n",
    "#     'cDays from Resection to Recurrence','cLiverMet','cLongTermRecurrer','cLongTermSurvivor',\n",
    "#     'cLungMet','cNeoadjuvant Treatment','cRapidRecurrer','cVitalStatus'] + df_patient.columns[df_patient.columns.str.contains('_x')].tolist() + df_patient.columns[df_patient.columns.str.contains('OPTR')].tolist()\n",
    "# se_col = df_patient.loc[:,~df_patient.columns.isin(ls_drop_columns)].columns\n",
    "# ls_col = se_col[~se_col.str.contains('Altered')].tolist() + se_col[se_col.str.contains('Altered')].tolist()\n",
    "# if not os.path.exists(s_out):\n",
    "#     print('saving')\n",
    "#     df_patient.loc[:,ls_col].to_csv(s_out)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11776f34-4f75-4ee7-baee-880ba917c2e3",
   "metadata": {},
   "source": [
    "## Primary versus Met  <a name=\"primet\"></a>\n",
    "\n",
    "plots\n",
    "\n",
    "met resection\n",
    "\n",
    "RNA, DNA, TCR datasets\n",
    "\n",
    "\n",
    "[contents](#contents)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01bfb10f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "# s_out = '20240105_Patient_Metadata.csv'#'20230921_Patient_Metadata.csv'\n",
    "# df_patient = pd.read_csv(f'annotation/{s_out}',index_col=0)\n",
    "# which patients are in which dataset\n",
    "df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "                          index_col=0)\n",
    "#load gsva\n",
    "df_gsva = pd.read_csv('data/results_GSVA_3_mets.csv',index_col=0)\n",
    "df_gsva = df_gsva[df_gsva.Public_Specimen_ID.notna()]\n",
    "df_gsva['Public_Patient_ID'] = [\"ST-\"+ item.split(\"-\")[1] for item in df_gsva.Public_Specimen_ID] \n",
    "df_gsva['dash_end'] = [item.split(\"-\")[-1] for item in df_gsva.Public_Specimen_ID] \n",
    "b_met = [item.find('M') > -1 for item in df_gsva['dash_end']]\n",
    "df_gsva['Specimen_Type'] = pd.NA\n",
    "df_gsva.loc[b_met,'Specimen_Type'] = 'Met'\n",
    "df_gsva.loc[:,'Specimen_Type'].fillna('Primary',inplace=True)\n",
    "\n",
    "sns.set_palette(\"tab10\")\n",
    "ls_site = [#['pORG_0.2_Primary', 'pORG_0.2_Met'],\n",
    "           ['pORG_allPrimary', 'pORG_allMet'],\n",
    "['pSUB_Primary', 'pSUB_allMet'],\n",
    " ['PurIST_Primary', 'PurIST_Met'] ]\n",
    "df_out = pd.DataFrame()\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    for tu_site in ls_site:\n",
    "        df_test = df_patient.loc[:,['Public_Patient_ID']+tu_site].dropna()\n",
    "        s_value = tu_site[0].replace(f\"_{tu_site[0].split('_')[-1]}\",'')\n",
    "        df_test.set_index('Public_Patient_ID',inplace=True)#.stack()\n",
    "        df_test.columns = [item.split('_')[-1].replace('all','') for item in df_test.columns]\n",
    "        df_long = df_test.stack().reset_index().rename({'level_1':'Specimen_Type',0:s_value},axis=1)\n",
    "        df_long['Met_Site'] = df_long.Public_Patient_ID.map(dict(zip(df_patient.Public_Patient_ID,df_patient.RNA_DNA_Specimen_Site_Met)))\n",
    "        df_long_gsva = df_long.merge(df_gsva,on=\"Public_Patient_ID\",how='left')\n",
    "        #'''#by site\n",
    "        fig,ax=plt.subplots(figsize=(3,2),dpi=300)\n",
    "        sns.pointplot(data=df_long,x='Specimen_Type',y=s_value,hue='Met_Site',\n",
    "                      dodge=0.2,ax=ax,alpha=0.5,linestyles='-.',linewidth=1)\n",
    "        ax.legend(bbox_to_anchor=(1,1.1),title='Met_Site',fontsize='x-small')\n",
    "        ax.set_title(s_value)\n",
    "        ax.set_ylabel('')\n",
    "        ax.set_xlabel(f'N={df_long.Public_Patient_ID.nunique()}')\n",
    "        plt.tight_layout()\n",
    "        fig.savefig(f'figures/Ext2_F_{s_value}.pdf')\n",
    "        # average\n",
    "        #\n",
    "        # fig,ax=plt.subplots(figsize=(2,2),dpi=300)\n",
    "        # sns.pointplot(data=df_long,x='Specimen_Type',y=s_value,ax=ax)\n",
    "        # ax.set_xlabel('')\n",
    "        # ax.set_title(ax.get_ylabel())\n",
    "        # ax.set_ylabel('')\n",
    "        # plt.tight_layout()\n",
    "        #by patient\n",
    "        fig,ax=plt.subplots(figsize=(3,2),dpi=300)\n",
    "        sns.pointplot(data=df_long,x='Specimen_Type',y=s_value,hue=\"Public_Patient_ID\",ax=ax)#color=\"C1\"\n",
    "        ax.set_xlabel(f'N={df_long.Public_Patient_ID.nunique()}')\n",
    "        statistic, pvalue = stats.wilcoxon(x=df_test.Primary, y=df_test.Met)\n",
    "        ax.set_title(f'{ax.get_ylabel()} P={pvalue:.2}')\n",
    "        ax.set_ylabel('')\n",
    "        ax.legend(bbox_to_anchor=(1,1.1),title='Patient',fontsize='x-small')\n",
    "        plt.tight_layout()\n",
    "        fig.savefig(f'figures/Ext2_E_{s_value}.pdf')\n",
    "        ##fig,ax=plt.subplots(figsize=(3.5,2.1),dpi=300)\n",
    "        # # # i need met gsvas #whatt\n",
    "        # y=\"HALLMARK_MYC_TARGETS_V1\"\n",
    "        # for s_type in ['Primary','Met']:\n",
    "        #     df_gsva_type = df_gsva[df_gsva.Specimen_Type==s_type]\n",
    "        #     d_y = dict(zip(df_gsva_type.loc[:,'Public_Patient_ID'],df_gsva_type.loc[:,y]))\n",
    "        #     df_long.loc[df_long.Specimen_Type==s_type,y]=df_long.loc[df_long.Specimen_Type==s_type].Public_Patient_ID.map(d_y)\n",
    "\n",
    "        # #sns.pointplot(data=df_long_gsva,x='Specimen_Type',y=s_value,hue=\"HALLMARK_MYC_TARGETS_V1\",color=\"C1\",ax=ax)\n",
    "        # #sns.pointplot(data=df_long,x=s_value,hue=\"Public_Patient_ID\",y=\"HALLMARK_MYC_TARGETS_V1\",ax=ax)\n",
    "        # ax.set_title(f'{ax.get_ylabel()} p={pvalue:.2}')\n",
    "        # ax.set_ylabel('')\n",
    "        # ax.legend(bbox_to_anchor=(1,1.1),title='Patient',fontsize='small')\n",
    "        # plt.tight_layout()#'''\n",
    "        #by cohort\n",
    "        df_long['Cohort'] = df_long.Public_Patient_ID.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Cohort)))\n",
    "        fig,ax=plt.subplots(figsize=(2,3),dpi=300)\n",
    "        sns.pointplot(data=df_long[df_long.Cohort!='nan'],x='Specimen_Type',y=s_value,hue='Cohort',\n",
    "                      ax=ax,palette=d_colorblind,dodge=0.1)\n",
    "        h, l = ax.get_legend_handles_labels()\n",
    "        ax.legend(handles=h,labels=[f'Liver{linebreak}(N=3)',f'Lung{linebreak}N=(4)'],\n",
    "                  bbox_to_anchor=(.4,-.2),title='Cohort',fontsize='small',ncol=2,loc='upper center')\n",
    "        ax.set_title(s_value)\n",
    "        ax.set_ylabel('')\n",
    "        ax.set_xlabel('')\n",
    "        plt.tight_layout()\n",
    "        fig.savefig(f'figures/Fig2_D_{s_value}.pdf')\n",
    "        d_fig2.update({f'Fig2_D_{s_value}':df_long[df_long.Cohort!='nan']})\n",
    "        df_out = pd.concat([df_out,df_long],axis=1)\n",
    "        #break#\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ec7a0a9-2064-4617-8d7f-1d74f5069570",
   "metadata": {},
   "outputs": [],
   "source": [
    "def excel_dict_from_csvs(s_find,s_prefix='ED_Fig2',s_dir='.'):\n",
    "    df = pd.DataFrame(sorted(os.listdir(s_dir)),columns=['Test'])\n",
    "    df_test = df[df.Test.str.contains(s_find)].copy()\n",
    "    df_test['Panel'] = [item.split(s_find)[1].split('.csv')[0] for item in df_test.Test]\n",
    "    df_test['Sheet'] = [f'{s_prefix}{item}' for item in df_test.Panel]\n",
    "    d_fig = {}\n",
    "    for ids, row in df_test.iterrows():\n",
    "        print(row.Test)\n",
    "        df_sheet = pd.read_csv(row.Test,index_col=0)\n",
    "        d_fig.update({row.Sheet:df_sheet})\n",
    "    return(d_fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd462b96-a845-4378-9555-002b05e803bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3732f783-d9b6-470a-90bd-5b29438a249e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# s_find = 'Source_Extended_Data_Figure2'\n",
    "# d_fig = excel_dict_from_csvs(s_find)\n",
    "# util.open_write_excel(d_fig,filename='Source_Extended_Data_Figure2.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c918d985-8187-424a-a193-9c653c42bb33",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_out.loc[:,~df_out.columns.duplicated()].to_csv(f'Source_Extended_Data_Figure2_G-H.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a2af7ec-9032-4321-8080-ccfea0dcec9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# d_fig2.keys()\n",
    "# util.open_write_excel(d_fig2,filename='Source_Data_Figure2.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d722f55-7ab8-457d-ab59-1b916e7cd479",
   "metadata": {},
   "outputs": [],
   "source": [
    "# which patients are in which dataset\n",
    "df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "                          index_col=0)\n",
    "df_tcr = pd.read_csv('Supplemental_Data/Supplemental_Dataset_8.csv',\n",
    "                          index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4674dbc-5754-43ec-8ee7-b932a2c1c23d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#plot pORG scores versus metasectomy\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    df_patient.columns[df_patient.columns.str.contains('pORG')]\n",
    "    df_patient['pORG_All'] = df_patient.pORG_allPrimary.fillna(df_patient.pORG_Met)\n",
    "    df_patient['pORG_All'] = df_patient.pORG_allPrimary.fillna(df_patient.pORG_allMet)\n",
    "    ls_col = ['Public_Patient_ID','pORG_allPrimary','pORG_allMet','Days from Resection to Recurrence',\n",
    "              'Days from Resection to FU','Resected','Vital Status at FU']\n",
    "    #df_patient.loc[df_patient.Public_Patient_ID.isin(se_lung_pub),ls_col]#.to_csv('Lung_meta.csv')\n",
    "    #df_patient.loc[df_patient.Public_Patient_ID.isin(se_liver),ls_col]#.to_csv('Liver_meta.csv')\n",
    "    \n",
    "    importlib.reload(util)\n",
    "    ls_groups = ['Liver','Liver_Met_Res', 'Lung', 'Lung_Met_Res']\n",
    "    for y in ['pORG_All','pORG_Primary','pORG_Met']:\n",
    "        plotting = {\"data\": df_patient,\"x\":\"Cohort_Met_Resection\",\"y\":y,\"order\":ls_groups}\n",
    "        util.annotated_stripplot(plotting,ls_groups,y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b427431a",
   "metadata": {},
   "outputs": [],
   "source": [
    "se_match = {'ST-00004898',\n",
    " 'ST-00006291',\n",
    " 'ST-00007307',\n",
    " 'ST-00010984',\n",
    " 'ST-00014524',\n",
    " 'ST-00015839',\n",
    " 'ST-00017804',\n",
    " 'ST-00017838',\n",
    " 'ST-00019601'}\n",
    "df_match = df_patient.loc[df_patient.Public_Patient_ID.isin(se_match),(df_patient.columns.str.contains('Altered_Primary')) | (df_patient.columns.str.contains('Altered_Met'))].copy()\n",
    "df_match = df_match.astype('Int64')#\n",
    "df_match = df_match.replace({True:1,False:0})\n",
    "ls_gene = sorted(set([item.split('_')[0] for item in df_match.columns]))\n",
    "df_match['Public_Patient_ID'] = df_patient.loc[df_patient.Public_Patient_ID.isin(se_match),'Public_Patient_ID']\n",
    "\n",
    "# df_match = df_match.replace(np.nan,pd.NA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "366c3acd-37ce-4adb-9ddf-cb69dd6cb60a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ls_genes = ['RNF43_Altered_Primary', 'RNF43_Altered_Met', 'RNF43_Altered', 'TP53_Altered_Primary', 'TP53_Altered_Met', \n",
    "            'TP53_Altered', 'GNAS_Altered_Primary', 'GNAS_Altered_Met', 'GNAS_Altered', 'KRAS_Altered_Primary', \n",
    "            'KRAS_Altered_Met', 'KRAS_Altered', 'APOB_Altered_Primary', 'APOB_Altered_Met', 'APOB_Altered', \n",
    "            'SMAD4_Altered_Primary', 'SMAD4_Altered_Met', 'SMAD4_Altered', 'BRCA1_Altered_Primary', 'BRCA1_Altered', \n",
    "            'CDKN2A_Altered_Primary', 'CDKN2A_Altered_Met', 'CDKN2A_Altered', 'NOTCH1_Altered_Primary', 'NOTCH1_Altered_Met',\n",
    "            'NOTCH1_Altered', 'BCOR_Altered_Primary', 'BCOR_Altered_Met', 'BCOR_Altered', 'STK11_Altered_Primary', \n",
    "            'STK11_Altered_Met', 'STK11_Altered', 'TGFBR2_Altered_Primary', 'TGFBR2_Altered_Met', 'TGFBR2_Altered', \n",
    "            'ARID1B_Altered_Primary', 'ARID1B_Altered_Met', 'ARID1B_Altered', 'TSC2_Altered_Primary', 'TSC2_Altered_Met', \n",
    "            'TSC2_Altered', 'ATRX_Altered_Primary', 'ATRX_Altered_Met', 'ATRX_Altered', 'KDM6A_Altered_Primary', \n",
    "            'KDM6A_Altered_Met', 'KDM6A_Altered', 'ATM_Altered_Primary', 'ATM_Altered_Met', 'ATM_Altered',\n",
    "            'LRP1B_Altered_Primary', 'LRP1B_Altered_Met', 'LRP1B_Altered', 'ARID1A_Altered_Primary', 'ARID1A_Altered_Met',\n",
    "            'ARID1A_Altered', 'CUX1_Altered_Primary', 'CUX1_Altered_Met', 'CUX1_Altered', 'KMT2A_Altered_Primary',\n",
    "            'KMT2A_Altered_Met', 'KMT2A_Altered', 'KMT2D_Altered_Primary', 'KMT2D_Altered_Met', 'KMT2D_Altered', \n",
    "            'PBRM1_Altered_Primary', 'PBRM1_Altered_Met', 'PBRM1_Altered', 'BRCA2_Altered_Primary', 'BRCA2_Altered_Met',\n",
    "            'BRCA2_Altered', 'GATA1_Altered_Primary', 'GATA1_Altered_Met', 'GATA1_Altered', 'KDM5C_Altered_Primary', \n",
    "            'KDM5C_Altered_Met', 'KDM5C_Altered', 'RBM10_Altered_Primary', 'RBM10_Altered_Met', 'RBM10_Altered',\n",
    "            'SMARCB1_Altered_Primary', 'SMARCB1_Altered_Met', 'SMARCB1_Altered', 'ELF3_Altered_Primary', 'ELF3_Altered_Met', \n",
    "            'ELF3_Altered', 'MTAP_Altered_Primary', 'MTAP_Altered_Met', 'MTAP_Altered', 'BRAF_Altered_Primary',\n",
    "            'BRAF_Altered_Met', 'BRAF_Altered', 'CREBBP_Altered_Primary', 'CREBBP_Altered_Met', 'CREBBP_Altered',\n",
    "            'CDKN2B_Altered_Primary', 'CDKN2B_Altered_Met', 'CDKN2B_Altered', 'ARID1aLOF_Altered']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "678d5546",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# genes\n",
    "%matplotlib inline\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "ls_all = []\n",
    "for s_gene in ls_gene:    \n",
    "    tu_site = [f'{s_gene}_Altered_Primary',f'{s_gene}_Altered_Met']\n",
    "    try:\n",
    "        df_test = df_match.loc[:,['Public_Patient_ID']+tu_site].dropna(how='any')\n",
    "        df_test.set_index('Public_Patient_ID',inplace=True)\n",
    "        df_test = df_test.astype('int')\n",
    "        if (df_test>0).any().any(): #df_test.mean(numeric_only=True).mean(numeric_only=True) > 0:\n",
    "            ls_all = ls_all + tu_site\n",
    "            s_value = tu_site[0].replace(f\"_{tu_site[0].split('_')[-1]}\",'')\n",
    "            #.stack()\n",
    "            df_test.columns = [item.split('_')[-1].replace('all','') for item in df_test.columns]\n",
    "            df_long = df_test.stack().reset_index().rename({'level_1':'Specimen_Type',0:s_value},axis=1)\n",
    "            df_long['Met_Site'] = df_long.Public_Patient_ID.map(dict(zip(df_patient.Public_Patient_ID,df_patient.RNA_DNA_Specimen_Site_Met)))\n",
    "            # average (rosie says dont need)\n",
    "            # stat, pvalue = scipy.stats.fisher_exact(pd.crosstab(df_long.Specimen_Type, df_long.loc[:,s_value]))\n",
    "            # fig,ax=plt.subplots(figsize=(2,2),dpi=200)\n",
    "            # sns.pointplot(data=df_long,x='Specimen_Type',y=s_value,ax=ax,errorbar=None)\n",
    "            # ax.set_xlabel('')\n",
    "            # ax.set_title(f\"{s_value.replace('_',' ')}\\nFisher's p={pvalue:.2}\",fontsize=10)\n",
    "            # ax.set_ylabel(f'Fraction Pts n={df_long.Public_Patient_ID.nunique()}')\n",
    "            # ax.set_ylim(-0.1,1.1)\n",
    "            # plt.tight_layout()\n",
    "            # fig.savefig(f'figures/Fraction_{s_value}_Primary_met_matched.png')\n",
    "            # if pvalue < 1.1:\n",
    "            stat, pvalue = statsmodels.sandbox.stats.runs.mcnemar(x=df_test.iloc[:,0], y=df_test.iloc[:,1], exact=True)\n",
    "            #stat, pvalue = scipy.stats.wilcoxon(x=df_test.iloc[:,0], y=df_test.iloc[:,1],)\n",
    "            # result = scipy.stats.binomtest(k=sum(df_test.iloc[:,0]!=df_test.iloc[:,1]),\n",
    "            #                                      n=len(df_test), p=0, alternative='greater')\n",
    "            # pvalue = result.pvalue\n",
    "            #by patient\n",
    "            fig,ax=plt.subplots(figsize=(3,2),dpi=300)\n",
    "            sns.pointplot(data=df_long,x='Specimen_Type',y=s_value,hue='Public_Patient_ID',ax=ax,alpha=0.7,\n",
    "                          linestyles='-.',dodge=.4)\n",
    "            ax.set_xlabel('')\n",
    "            ax.set_title(f\"{s_value.replace('_',' ')}\\nMcNemar's P={pvalue:.2} N={df_long.Public_Patient_ID.nunique()}\",fontsize=10)\n",
    "            ax.set_ylabel('')\n",
    "            ax.set_yticklabels('')\n",
    "            ax.legend(bbox_to_anchor=(1,1.3),title='Patient',fontsize='x-small')\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/Mcnemar_primary_met_{s_gene}.pdf')    \n",
    "\n",
    "        else:\n",
    "            print(f'{s_gene} not altered')\n",
    "    except:\n",
    "        print(f'{s_gene} failed')\n",
    "\n",
    "# alterations primary to met\n",
    "df_pri = df_match.loc[:,ls_all].loc[:,::2]\n",
    "df_pri.columns = [item.split('_')[0] for item in df_pri.columns]\n",
    "df_met = df_match.loc[:,ls_all].iloc[:,1::2]\n",
    "df_met.columns = [item.split('_')[0] for item in df_met.columns]\n",
    "fig,ax=plt.subplots(dpi=300,figsize=(3,2))\n",
    "se_diff = (df_pri != df_met).sum(axis=1)\n",
    "se_diff.plot(kind='hist',ax=ax)\n",
    "ax.set_title(f'Matched Primary vs. Met\\nMean changes={se_diff.mean():.2}')\n",
    "ax.set_xlabel('No. Alterations Different')\n",
    "ax.set_ylabel('No. Alterations Different')\n",
    "plt.tight_layout()\n",
    "fig.savefig(f'figures/hist_matched_primary_met_gene_alt.pdf') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad437c00-9139-4552-9249-2ffe401487c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #se_diff.sum()\n",
    "# s_pub = 'Public_Patient_ID'\n",
    "# df_match.loc[:,[s_pub] + ls_all].to_csv(f'Source_Extended_Data_Figure4_F-G.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c825425b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# genes - not matched - chort and primary versus met fishers exact\n",
    "\n",
    "ls_pvalue = []\n",
    "ls_gene_correct = []\n",
    "dd_fdr = {'Specimen_Type':{'AMER1': 0.8853937569128851,'APOB': 0.89561556844377,'ARID1A': 0.8853937569128851,'ARID1B': 0.5299424678780686,\n",
    " 'ATRX': 0.89561556844377,'BCOR': 0.89561556844377,'BRCA2': 1.0,'CDKN2A': 0.5299424678780686,'CDKN2B': 0.349877752253849,\n",
    " 'CUX1': 0.89561556844377,'GATA1': 0.349877752253849,'GNAS': 0.561191957075014,'IRS2': 0.561191957075014,'KDM5C': 0.5299424678780686,\n",
    " 'KDM6A': 1.0,'KMT2D': 0.8183021772923266, 'KRAS': 0.89561556844377,'LRP1B': 1.0,\n",
    " 'MTAP': 0.06251890221846075,'NOTCH1': 0.89561556844377,'NOTCH3': 0.8853937569128851,'RBM10': 0.561191957075014,\n",
    " 'RNF43': 1.0,'SMAD4': 0.8853937569128851,'SMARCB1': 0.06392308434351372,'TGFBR2': 0.89561556844377,'TP53': 0.89561556844377,\n",
    " 'TSC2': 0.8674093953492792}}# michael\n",
    "dd_fdr= {'Specimen_Type':{'KDM6A': 0.8366401256596958,'RNF43': 0.8366401256596958,'TGFBR2': 0.8667385017597867,'ARID1B': 0.9101123704474682,\n",
    " 'KMT2A': 0.7387547787208116,'KRAS': 1.0,'ATM': 0.8667385017597867,'GATA1': 0.11746700480125209,\n",
    " 'BRAF': 0.7387547787208116,'CREBBP': 0.7387547787208116,'ATRX': 0.9101123704474682,'CUX1': 0.9246912476384586,\n",
    " 'GNAS': 0.8366401256596958,'BCOR': 1.0,'SMARCB1': 0.11746700480125209,'ELF3': 0.9101123704474682,\n",
    " 'MTAP': 0.07677475369566632,'APOB': 0.7387547787208116,\n",
    " 'ARID1A': 0.8366401256596958, 'CDKN2A': 0.8078087242382573,'NOTCH1': 1.0,\n",
    " 'KMT2D': 0.7387547787208116,'STK11': 0.7397789776431964,\n",
    " 'CDKN2B': 0.5248250565713328,'SMAD4': 0.8366401256596958,'PBRM1': 0.6918495424837988,\n",
    " 'TP53': 0.8366401256596958,'TSC2': 0.8366401256596958,'BRCA2': 1.0,\n",
    " 'LRP1B': 0.8366401256596958,'KDM5C': 0.11746700480125209,'RBM10': 0.43577065365578377},\n",
    " 'Cohort':{'ARID1A': 1.0,'SMAD4': 1.0,'NOTCH1': 1.0,'BCOR': 1.0,'ELF3': 1.0,'GATA1': 0.8990812819973347,\n",
    " 'KDM5C': 0.8990812819973347,'CDKN2A': 1.0,'KMT2D': 1.0,'PBRM1': 1.0,'TGFBR2': 1.0,'ATRX': 1.0,\n",
    " 'CDKN2B': 1.0,'STK11': 1.0,'BRCA2': 1.0,'CUX1': 1.0,'APOB': 1.0,'RBM10': 1.0,'TP53': 0.8990812819973347,\n",
    " 'ARID1B': 1.0,'BRAF': 1.0,'ATM': 0.8990812819973347,'CREBBP': 1.0,'KDM6A': 1.0,'LRP1B': 1.0,\n",
    " 'GNAS': 0.8990812819973347,'MTAP': 1.0,'SMARCB1': 0.8990812819973347,'TSC2': 1.0,'KRAS': 0.8990812819973347,\n",
    " 'RNF43': 1.0,'KMT2A': 1.0}}#tempus\n",
    "ls_gene = set([item.split('_')[0] for item in ls_genes]) - {'BRCA1','ARID1aLOF'}\n",
    "for s_type in ['Cohort','Specimen_Type']:\n",
    "    d_fdr = dd_fdr[s_type]\n",
    "    df_result = pd.DataFrame()\n",
    "    for s_gene in ls_gene:\n",
    "        try:\n",
    "            tu_site = [f'{s_gene}_Altered_Primary',f'{s_gene}_Altered_Met']\n",
    "            df_test = df_patient.loc[:,['Public_Patient_ID']+tu_site].dropna(how='all')\n",
    "            s_value = tu_site[0].replace(f\"_{tu_site[0].split('_')[-1]}\",'')\n",
    "            df_test.set_index('Public_Patient_ID',inplace=True)#.stack()\n",
    "            df_test.columns = [item.split('_')[-1].replace('all','') for item in df_test.columns]\n",
    "            df_long = df_test.stack().reset_index().rename({'level_1':s_type,0:s_value},axis=1)\n",
    "            if s_type == 'Cohort':\n",
    "                s_value = f'{s_gene}_Altered'\n",
    "                df_long = df_patient.loc[:,['Public_Patient_ID',s_value,s_type]].dropna(how='all')\n",
    "            # average\n",
    "            stat, pvalue = scipy.stats.fisher_exact(pd.crosstab(df_long.loc[:,s_type], df_long.loc[:,s_value]))\n",
    "            fig,ax=plt.subplots(figsize=(2,2),dpi=200)\n",
    "            sns.pointplot(data=df_long,x=s_type,y=s_value,ax=ax,errorbar=None)\n",
    "            ax.set_xlabel('')\n",
    "            ax.set_title(f\"{s_value.replace('_',' ')}\\nFisher's P={pvalue:.2}\\nFDR={d_fdr[s_gene]:.2}\",fontsize=10)\n",
    "            ax.set_ylabel(f'Fraction Pts N={df_long[df_long.loc[:,s_type].notna()].Public_Patient_ID.nunique()}')\n",
    "            ax.set_ylim(-0.1,1.1)\n",
    "            plt.tight_layout()\n",
    "            if pvalue > 0.05:\n",
    "               plt.close(fig)\n",
    "            else:\n",
    "                fig.savefig(f'figures/Fraction_{s_value}_{s_type}_all.pdf')\n",
    "            ls_gene_correct.append(s_gene)\n",
    "            ls_pvalue.append(pvalue)\n",
    "            #update results\n",
    "            se_result = df_long.groupby(s_type).mean(numeric_only=True).unstack().unstack()\n",
    "            se_result.columns = [f'Percent_Altered_in_{item}' for item in se_result.columns]\n",
    "            se_result[\"pvalue (Fisher's Exact)\"] = pvalue\n",
    "            se_result['FDR'] = d_fdr[s_gene]\n",
    "            df_result = pd.concat([df_result,se_result])\n",
    "        except:\n",
    "            print(f'skipped {s_gene}')\n",
    "        df_result.sort_values(by=\"pvalue (Fisher's Exact)\").to_csv(f'results/Gene_Alteration_Fraction_{s_type}.csv')\n",
    "    #break\n",
    "# #correct pvalues\n",
    "__, corrected = statsmodels.stats.multitest.fdrcorrection(np.array(ls_pvalue))\n",
    "#dict(zip(ls_gene_correct,corrected))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e30fc84a-10b3-4a4c-9652-b3dc48773bf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_result.to_csv(f'Source_Extended_Data_Figure4_E.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e405b03-0750-4b80-b6bc-4f6d91ca7de9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_table.Big.dropna().unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdc14281-85bc-4d4d-9bce-851d60b734a5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# #load table1\n",
    "# df_table = pd.read_excel(f'annotation/Aug2024-demo characteristics-dk.xlsx')\n",
    "\n",
    "# # average\n",
    "# #stat, pvalue = scipy.stats.fisher_exact(pd.crosstab(df_long.loc[:,s_type], df_long.loc[:,s_value]))\n",
    "# lls_pts = [['Liver_Cohort','Lung_Cohort'],['High_pORG','Low_pORG']]\n",
    "# ls_bigs = ['Clinical Characteristics','Treated by Resection',\n",
    "#        'Histology Analysis Performed']\n",
    "# for idx,s_big in enumerate(ls_bigs):\n",
    "#     #print(s_big)\n",
    "#     df_test = df_table.loc[df_table.Big==idx].copy()\n",
    "#     for ls_pts in lls_pts:\n",
    "#         d_pts = {}\n",
    "#         for s_pt in ls_pts:\n",
    "#             #print(s_pt)\n",
    "#             n_pt = int(df_table.loc[df_table.Big==s_big,s_pt].iloc[0].split('N=')[1].split(' ')[0])\n",
    "#             d_pts.update({s_pt:n_pt})\n",
    "#         for s_group in df_test.Group.dropna().unique():\n",
    "#             df_group = df_test.loc[df_test.Group==s_group,['Type']+ls_pts].copy()\n",
    "#             df_group = df_group.set_index('Type')\n",
    "#             for s_type in df_group.index.dropna().unique():\n",
    "#                 se_yes = df_group.loc[s_type].astype(int)\n",
    "#                 se_no = pd.Series(d_pts)-df_group.loc[s_type].astype('int')\n",
    "#                 df_con = pd.concat([se_yes,se_no],axis=1)\n",
    "#                 df_con.columns = ['Yes','No']\n",
    "#                 stat, pvalue = scipy.stats.fisher_exact(df_con)\n",
    "#                 if pvalue < 0.05:\n",
    "#                     print(f'{s_group} {ls_pts} {s_type} P={pvalue:.03}')\n",
    "#                     print(df_con)\n",
    "#                     print(df_con.div(df_con.sum(axis=1).values,axis=0))\n",
    "#         #     break\n",
    "            \n",
    "#         # break\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c20f395-e5a5-45ae-bda9-c2dcc63bb602",
   "metadata": {},
   "source": [
    "## Gene alterations vs GSVA  <a name=\"genes\"></a>\n",
    "\n",
    "\n",
    "\n",
    "[contents](#contents)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cfb56e3-576f-48e6-9b05-e1490c84f5be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# which patients are in which dataset\n",
    "df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "                          index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b9bf0df-9a6c-4ba2-8967-e0cb0ae55efb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#new, altered vs wt; my gene level analysis\n",
    "d_fdr = {'pORG_Primary': {'ARID1A': 0.48617899229664635,'SMAD4': 0.09110462199827897,'NOTCH1': 0.269701902132145,'BCOR': 0.18980268443777357,\n",
    "  'ELF3': 0.068428196785162,'GATA1': 0.068428196785162,'KDM5C': 0.14332513046948206,'CDKN2A': 0.0005185416797667346,\n",
    "  'KMT2D': 0.7935733317589467,'PBRM1': 0.9978813766139409,'TGFBR2': 0.9570927059606049,'ATRX': 0.7935733317589467,\n",
    "  'CDKN2B': 0.068428196785162,'STK11': 0.7935733317589467,'BRCA2': 0.6065665462681668,\n",
    "  'CUX1': 0.14332513046948206,'APOB': 0.16323222392911738,'RBM10': 0.9978813766139409,\n",
    "  'TP53': 9.271095789895771e-13,'ARID1B': 0.7935733317589467,'BRAF': 0.6510263072121756,\n",
    "  'ATM': 0.14678185036349684,'CREBBP': 0.6656859632307156,'KDM6A': 0.9978813766139409,'LRP1B': 0.3853372172236763,\n",
    "  'GNAS': 0.6510263072121756,'MTAP': 0.25270395066023793,'SMARCB1': 0.7935733317589467,'TSC2': 0.7935733317589467,'KRAS': 0.09110462199827897,\n",
    "  'RNF43': 0.7935733317589467,'KMT2A': 0.6065665462681668},\n",
    " 'pSUB_Primary': {'ARID1A': 0.7642071598173099,'SMAD4': 0.7642071598173099,'NOTCH1': 0.7741553663674001,'BCOR': 0.7642071598173099,\n",
    "  'ELF3': 0.9261043757205508,'GATA1': 0.7104667208430149,'KDM5C': 0.7642071598173099,'CDKN2A': 0.9261043757205508,\n",
    "  'KMT2D': 0.8361484568451376,'PBRM1': 0.8361484568451376,'TGFBR2': 0.7642071598173099,'ATRX': 0.8361484568451376,\n",
    "  'CDKN2B': 0.8361484568451376,'STK11': 0.7642071598173099,'BRCA2': 0.5875349311964698,'CUX1': 0.7741553663674001,\n",
    "  'APOB': 0.9261043757205508,'RBM10': 0.8361484568451376,'TP53': 0.8220600811875246,'ARID1B': 0.7642071598173099,\n",
    "  'BRAF': 0.7642071598173099,'ATM': 0.8361484568451376,'CREBBP': 0.7642071598173099,'KDM6A': 0.8361484568451376,'LRP1B': 0.8220600811875246,\n",
    "  'GNAS': 0.7642071598173099,'MTAP': 0.7642071598173099,'SMARCB1': 0.9261043757205508,'TSC2': 0.5875349311964698,\n",
    "  'KRAS': 0.7642071598173099,'RNF43': 0.7642071598173099,'KMT2A': 0.7642071598173099},\n",
    " 'PurIST_Primary': {'ARID1A': 0.9209640579424575,'SMAD4': 0.8883973415028421,'NOTCH1': 0.6558250857380629,\n",
    "  'BCOR': 0.8542383497040145,'ELF3': 0.7866250472890529,'GATA1': 0.7866250472890529,'KDM5C': 0.7921369773641237,\n",
    "  'CDKN2A': 0.8542383497040145,'KMT2D': 0.8542383497040145,'PBRM1': 0.6558250857380629,'TGFBR2': 0.18734445150072912,\n",
    "  'ATRX': 0.8542383497040145,'CDKN2B': 0.8542383497040145,'STK11': 0.8542383497040145,'BRCA2': 0.6558250857380629,\n",
    "  'CUX1': 0.6806846818672203,'APOB': 0.7866250472890529,'RBM10': 0.8542383497040145,'TP53': 0.9209640579424575,\n",
    "  'ARID1B': 0.6558250857380629,'BRAF': 0.8542383497040145,'ATM': 0.21014498610428547,'CREBBP': 0.7866250472890529,\n",
    "  'KDM6A': 0.8542383497040145,'LRP1B': 0.9209640579424575,'GNAS': 0.18734445150072912,'MTAP': 0.8542383497040145,\n",
    "  'SMARCB1': 0.8542383497040145,'TSC2': 0.9209640579424575,'KRAS': 0.8542383497040145,'RNF43': 0.9396759291714212,'KMT2A': 0.8542383497040145},\n",
    "'pORG_Met': {'ARID1A': 0.7992223964064106,'SMAD4': 0.8158385100954274,'NOTCH1': 0.9566495471426233,\n",
    "  'BCOR': 0.8528607126790245,'ELF3': 0.8158385100954274,'GATA1': 0.8158385100954274,'KDM5C': 0.8158385100954274,\n",
    "  'CDKN2A': 0.7992223964064106,'KMT2D': 0.8528607126790245,'PBRM1': 0.8158385100954274,'TGFBR2': 0.953053484174483,\n",
    "  'ATRX': 0.953053484174483,'CDKN2B': 0.8158385100954274,'STK11': 0.8158385100954274,'BRCA2': 0.8174660409364153,\n",
    "  'CUX1': 0.8158385100954274,'APOB': 0.8158385100954274,'RBM10': 0.953053484174483,'TP53': 0.00048434216457797074,\n",
    "  'ARID1B': 0.9023526118594072,'BRAF': 0.8158385100954274,'ATM': 0.8158385100954274,'CREBBP': 0.9023526118594072,\n",
    "  'KDM6A': 0.8158385100954274,'LRP1B': 0.7264937773913535, 'GNAS': 0.953053484174483,'MTAP': 0.7992223964064106,\n",
    "  'SMARCB1': 0.8158385100954274,'TSC2': 0.8158385100954274,'KRAS': 0.8158385100954274,'RNF43': 0.7992223964064106,\n",
    "  'KMT2A': 0.8158385100954274},\n",
    "'PurIST_Met': {'ARID1A': 0.7380743812841982,'SMAD4': 0.9546982063202067,'NOTCH1': 0.31880329903072585,\n",
    "  'BCOR': 0.7380743812841982,'ELF3': 0.7380743812841982,'GATA1': 0.9660177083900605,'KDM5C': 0.9546982063202067,\n",
    "  'CDKN2A': 0.0190917574182068,'KMT2D': 0.9660177083900605,'PBRM1': 0.7380743812841982,'TGFBR2': 0.7380743812841982,\n",
    "  'ATRX': 0.7380743812841982,'CDKN2B': 0.0022142918259284894,'STK11': 0.39327017037707046,'BRCA2': 0.7380743812841982,\n",
    "  'CUX1': 0.12558426396976066,'APOB': 0.7380743812841982,'RBM10': 0.7380743812841982,'TP53': 0.7113905791758974,\n",
    "  'ARID1B': 0.7380743812841982,'BRAF': 0.39327017037707046, 'ATM': 0.7380743812841982,'CREBBP': 0.9546982063202067,\n",
    "  'KDM6A': 0.9546982063202067,'LRP1B': 0.9546982063202067,'GNAS': 0.7380743812841982,'MTAP': 0.014293658019648827,\n",
    "  'SMARCB1': 0.7402092085135089,'TSC2': 0.7380743812841982,'KRAS': 0.7380743812841982,'RNF43': 0.9660177083900605,\n",
    "  'KMT2A': 0.9546982063202067}}\n",
    "order = [1,0]\n",
    "sns.set_palette(pal_porg)\n",
    "df_patient['PurIST_All'] = df_patient.loc[:,'PurIST_Primary'].fillna(df_patient.loc[:,'PurIST_Met'])\n",
    "df_out = pd.DataFrame()\n",
    "import warnings\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    d_all = {}\n",
    "    d_all_p = {}\n",
    "    cutoff = 0.1\n",
    "    for s_gsva_name in ['pORG_Primary','pSUB_Primary','PurIST_Primary','pORG_Met','PurIST_Met']:\n",
    "    #for s_gsva_name in ['pORG_Met','PurIST_Met']: #'pSUB_Met',\n",
    "        print(s_gsva_name)\n",
    "        b_rna = df_patient.loc[:,s_gsva_name].notna()\n",
    "        ls_pvalue = []\n",
    "        ls_gene_correct = []\n",
    "        for s_col in ls_gene:\n",
    "            s_gene = f'{s_col}_Altered_Primary'\n",
    "            if s_gsva_name.find('_Met')>-1:\n",
    "                s_gene = f'{s_col}_Altered_Met'\n",
    "            try:\n",
    "                b_dna = df_patient.loc[:,s_gene].notna()\n",
    "            except:\n",
    "                continue\n",
    "            df_plot = df_patient.loc[b_rna & b_dna,['Public_Patient_ID',s_gene,s_gsva_name]].copy()\n",
    "            a = df_plot.loc[df_plot.loc[:,s_gene]==0,s_gsva_name]\n",
    "            b = df_plot.loc[df_plot.loc[:,s_gene]==1,s_gsva_name]\n",
    "            try:\n",
    "                stat, pvalue = scipy.stats.ttest_ind(a, b)\n",
    "            except:\n",
    "                    print('error')\n",
    "                    pvalue = 1\n",
    "            ls_gene_correct.append(s_col)\n",
    "            ls_pvalue.append(pvalue)\n",
    "            #plot\n",
    "            fig,ax=plt.subplots(figsize=(1.9,2.2),dpi=300)\n",
    "            sns.stripplot(data=df_plot,x=s_gene,y=s_gsva_name,ax=ax,order=order,s=2,alpha=0.8)\n",
    "            sns.boxplot(data=df_plot,x=s_gene,y=s_gsva_name,ax=ax,order=order,\n",
    "                       whiskerprops={'visible': False},showcaps=False,\n",
    "                       #meanline=True,showmeans=True,medianprops={'visible': False},\n",
    "                       medianprops={'color': 'k', 'ls': '-', 'lw': 1},#meanprops\n",
    "                showfliers=False,showbox=False)\n",
    "            ax.set_ylabel(s_gsva_name.replace(\"_\",\" \"))\n",
    "            ax.set_xlabel(\"\")\n",
    "            if len(order)==2:\n",
    "               xlabels = [f'{order[0]}{linebreak}N={(df_plot.loc[:,s_gene].value_counts()[order[0]])}',\n",
    "                f'{order[1]}{linebreak}N={(df_plot.loc[:,s_gene].value_counts()[order[1]])}']\n",
    "               ax.set_xticklabels(xlabels)\n",
    "            ax.set_xticklabels([item.get_text().replace(f'1{linebreak}',f'Altered{linebreak}').replace(f'0{linebreak}',f'WT{linebreak}') for item in ax.get_xticklabels()])\n",
    "            try:\n",
    "                p_corr = d_fdr[s_gsva_name][s_col]\n",
    "                ax.set_title(f'{s_gene.replace(\"_Altered_\",\" \")}\\nFDR={p_corr:.2}',fontsize=10)\n",
    "                if p_corr > 0.1:\n",
    "                    plt.close(fig)\n",
    "                else:\n",
    "                   df_out = pd.concat([df_out,df_plot],axis=1)\n",
    "                   plt.tight_layout()\n",
    "                   fig.savefig(f'figures/Ttest_FDR_{s_col}_{s_gsva_name}.pdf')\n",
    "                if s_gsva_name == 'pORG_Primary':\n",
    "                    if s_col == 'TP53':\n",
    "                        d_fig2.update({f'Figure2_J_TP53':df_plot})\n",
    "                    elif s_col == 'CDKN2A':\n",
    "                        d_fig2.update({f'Figure2_J_CDKN2A':df_plot})\n",
    "                   #break\n",
    "            except:\n",
    "                ax.set_title(f'{s_gene.replace(\"_Altered_\",\" \")}\\nT-test p={pvalue:.2}',fontsize=10)\n",
    "                if pvalue > 0.05:\n",
    "                    plt.close(fig)\n",
    "                else:\n",
    "                   plt.tight_layout()\n",
    "                fig.savefig(f'figures/Ttest_{s_col}_{s_gsva_name}.pdf')\n",
    "            #break\n",
    "                    #correct pvalues\n",
    "            __, corrected = statsmodels.stats.multitest.fdrcorrection(np.array(ls_pvalue))\n",
    "            d_pval = dict(zip(ls_gene_correct,ls_pvalue))\n",
    "            d_all_p.update({f'{s_gsva_name}':d_pval})\n",
    "            d_qval = dict(zip(ls_gene_correct,corrected))\n",
    "            d_all.update({f'{s_gsva_name}':d_qval})\n",
    "            \n",
    "        \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "292b6178-f174-4499-8d8c-1a66d2bf97b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# util.open_write_excel(d_fig2,filename='Source_Data_Figure2.xlsx')\n",
    "#df_out.loc[:,~df_out.columns.duplicated()].to_csv(f'Source_Extended_Data_Figure4_A-C.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78230565-e23f-4956-aa5d-4f12d0e729dd",
   "metadata": {},
   "source": [
    "## Tumor mutation burden vs GSVA, Cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "701b3f8b-a719-4b64-91e9-a4d4293b41ad",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# # #pORG primary versus TMB\n",
    "# d_fdr = {'pORG_Primary':{'KRAS' : 0.38287593347963944,\n",
    "# 'GATA1' : 0.1855045586524062,\n",
    "# 'CDKN2A' : 0.06609676055955013,\n",
    "# 'TP53' : 1.038674992908113e-09,\n",
    "# 'KDM5C' : 0.4251464077159879,},\n",
    "#         'pSUB_Primary':{'TSC2' : 1.0,},'Cohort':{'ATM' : 0.8990812819973347,\n",
    "# 'GATA1' : 0.8990812819973347,},\n",
    "#         'PurIST_Primary':{'TGFBR2' : 0.284384813956165,\n",
    "# 'ATM' : 0.284384813956165,\n",
    "# 'GNAS' : 0.284384813956165,}}\n",
    "# se_normal = {'ST-00007288', 'ST-00020154', 'ST-00007313', 'ST-00021102', 'ST-00017838', 'ST-00020118', 'ST-00007183', 'ST-00015332', 'ST-00017966', 'ST-00020119', 'ST-00020102', 'ST-00013013', 'ST-00007168', 'ST-00007195', 'ST-00007067', 'ST-00011356', 'ST-00018260', 'ST-00007270', 'ST-00020092', 'ST-00012931', 'ST-00020105', 'ST-00020117', 'ST-00018988', 'ST-00020123', 'ST-00010984', 'ST-00019171', 'ST-00007180', 'ST-00014550', 'ST-00007595', 'ST-00020098', 'ST-00013861', 'ST-00011343', 'ST-00018179', 'ST-00015973', 'ST-00020213', 'ST-00020175', 'ST-00020168', 'ST-00013291', 'ST-00020149', 'ST-00020190', 'ST-00007279', 'ST-00006291', 'ST-00017310', 'ST-00020070', 'ST-00008043', 'ST-00020181', 'ST-00007303', 'ST-00007302', 'ST-00007307', 'ST-00017381', 'ST-00017762', 'ST-00020215', 'ST-00017340', 'ST-00018893', 'ST-00020085', 'ST-00012327', 'ST-00006650', 'ST-00013158', 'ST-00007178', 'ST-00020127', 'ST-00010924', 'ST-00007298', 'ST-00018525', 'ST-00013084', 'ST-00020077', 'ST-00013253', 'ST-00017297', 'ST-00020232', 'ST-00020205', 'ST-00007389', 'ST-00007585', 'ST-00018301', 'ST-00020236', 'ST-00020207', 'ST-00014391', 'ST-00007257', 'ST-00015839', 'ST-00007239', 'ST-00019368', 'ST-00019466', 'ST-00018002', 'ST-00016623', 'ST-00020069', 'ST-00020148', 'ST-00020164', 'ST-00012967', 'ST-00006625', 'ST-00020195', 'ST-00012742', 'ST-00018269', 'ST-00014524', 'ST-00019043', 'ST-00020285', 'ST-00012818', 'ST-00018584', 'ST-00007071', 'ST-00013526', 'ST-00014104', 'ST-00020221', 'ST-00020281', 'ST-00007146', 'ST-00020223', 'ST-00014575', 'ST-00020144', 'ST-00020824', 'ST-00018268', 'ST-00017227', 'ST-00007165', 'ST-00017549', 'ST-00007266', 'ST-00020203', 'ST-00021351', 'ST-00013091', 'ST-00018236', 'ST-00003662', 'ST-00020153', 'ST-00007208', 'ST-00020057', 'ST-00015184', 'ST-00017945', 'ST-00020106', 'ST-00020126', 'ST-00007174', 'ST-00007213', 'ST-00019367', 'ST-00020341', 'ST-00020291', 'ST-00017261', 'ST-00020064', 'ST-00012908', 'ST-00007284', 'ST-00017804', 'ST-00019682', 'ST-00020209', 'ST-00020308', 'ST-00020112', 'ST-00007234', 'ST-00020323', 'ST-00008056', 'ST-00016028', 'ST-00010919', 'ST-00019056', 'ST-00020165', 'ST-00011168', 'ST-00018362', 'ST-00012890', 'ST-00020206', 'ST-00013415', 'ST-00012975', 'ST-00017080', 'ST-00007380', 'ST-00020225', 'ST-00013971', 'ST-00007296', 'ST-00014023', 'ST-00007099', 'ST-00007381', 'ST-00007216', 'ST-00007300', 'ST-00009790', 'ST-00019945', 'ST-00020217', 'ST-00007120', 'ST-00017449', 'ST-00016968', 'ST-00018136', 'ST-00020218', 'ST-00020131', 'ST-00017985', 'ST-00020109', 'ST-00010889', 'ST-00008601', 'ST-00015136', 'ST-00020185', 'ST-00007590', 'ST-00018628', 'ST-00012319', 'ST-00018257', 'ST-00018955', 'ST-00016289', 'ST-00005536', 'ST-00017074', 'ST-00018360', 'ST-00009788', 'ST-00020045', 'ST-00015777', 'ST-00017440', 'ST-00013160', 'ST-00015760', 'ST-00020093', 'ST-00003635', 'ST-00016499', 'ST-00016806', 'ST-00018232', 'ST-00018426', 'ST-00007613', 'ST-00015114', 'ST-00020089', 'ST-00018963', 'ST-00014537', 'ST-00007377', 'ST-00018932', 'ST-00016742', 'ST-00017041', 'ST-00020178', 'ST-00007402', 'ST-00007203', 'ST-00007382', 'ST-00020108', 'ST-00007085', 'ST-00015163', 'ST-00020196', 'ST-00007267', 'ST-00012878', 'ST-00020186', 'ST-00017915', 'ST-00019006', 'ST-00019598', 'ST-00019590', 'ST-00017975', 'ST-00019458', 'ST-00020133', 'ST-00019194', 'ST-00020251', 'ST-00016257', 'ST-00018616', 'ST-00017967', 'ST-00019601', 'ST-00017003', 'ST-00020208', 'ST-00020081', 'ST-00005813', 'ST-00020072', 'ST-00007175', 'ST-00016911'}\n",
    "\n",
    "\n",
    "# df_patient['PurIST_All'] = df_patient.loc[:,'PurIST_Primary'].fillna(df_patient.loc[:,'PurIST_Met'])\n",
    "# import warnings\n",
    "# with warnings.catch_warnings():\n",
    "#     warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "#     d_all = {}\n",
    "#     d_all_p = {}\n",
    "#     cutoff = 0.1\n",
    "#     for s_gsva_name in ['pORG','Cohort','pSUB','PurIST']:\n",
    "#         print(s_gsva_name)\n",
    "#         for s_type in ['Primary','Met',]: #'All',\n",
    "#             if s_gsva_name == 'Cohort':\n",
    "#                 s_gsva = 'Cohort'\n",
    "#             elif s_type == 'All' and s_gsva_name !='PurIST':\n",
    "#                 ls = df_patient.columns[(df_patient.columns.str.contains(s_gsva_name)) & (df_patient.columns.str.contains(f'_all'))]\n",
    "#                 s_gsva = f\"{ls[0].split('_all')[0]}_All\"\n",
    "#                 df_patient[s_gsva] = df_patient.loc[:,ls[0]].fillna(df_patient.loc[:,ls[1]])\n",
    "#             else:\n",
    "#                 s_gsva = df_patient.columns[(df_patient.columns.str.contains(s_gsva_name)) & (df_patient.columns.str.contains(f'_{s_type}'))][0]\n",
    "#             b_rna = df_patient.loc[:,s_gsva].notna()\n",
    "#             # define quartiles\n",
    "#             ls_pvalue = []\n",
    "#             ls_gene_correct = []\n",
    "#             for s_col in ls_gene:\n",
    "#                 s_gene = f'{s_col}_Altered'\n",
    "#                 try:\n",
    "#                     b_dna = df_patient.loc[:,s_gene].notna()\n",
    "#                 except:\n",
    "#                     continue\n",
    "#                 df_plot = df_patient.loc[b_rna & b_dna,['Public_Patient_ID',s_gene,s_gsva]].copy()\n",
    "#                 if not s_gsva == 'Cohort':\n",
    "#                     #df_plot = df_patient.loc[b_rna & b_dna,['Public_Patient_ID',s_gene,s_gsva]].copy()\n",
    "#                     i_quart = df_plot.loc[:,s_gsva].notna().sum()//4\n",
    "#                     se_low = df_plot.loc[:,['Public_Patient_ID',s_gsva]].sort_values(by=s_gsva)[0:i_quart].Public_Patient_ID\n",
    "#                     se_high = df_plot.loc[:,['Public_Patient_ID',s_gsva]].sort_values(by=s_gsva)[-i_quart::].Public_Patient_ID\n",
    "#                     #df_plot['quartiles'] = pd.qcut(df_plot.loc[:,s_gsva],q=4,labels=['low','medX','med','high'])\n",
    "#                     #df_long = df_plot.loc[~df_plot.quartiles.str.contains('med'),[s_gene,'quartiles']]\n",
    "#                     df_plot['quartiles'] = pd.NA\n",
    "#                     df_plot.loc[df_plot.Public_Patient_ID.isin(se_low),'quartiles'] = 'low'\n",
    "#                     df_plot.loc[df_plot.Public_Patient_ID.isin(se_high),'quartiles'] = 'high'\n",
    "#                     #print(df_plot.quartiles.value_counts())\n",
    "#                     df_long = df_plot.dropna()\n",
    "#                     break\n",
    "#             #pORG primary versus TMB\n",
    "#             df_high_low = df_patient.merge(df_plot.loc[:,['Public_Patient_ID','quartiles']],\n",
    "#                              on='Public_Patient_ID',how='inner').copy()\n",
    "#             df_high_low = df_high_low[df_high_low.quartiles.notna()]\n",
    "#             for b_match in [True,False]:\n",
    "#                 if b_match:\n",
    "#                     df = df_high_low[df_high_low.Public_Patient_ID.isin(se_normal)]\n",
    "#                 else:\n",
    "#                     df = df_high_low\n",
    "#                 ls_groups = ['low','high']\n",
    "#                 if s_type == 'Met':\n",
    "#                     y = 'Tumor_Mutation_Burden_Met'\n",
    "#                 elif s_type == 'Primary':\n",
    "#                     y = 'Tumor_Mutation_Burden_Primary'\n",
    "#                 plotting = {\"data\": df,\"x\":\"quartiles\",\"y\":y,\"order\":ls_groups}\n",
    "#                 fig,ax,test_results = util.annotated_stripplot(plotting,ls_groups,y)\n",
    "#                 for res in test_results:\n",
    "#                     pvalue=res.data.pvalue\n",
    "#                 if b_match:\n",
    "#                     ax.set_title(f'Samples w/ Matched Normal\\nTMB {y.split(\"_\")[-1]} vs {s_gsva_name}\\np={pvalue:.2} n={len(df)}')\n",
    "#                 else:\n",
    "#                     ax.set_title(f'TMB {y.split(\"_\")[-1]} vs {s_gsva_name}\\np={pvalue:.2} n={len(df)}')\n",
    "#         break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef69793d-000b-4bba-8993-82966808d4f4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# 7. cohort vs TMB\n",
    "# c. Authors suggest in 6B that an increase in shared clonal TCRs in the blood of liver cohort indicates \n",
    "#response to a limited set of tumor-initiating neoepitopes. This should be validated with comparison of TMB in each cohort.\n",
    " \n",
    "# And\n",
    "# 8. Figure 3E- DDR gene alteration+ tumors likely have high TMB- should be discussed if\n",
    "#there are differences in liver and lung TMB in primary or met that confound the observation here. \n",
    " \n",
    "# Can you graph differences in TMB between cohorts and check if DDR gene altered have higher TMB?\n",
    "ls_groups = ['Lung','Liver']\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    for b_match in [True,False]:\n",
    "        for y in ['Tumor_Mutation_Burden_Primary','Tumor_Mutation_Burden_Met']:\n",
    "            if b_match:\n",
    "                df = df_patient[(df_patient.Public_Patient_ID.isin(se_normal)) & (df_patient.loc[:,y].notna())]\n",
    "            else:\n",
    "                df = df_patient[df_patient.loc[:,y].notna()]\n",
    "            plotting = {\"data\": df,\"x\":\"Cohort\",\"y\":y,\"order\":ls_groups}\n",
    "            fig,ax,test_results = util.annotated_stripplot(plotting,ls_groups,y)\n",
    "            for res in test_results:\n",
    "                pvalue=res.data.pvalue\n",
    "            if b_match:\n",
    "                ax.set_title(f'Samples w/ Matched Normal\\nTMB {y.split(\"_\")[-1]} vs Cohort\\np={pvalue:.2} n={len(df)}')\n",
    "            else:    \n",
    "                ax.set_title(f'TMB {y.split(\"_\")[-1]} vs Cohort\\np={pvalue:.2} n={len(df)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7df327e-a675-4384-a7f5-e2e180c9260d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# d_fdr = {'pORG_Primary':{'KRAS' : 0.38287593347963944,\n",
    "# 'GATA1' : 0.1855045586524062,\n",
    "# 'CDKN2A' : 0.06609676055955013,\n",
    "# 'TP53' : 1.038674992908113e-09,\n",
    "# 'KDM5C' : 0.4251464077159879,},\n",
    "#         'pSUB_Primary':{'TSC2' : 1.0,},'Cohort':{'ATM' : 0.8990812819973347,\n",
    "# 'GATA1' : 0.8990812819973347,},\n",
    "#         'PurIST_Primary':{'TGFBR2' : 0.284384813956165,\n",
    "# 'ATM' : 0.284384813956165,\n",
    "# 'GNAS' : 0.284384813956165,}}\n",
    "# for big_key, big_item in d_all_p.items():\n",
    "#     print(big_key)\n",
    "#     for key, item in big_item.items():\n",
    "#         if item < 0.1:\n",
    "#             print(f\"'{key}' : {d_all[big_key][key]},\")\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8554d746-cb57-41df-878a-1e76f3ff92b9",
   "metadata": {},
   "source": [
    "## Venn diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec689f68-141f-4fc9-abe0-3a8ff0f0222a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "import venn\n",
    "df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "                          index_col=0)\n",
    "df_merge = pd.read_csv('Supplemental_Data/Supplemental_Dataset_8.csv',\n",
    "                          index_col=0)\n",
    "b_tcr_pts = (df_merge.Clonality_Blood.notna() | df_merge.Clonality_Tumor.notna())\n",
    "df = df_patient.merge(df_merge,on='Public_Patient_ID',how='left',suffixes=('','_x'))\n",
    "df_patient['PurIST_All'] = df_patient.loc[:,'PurIST_Primary'].fillna(df_patient.loc[:,'PurIST_Met'])\n",
    "d_pat = { #'CT':((df_patient.Cohort.notna()) | (df_patient.Recurrence_Sites_4 == 'Other_site')),\n",
    "         'RNA':df_patient.PurIST_All.notna(),\n",
    "         'DNA':df_patient.BRAF_Altered.notna(),\n",
    "         'TCR Tumor':df.Productive_Rearrangements.notna().values,\n",
    "         'TCR Blood':df.loc[:,\"Simpson's Evenness blood\"].notna().values,\n",
    "         #'TCR':(df.loc[:,\"Simpson's Evenness blood\"].notna()) | (df.Productive_Rearrangements.notna())\n",
    "        }\n",
    "s_col = 'Public_Patient_ID'\n",
    "d_sets = {}\n",
    "sets = {}\n",
    "for s_key, b_key in d_pat.items():\n",
    "        print(s_key)\n",
    "        print(b_key.sum())\n",
    "        d_sets.update({s_key:df_patient.loc[b_key,['Public_Patient_ID']]})\n",
    "        sets.update({s_key:set(df_patient.loc[b_key,'Public_Patient_ID'])})\n",
    "if len(d_pat) == 5:\n",
    "    labels = venn.get_labels([set(item.loc[:,s_col]) for key, item in d_sets.items()])\n",
    "    fig,ax = venn.venn5(labels, names=[key for key, item in d_sets.items()])#,ax=ax\n",
    "elif len(d_pat) == 4:\n",
    "    labels = venn.get_labels([set(item.loc[:,s_col]) for key, item in d_sets.items()])\n",
    "    fig,ax = venn.venn4(labels, names=[key for key, item in d_sets.items()])#,ax=ax\n",
    "elif len(d_pat) == 3:\n",
    "    labels = venn.get_labels([set(item.loc[:,s_col]) for key, item in d_sets.items()])\n",
    "    fig,ax = venn.venn3(labels, names=[key for key, item in d_sets.items()])#,ax=ax\n",
    "elif len(d_pat) == 2:\n",
    "    labels = venn.get_labels([set(item.loc[:,s_col]) for key, item in d_sets.items()])\n",
    "    fig,ax = venn.venn2(labels, names=[key for key, item in d_sets.items()])#,ax=ax\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8db21682-f8f0-4088-b4af-1479ce2e6f85",
   "metadata": {},
   "outputs": [],
   "source": [
    "from venny4py.venny4py import *\n",
    "fig = venny4py(sets=sets,size=6, dpi=600, out='figures', ext='pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6268c7b-c8d9-4b8e-aa7b-3df24887e110",
   "metadata": {},
   "source": [
    "# DDR vs TMB <a name=\"DDR\"></a>\n",
    "\n",
    "\n",
    "I will check if DDR gene altered have higher TMB.\n",
    "\n",
    "[contents](#contents)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef528e73-c7e1-4e53-b860-06635b02117a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "s_out = 'Supplemental_Dataset_1.xlsx'#'20230921_Patient_Metadata.csv'\n",
    "d_patient = pd.read_excel(f'Supplemental_Data/{s_out}',sheet_name=None)\n",
    "print(d_patient.keys())\n",
    "df_patient = d_patient['Patient_Metadata']\n",
    "df_spec = d_patient['DNA_Specimen_Metadata']\n",
    "df_spec.columns = [item.replace('.','_').replace('_Alterations','') for item in df_spec.columns]\n",
    "df_spec['Type'] = [item.split('-')[-1] for item in df_spec.Public_Specimen_ID]\n",
    "\n",
    "#add pORG all\n",
    "df_rna = d_patient['RNA_Specimen_Metadata'] #rna metadata update\n",
    "df_spec = df_spec.merge(df_rna, on ='Public_Patient_ID',how='left',suffixes=('','_x'))\n",
    "\n",
    "df_dna = d_patient['DNA_Specimen_Metadata'][~d_patient['DNA_Specimen_Metadata'].Type.str.contains('M')].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "019622c0-e9c1-4d4e-b783-b27137810000",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7494495a-8bc2-43e8-9508-3b9cc703a634",
   "metadata": {},
   "outputs": [],
   "source": [
    "cutoff = 0.0249109\n",
    "df_dna['pORG_Pri']=pd.NA\n",
    "df_dna.loc[df_dna.GSVA_pORG_Up_55_Primaries >= cutoff,'pORG_Pri'] = 'pORG high' \n",
    "df_dna.loc[df_dna.GSVA_pORG_Up_55_Primaries < cutoff,'pORG_Pri'] = 'pORG low' \n",
    "\n",
    "df_spec['pORG_Pri']=pd.NA\n",
    "df_spec.loc[df_spec.pORG_Up_55_Primaries >= cutoff,'pORG_Pri'] = 'pORG high' \n",
    "df_spec.loc[df_spec.pORG_Up_55_Primaries < cutoff,'pORG_Pri'] = 'pORG low' \n",
    "s_porg = 'pORG_Up_55_Primaries'\n",
    "df_spec = util.add_quartiles(df_spec,s_porg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "069bda46-1af9-44ba-936d-111f65f0db3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "d_cohort =   dict(zip(df_patient.Public_Patient_ID,df_patient.Cohort))\n",
    "df_spec['Cohort'] = df_spec.Public_Patient_ID.map(d_cohort)\n",
    "\n",
    "d_cohort =  dict(zip(df_patient.Public_Patient_ID,df_patient.DDR_Alteration_Primary))\n",
    "df_spec['DDR_Alteration_Primary'] = df_spec.Public_Patient_ID.map(d_cohort)\n",
    "df_spec['DDR_Alteration_Primary'] = df_spec.DDR_Alteration_Primary.fillna(False).replace({1:True})\n",
    "df_spec['DDR_Alteration_Primary'].unique()\n",
    "\n",
    "df_dna['Alive_30_days_post_surgery'] = df_dna.Public_Patient_ID.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Alive_30_days_post_surgery)))\n",
    "df_dna['Survival'] = df_dna.Public_Patient_ID.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Survival)))\n",
    "d_cohort =  dict(zip(df_patient.Public_Patient_ID,df_patient.DDR_Alteration_Primary.fillna(df_patient.DDR_Alteration_Met)))\n",
    "df_dna['DDR_Alteration_'] = df_dna.Public_Patient_ID.map(d_cohort)\n",
    "df_dna.DDR_Alteration_ = df_dna.DDR_Alteration_.fillna(False).replace({1:True})\n",
    "df_dna.columns#['Alive_30_days_post_surgery'] = df_dna.Public_Patient_ID.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Alive_30_days_post_surgery)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68a77d1d-d9a1-4238-ac0d-767aeb3a08d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "ls_groups = ['pORG high','pORG low']#['Liver','Lung']##\n",
    "ls_groups = [True,False]\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    d_replace ={'blood':True, 'tumor only':False, 'spleen':True, 'colon':True}\n",
    "    df_spec['Matched_Normal']=df_spec.NormalSampleSource.replace(d_replace)\n",
    "    x = 'DDR_Alteration_Primary'#'pORG_Pri'#'Cohort'#'Matched_Normal'\n",
    "    for y in ['Tumor_Mutation_Burden']:\n",
    "            plotting = {\"data\": df_spec[df_spec.loc[:,'pORG_Pri']=='pORG low'],\"x\":x,\"y\":y,\"order\":ls_groups}\n",
    "            fig,ax,test_results = util.annotated_stripplot(plotting,ls_groups,y)\n",
    "            h, l = ax.get_legend_handles_labels()\n",
    "            for res in test_results:\n",
    "                pvalue=res.data.pvalue\n",
    "            ax.set_title(f'\\np={pvalue:.2}')\n",
    "            plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95482ef7-f462-4c26-b65c-2fa94642c355",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_patient['DDR_Alteration_Pri'] = df_patient.DDR_Alteration_Primary.replace({1:'Yes'}).fillna(df_patient.TP53_Altered_Primary).replace({1:'No',0:'No'})\n",
    "# df_patient['DDR_Alteration_Mets'] = df_patient.DDR_Alteration_Met.replace({1:'Yes'}).fillna(df_patient.TP53_Altered_Met).replace({1:'No',0:'No'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b1d0a6e-2bc7-4c91-84f4-2a0c6210e332",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "linebreak = '\\n'\n",
    "df_patient.columns\n",
    "df_patient['pORG_Pri']=pd.NA\n",
    "df_patient.loc[df_patient.pORG_Primary > cutoff,'pORG_Pri'] = 'pORG high' \n",
    "df_patient.loc[df_patient.pORG_Primary < cutoff,'pORG_Pri'] = 'pORG low' \n",
    "tu_time_censor = (('Days_from_Diagnosis_to_FU','Survival'),\n",
    "                  ('Days from Resection to Recurrence', 'Recurrence')\n",
    "                 )\n",
    "#df_patient.DDR_Alteration_Primary.replace({1:'Yes'}).fillna(df_patient.TP53_Altered_Primary)\n",
    "for (s_time, s_censor) in tu_time_censor:\n",
    "    sns.set_palette(['gray','black'])\n",
    "    for s_pORG in ['pORG high','pORG low']:\n",
    "        for s_col in ['HR_DDR_Altered_BioRel','HR_DDR_Altered_All']:\n",
    "            df = df_dna[(df_dna.Alive_30_days_post_surgery) & (df_dna.pORG_Pri==s_pORG)& (df_dna.Type.isin(['T','F']))].copy()\n",
    "            df.loc[:,s_col] = df.loc[:,s_col].replace({True:'Yes',False:'No'})\n",
    "            fig, ax, __ =  util.km_plot(df,s_col,s_time, s_censor)\n",
    "            ax.set_title(f'{s_pORG}{linebreak}{s_col}')\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/KM_{s_pORG}_{s_col}.pdf')\n",
    "            #plt.close()\n",
    "    for s_pORG in [True,False]:\n",
    "        for s_col in ['HR_DDR_Altered_BioRel','HR_DDR_Altered_All']:\n",
    "            df = df_dna[(df_dna.Alive_30_days_post_surgery) & (df_dna.loc[:,s_col]==s_pORG)& (df_dna.Type.isin(['T','F']))].copy()\n",
    "            #df.loc[:,s_col] = df.loc[:,s_col].replace({True:'Yes',False:'No'})\n",
    "            fig, ax, __ =  util.km_plot(df,'pORG_Pri',s_time, s_censor)\n",
    "            ax.set_title(f\"{str(s_pORG).replace('True','Altered').replace('False','WT')}{linebreak}{s_col}\")\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f\"figures/KM_{str(s_pORG).replace('True','Altered').replace('False','WT')}_{s_col}.pdf\")\n",
    "            #plt.close()\n",
    "    sns.set_palette(['blue','purple','red','green'])\n",
    "    for s_col in ['HR_DDR_Altered_BioRel','HR_DDR_Altered_All','DDR_Alteration_']:\n",
    "            df = df_dna[(df_dna.Alive_30_days_post_surgery) & (df_dna.Type.isin(['T','F']))].copy()\n",
    "            df.loc[:,f'{s_col}_pORG'] = df.loc[:,s_col].replace({True:'Yes',False:'No'}) + ' ' + df.pORG_Pri\n",
    "            fig, ax, __ =  util.km_plot(df,f'{s_col}_pORG',s_time, s_censor,fontsize='xx-small',b_pval=False)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ba19119-f2f5-4a4e-9528-5a08ba4956e4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#figure 3H\n",
    "linebreak = '\\n'\n",
    "df_spec['Type'] = df_spec.Type.replace('M2','M').replace('T','Pri').replace('T2','Pri').replace('M','Met')\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    for s_cohort in ['Liver','Lung']:\n",
    "        df = df_patient[df_patient.Cohort==s_cohort].copy()\n",
    "        '''\n",
    "        df['DDR_Alteration_'] = df['DDR_Alteration_Primary'].fillna(df['DDR_Alteration_Met'])\n",
    "        for x in ['DDR_Alteration_','DDR_Alteration_Primary', 'DDR_Alteration_Met']:\n",
    "            s=x.split('_')[-1]\n",
    "            if s=='': \n",
    "                s_col = f'TP53_Altered'\n",
    "                y = f\"pORG_All\"\n",
    "                df[\"pORG_All\"]=df.pORG_Primary.fillna(df.pORG_Met)\n",
    "            else:\n",
    "                s_col = f'TP53_Altered_{s}'\n",
    "                y = f\"pORG_{s}\"\n",
    "            b_dna = df.loc[:,s_col].notna()\n",
    "            df.loc[:,x].fillna(0,inplace=True)\n",
    "            df.loc[~b_dna,x] = pd.NA\n",
    "            plotting = {\"data\": df,\"x\":x,\"y\":y,\"order\":ls_groups}\n",
    "            fig,ax,test_results = util.annotated_stripplot(plotting,ls_groups,b_correct=False,figsize=(2.4,3))\n",
    "            for res in test_results:\n",
    "                pvalue=res.data.pvalue\n",
    "            ax.set_title(f'{s_cohort} Cohort Patients\\np={pvalue:.2}')\n",
    "            ax.set_xlabel(f'{x.replace(\"_\",\" \")}')\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/{s_cohort}_{y}_vs_{x}.pdf')\n",
    "            plt.close()'''\n",
    "        #specimens\n",
    "        df2 = df_spec[df_spec.Public_Patient_ID.isin(df.Public_Patient_ID)].copy()\n",
    "        y = 'pORG_Up_55_All'\n",
    "        sns.set_palette(['gray','black'])\n",
    "        for x in ['HR_DDR_Altered_BioRel','HR_DDR_Altered_All']:\n",
    "                plotting = {\"data\": df2,\"x\":x,\"y\":y,\"order\":ls_groups}\n",
    "                fig,ax,test_results = annotated_stripplot_color(plotting,ls_groups,color='Type',figsize=(2.4,4))\n",
    "                h, l = ax.get_legend_handles_labels()\n",
    "                print(df2.loc[:,x].value_counts()[ls_groups[0]])\n",
    "                xlabels = [f'DDR Altered{linebreak}(N={(df2.loc[:,x].value_counts()[ls_groups[0]])})',\n",
    "                           f'DDR WT{linebreak}(N={(df2.loc[:,x].value_counts()[ls_groups[1]])})']\n",
    "                ax.legend(handles=h, labels=l,\n",
    "                          bbox_to_anchor=(1,-0.2),title='Specimen Type',ncol=2,fontsize='small')\n",
    "                for res in test_results:\n",
    "                    pvalue=res.data.pvalue\n",
    "                ax.set_title(f'{s_cohort} Cohort')\n",
    "                ax.set_ylabel('pORG All GSVA Score')\n",
    "                ax.set_xlabel('')\n",
    "                ax.set_xticklabels(xlabels)\n",
    "                plt.tight_layout()\n",
    "                s_save = f'Figure3_H_{x}_{s_cohort}'\n",
    "                if x == 'HR_DDR_Altered_BioRel':\n",
    "                    if s_cohort == 'Liver':\n",
    "                        try:\n",
    "                            d_result_fig.update({s_save:df2})\n",
    "                            #util.open_write_excel(d_result_fig,filename='Source_Data_Figure3.xlsx') \n",
    "                        except:\n",
    "                            pass\n",
    "                fig.savefig(f'figures/{s_save}.pdf')\n",
    "                break\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23170dec-40ab-44dd-b53c-413841aa88b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2.Public_Patient_ID.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ba9db5e-ea8f-4bc5-bf6c-89a3671596c0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "d_cohort =   dict(zip(df_patient.Public_Patient_ID,df_patient.Cohort))\n",
    "df_spec['Cohort'] = df_spec.Public_Patient_ID.map(d_cohort)\n",
    "d_tumor = {'Primary':['T', 'F'],'Met':['M2','M',]}#'T2',\n",
    "y = 'Tumor_Mutation_Burden'\n",
    "\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    for key, items in d_tumor.items():\n",
    "        for b_matched in [True,False]:\n",
    "            print(df_spec.loc[df_spec.Type.isin(items), 'Public_Patient_ID'].duplicated().any())\n",
    "            print(len(df_spec.loc[df_spec.Type.isin(items), 'Public_Patient_ID']))\n",
    "            if b_matched:\n",
    "                df_tum = df_spec.loc[(df_spec.Type.isin(items)) & (df_spec.Matched_Normal)]\n",
    "            else:\n",
    "                df_tum = df_spec.loc[(df_spec.Type.isin(items))]\n",
    "            for x in ['HR_DDR_Altered_BioRel','quartiles','pORG_Pri','Cohort','HR_DDR_Altered_All']:\n",
    "                df = df_tum[df_tum.loc[:,x].notna()]\n",
    "                ls_groups = df_tum.loc[:,x].dropna().unique()\n",
    "                plotting = {\"data\": df,\"x\":x,\"y\":y,\"order\":ls_groups}\n",
    "                fig,ax,test_results = util.annotated_stripplot(plotting,ls_groups,y)\n",
    "                for res in test_results:\n",
    "                    pvalue=res.data.pvalue\n",
    "                if b_matched:\n",
    "                    ax.set_title(f'Samples w/ Matched Normal\\n{key} HR/DDR\\np={pvalue:.2} n={len(df)}')#\\n{x.split(\"DDR\")[1].replace(\" _\",\" \").replace(\"_\",\" \")}')\n",
    "                else:\n",
    "                    ax.set_title(f'All Samples\\n{key} HR/DDR\\np={pvalue:.2} n={len(df)}')#\\n{x.split(\"DDR\")[1].replace(\" _\",\" \").replace(\"_\",\" \")}')\n",
    "                plt.tight_layout()\n",
    "                fig.savefig(f'figures/TMB_vs_{x}_{key}.png')\n",
    "                break\n",
    "        #break'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "195acfb7-83c1-4029-9813-98ff1f5f4302",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_tum.loc[:,x].dropna().unique()#.value_counts()\n",
    "# df_tum.Cohort.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d67689c4",
   "metadata": {},
   "source": [
    "# Survival Analysis - patients in paper <a name=\"km\"></a>\n",
    "\n",
    "Kaplan Meier, Cox Propotional Hazards\n",
    "\n",
    "\n",
    "[contents](#contents)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9eece8d8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## all combos of liver, lung, liver w/ lung\n",
    "df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "                          index_col=0)\n",
    "df_patient['Lung'] = df_patient.loc[:,'Lung Met Present'].replace('YES','Lung').fillna('')\n",
    "df_patient['Liver'] = df_patient.loc[:,'Liver Met Present'].replace('YES','Liver').fillna('')#\n",
    "df_patient['Liver_Lung'] = df_patient['Liver'] + ' ' + df_patient['Lung']\n",
    "df_patient['Liver_Lung'].replace('Liver Lung','Liver w/ Lung',inplace=True)\n",
    "df_patient['Liver_Lung'].replace('Liver ','Liver w/o Lung',inplace=True)\n",
    "df_patient['Liver_Lung'].replace(' ',pd.NA,inplace=True)\n",
    "df_patient['Liver_Lung'].replace(' Lung','Lung w/o Liver',inplace=True)\n",
    "df_patient['Liver_Only'] = df_patient['Liver_Lung'].replace('Lung w/o Liver',pd.NA)\n",
    "df_patient['Liver_vs_Lung'] = df_patient['Liver_Lung'].replace('Liver w/ Lung',pd.NA)\n",
    "df_patient['Liver_and_Lung'] = df_patient['Liver_Lung'].replace('Liver w/o Lung',pd.NA)\n",
    "#df_patient['Liver_Lung'] = df_patient['Liver_Lung'].replace('Liver','Liver w/o Lung').replace('Lung','Lung w/o Liver')\n",
    "df_patient['Recurrence_Sites'] = df_patient['Recurrence_Sites_4'].replace('No_Resection',pd.NA).replace('No_Doc_Recur','No Documented\\nRecurrence').replace('Other_site',\n",
    "                            'Other Recurrence\\nSite').replace('Lung','Lung Cohort').replace('Liver','Liver Cohort')\n",
    "\n",
    "df_patient['Other_vs_no_doc'] = df_patient.Recurrence_Sites.replace({'Lung Cohort':pd.NA,'Liver Cohort':pd.NA})\n",
    "df_patient['Other_vs_Lung'] = df_patient.Recurrence_Sites.replace({'No Documented\\nRecurrence':pd.NA,'Liver Cohort':pd.NA})\n",
    "df_patient['Other_vs_Liver'] = df_patient.Recurrence_Sites.replace({'No Documented\\nRecurrence':pd.NA,'Lung Cohort':pd.NA})\n",
    "\n",
    "df_patient.Recurrence_Sites.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f27f2a8f-7910-4a1a-9a2b-9986bb85275b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_patient.columns[df_patient.columns.str.contains('Days')]\n",
    "df_patient['Recurrence Free Survival (Days)'] = df_patient['Days from Diagnosis to FU'] - df_patient['Days from Earliest Recur to FU']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a4a2f5b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#KM\n",
    "d_fig1 = {}\n",
    "%matplotlib inline\n",
    "importlib.reload(util)\n",
    "df_patient.columns[df_patient.columns.str.contains('_Youden')]\n",
    "ls_km = [#'Other_vs_Liver','Other_vs_Lung',#\n",
    "    #'Other_vs_no_doc',#\n",
    "    'Liver_Lung',\n",
    "        # 'Liver_Only','Liver_vs_Lung','Liver_and_Lung',#'Cohort_Met_Resection',\n",
    "    'Recurrence_Sites',#'Cohort2',\n",
    "       #   'Classical_Cohort',\n",
    "       #   'PurIST_Subtype',#\n",
    "        'Cohort',\n",
    "       #         'Other_Recurrence', 'No_Doc_Recur',\n",
    "       #      'Lung_Cohort','Liver_Cohort',\n",
    "        ]\n",
    "pal_liv = ('#D55E00','#0072B2')\n",
    "pal_porg_r = ('#E69F00','#56B4E9')\n",
    "pal_three = ('#d62728','#bcbd22',#'#D55E00',#(1.0, 0.4980392156862745, 0.054901960784313725),\n",
    "             #'#d62728',#(0.17254901960784313, 0.6274509803921569, 0.17254901960784313),\n",
    "             '#0072B2'#(0.12156862745098039, 0.4666666666666667, 0.7058823529411765)\n",
    "             )\n",
    "pal_four = ('#D55E00','#0072B2','#2ca02c','#9467bd','#2ca02c',\n",
    "             )\n",
    "pal_four_r = ('#D55E00','#0072B2','#9467bd','#2ca02c','#2ca02c',\n",
    "             )\n",
    "tu_time_censor = (('Days from Diagnosis to FU','Survival'),\n",
    "                  #('Recurrence Free Survival (Days)','Survival'),\n",
    "                  ('Days from Earliest Recur to FU','Survival'),\n",
    "                  ('Days from Resection to FU','Survival'),\n",
    "                  ('Days from Resection to Recurrence', 'Recurrence')\n",
    "                 )\n",
    "for (s_time, s_censor) in tu_time_censor:\n",
    "    for s_km in ls_km:\n",
    "        loc='upper center'\n",
    "        if s_km == 'Liver_Lung':\n",
    "            sns.set_palette(pal_three)\n",
    "        elif s_km.find('Liver') > -1:\n",
    "            sns.set_palette(pal_liv)\n",
    "        elif s_km.find('Cohort') > -1:\n",
    "            sns.set_palette(pal_liv)\n",
    "        elif s_km == 'Recurrence_Sites':\n",
    "            sns.set_palette(pal_four)\n",
    "            if s_censor == 'Recurrence':\n",
    "                sns.set_palette(pal_four_r)\n",
    "            if s_censor == 'Survival':\n",
    "                loc='upper right'\n",
    "        else:\n",
    "            sns.set_palette(pal_porg_r)\n",
    "        df_km = df_patient.loc[df_patient.Alive_30_days_post_surgery,['Public_Patient_ID',s_km,s_time,s_censor]].dropna().copy()\n",
    "        df_km[s_km] = df_km[s_km].astype('str')\n",
    "        if s_time=='Days from Earliest Recur to FU':\n",
    "            b_pval = True\n",
    "        else:\n",
    "            b_pval = False\n",
    "        fig,ax,ls_order = util.km_plot(df_km,s_km,s_time,s_censor,fontsize='small',loc=loc,\n",
    "                                      figsize=(4,3.5),b_pval=b_pval)\n",
    "        ax.set_ylabel(f'Probability of {s_censor}')\n",
    "        ax.set_xlabel(f'{s_time}')\n",
    "        if s_censor == 'Survival':\n",
    "            if s_km == 'Liver_Lung':\n",
    "                if s_time == 'Days from Diagnosis to FU':\n",
    "                    # ax,x,top, bottom,ytext\n",
    "                    util.demo_con_style(ax,0.68,0.93,0.67,0.93, \"P=0.005\") # red vs blue\n",
    "                    util.demo_con_style(ax,0.74,0.8,0.67,0.81, \"P=0.0007\") # yellow vs blue\n",
    "                    ax.set_title('Overall Survival',fontweight='bold')\n",
    "            if s_km == 'Recurrence_Sites':\n",
    "                if s_time == 'Days from Diagnosis to FU':\n",
    "                    util.demo_con_style(ax,0.87,0.94,0.8,0.89, \"P=0.008\") # liver vs lung\n",
    "                    util.demo_con_style(ax,0.95,0.7,0.5,0.7, \"P=0.003\") # other vs no doc\n",
    "                    ax.set_title('OS Resected Patients',fontweight='bold')\n",
    "                elif s_time == 'Days from Resection to FU':\n",
    "                    util.demo_con_style(ax,0.87,0.94,0.82,0.89, \"P=0.003\") # liver vs lung\n",
    "                    util.demo_con_style(ax,0.95,0.73,0.55,0.7, \"P=0.005\") # other vs no doc\n",
    "                    ax.set_title('')\n",
    "        if s_censor == 'Recurrence':\n",
    "            if s_km == 'Recurrence_Sites':\n",
    "                util.demo_con_style(ax,0.8,0.94,0.8,0.93, \"P<0.001\") # liver vs lung\n",
    "                util.demo_con_style(ax,0.72,0.94,0.7,0.78, \"P<0.001\") # liver vs other\n",
    "                ax.set_title('Resection to Recurrence',fontweight='bold')\n",
    "        plt.tight_layout()\n",
    "        if f'KM_{s_km}_{s_time}'=='KM_Liver_Lung_Days from Diagnosis to FU':\n",
    "            d_fig1.update({'Figure1_A':df_km}) \n",
    "        elif f'KM_{s_km}_{s_time}'=='KM_Recurrence_Sites_Days from Diagnosis to FU':\n",
    "            d_fig1.update({'Figure1_B':df_km}) \n",
    "        elif f'KM_{s_km}_{s_time}'=='KM_Recurrence_Sites_Days from Resection to Recurrence':\n",
    "            d_fig1.update({'Figure1_D':df_km}) \n",
    "        else:\n",
    "            print(f'{linebreak}')\n",
    "            print(f'KM_{s_km}_{s_time}')\n",
    "        fig.savefig(f'figures/KM_{s_km}_{s_time}.pdf')\n",
    "        #break\n",
    "    #break  \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb22988d-3582-45eb-93d7-050ed1adc15c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "importlib.reload(util)\n",
    "pal_4 = ('#000000','#cc79a7','#D55E00','#0072B2',)\n",
    "sns.set_palette(pal_4)\n",
    "df_patient['Classical_Cohort'] = df_patient['Classical_Cohort'].replace('Liver','Liver Classical').replace('Lung','Lung Classical')\n",
    "df_patient['PurIST_Subtype'] = df_patient['PurIST_Subtype'].replace('basal-like','Basal-like').replace('classical','Classical')\n",
    "\n",
    "df = df_patient.loc[df_patient.Alive_30_days_post_surgery]\n",
    "for (s_time, s_censor) in tu_time_censor:\n",
    "    ls_col = ['PurIST_Subtype', 'Classical_Cohort']\n",
    "    fig,ax,ls_order = util.km_plot_overlay(df,ls_col,s_time,s_censor,fontsize='small',loc='upper center',\n",
    "                                          figsize=(4,3.5))\n",
    "    ax.set_ylabel(f'Probability of {s_censor}')\n",
    "    ax.set_xlabel(f'{s_time}')\n",
    "    if s_time == 'Days from Diagnosis to FU':\n",
    "        # ,x,top, bottom,ytext\n",
    "        util.demo_con_style(ax,0.72,0.94,0.79,0.93, \"P=0.0003\") # purist\n",
    "        util.demo_con_style(ax,0.72,0.7,0.56,0.7, \"P=0.002\") # ll classical\n",
    "    if s_censor == 'Recurrence':\n",
    "        util.demo_con_style(ax,0.72,0.97,0.81,0.93, \"P=0.01\") # purist\n",
    "        util.demo_con_style(ax,0.72,0.77,0.6,0.73, \"P=0.003\") # ll \n",
    "    plt.tight_layout()\n",
    "    if f'KM_{\".\".join(ls_col)}_{s_time}'=='KM_PurIST_Subtype.Classical_Cohort_Days from Diagnosis to FU':\n",
    "            d_fig1.update({'Figure1_G':df_km}) \n",
    "    else:\n",
    "        print(f'{linebreak}')\n",
    "        print(f'KM_{\".\".join(ls_col)}_{s_time}')\n",
    "    fig.savefig(f'figures/KM_{\".\".join(ls_col)}_{s_time}.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "717cef93-54bb-4597-8c7a-744d1f124f99",
   "metadata": {},
   "outputs": [],
   "source": [
    "#util.open_write_excel(d_fig1,filename='Source_Data_Figure1.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eed226c7-afc6-417f-b2a3-d815bf0b1c2e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#supp fig 2\n",
    "importlib.reload(util)\n",
    "pal_gray = ('#000000','#009E73')#cc79a7','#D55E00','#0072B2',)\n",
    "sns.set_palette(pal_gray)\n",
    "df = df_patient[df_patient.Neoadjuvant.notna()]\n",
    "for (s_time, s_censor) in tu_time_censor:\n",
    "    for s_col in ['PurIST_Subtype', 'Cohort']:\n",
    "        for s_group in df.loc[:,s_col].dropna().unique():\n",
    "            df2 = df[df.loc[:,s_col]==s_group].copy()\n",
    "            s_km = f'{s_col} {s_group}'\n",
    "            df2[s_km] = df2.loc[:,s_col] + '_Neo_' + df2.Neoadjuvant.replace({'No':'N','Yes':'Y'})\n",
    "            df_km = df2.loc[df2.Alive_30_days_post_surgery,['Public_Patient_ID',s_km,s_time,s_censor]].dropna().copy()\n",
    "            fig,ax,ls_order = util.km_plot(df_km,s_km,s_time,s_censor,fontsize='small',loc='upper center',figsize=(4,3))\n",
    "            ax.set_ylabel(f'Probability of {s_censor}')\n",
    "            ax.set_xlabel(f'{s_time}')\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/KM_neoadjuvant_{s_km}.pdf')\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7814dd34-0cd2-40bb-9ceb-a0b5dbc91997",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#censors\n",
    "#days survival after resection vs days survival after recurrence\n",
    "linebreak = '\\n'\n",
    "df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "                          index_col=0)\n",
    "df_patient.loc[:,'Days from Resection to FU'] = df_patient.loc[:,'Days from Resection to Recurrence'] + df_patient.loc[:,'Days from Earliest Recur to FU']\n",
    "df_patient.loc[:,'Survival after recurrence (days)'] = df_patient.loc[:,'Days from Earliest Recur to FU']\n",
    "df_patient.loc[:,'Days from Diagnosis to Resection'] = df_patient.loc[:,'Days from Diagnosis to FU'] - df_patient.loc[:,'Days from Resection to FU']\n",
    "\n",
    "sx='Days from Earliest Recur to FU'#'Survival after recurrence (days)'\n",
    "l_sy= ['Days from Resection to FU','Days from Resection to Recurrence']\n",
    "s_censors = 'Survival'\n",
    "d_color_alive = {'Dead':'gray','Alive':'black'}\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    for sy in l_sy:\n",
    "        fig,ax=plt.subplots(2,1,sharex=True,figsize=(2.5,3.5),dpi=300)\n",
    "        for idx, s_cohort in enumerate(['Liver','Lung']):\n",
    "            print(s_cohort)\n",
    "            df_plot = df_patient[(df_patient.Cohort==s_cohort) & (df_patient.loc[:,sy].notna()) & (df_patient.loc[:,sx].notna())]\n",
    "            tau,p= util.patient_kendall(df_plot,sx,sy,s_censors)\n",
    "            r, pval = util.patient_pearson(df_plot,sx,sy)\n",
    "            sns.scatterplot(data=df_plot,x=sx,y=sy,hue=df_plot.loc[:,'Vital Status at FU'],\n",
    "                            ax=ax[idx],alpha=0.8,s=5)\n",
    "            ax[idx].set_title(f'{s_cohort} Cohort\\nKendall tau={tau:.2}\\nP={p:.2} N={len(df_plot)}',\n",
    "                             fontsize='small')\n",
    "            ax[idx].set_ylabel(ax[idx].get_ylabel().replace(' to ',f'{linebreak} to '),fontsize='small')\n",
    "            ax[idx].set_xlabel(ax[idx].get_xlabel(),fontsize='small')\n",
    "        ax[0].legend(fontsize='x-small',markerscale=0.5,framealpha=0.5,facecolor=None)\n",
    "        ax[1].legend().remove()\n",
    "        plt.tight_layout()\n",
    "        fig.savefig(f'figures/scatterplot_{sy}.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf5c4f57-3f6e-4572-9187-68d444b9f0dd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sns.set_palette(pal_liv)\n",
    "for s_km in ['Cohort']:\n",
    "    for s_time in ['Days from Earliest Recur to FU','Days from Resection to FU','Days from Resection to Recurrence','Days from Diagnosis to Resection']:\n",
    "        if s_time.find('Days from R') > -1:\n",
    "            s_censors = 'Recurrence'#'Recurrence'\n",
    "        elif s_time.find('Days from D') > -1:\n",
    "            df_patient['Resection'] = 1\n",
    "            s_censors = 'Resection'\n",
    "        else:\n",
    "            s_censors = 'Survival'\n",
    "        print(s_censors)\n",
    "        df_km = df_patient.loc[:,['Public_Patient_ID',s_km,s_time,s_censors]].dropna().copy()\n",
    "        fig,ax,ls_order = util.km_plot(df_km,s_km,s_time,s_censors,fontsize='small',loc='upper center',figsize=(4,3.5))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8443870b-7abf-48ff-ba2c-ee5e9e8252b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #supp figure 1 data\n",
    "# ls_col = ['Public_Patient_ID','Days from Resection to FU', 'Days from Earliest Recur to FU','Days from Resection to Recurrence',\n",
    "#          'PurIST_Subtype','Classical_Cohort','Cohort','DNAseq_Patient','RNAseq_Patient',\n",
    "#           'TCRseq_Tumor_Patient', 'TCRseq_Blood_Patient','Neoadjuvant',\n",
    "#           'Age','Sex','Stage','Grade','LN_Pos','LV_Invasion','Resected','Alive_30_days_post_surgery',\n",
    "#          ]\n",
    "# df_patient.columns[df_patient.columns.str.contains('Pub')]\n",
    "# #df_patient.Neoadjuvant.value_counts()\n",
    "# df_patient.loc[:,ls_col].to_excel('Source_Data_EDFigure1.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7a66126-dfd5-4f1b-ab9a-0de570e7a151",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # supp fig 2 censored correlation\n",
    "# #days survival after recurrence vs days from resection to recurrence\\\n",
    "# sx = 'days survival after recurrence'\n",
    "# sy = 'Days from Resection to Recurrence'\n",
    "# s_censors = 'Survival'\n",
    "# tau,p=patient_kendall(df_patient,sx,sy,s_censors)\n",
    "# r, pval= patient_pearson(df_patient,sx,sy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20be5b56-a8a3-4ead-84d7-c918c015de2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from kendall import *\n",
    "# # gen random data\n",
    "# np.random.seed(123)\n",
    "# x = np.random.rand(89)\n",
    "# y = x+0.1*np.random.randn(len(x))\n",
    "# # gen random \"upper limits\"\n",
    "# censors = np.ones((2,len(x)))\n",
    "# uplims = np.unique(np.random.randint(0,high=len(x)-1,size=60))\n",
    "# censors[1,uplims]=0\n",
    "# uncens = [i for i in range(len(x)) if i not in uplims]\n",
    "# # faux uncertainties, accounting for upper limits\n",
    "# x_err = 0.1*np.random.rand(len(x))*censors[0]\n",
    "# y_err = 0.1*np.random.rand(len(y))*censors[1]\n",
    "# # Kendall tau\n",
    "# tau,p = util.kendall(x,y,censors=censors)\n",
    "# tau_lo,tau_up = util.tau_conf(x,y,censors,method='bootstrap')\n",
    "# print(f'censored Kendall tau: {tau:.3f}-{tau_lo:.3f}+{tau_up:.3f}, p: {p:.3e}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccd12634-763f-4165-b0e9-3cf4d2dbb09c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_patient['RFS'] = df_patient['Survival']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f02a11b3-b5d9-4489-9107-52de65e3ea10",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#figure 2 The new Youden cutoffs are:\n",
    "d_fig2 = {}\n",
    "#pORG: 0.02491 (same)\n",
    "#pSUB: -0.31836 (same)\n",
    "#PuRIST: new: 0.01340\n",
    "importlib.reload(util)\n",
    "sns.set_palette(pal_porg_r)\n",
    "d_cutoff = {'pORG Primary':0.02491, 'pSUB Primary':-0.31836, 'PurIST Primary':0.0134}\n",
    "tu_time_censor = (('Days from Diagnosis to FU','Survival'),\n",
    "                  ('Recurrence Free Survival (Days)','RFS'),\n",
    "                 )\n",
    "df_patient['Recurrence Free Survival (Days)'] = df_patient.loc[:,'Days from Diagnosis to FU'] - df_patient.loc[:,'Days from Earliest Recur to FU'].fillna(0)\n",
    "df_result = pd.DataFrame()\n",
    "for s_col, cutoff in d_cutoff.items():\n",
    "    df_patient[s_col] = pd.NA\n",
    "    df_patient.loc[df_patient.loc[:,f'{s_col.replace(\" \",\"_\")}'] > cutoff,s_col] = f'High' #\n",
    "    df_patient.loc[df_patient.loc[:,f'{s_col.replace(\" \",\"_\")}'] <= cutoff,s_col] = f'Low'\n",
    "    df = df_patient[(df_patient.Alive_30_days_post_surgery) & (df_patient[s_col].notna())]\n",
    "    for (s_time, s_censor) in tu_time_censor:\n",
    "        figsize=(3.5,3)\n",
    "        lower=0.73\n",
    "        if s_time == 'Recurrence Free Survival (Days)':\n",
    "            figsize=(3.3,3.5)\n",
    "            lower = 0.76\n",
    "        df_km = df.loc[:,[s_col,s_time,s_censor]].dropna()\n",
    "        fig,ax,pvalue = util.km_plot(df_km,s_col,s_time,s_censor,fontsize='small',loc='upper center',figsize=figsize,b_pval=False)\n",
    "        util.demo_con_style(ax,0.7,0.88,lower,0.85, f\"P={pvalue:.2}\") #ax,x,top, bottom,ytext,\n",
    "        ax.set_ylabel(f'Probability of {s_censor}')\n",
    "        ax.set_xlabel(f'{s_time}')\n",
    "        ax.set_title(f'OHSU: {s_col.replace(\"Primary\",\"Score\")}')\n",
    "        #fig.suptitle(f'{s_col.replace(\"Primary\",\"Score\")}',fontweight='bold',fontsize='x-large',y=.94,x=.56)\n",
    "        plt.tight_layout()\n",
    "        fig.savefig(f'figures/km_{s_col}_{s_time}.pdf')\n",
    "        #cph - # linear\n",
    "        df = df_patient[(df_patient.Alive_30_days_post_surgery) & (df_patient[s_col.replace(\" \",\"_\")].notna())]\n",
    "        fig,cph = util.cph_plot(df.loc[:,[s_col.replace(\" \",\"_\"),s_time,s_censor]],s_col.replace(\" \",\"_\"),s_time,s_censor,figsize=(3,3))\n",
    "        plt.close(fig)\n",
    "        df_cph = cph.summary.loc[:,['exp(coef)','exp(coef) lower 95%','exp(coef) upper 95%','p']]\n",
    "        df_cph['n'] = len(df)\n",
    "        df_cph['time'] = s_time\n",
    "        df_cph.index = [f'OHSU_{s_col}']\n",
    "        if s_censor == 'Survival':\n",
    "            d_fig2.update({f'Figure2_A_{s_col}_{s_censor}':df.loc[:,['Public_Patient_ID',s_col,s_time,s_censor]]})\n",
    "        df_result=pd.concat([df_result,df_cph])\n",
    "    #     break\n",
    "    # break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b8b4884-0b57-446c-b975-b60f5003113b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_out = df_patient.loc[:,['Public_Patient_ID','Recurrence Free Survival (Days)','RFS'] + list(d_cutoff.keys())].dropna()\n",
    "df_out.to_csv(f'Source_Extended_Data_Figure2_E.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "118fdd50-4acc-4947-ba53-0666c8ae325a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#util.open_write_excel(d_fig2,filename='Source_Data_Figure2.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32967725-ebe8-4347-bb33-3b4558a05751",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_result.sort_values(by='time')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55ef5d02-6109-4ac0-a71f-2a19146816b4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#neoadjuvant\n",
    "importlib.reload(util)\n",
    "pal_gray = ('#000000','#009E73')#cc79a7','#D55E00','#0072B2',)\n",
    "sns.set_palette(pal_gray)\n",
    "df = df_patient[df_patient.Neoadjuvant.notna()]\n",
    "for (s_time, s_censor) in tu_time_censor:\n",
    "    for s_col in ['pORG Primary', 'pSUB Primary', 'PurIST Primary']:\n",
    "        for s_group in df.loc[:,s_col].dropna().unique():\n",
    "            df2 = df[df.loc[:,s_col]==s_group].copy()\n",
    "            s_km = f'{s_col} {s_group}'\n",
    "            df2[s_km] = df2.loc[:,s_col] + '_Neo_' + df2.Neoadjuvant.replace({'No':'N','Yes':'Y'})\n",
    "            df_km = df2.loc[df2.Alive_30_days_post_surgery,['Public_Patient_ID',s_km,s_time,s_censor]].dropna().copy()\n",
    "            fig,ax,ls_order = util.km_plot(df_km,s_km,s_time,s_censor,fontsize='small',loc='upper center',figsize=(4,3))\n",
    "            ax.set_ylabel(f'Probability of {s_censor}')\n",
    "            ax.set_xlabel(f'{s_time}')\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/KM_neoadjuvant_{s_km}.pdf')\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b098151-a3f3-442d-be50-6cd0f20ca10a",
   "metadata": {},
   "source": [
    "## TCGA and ICGC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a4bf5e9-d65b-42e1-915f-434a194b1356",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the external datasets\n",
    "\n",
    "#'ICGC_APGI_Survival_Metadata.tsv', 'ICGC_KM_TestData.tsv',\n",
    "#  'TCGA_KM_TestData.tsv','CRP_YoudenBasedCutoffs.tsv   TCGA_ListToUse.tsv\n",
    "s_load =  'data/CRP_YoudenBasedCutoffs.tsv'\n",
    "df_cutoff = pd.read_csv(s_load,index_col=0,sep='\\t',header=None)\n",
    "df_cutoff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09fe2188-599c-4362-9529-ce1f01b3606e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "s_time = 'Survival'\n",
    "s_censor = 'Event'\n",
    "ls_load = [#'ICGC_APGI_Survival_Metadata.tsv','TCGA_ListToUse.tsv'\n",
    "          \n",
    "  'TCGA_KM_TestData.tsv', 'ICGC_KM_TestData.tsv',\n",
    " ]\n",
    "df_result=pd.DataFrame()\n",
    "ls_marker = ['PurIST_Score','pORG_Up_55','pSUB_Up_51']\n",
    "for s_load in ls_load:\n",
    "    df_ext = pd.read_csv(f'data/{s_load}',index_col=0,sep='\\t')\n",
    "    figsize=(3.5,3)\n",
    "    lower=0.73\n",
    "    if s_load == 'ICGC_KM_TestData.tsv':\n",
    "        figsize=(3.3,3.5)\n",
    "        lower = 0.76\n",
    "    for s_marker in ls_marker:\n",
    "        cutoff = df_cutoff.loc[f'{s_marker}_Cutoff',1]\n",
    "        s_col = s_marker.split('_')[0]\n",
    "        df_ext[s_col] = pd.NA\n",
    "        df_ext.loc[df_ext.loc[:,s_marker] > cutoff,s_col] = f'High' \n",
    "        df_ext.loc[df_ext.loc[:,s_marker] <= cutoff,s_col] = f'Low'\n",
    "        fig,ax,pvalue = util.km_plot(df_ext,s_col,s_time,s_censor,fontsize='small',loc='upper center',\n",
    "                                     figsize=figsize,b_pval=False)\n",
    "        ax.set_title(f'{s_load.split(\"_\")[0]}: {s_col} Score')\n",
    "        util.demo_con_style(ax,0.70,0.88,lower,0.85, f\"P={pvalue:.2}\")\n",
    "        ax.set_ylabel(f'Probability of Survival')\n",
    "        ax.set_xlabel(f'Overall Survival (days)')\n",
    "        if s_load=='ICGC_KM_TestData.tsv':\n",
    "            ax.set_ylabel(f'Probability of Survival')\n",
    "            ax.set_xlabel(f'Recurrence-Free Survival (days)')\n",
    "        plt.tight_layout()\n",
    "        fig.savefig(f'figures/km_{s_load.split(\"_\")[0]}_{s_col}_{s_censor}.pdf')\n",
    "        if f'{s_load.split(\"_\")[0]}'=='TCGA':\n",
    "            d_fig2.update({f'Figure2_B_{s_col}':df_ext.loc[:,[s_col,s_time,s_censor]]})\n",
    "        #cph\n",
    "        df = df_ext[df_ext.loc[:,s_col].notna()].copy()\n",
    "        # if df.loc[:,s_col].dtype=='O':\n",
    "        #     df_dummy = pd.get_dummies(df.loc[:,[s_col]],drop_first=True)\n",
    "        #     if df_dummy.shape[1]>0:\n",
    "        #         print(s_col)\n",
    "        #         df.drop(s_col,axis=1,inplace=True)\n",
    "        #         df[s_col] = df_dummy\n",
    "        #         fig,cph = util.cph_plot(df.loc[:,[s_col,s_time,s_censor]],s_col,s_time,s_censor,figsize=(3,3))\n",
    "        #         df_cph = cph.summary.loc[:,['exp(coef)','exp(coef) lower 95%','exp(coef) upper 95%','p']]\n",
    "        #         df_cph['n'] = len(df)\n",
    "        #         df_cph.index = [f'{s_load.split(\"_\")[0]}_{df_dummy.columns[0]}']\n",
    "        #         df_result=pd.concat([df_result,df_cph])\n",
    "        # linear\n",
    "        fig,cph = util.cph_plot(df.loc[:,[s_marker,s_time,s_censor]],s_marker,s_time,s_censor,figsize=(3,3))\n",
    "        df_cph = cph.summary.loc[:,['exp(coef)','exp(coef) lower 95%','exp(coef) upper 95%','p']]\n",
    "        df_cph['n'] = len(df)\n",
    "        df_cph.index = [f'{s_load.split(\"_\")[0]}_{s_marker}']\n",
    "        df_result=pd.concat([df_result,df_cph])\n",
    "    #break\n",
    "df_result.to_csv(f'results/results_single_CPH_external.csv')\n",
    "#df_ext.to_csv(f'Source_Extended_Data_Figure2_F.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e27b100d-cfd6-4184-9473-6eb5511f6a19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# os.listdir('annotation')\n",
    "# df = pd.read_excel('annotation/APGI_Transcriptome_Cohort_PoR_M0_ROSIES.xlsx')\n",
    "# df.columns\n",
    "#util.open_write_excel(d_fig2,filename='Source_Data_Figure2.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7005c8a3-71bd-4961-b143-6463e3fad771",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pandas as pd\n",
    "# import sys\n",
    "# import os\n",
    "\n",
    "\n",
    "# with pd.ExcelWriter('Supplemental_Table_2.xlsx') as writer:\n",
    "#     for csvfilename in sorted(os.listdir('results/CPH')):\n",
    "#         df = pd.read_csv(f'results/CPH/{csvfilename}',index_col=0)\n",
    "#         sheet_name = csvfilename.split(\"results_\")[1][:31]\n",
    "#         df.to_excel(writer, sheet_name=sheet_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b4c5294-d859-4a00-9e27-8103fe96b492",
   "metadata": {},
   "source": [
    "## Liver vs lung comparison \n",
    "\n",
    "supplemental \n",
    "\n",
    " <a name=\"livlungclin\"></a>\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9856a52a-dfba-421f-ae0a-c87447511cf7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#clinical covariates\n",
    "df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "                          index_col=0)\n",
    "\n",
    "df_hist = pd.read_csv('histology_for_Dove.csv')\n",
    "for s_type in ['T','M']:\n",
    "    for s_col in ['Acute Inflammation', 'Chronic Inflammation','Plasmacytoid Inflammation']:\n",
    "        df_test = df_hist[(df_hist.Tumor_Type==s_type) & (df_hist.Public_Patient_ID.notna())]\n",
    "        d_hist = dict(zip(df_test.Public_Patient_ID,df_test.loc[:,s_col]))\n",
    "        if s_type == 'M':\n",
    "            s_col_name = f'{s_col}_Met'\n",
    "        else:\n",
    "            s_col_name = s_col\n",
    "        df_patient[s_col_name] = df_patient.Public_Patient_ID.map(d_hist)\n",
    "\n",
    "#Lung Cohort TCR versus overall vs. Clinical Variables (chi squared)\n",
    "ls_clin = ['Desmoplasia (Y/N)','Desmoplasia (Y/N)_Met','ALI (Y/N)','ALI (Y/N)_Met','PNI','PNI_Met',\n",
    "    #'Resected','Grade',\n",
    "     #      'Age','Sex',\n",
    "      #     'Neoadjuvant',#'Stage',\n",
    "          #  'LV_Invasion','LN_Pos',\n",
    "           'TLS', #'Inflammation', #\n",
    "           'Acute Inflammation', 'Chronic Inflammation','Plasmacytoid Inflammation',\n",
    "     'TLS_Met', #'Inflammation_Met', #\n",
    "           'Acute Inflammation_Met', 'Chronic Inflammation_Met','Plasmacytoid Inflammation_Met',\n",
    "       \n",
    "          ]\n",
    "ls_cohort = ['Cohort',\n",
    "             #'Lung_Cohort',#'No_Recurrence',\n",
    "             #'Liver_Cohort',#'Other_Recurrence',\n",
    "             ]\n",
    "d_result = {}\n",
    "for s_clin in ls_clin:\n",
    "    for s_cohort in ls_cohort:\n",
    "        crosstab = pd.crosstab(df_patient.loc[:,s_clin],df_patient.loc[:,s_cohort])\n",
    "        df = crosstab.T.unstack().reset_index().rename({0:'Total Pts'},axis=1)\n",
    "        sheet_name = f'{s_clin.replace(\"/\",\"\")}_vs_{s_cohort}'\n",
    "        df.index.name = sheet_name\n",
    "        statistic,pvalue, dof, expected_freq = stats.chi2_contingency(crosstab)\n",
    "        d_result.update({sheet_name:df})\n",
    "        if pvalue < 1.05:\n",
    "            fig, ax = plt.subplots(figsize=(2.1,2),dpi=300)\n",
    "            #sns.heatmap(crosstab - expected_freq,ax=ax,annot=True,cbar_kws={'label':'Obs - Exp'})\n",
    "            sns.heatmap(crosstab/crosstab.sum(),ax=ax,annot=True,cbar_kws={'label':'Frac. Pts. in Cohort'})\n",
    "            ax.set_title(f'P={pvalue:.2} N={df.loc[:,\"Total Pts\"].sum()}',fontsize='medium')#{s_clin}\\nvs. {s_cohort} \n",
    "            ax.set_ylabel(ax.get_ylabel().replace(\"_\",\" \"))\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/heatmap_{sheet_name}.pdf')\n",
    "        #break\n",
    "        #break\n",
    "    #break\n",
    "# with pd.ExcelWriter('Supplemental_Table_10.xlsx') as writer:\n",
    "#     for sheet_name, df in d_result.items():\n",
    "#         df.to_excel(writer, sheet_name=sheet_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23e0f4d0-b11d-45b6-9a36-077b0fefd814",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_patient.to_csv('Patient_Metadata_w_Hist.csv')\n",
    "df_patient.loc[:,['Public_Patient_ID','Cohort']+ls_clin].to_csv('Source_Extended_Data_Fgiure2_A-D.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a26de4a9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#effect of tumor cellularity\n",
    "d_group = {'PurIST_Subtype':['classical','basal-like',],\n",
    "           'Cohort':['Lung','Liver',],\n",
    "            'pORG_Primary_quartiles':['low','high',], #'pORG_Met_quartiles', 'pORG_allMet_quartiles' #'pORG_allPrimary_quartiles',\n",
    "            'pSUB_Primary_quartiles':['low','high',],\n",
    "           'pSUB_Met_quartiles':['low','high',],\n",
    "          'pORG_Met_quartiles':['low','high',]\n",
    " }\n",
    "ls_col = ['Tumor_Cellularity_by_DNA_Met','Tumor_Cellularity_by_DNA_Primary',]\n",
    "sns.set_palette('tab10')\n",
    "linebreak='\\n'\n",
    "df_out = pd.DataFrame()\n",
    "import warnings\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    for s_col in ls_col:\n",
    "        for s_group, order  in d_group.items():\n",
    "            try:\n",
    "                df = df_patient.loc[:,[s_col,s_group]].dropna()\n",
    "            except:\n",
    "                df_patient[s_group] = pd.qcut(df_patient.loc[:,s_group.split('_quartiles')[0]],4,labels=['low','med','medx','high'])\n",
    "                df = df_patient.loc[:,[s_col,s_group]].dropna()\n",
    "            fig,ax = plt.subplots(figsize=(2.2,2),dpi=300)\n",
    "            sns.stripplot(data=df,x=s_group,y=s_col,ax=ax,s=3,alpha=0.8,order=order)\n",
    "            sns.boxplot(data=df,x=s_group,y=s_col,ax=ax,showmeans=True,medianprops={'visible': False},\n",
    "                           whiskerprops={'visible': False},meanline=True,showcaps=False,order=order,\n",
    "                           meanprops={'color': 'k', 'ls': '-', 'lw': 2},showfliers=False,showbox=False)\n",
    "            a = df.loc[df.loc[:,s_group]==order[0],s_col].values\n",
    "            b = df.loc[df.loc[:,s_group]==order[1],s_col].values\n",
    "            stat, pvalue = stats.ttest_ind(a, b)\n",
    "            n = len(df[df.loc[:,s_group].isin(order)])\n",
    "            ax.set_title(f'P={pvalue:.2} N={n}')\n",
    "            ax.set_xlabel(ax.get_xlabel().replace(\"_\",\" \"))\n",
    "            ax.set_ylabel(s_col.replace(\"_\",\" \").replace(\"Met\",f\"{linebreak}(Mets)\").replace(\"Primary\",f\"{linebreak}(Primaries)\"),fontsize='small')\n",
    "            plt.tight_layout()\n",
    "            if n > 6:\n",
    "                fig.savefig(f'figures/dotplot_{s_col}_{s_group}.pdf')\n",
    "                df_out = pd.concat([df_out,df],axis=1)\n",
    "            else:\n",
    "                plt.close()\n",
    "                 \n",
    "            #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38dea375-7669-4308-9e09-ca5348f770ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_out.loc[:,~df_out.columns.duplicated()]#.to_csv(f'Source_Extended_Data_Figure4_H-I.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05871bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # matches kevin, 58 liver metastasis 18 lung metastasis 83 no documented recurrence \n",
    "# #53 other recurrence site 6 no resection\n",
    "# print(df_patient[~df_patient.PurIST_Primary.isna()].Recurrence_Sites_4.value_counts())\n",
    "# print(df_patient[~df_patient.PurIST_Primary_T2.isna()].Recurrence_Sites_4.value_counts())\n",
    "# 58 + 18 + 83 + 53 + 6\n",
    "# print(df_patient[~df_patient.loc[:,'pORG_0.2_Met'].isna()].Recurrence_Sites_4.value_counts().sum())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f95e47ad-1346-47d4-b4b5-15a358900d75",
   "metadata": {},
   "source": [
    "## Cox Propotional Hazards <a name=\"clin\"></a>\n",
    "\n",
    "\n",
    "[contents](#contents)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0779a81a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CPH single variable\n",
    "\n",
    "tu_time_censor = (('Days from Diagnosis to FU','Survival'),\n",
    "                  ('Days from Resection to Recurrence', 'Recurrence'))\n",
    "for (s_time, s_censor) in tu_time_censor:\n",
    "    print(s_censor)\n",
    "    df_result=pd.DataFrame()\n",
    "    se_recur = {'Other_Recurrence', 'No_Doc_Recur','Lung_Cohort','Liver_Cohort'} #don't compare to the 74 with no resection\n",
    "    ls_vital = ['Cohort','pORG_Primary','pORG_Met','Neoadjuvant','Age','Sex','Stage', 'Grade','LV_Invasion','LN_Pos', #'Age at Diagnosis',\n",
    "          'Other_Recurrence', 'No_Doc_Recur',\n",
    "                'PurIST_Subtype',\n",
    "                'Classical_Cohort',\n",
    "                'Lung_Cohort','Liver_Cohort',\n",
    "               'pSUB_Primary', 'pSUB_Met','Classical_Cohort',\n",
    "           'PurIST_Primary', 'PurIST_Met']\n",
    "    for s_vital in ls_vital:\n",
    "        print(s_vital)\n",
    "        if s_vital == 'Neoadjuvant':\n",
    "            df = df_patient.loc[(df_patient.Alive_30_days_post_surgery) &(df_patient.Resected),[s_vital,s_time,s_censor]].copy()\n",
    "        else:\n",
    "            df = df_patient.loc[df_patient.Alive_30_days_post_surgery,[s_vital,s_time,s_censor]].copy()#.dropna(how='any')\n",
    "        if len(se_recur.intersection(set([s_vital]))):\n",
    "            df.loc[df_patient.Recurrence_Sites_4=='No_Resection',s_vital] = np.nan\n",
    "        df = df.dropna()    \n",
    "        print(len(df))\n",
    "        if df.columns.isin(['Stage']).any():\n",
    "            df.Stage = df.Stage.replace({'I':1,'II':2,'III':3,'IV':4}).astype('int')\n",
    "        if df.columns.isin(['Grade']).any():\n",
    "            df.Grade = df.Grade.astype('int')\n",
    "        if df.loc[:,s_vital].dtype=='O':\n",
    "            df_dummy = pd.get_dummies(df.loc[:,[s_vital]],drop_first=True)\n",
    "            if df_dummy.shape[1]>0:\n",
    "                print(s_vital)\n",
    "                df.drop(s_vital,axis=1,inplace=True)\n",
    "                df[s_vital] = df_dummy\n",
    "            else:\n",
    "                continue\n",
    "        try: \n",
    "            fig, cph = util.cph_plot(df,s_vital,s_time,s_censor,figsize=(3.8,1.7))\n",
    "            pvalue = cph.summary.p[s_vital]\n",
    "            fig.suptitle(s_censor,y=0.85,x=0.6)\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/CPH_single_{s_vital}_{pvalue:.2}_{s_censor}.png')\n",
    "            if s_vital == 'Neoadjuvant':\n",
    "                pass\n",
    "            elif pvalue > 0.09:\n",
    "                plt.close(fig)\n",
    "            df_cph = cph.summary.loc[:,['exp(coef)','exp(coef) lower 95%','exp(coef) upper 95%','p']]\n",
    "            df_cph['n'] = len(df)\n",
    "            df_result=pd.concat([df_result,df_cph])\n",
    "        except:\n",
    "            print('')\n",
    "        #break\n",
    "    #'''\n",
    "    #all the recurrence combos\n",
    "    s_vital = 'Recurrence_Sites_4'\n",
    "    for tu_combo in combinations([\"Lung\",\"Liver\",'Other_site','No_Doc_Recur','No_Resection'],r=2):#combinations(df_patient.Recurrence_Sites_4.unique(),r=2):\n",
    "        print(tu_combo)\n",
    "        df = df_patient.loc[(df_patient.Alive_30_days_post_surgery) &(df_patient.Resected),[s_vital,s_time,s_censor]]#.dropna(how='any')\n",
    "        # df['Recurrence_Sites_4'] = df['Recurrence_Sites_4'].replace({'Liver':'liver'})\n",
    "        print(len(df))\n",
    "        df = df[df.loc[:,s_vital].isin(tu_combo)].dropna()\n",
    "        #order them\n",
    "        df.Recurrence_Sites_4 = df.Recurrence_Sites_4.astype('category')\n",
    "        df.Recurrence_Sites_4 = df.Recurrence_Sites_4.cat.set_categories(list(tu_combo))\n",
    "        #break\n",
    "        df_dummy = pd.get_dummies(df.loc[:,[s_vital]],drop_first=False)\n",
    "        df.drop(s_vital,axis=1,inplace=True)\n",
    "        s_compare = \" to \".join((tu_combo))\n",
    "        df[s_compare] = df_dummy.iloc[:,0]\n",
    "        print(len(df))\n",
    "        try:\n",
    "            fig, cph = util.cph_plot(df,s_compare,s_time,s_censor,figsize=(4,1.5))\n",
    "            plt.tight_layout()\n",
    "            plt.close(fig)\n",
    "            df_cph = cph.summary.loc[:,['exp(coef)','exp(coef) lower 95%','exp(coef) upper 95%','p']]\n",
    "            df_cph['n'] = len(df)\n",
    "            df_result=pd.concat([df_result,df_cph])\n",
    "        except:\n",
    "            print('cph error')\n",
    "        break\n",
    "        #save results\n",
    "    #df_result.to_csv(f'results/results_single_CPH_{s_censor}.csv')\n",
    "    df = df_patient.loc[(df_patient.Alive_30_days_post_surgery) &(df_patient.Resected),[s_vital,s_time,s_censor]]\n",
    "    for s_recur in df.Recurrence_Sites_4.unique():\n",
    "        print(f'{s_recur}: Median {s_censor} = {df.loc[df.Recurrence_Sites_4==s_recur,s_time].median()}')\n",
    "    #'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db968bd0-6cec-4a9b-9ab4-bbf6868814d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cph.summary.loc[:,'p']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41b458e1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#GENES\n",
    "ls_vital = ls_genes\n",
    "\n",
    "#CPH single variable\n",
    "\n",
    "tu_time_censor = (('Days from Diagnosis to FU','Survival'),\n",
    "                  ('Days from Resection to Recurrence', 'Recurrence'))\n",
    "for (s_time, s_censor) in tu_time_censor:\n",
    "    print(s_censor)\n",
    "    df_result=pd.DataFrame()\n",
    "    se_recur = {'Other_Recurrence', 'No_Recurrence','Lung_Cohort','Liver_Cohort','PurIST_All','pORG_All','pSUB_All',}\n",
    "    for s_vital in ls_vital:\n",
    "        print(s_vital)\n",
    "        df = df_patient.loc[df_patient.Alive_30_days_post_surgery,[s_vital,s_time,s_censor]].copy()#.dropna(how='any')\n",
    "        df = df.dropna()    \n",
    "        print(len(df))\n",
    "        if df.loc[:,s_vital].dtype=='O':\n",
    "            df_dummy = pd.get_dummies(df.loc[:,[s_vital]],drop_first=True)\n",
    "            if df_dummy.shape[1]>0:\n",
    "                df.drop(s_vital,axis=1,inplace=True)\n",
    "                df[s_vital] = df_dummy\n",
    "            else:\n",
    "                continue\n",
    "        try: \n",
    "            fig, cph = util.cph_plot(df,s_vital,s_time,s_censor,figsize=(3.8,1.5))\n",
    "            pvalue = cph.summary.p[s_vital]\n",
    "            plt.tight_layout()\n",
    "            #ig.savefig(f'figures/CPH_single_{s_vital}_{pvalue:.2}_{s_censor}.png')\n",
    "            #if pvalue > 0.09:\n",
    "            plt.close(fig)\n",
    "            df_cph = cph.summary.loc[:,['exp(coef)','exp(coef) lower 95%','exp(coef) upper 95%','p']]\n",
    "            df_cph['n'] = len(df)\n",
    "            df_result=pd.concat([df_result,df_cph])\n",
    "        except:\n",
    "            print('')\n",
    "        #save results\n",
    "    df_result.to_csv(f'results/results_single_CPH_{s_censor}_gene_alterations.csv')\n",
    "    #'''\n",
    "    #break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4e136a9",
   "metadata": {},
   "source": [
    "## multivariable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d33d410f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CPH multivariable\n",
    "for s_porg in ['pORG','pSUB','PurIST']:\n",
    "    if s_porg == 'PurIST':\n",
    "        df_patient[f'{s_porg}_All'] = df_patient.loc[:,f'{s_porg}_Primary'].fillna(df_patient.loc[:,f'{s_porg}_Met'])\n",
    "    else:\n",
    "        df_patient[f'{s_porg}_All'] = df_patient.loc[:,f'{s_porg}_allPrimary'].fillna(df_patient.loc[:,f'{s_porg}_allMet'])\n",
    "    #break\n",
    "# CDKN2A_Altered# ARID1B_Altered # KMT2D_Altered# MTAP_Altered TP53_Altered\n",
    "# CDKN2A_Altered_Primary  MTAP_Altered_Primary # ARID1B_Altered_Primary # KMT2D_Altered_Primary\n",
    "# KDM6A_Altered_Met # CDKN2A_Altered_Met\n",
    "b_primary = False\n",
    "for (s_time, s_censor) in tu_time_censor:\n",
    "    print(s_censor)\n",
    "    for b_genes in [True,False,]:\n",
    "        if b_genes:\n",
    "            s_covar = 'gene_alterations'\n",
    "            figsize = (5,3)\n",
    "        else:\n",
    "            s_covar = 'clinical_covariates'\n",
    "            figsize = (4,3)\n",
    "        df_result_multi = pd.DataFrame()\n",
    "        ls_multi = ['PurIST_Primary','pORG_Primary','pSUB_Primary', #'pORG_0.2_Met', 'pORG_0.2_All',,'PurIST_Met',\n",
    "                'PurIST_All','pORG_All','pSUB_All',\n",
    "                    'Liver_Cohort','Lung_Cohort',#'pSUB1e-04_Met',#'pSUB1e-04_All'\n",
    "            'Other_Recurrence', 'No_Doc_Recur',\n",
    "                     'Classical_Cohort',\n",
    "               ]\n",
    "        ls_cats = ['LV_Invasion','LN_Pos'] #categorical,'Neoadjuvant'\n",
    "        if b_genes:\n",
    "            ls_cats =['ARID1B_Altered_Primary','KMT2D_Altered_Primary','MTAP_Altered_Primary','CDKN2A_Altered_Primary']# ['CDKN2A_Altered_Primary','ARID1B_Altered_Primary','KMT2D_Altered_Primary','MTAP_Altered_Primary','TP53_Altered_Primary']#\n",
    "        for s_multi in ls_multi:\n",
    "            if not b_primary:\n",
    "                if len(se_recur.intersection(set([s_multi]))) and b_genes:\n",
    "                    ls_cats = ['CDKN2A_Altered','ARID1B_Altered','KMT2D_Altered','MTAP_Altered','TP53_Altered']\n",
    "            print(s_multi)\n",
    "            df = df_patient.loc[df_patient.Alive_30_days_post_surgery,[s_multi,s_time,s_censor,'Grade','Stage','Public_Patient_ID']].copy()#.dropna() #\n",
    "            if ls_cats[0].find('Altered') > -1:\n",
    "                df = df_patient.loc[df_patient.Alive_30_days_post_surgery,[s_multi,s_time,s_censor,'Public_Patient_ID']].copy()#.dropna()\n",
    "            #if len(se_recur.intersection(set([s_multi]))):\n",
    "            #    df.loc[df_patient.Recurrence_Sites_4=='No_Resection',s_multi] = np.nan\n",
    "            df = df.dropna()    \n",
    "            if df.loc[:,s_multi].dtype=='O':\n",
    "                df_dummy = pd.get_dummies(df.loc[df_patient.Alive_30_days_post_surgery,[s_multi]],drop_first=True)\n",
    "                if df_dummy.shape[1]>0:\n",
    "                    df.drop(s_multi,axis=1,inplace=True)\n",
    "                    s_multi = df_dummy.columns[0]\n",
    "                    df[s_multi] = df_dummy\n",
    "                else:\n",
    "                    continue\n",
    "            df_dummy = pd.get_dummies(df_patient.loc[df_patient.Alive_30_days_post_surgery,ls_cats+['Public_Patient_ID']].dropna().set_index('Public_Patient_ID'),drop_first=True) #\n",
    "            df = df.merge(df_dummy.reset_index(),on='Public_Patient_ID',how='left').set_index('Public_Patient_ID')\n",
    "            df = df.dropna()\n",
    "            print(len(df))\n",
    "            try:\n",
    "                fig, cph = util.cph_plot(df,s_multi,s_time,s_censor,figsize=figsize)\n",
    "                fig.savefig(f'figures/CPH_multi_{s_multi}_{s_covar}_{s_censor}.png')\n",
    "                #plt.close(fig)\n",
    "                df_result_model = cph.summary.loc[:,['exp(coef)','exp(coef) lower 95%','exp(coef) upper 95%','p']].reset_index()\n",
    "                df_result_model['model'] = s_multi\n",
    "                df_result_model['n'] = len(df)\n",
    "                df_result_multi=pd.concat([df_result_multi,df_result_model])\n",
    "            except:\n",
    "                print('cph error')\n",
    "        #     break\n",
    "    #     break\n",
    "    # break\n",
    "    #all the recurrence combos - multi\n",
    "    #'''\n",
    "        s_vital = 'Recurrence_Sites_4'\n",
    "        figsize=(5,3)\n",
    "        for tu_combo in combinations(df_patient.Recurrence_Sites_4.unique(),r=2):\n",
    "            print(tu_combo)\n",
    "            if b_genes:\n",
    "                df = df_patient.loc[df_patient.Alive_30_days_post_surgery,[s_vital,s_time,s_censor]]\n",
    "            else:\n",
    "                df = df_patient.loc[df_patient.Alive_30_days_post_surgery,[s_vital,s_time,s_censor,'Stage','Grade']]#.dropna(how='any')\n",
    "            df = df[df.loc[:,s_vital].isin(tu_combo)].dropna()\n",
    "            #order them\n",
    "            df.Recurrence_Sites_4 = df.Recurrence_Sites_4.astype('category')\n",
    "            df.Recurrence_Sites_4 = df.Recurrence_Sites_4.cat.set_categories((tu_combo))\n",
    "            df_dummy = pd.get_dummies(df.loc[df_patient.Alive_30_days_post_surgery,[s_vital]],drop_first=False)\n",
    "            df.drop(s_vital,axis=1,inplace=True)\n",
    "            s_compare = \" to \".join((tu_combo))\n",
    "            df[s_compare] = df_dummy.iloc[:,0]\n",
    "            df_dummy = pd.get_dummies(df_patient.loc[:,ls_cats].dropna(),drop_first=True) \n",
    "            df = pd.concat([df,df_dummy],axis=1)\n",
    "            df = df.dropna()\n",
    "            print(len(df))\n",
    "            try:\n",
    "                fig, cph = util.cph_plot(df,s_compare,s_time,s_censor,figsize=figsize)\n",
    "                plt.tight_layout()\n",
    "                #fig.savefig(f'figures/CPH_mutli_{s_vital}_{s_compare}_{s_covar}_{s_censor}.png')\n",
    "                plt.close(fig)\n",
    "                df_result_model = cph.summary.loc[:,['exp(coef)','exp(coef) lower 95%','exp(coef) upper 95%','p']].reset_index()\n",
    "                df_result_model['model'] = s_compare\n",
    "                df_result_model['n'] = len(df)\n",
    "                df_result_multi=pd.concat([df_result_multi,df_result_model])\n",
    "            except:\n",
    "                print('cph error')\n",
    "        \n",
    "        df_result_multi.reset_index(drop=True).to_csv(f'results/results_multi_CPH_{s_covar}_{s_censor}.csv')\n",
    "        #'''\n",
    "        \n",
    "    #     break\n",
    "    # break\n",
    "# save with edits\n",
    "#df_patient.to_csv('annotation/20231206_Patient_Metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "916b81e9-6371-41e3-b6e4-674ee257dc4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_result_multi.reset_index(drop=True).to_csv(f'results/results_multi_CPH_{s_covar}_{s_censor}_All.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0c79317",
   "metadata": {},
   "source": [
    "## Cox PH plots <a name=\"cphplot\"></a>\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "270c0eee",
   "metadata": {},
   "outputs": [],
   "source": [
    "def forrest_plot(df_plot,s_title,figsize=(4,2.5),ymargin=0.1):\n",
    "    '''\n",
    "    plot hazard ratios as error bars\n",
    "    df_plot: dataframe with HR 'exp(coef)', confidence 'ci', pvalue 'p'\n",
    "    s_title: axis title string\n",
    "    '''\n",
    "    #colors\n",
    "    fig,ax = plt.subplots(figsize=figsize,dpi=300)\n",
    "    if len(df_plot) > 0:\n",
    "        df_plot['color'] = 'k'\n",
    "        df_plot.loc[((df_plot.loc[:,'exp(coef)']<1)&(df_plot.p<0.05)),'color'] = '#3D65A5'#'limegreen'\n",
    "        df_plot.loc[((df_plot.loc[:,'exp(coef)']>1)&(df_plot.p<0.05)),'color'] = '#F05039'\n",
    "        for idx, row in df_plot.iterrows():\n",
    "            ax.errorbar(x=row['exp(coef)'], y=row.name, \n",
    "                        xerr=np.array([row.ci[0],row.ci[1]]).reshape(2,1), #row.ci,#[0],\n",
    "                    ecolor=\"k\", capsize=3, linestyle='None', linewidth=1, marker=\"s\", \n",
    "                             markersize=5, mfc=row.color, mec=row.color)\n",
    "        ax.axvline(x=1, linewidth=0.8, linestyle='--', color='black')\n",
    "        ax.set_xlabel('Hazard Ratio')\n",
    "        ax.set_title(s_title)\n",
    "        # for multivariable\n",
    "        if df_plot.columns.isin(['model']).any():\n",
    "            s_n = f\"{df_plot.loc[df_plot.covariate==s_title,'n'][0]}\"\n",
    "            s_p = f\"{df_plot.loc[df_plot.covariate==s_title,'p'][0]:.2}\"\n",
    "            s_hr = f\"{df_plot.loc[df_plot.covariate==s_title,'exp(coef)'][0]:.2}\"\n",
    "            ax.set_title(f'{s_title}\\nHR={s_hr} p={s_p} n={s_n}')\n",
    "        ax.margins(y=ymargin)\n",
    "        plt.tight_layout()\n",
    "    return(fig,ax,df_plot)\n",
    "\n",
    "def make_ci(df):\n",
    "    df['upper'] = df['exp(coef) upper 95%'] - df['exp(coef)']\n",
    "    df['lower'] = df['exp(coef)'] - df['exp(coef) lower 95%'] \n",
    "    df['ci'] = list(zip(df['lower'], df['upper']))\n",
    "    return(df)\n",
    "\n",
    "def bold_title(ax,s_title):\n",
    "    labels = ax.get_yticklabels()\n",
    "    for label in labels:\n",
    "        if label.get_text() == s_title:\n",
    "            label.set_fontweight('bold')\n",
    "        else:\n",
    "            continue\n",
    "def title_replace(s_model): #\n",
    "    s_new = s_model.replace('Classical_Cohort_Lung','Classical_Lung_vs_Classical_Liver').replace('_',' ').replace('True','').replace(' YES','').replace(' Classical ',' ')\n",
    "    return(s_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b73b09d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load patient metadata#fix this #start here\n",
    "%matplotlib inline\n",
    "# s_out = 'annotation/20231219_Patient_Metadata.csv'#'annotation/20230921_Patient_Metadata.csv'\n",
    "# df_patient = pd.read_csv(s_out,index_col=0)\n",
    "df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "                          index_col=0)\n",
    "# #add recurrence censor\n",
    "# df_patient.loc[df_patient.loc[:,'Days from Resection to Recurrence'].notna(),'Recurrence'] = 1 #event observed\n",
    "# df_patient.loc[df_patient.No_Recurrence.fillna(False),'Recurrence'] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55ef7e84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load saved CPH results\n",
    "# for s_file in os.listdir('results'):\n",
    "#     if s_file.find('results_multi') > -1:\n",
    "#         print(s_file)\n",
    "d_cph = pd.read_excel('Supplemental_Data/Supplemental_Dataset_2.xlsx',sheet_name=None)\n",
    "list(d_cph.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fa6d0b5-754f-4fcd-83ec-40076252ce80",
   "metadata": {},
   "outputs": [],
   "source": [
    "#single compare resected\n",
    "#df = pd.read_csv(f'results/results_single_CPH_Survival.csv',index_col=0)\n",
    "df = d_cph['single_CPH_Survival.csv']\n",
    "df=make_ci(df)\n",
    "df.set_index('covariate',inplace=True)\n",
    "df.rename({'Classical_Cohort':'Classical_Lung_vs_Liver'},axis=0,inplace=True)\n",
    "# ls_row = ['Liver to No_Doc_Recur', 'Liver to Lung', 'Liver to Other_site',\n",
    "#        'Lung to No_Doc_Recur', 'No_Doc_Recur to Other_site',\n",
    "#        'Lung to Other_site']\n",
    "# ls_row = ['Lung to Liver', 'Lung to Other_site',\n",
    "#        'Lung to No_Doc_Recur', 'Liver to Other_site',\n",
    "#        'Liver to No_Doc_Recur', 'Other_site to No_Doc_Recur']\n",
    "ls_row = df.index[-6::]\n",
    "df_plot = df.loc[ls_row]\n",
    "#df_plot.index = [item.replace('_',' ').replace('Lung','Lung Cohort').replace('Liver','Liver Cohort') for item in df_plot.index]\n",
    "df_plot.index = [item.replace('_',' ') for item in df_plot.index]\n",
    "#df_plot.sort_index(ascending=False,inplace=True)\n",
    "# # plot\n",
    "s_title='Resected Patients'\n",
    "s_title='Single Variable CPH\\nSurvival'\n",
    "fig,ax, __ = forrest_plot(df_plot,s_title,figsize=(3.5,2.5))\n",
    "plt.tight_layout()\n",
    "fig.savefig(f'figures/cph_final_cohorts_compare.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4809a0b9-02e7-43b1-9c23-e5eeabbff8c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_fig1.update({'Figure1_B_cohorts':df_plot})\n",
    "d_fig1.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5044206a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "#########################################################\n",
    "#compare to all patients\n",
    "\n",
    "lls_row = [['PurIST_Subtype','Classical_Lung_vs_Liver'], #,'Classical_Cohort'\n",
    "           [ 'Liver_Cohort','Other_Recurrence','No_Doc_Recur','Lung_Cohort',\n",
    "         ]]\n",
    "for ls_row in lls_row:\n",
    "    df_plot = df.loc[ls_row].copy()\n",
    "    df_plot.index = [item.replace('_',' ') for item in df_plot.index]\n",
    "    df_plot = df_plot.iloc[::-1]\n",
    "    # # plot\n",
    "    s_title='Single Variable CPH: Survival'\n",
    "    if len(ls_row)==2:\n",
    "        fig,ax,__ = forrest_plot(df_plot,s_title,figsize=(1,0.8),ymargin=0.2)\n",
    "    else:\n",
    "        fig,ax,__ = forrest_plot(df_plot,s_title,figsize=(2.5,2),ymargin=0.1)\n",
    "    plt.tight_layout()\n",
    "    fig.savefig(f'figures/cph_final_cohorts_{len(ls_row)}.pdf')\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52725b19",
   "metadata": {},
   "outputs": [],
   "source": [
    "#single compare resected\n",
    "#df = pd.read_csv(f'results/results_single_CPH_Recurrence.csv',index_col=0)\n",
    "df = d_cph['single_CPH_Recurrence.csv']\n",
    "df=make_ci(df)\n",
    "df.set_index('covariate',inplace=True)\n",
    "\n",
    "#compare to all patients\n",
    "ls_row = df.index[-3::]#['Liver to Lung', 'Liver to Other_site', 'Lung to Other_site']\n",
    "df_plot = df.loc[ls_row].copy()\n",
    "df_plot.index = [item.replace('_',' ') for item in df_plot.index]\n",
    "df_plot = df_plot.iloc[::-1]\n",
    "# # plot\n",
    "s_title='Single Variable CPH: Recurrence'\n",
    "if len(ls_row)==2:\n",
    "    fig,ax,__ = forrest_plot(df_plot,s_title,figsize=(1,0.8),ymargin=0.2)\n",
    "else:\n",
    "    fig,ax,__ = forrest_plot(df_plot,s_title,figsize=(3,2),ymargin=0.1)\n",
    "plt.tight_layout()\n",
    "fig.savefig(f'figures/cph_final_cohorts_recur.pdf')\n",
    "#df.iloc[-3::,0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e98e930b-fbe2-4a86-84af-e3bd0245172e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted(os.listdir('results/CPH'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e560f1b-42f4-4e5e-8ea9-c523ab28d89c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.model.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dc04ac1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# clinical covariates/genetics\n",
    "#results_multi_CPH_clinical_covariates_Recurrence_JE.csv\n",
    "['multi_CPH_clinical_covariates_R',\n",
    " 'multi_CPH_clinical_covariates_S',\n",
    " 'multi_CPH_gene_alterations_Recu',\n",
    " 'multi_CPH_gene_alterations_Surv',\n",
    " 'single_CPH_Recurrence.csv',\n",
    " 'single_CPH_Recurrence_gene_alte',\n",
    " 'single_CPH_Survival.csv',\n",
    " 'single_CPH_Survival_gene_altera',\n",
    " 'single_CPH_TCR_Recurrence.csv',\n",
    " 'single_CPH_TCR_Survival.csv']\n",
    "ls_model = ['PurIST_Primary',  'pSUB_Primary', 'PurIST_All', 'pSUB_All',]\n",
    "df_out = pd.DataFrame()\n",
    "for s_variable in ['Survival' , 'Recurrence']:\n",
    "    for s_covar in ['gene_alterations','clinical_covariates',]:#'gene_alterations_all',\n",
    "        if s_covar == 'clinical_covariates':\n",
    "            s_file = f'multi_CPH_{s_covar}_{s_variable[0:1]}'\n",
    "        elif s_covar == 'gene_alterations_all':\n",
    "            s_file = f'results_multi_CPH_{s_covar.split(\"_all\")[0]}_{s_variable}_All.csv'\n",
    "        else:\n",
    "            s_file = f'multi_CPH_{s_covar}_{s_variable[0:4]}'\n",
    "        #df = pd.read_csv(f'results/{s_file}',index_col=0)\n",
    "        df = d_cph[s_file]\n",
    "        df = make_ci(df)\n",
    "        d_rename = {'pORG_0.2_Primary':'pORG_Primary', 'pSUB1e-04_Primary':'pSUB_Primary',}\n",
    "        for s_col in ['covariate','model']:\n",
    "            df[s_col] = df[s_col].replace(d_rename)\n",
    "        with warnings.catch_warnings():\n",
    "            warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "            for s_model in ls_model:#df.model.unique(): #\n",
    "                print(s_model)\n",
    "                df_plot = df[df.model==s_model].copy()\n",
    "                df_plot.rename(d_rename,inplace=True)\n",
    "                df_plot.covariate = [title_replace(item) for item in df_plot.covariate]\n",
    "                df_plot.index = df_plot.covariate\n",
    "                s_title = title_replace(s_model)\n",
    "                df_plot.sort_values(by='exp(coef)',inplace=True)\n",
    "                print(s_title)\n",
    "                if s_model.find('to') > -1:\n",
    "                    figsize=(3.5,2.5)\n",
    "                elif s_model.find('vs') > -1:\n",
    "                    figsize=(3,2.5)\n",
    "                else:\n",
    "                    figsize=(2.8,2.5)\n",
    "                    #figsize = (2.5,2)\n",
    "                fig,ax,__ = forrest_plot(df_plot,s_title,figsize=figsize)\n",
    "                bold_title(ax,s_title)\n",
    "                plt.tight_layout()\n",
    "                fig.savefig(f'figures/cph_final_{s_title}_{s_variable}_{s_covar}.pdf')\n",
    "                df_out = pd.concat([df_out,df_plot])\n",
    "                if s_covar== 'clinical_covariates':\n",
    "                    if s_variable == 'Survival':\n",
    "                        if s_model == 'Liver_Cohort':\n",
    "                            d_fig1.update({f'Figure1_C_Liver':df_plot})\n",
    "                        elif s_model == 'Lung_Cohort':\n",
    "                            d_fig1.update({f'Figure1_C_Lung':df_plot})\n",
    "                        elif s_model == 'Other_Recurrence_True':\n",
    "                            d_fig1.update({f'Figure1_C_Other':df_plot})\n",
    "                        elif s_model == 'No_Doc_Recur_True':\n",
    "                            d_fig1.update({f'Figure1_C_NoDoc':df_plot})\n",
    "                        elif s_model =='Classical_Cohort_Lung':\n",
    "                            d_fig1.update({f'Figure1_H':df_plot})\n",
    "                        elif s_model =='PurIST_Primary':\n",
    "                            d_fig2.update({f'Figure2_C_PurIST':df_plot})\n",
    "                        elif s_model =='pORG_Primary':\n",
    "                            d_fig2.update({f'Figure2_C_pORG_Primary':df_plot})\n",
    "                        elif s_model =='pSUB_Primary':\n",
    "                            d_fig2.update({f'Figure2_C_pSUB_Primary':df_plot})\n",
    "                    elif s_variable == 'Recurrence':\n",
    "                        if s_model == 'Liver_Cohort':\n",
    "                            d_fig1.update({f'Figure1_E_Liver':df_plot})\n",
    "                        elif s_model == 'Lung_Cohort':\n",
    "                            d_fig1.update({f'Figure1_E_Lung':df_plot})\n",
    "                        elif s_model == 'Other_Recurrence_True':\n",
    "                            d_fig1.update({f'Figure1_E_Other':df_plot})\n",
    "                elif s_covar== 'gene_alterations':\n",
    "                    if s_variable == 'Survival':\n",
    "                        if s_model == 'pORG_Primary':\n",
    "                            d_fig2.update({f'Figure2_K_pORG_Pri':df_plot})\n",
    "                        elif s_model == 'pORG_All':\n",
    "                            d_fig2.update({f'Figure2_K_pORG_all':df_plot})\n",
    "        break #clinical/genomics\n",
    "    break #os, recurrence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7672d18-6e36-4f99-9afb-4a7b9037290c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# d_fig2.keys()\n",
    "# df_out.drop('Unnamed: 0',axis=1).to_csv('Source_Extended_Data_Fgiure4_D.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3bd52ff-f21f-4fdf-a463-f81557c57454",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # #util.open_write_excel(d_fig1,filename='Source_Data_Figure1.xlsx')\n",
    "# # util.open_write_excel(d_fig2,filename='Source_Data_Figure2.xlsx')\n",
    "# s_find = 'Source_Extended_Data_Figure4'\n",
    "# d_fig = excel_dict_from_csvs(s_find,s_prefix='ED_Fig4')\n",
    "# util.open_write_excel(d_fig,filename='Source_Extended_Data_Figure4.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8c480d1",
   "metadata": {},
   "source": [
    "# Section 6 <a name=\"geneexp\"></a>\n",
    "\n",
    "gene expresison correlation\n",
    "/\n",
    "The 7 genes used for the IRDS signature were: STAT1, IFI44, IFIT3, OAS1,208\n",
    "IFIT1, G1P2, and MX1\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c288c22e-d31e-4aae-ba47-6efb1b7b8192",
   "metadata": {},
   "source": [
    "## Correlation heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5e75232",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "                          index_col=0)\n",
    "#download from synapse https://www.synapse.org/Synapse:syn53977850 \n",
    "df_vst = pd.read_csv('data/VST_Genes_Link.csv',index_col=0) #download from synapse https://www.synapse.org/Synapse:syn53977850\n",
    "df_rna = df_vst.T.copy()\n",
    "df_rna.index = [item[0:-2] for item in df_rna.index]\n",
    "print(len(df_rna))\n",
    "\n",
    "ls_add = ['pORG_Primary','pORG_Met','pSUB_Primary','PurIST_Primary']\n",
    "for s_add in ls_add:\n",
    "    d_map = dict(zip(df_patient.loc[:,'Public_Patient_ID'],df_patient.loc[:,s_add]))\n",
    "    df_rna[s_add] = df_rna.index.map(d_map)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c00d73f7-01bf-49f6-80e6-04c47a1e05de",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_rna.columns[df_rna.columns.str.contains('USP18')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db8c2fde",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "d_rename={'txi_pORG_Up_42_Genes':'pORG_42',\n",
    "          'trim_padj_0.2_pORG_Up_55_Genes':'pORG_55',\n",
    "          'pORG.14':'pORG_14', 'pORG.15':'pORG_15',\n",
    "          'pORG_0.2_Primary':'pORG_Primary',\n",
    "          'pSUB1e-04_Primary':'pSUB1e-04',\n",
    "          'MRC1':'MRC1 (CD206)',\n",
    "          'OLR1':'OLR1 (LOX1)',\n",
    "          'MS4A1':'MS4A1 (CD20)'}\n",
    "from scipy.stats import pearsonr\n",
    "with warnings.catch_warnings():\n",
    "    warnings.filterwarnings(\"ignore\",category=FutureWarning)\n",
    "    dim = (5,4)\n",
    "    for s_add in ls_add:\n",
    "        ls_marker = ['CD3E','CD4', 'CD8A', 'MS4A1',#'LAMP3','ITGAX','CD209',\n",
    "                     'CD68','OLR1','MRC1','MX1','STAT1',\n",
    "                     'IFNA1','IFNB1','IFNG','IFNAR1',\n",
    "                     s_add] #\n",
    "        df_all = df_rna.loc[:,ls_marker].corr().rename(d_rename,axis=1).rename(d_rename,axis=0)\n",
    "        print(len(df_rna))\n",
    "        g = sns.clustermap(df_all)\n",
    "        plt.close()\n",
    "        categories_order = df_all.iloc[g.dendrogram_col.reordered_ind,:].index.tolist()\n",
    "        df_all = df_all.loc[categories_order,categories_order]\n",
    "        rho = df_rna.loc[:,ls_marker].corr() #df_all.corr()\n",
    "        pval = df_rna.loc[:,ls_marker].corr(method=lambda x, y: pearsonr(x, y)[1]) - np.eye(*rho.shape)\n",
    "        p_vals = pval.applymap(lambda x: ''.join(['*' for t in [0.001,0.005,0.05] if x<=t]))\n",
    "        p_vals = p_vals.rename(d_rename,axis=1).rename(d_rename,axis=0)\n",
    "        p_vals = p_vals.loc[categories_order,categories_order]\n",
    "        #fdr correct\n",
    "        pvalues = pval.loc[s_add,~pval.columns.isin([s_add])]\n",
    "        reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='fdr_bh')\n",
    "        df_corrected = pd.DataFrame(np.ones_like(pval),index=pval.index,columns=pval.columns)\n",
    "        df_corrected.loc[s_add,~df_corrected.columns.isin([s_add])] = corrected\n",
    "        df_corrected.loc[~df_corrected.columns.isin([s_add]),s_add] = corrected\n",
    "        p_vals = pval.applymap(lambda x: ''.join(['*' for t in [0.001,0.005,0.05] if x<=t]))\n",
    "        p_vals = df_corrected.applymap(lambda x: ''.join(['*' for t in [0.001,0.005,0.05] if x<=t]))\n",
    "        p_vals = p_vals.rename(d_rename,axis=1).rename(d_rename,axis=0)\n",
    "        p_vals = p_vals.loc[categories_order,categories_order]\n",
    "        fig, ax = plt.subplots(figsize=dim,dpi=300)\n",
    "        sns.heatmap(df_all, vmin=-1, vmax=1, annot=p_vals, fmt = '', cmap='RdBu_r',ax=ax,\n",
    "                   cbar_kws={'shrink':0.85,'label':'Pearson Correlation'})\n",
    "        ax.set_title(f'N={len(df_rna)}')\n",
    "        # #triangle layout\n",
    "        # plt.title(f'{s_add} n={len(df_rna)}')\n",
    "        # fig, ax = plt.subplots(figsize=dim,dpi=300)\n",
    "        # matrix = np.triu(np.ones_like(rho))\n",
    "        # np.fill_diagonal(matrix, val=0)\n",
    "        # np.fill_diagonal(p_vals.values,'')\n",
    "        # sns.heatmap(df_all, vmin=-1, vmax=1, annot=p_vals, fmt = '', cmap='RdBu_r',\n",
    "        #             ax=ax,mask=matrix,cbar_kws={'shrink':0.85,'label':'Pearson Correlation'},\n",
    "        #            ) #'anchor':(-1.4,0.0)\n",
    "        if s_add == 'pORG_Primary':\n",
    "            ax.set_ylabel('Neutrophil Macrophage Intrinsic IFN   IFN Ligands   Lympocytes                      ',fontsize='xx-small')\n",
    "            trans = ax.get_xaxis_transform()\n",
    "            ls_lest = [-6.2,-6.2]\n",
    "            ax.plot(ls_lest,[0.01,0.08], color=\"k\", transform=trans, clip_on=False,linewidth=1)\n",
    "            ax.plot(ls_lest,[0.1,0.22], color=\"k\", transform=trans, clip_on=False,linewidth=1)\n",
    "            ax.plot(ls_lest,[0.25,0.42], color=\"k\", transform=trans, clip_on=False,linewidth=1)\n",
    "            ax.plot(ls_lest,[0.5,0.64], color=\"k\", transform=trans, clip_on=False,linewidth=1)\n",
    "            ax.plot(ls_lest,[0.66,.92], color=\"k\", transform=trans, clip_on=False,linewidth=1)\n",
    "        plt.tight_layout()\n",
    "        fig.savefig(f'figures/heatmap_genes_{s_add}.pdf')\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41babe86-6821-4b6c-bb68-c95ecae11db4",
   "metadata": {},
   "outputs": [],
   "source": [
    "[print(f'{item:.2}') for item in df_corrected.rename(d_rename).loc[categories_order].pORG_Primary]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcd23be7-b7aa-40b3-b8c9-d0c2d27de9a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_all.to_csv(f'Source_Extended_Data_Figure6_G.csv')\n",
    "#p_vals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29f7d0c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# viper\n",
    "df_viper = pd.read_excel(f'Supplemental_Data/Supplemental_Dataset_6.xlsx',index_col=0)\n",
    "df_viper.columns = [item[0:-2] for item in df_viper.columns]\n",
    "df_viper = df_viper.T\n",
    "#ls_add = ['pORG_0.2_Primary','pSUB1e-04_Primary','PurIST_Primary']\n",
    "for s_add in ls_add:\n",
    "    d_map = dict(zip(df_patient.loc[:,'Public_Patient_ID'],df_patient.loc[:,s_add]))\n",
    "    df_viper[s_add] = df_viper.index.map(d_map)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c78a859f-084e-49ae-a123-ca6c01571b66",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_viper.columns[df_viper.columns.str.contains('IFN')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a854a044",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "d_rename={'txi_pORG_Up_42_Genes':'pORG_42',\n",
    "          'trim_padj_0.2_pORG_Up_55_Genes':'pORG_55',\n",
    "          'pORG.14':'pORG_14', 'pORG.15':'pORG_15',\n",
    "          'pORG_0.2_Primary':'pORG_0.2',\n",
    "          'pSUB1e-04_Primary':'pSUB1e-04',\n",
    "          'MRC1':'MRC1 (CD206)',\n",
    "          'OLR1':'OLR1 (LOX1)',\n",
    "          'MS4A1':'MS4A1 (CD20)',#'CD274':'CD274 (PD-L1)'\n",
    "         }\n",
    "from scipy.stats import pearsonr\n",
    "dim = (5,4)\n",
    "with warnings.catch_warnings():\n",
    "    warnings.filterwarnings(\"ignore\",category=FutureWarning)\n",
    "    for s_add in ls_add:\n",
    "        ls_marker = ['CD3G', 'MS4A1',#'CD274',#'ITGAX',#'CD209','CD4', 'CD8A','LAMP3',\n",
    "                    'IFNAR1','IFNAR2', 'IFNGR2','IFNG','IFNB1',#'CD164',#'FOXP3',#'OLR1','MRC1',\n",
    "                     'MX1','STAT1',#'CD68',\n",
    "                     s_add] #\n",
    "        df_all = df_viper.loc[:,ls_marker].corr().rename(d_rename,axis=1).rename(d_rename,axis=0)\n",
    "        print(len(df_viper))\n",
    "        g = sns.clustermap(df_all)\n",
    "        plt.close()\n",
    "        categories_order = df_all.iloc[g.dendrogram_col.reordered_ind,:].index.tolist()\n",
    "        df_all = df_all.loc[categories_order,categories_order]\n",
    "        rho = df_viper.loc[:,ls_marker].corr() #df_all.corr()\n",
    "        pval = df_viper.loc[:,ls_marker].corr(method=lambda x, y: pearsonr(x, y)[1]) - np.eye(*rho.shape)\n",
    "        #fdr correct\n",
    "        pvalues = pval.loc[s_add,~pval.columns.isin([s_add])]\n",
    "        reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='fdr_bh')\n",
    "        df_corrected = pd.DataFrame(np.ones_like(pval),index=pval.index,columns=pval.columns)\n",
    "        df_corrected.loc[s_add,~df_corrected.columns.isin([s_add])] = corrected\n",
    "        df_corrected.loc[~df_corrected.columns.isin([s_add]),s_add] = corrected\n",
    "        p_vals = pval.applymap(lambda x: ''.join(['*' for t in [0.001,0.005,0.05] if x<=t]))\n",
    "        p_vals = df_corrected.applymap(lambda x: ''.join(['*' for t in [0.001,0.005,0.05] if x<=t]))\n",
    "        p_vals = p_vals.rename(d_rename,axis=1).rename(d_rename,axis=0)\n",
    "        p_vals = p_vals.loc[categories_order,categories_order]\n",
    "        fig, ax = plt.subplots(figsize=dim,dpi=300)\n",
    "        sns.heatmap(df_all, vmin=-1, vmax=1, annot=p_vals, fmt = '', cmap='RdBu_r',ax=ax,\n",
    "                   cbar_kws={'shrink':0.85,'label':'Pearson Correlation'})\n",
    "        ax.set_title(f'N={len(df_viper)}')\n",
    "        plt.tight_layout()\n",
    "        fig.savefig(f'figures/Figure3_I_viper_IFN_{s_add}.pdf')\n",
    "        #d_result_fig.update({f'Figure3_I_viper_IFN_{s_add}':df_all})\n",
    "        #util.open_write_excel(d_result_fig,filename='Source_Data_Figure3.xlsx')\n",
    "        # #triangle\n",
    "        # plt.title(f'{s_add} n={len(df_viper)}')\n",
    "        # fig, ax = plt.subplots(figsize=dim,dpi=300)\n",
    "        # matrix = np.triu(np.ones_like(rho))\n",
    "        # np.fill_diagonal(matrix, val=0)\n",
    "        # np.fill_diagonal(p_vals.values,'')\n",
    "        # sns.heatmap(df_all, vmin=-1, vmax=1, annot=p_vals, fmt = '', cmap='RdBu_r',\n",
    "        #             ax=ax,mask=matrix,cbar_kws={'shrink':0.85,'label':'Pearson Correlation'},\n",
    "        #            ) #'anchor':(-1.4,0.0)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e7afcc4-62a9-4149-b3bc-6c925b674a6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "[print(f'{item:.2}') for item in df_corrected.rename(d_rename).loc[categories_order].pORG_Primary]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "576aa441-e58a-42b8-895b-07c4b6b30eef",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_viper['Cohort']=df_viper.index.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Cohort)))\n",
    "df_viper.columns[df_viper.columns.str.contains('MYC')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed15b5b8-7151-4889-a4a1-2377721efda1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_viper.loc[df_viper.pORG_Primary > 0.02,'pORG_Pri'] = 'pORG high' \n",
    "df_viper.loc[df_viper.pORG_Primary < 0.02,'pORG_Pri'] = 'pORG low' \n",
    "df_viper.loc[df_viper.PurIST_Primary > 0.5,'PurIST'] = 'basal-like'\n",
    "df_viper.loc[df_viper.PurIST_Primary < 0.5,'PurIST'] = 'classical'\n",
    "df_viper['quadrant']=df_viper.pORG_Pri + '/'+ df_viper.PurIST\n",
    "#df_viper.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12603818-f960-4905-a959-94d7724e027d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_gsva.reset_index(drop=True).to_csv('data/results_GSVA_3_mets.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8e689cc-e1ff-4a0d-8a5d-f9b78286145e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ##\n",
    "# # df_id = pd.read_csv('../annotation/Patient_IDs.csv')\n",
    "# # d_id = dict(zip(df_id.loc[:,'OPTR.Specimen.ID'],df_id.loc[:,'Biolibrary.Subject.ID']))\n",
    "# # df_gsva['Public_Patient_ID'] = df_gsva.index.map(d_id)\n",
    "# df_gsva = pd.read_csv('../R/results_GSVA_3_mets.csv',index_col=0) #results_GSVA_3_mets\n",
    "# df_gsva.index = [item.replace('X','').replace('.','-') for item in df_gsva.index]\n",
    "# d_id = dict(zip(df_id.loc[:,'OPTR.Specimen.ID'],df_id.loc[:,'Public.Specimen.ID']))\n",
    "# df_gsva['Public_Specimen_ID'] = df_gsva.index.map(d_id)\n",
    "# df_gsva.reset_index(drop=True).to_csv('data/results_GSVA_3_mets.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f93064ab-e23c-4810-9cea-678bf1a8da82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## load hallmarks\n",
    "# # df_gsva.reset_index(drop=True).to_csv('data/results_GSVA_2.csv')\n",
    "# # (df_id.OPTR==3084).any()\n",
    "# # ##\n",
    "# # df_id = pd.read_csv('../annotation/Patient_IDs.csv')\n",
    "# # d_id = dict(zip(df_id.loc[:,'OPTR.Specimen.ID'],df_id.loc[:,'Biolibrary.Subject.ID']))\n",
    "# # df_gsva['Public_Patient_ID'] = df_gsva.index.map(d_id)\n",
    "# #df_gsva = pd.read_csv('../R/results_GSVA_2.csv',index_col=0) #results_GSVA_3_mets\n",
    "# df_gsva = pd.read_csv('data/results_GSVA_2.csv',index_col=0)\n",
    "# #df_gsva.index = [item.replace('X','').replace('.','-') for item in df_gsva.index]\n",
    "# df_gsva.rename({'neutrophil activation involved in immune response':'neut. act. invol. in imm. resp.',\n",
    "#                'T cell activation involved in immune response':'T cell act. invol. in imm. resp.'},axis=1,inplace=True)\n",
    "# df_viper = df_viper.reset_index().rename({'index':'Public_Patient_ID'},axis=1)\n",
    "# df_viper = df_viper.merge(df_gsva,on='Public_Patient_ID',how='left',suffixes=('','_x'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e5bbdb5-95e1-4cdf-8c0e-296ff21c39b4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# #quadrant\n",
    "# ls_hue =  ['HALLMARK_MYC_TARGETS_V1','MYC','IFNAR1']\n",
    "# with warnings.catch_warnings():\n",
    "#     warnings.filterwarnings(\"ignore\",category=FutureWarning)\n",
    "#     for hue in ls_hue:\n",
    "#         fig, ax = plt.subplots(dpi=200)\n",
    "#         sns.scatterplot(data=df_viper,style='Cohort',\n",
    "#                         x='pORG_Primary',y='PurIST_Primary',hue=hue,ax=ax,alpha=.7,\n",
    "#                        palette='YlGnBu')\n",
    "#         ax.axvline(0.02,color='k',alpha=0.7,linestyle='--')\n",
    "#         ax.axhline(0.5,color='k',alpha=0.7,linestyle='--')\n",
    "#         ax.legend(bbox_to_anchor=(1,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a0668ef-319f-4078-9ba2-c75006450c82",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# x = 'quadrant'\n",
    "# with warnings.catch_warnings():\n",
    "#     warnings.filterwarnings(\"ignore\",category=FutureWarning)\n",
    "#     for y in ls_hue:\n",
    "#         ls_groups =['pORG low/classical', 'pORG high/classical',\n",
    "#            'pORG high/basal-like', 'pORG low/basal-like',] #df_viper.loc[:,x].dropna().unique()\n",
    "#         plotting = {\"data\": df_viper,\"x\":x,\"y\":y,\"order\":ls_groups}\n",
    "#         fig,ax,test_results = util.annotated_stripplot(plotting,ls_groups,y)\n",
    "#         ax.set_xticklabels(ax.get_xticklabels(),rotation=45)\n",
    "#         if y.find('_') == -1:\n",
    "#             ax.set_ylabel(f\"{y} Regulon\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9adbad5-4a55-4ebf-a55a-833334bfc990",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #quadrant\n",
    "# with warnings.catch_warnings():\n",
    "#     warnings.filterwarnings(\"ignore\",category=FutureWarning)\n",
    "#     for hue in ['Cohort']:\n",
    "#         fig, ax = plt.subplots(dpi=200)\n",
    "#         sns.scatterplot(data=df_patient,#[df_patient.Clonality_Tumor.notna()],\n",
    "#                         x='pORG_Primary',y='PurIST_Primary',hue=hue,ax=ax,alpha=.7)\n",
    "#         ax.axvline(0.02,color='k',alpha=0.7,linestyle='--')\n",
    "#         ax.axhline(0.5,color='k',alpha=0.7,linestyle='--')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b0ce485-4e4f-4788-94f1-e8f83c4bdda7",
   "metadata": {},
   "source": [
    "## Quadrant plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9ddfa52-97b4-477f-8ce8-498a81eb360c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pal_liv_r\n",
    "df_spec.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36c38c64-87ba-4ca0-8b29-c1e3096866ce",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sns.set_palette(('#0072B2','#D55E00'))\n",
    "df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "                          index_col=0)\n",
    "df_spec =  pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name='RNA_Specimen_Metadata',\n",
    "                          index_col=0)\n",
    "df_spec['Cohort'] = df_spec.index.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Cohort)))\n",
    "df_spec.rename({'Tumor_Type':'Type'},axis=1,inplace=True)\n",
    "with warnings.catch_warnings():\n",
    "    warnings.filterwarnings(\"ignore\",category=FutureWarning)\n",
    "    fig, ax = plt.subplots(figsize=(4.5,2.8),dpi=300)\n",
    "    sns.scatterplot(data=df_spec,style='Type',\n",
    "                    x='pORG_Up_55_All',y='PurIST_Score',hue='Cohort',ax=ax,alpha=.7,\n",
    "                   )\n",
    "    ax.axvline(0.02,color='k',alpha=0.7,linestyle='--')\n",
    "    ax.axhline(0.5,color='k',alpha=0.7,linestyle='--')\n",
    "    ax.set_xlabel(f\"{ax.get_xlabel().split('_')[0]} Score\")\n",
    "    ax.set_ylabel(f\"{ax.get_ylabel().split('_')[0]} Score\")\n",
    "    ax.legend(bbox_to_anchor=(1,1))\n",
    "    plt.tight_layout()\n",
    "    fig.savefig(f'figures/Fig2_E_quagrant.pdf')\n",
    "    d_fig2.update({f'Fig2_E_quagrant':df_spec.loc[:,['Public_Specimen_ID','pORG_Up_55_All','PurIST_Score','Cohort']]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb4ced64-a408-4efc-a068-256137e0123e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #print(df_spec[df_spec.quadrant.notna()].Public_Specimen_ID.nunique())\n",
    "# util.open_write_excel(d_fig2,filename='Source_Data_Figure2.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fe53ee9-5223-4bc2-a115-5a0b1a8e9f1b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# #quadrant survival\n",
    "# df_spec.loc[df_spec.pORG_Up_55_All > 0.02,'pORG_All'] = 'pORG high' \n",
    "# df_spec.loc[df_spec.pORG_Up_55_All < 0.02,'pORG_All'] = 'pORG low' \n",
    "# df_spec.loc[df_spec.PurIST_Score > 0.5,'PurIST'] = 'basal-like'\n",
    "# df_spec.loc[df_spec.PurIST_Score < 0.5,'PurIST'] = 'classical'\n",
    "# df_spec['quadrant']=df_spec.pORG_All + '/'+ df_spec.PurIST\n",
    "# s_time='Days_from_Diagnosis_to_FU'\n",
    "# s_censor = 'Survival'\n",
    "# df_spec['Survival']=df_spec.Vital_Status_at_FU.replace({'Dead':1,'Alive':0})\n",
    "# for s_cohorts in ['Liver','Lung']:\n",
    "#     df2 = df_spec.loc[(df_spec.Cohort==s_cohorts) & (~df_spec.Public_Specimen_ID.str.endswith('2')),:]\n",
    "#     fig, ax, __=util.km_plot(df2,'pORG_All',s_time,s_censor)\n",
    "#     fig.suptitle(f'{s_cohorts}: Primary + Met',y=1.02)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5552d356-4169-4bba-841c-22b22e1e366d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# #met survival pORG All\n",
    "# s_time='Days_from_Diagnosis_to_FU'\n",
    "# # for s_cohorts in ['Liver','Lung']:\n",
    "# #     df2 = df_spec[(df_spec.Cohort==s_cohorts) & (~df_spec.Public_Specimen_ID.str.endswith('2')) & (~df_spec.Public_Specimen_ID.str.endswith('M'))]\n",
    "# #     fig, ax, __=util.km_plot(df2,'pORG_All',s_time,s_censor)\n",
    "# #     fig.suptitle(f'{s_cohorts}: Primary',y=1.02)\n",
    "\n",
    "# for s_cohorts in ['Liver','Lung']:\n",
    "#     df2 = df_spec[(df_spec.Cohort==s_cohorts) & (df_spec.Public_Specimen_ID.str.endswith('M'))]\n",
    "#     fig, ax, __=util.km_plot(df2,'pORG_All',s_time,s_censor)\n",
    "#     fig.suptitle(f'{s_cohorts}: Mets',y=1.02)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37fc1daf-4c4f-4710-be2c-b0e90ae4aa74",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# #met survival pORG met\n",
    "# s_time='Days from Diagnosis to FU'\n",
    "# df_patient['pORG_Pri'] = pd.NA\n",
    "# df_patient.loc[df_patient.pORG_Primary > 0.02,'pORG_Pri'] = 'pORG high' \n",
    "# df_patient.loc[df_patient.pORG_Primary < 0.02,'pORG_Pri'] = 'pORG low' \n",
    "# # for s_cohorts in ['Liver','Lung']:\n",
    "# #     df2 = df_patient.loc[(df_patient.Cohort==s_cohorts),['pORG_Pri',s_time,s_censor]].dropna()\n",
    "# #     fig, ax, __=util.km_plot(df2,'pORG_Pri',s_time,s_censor)\n",
    "# #     fig.suptitle(f'{s_cohorts}: Primary',y=1.02)\n",
    "# df_patient['pORG Met'] = pd.NA\n",
    "# df_patient.loc[df_patient.pORG_Met > 0.02,'pORG Met'] = 'pORG high' \n",
    "# df_patient.loc[df_patient.pORG_Met < 0.02,'pORG Met'] = 'pORG low' \n",
    "# for s_cohorts in ['Liver','Lung']:\n",
    "#     df2 = df_patient.loc[(df_patient.Cohort==s_cohorts),['pORG Met',s_time,s_censor]].dropna()\n",
    "#     fig, ax, __=util.km_plot(df2,'pORG Met',s_time,s_censor)\n",
    "#     fig.suptitle(f'{s_cohorts}: Mets',y=1.02)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3620f46-5031-4dac-8084-cf656db2e512",
   "metadata": {},
   "outputs": [],
   "source": [
    "#quadrant chi\n",
    "linebreak = '\\n'\n",
    "for s_cohorts in ['Liver','Lung']:\n",
    "        #for s_fill in ['Primary','Met']:\n",
    "        s_clin = 'quadrant' #_Primary\n",
    "        s_cohort = 'Type'#'Cohort'\n",
    "        df2 = df_spec.loc[(df_spec.Cohort==s_cohorts),[s_clin,s_cohort]].dropna()\n",
    "        crosstab= pd.crosstab(df2[s_clin], df2[s_cohort]) \n",
    "        annot = pd.crosstab(df2[s_clin], df2[s_cohort],normalize='columns') \n",
    "        try:\n",
    "            statistic,pvalue, dof, expected_freq = stats.chi2_contingency(crosstab)\n",
    "            # if pvalue < 0.99:\n",
    "            fig, ax = plt.subplots(dpi=300,figsize=(2.7,2.6))\n",
    "            sns.heatmap(crosstab/crosstab.sum(),ax=ax,annot=True,cbar_kws={'label':'Frac. Pts.'},\n",
    "                       cmap='Blues') #/crosstab.sum()\n",
    "            ax.set_title(f'{s_cohorts}\\n$\\chi^2$ P={pvalue:.2} N={len(df2)}')\n",
    "            ax.set_yticklabels([item.get_text().replace('/',f\"{linebreak}\") for item in ax.get_yticklabels()])\n",
    "        except:\n",
    "            print('error')\n",
    "        plt.tight_layout()\n",
    "        fig.savefig(f'figures/Fig2_F_chi_{s_cohorts}.pdf')\n",
    "        d_fig2.update({f'Fig2_F_chi_{s_cohorts}':crosstab})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0a08e12-94b8-4133-9e54-2893d0ff40a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_patient.columns.tolist()\n",
    "df_patient['DDR_Alteration'] = df_patient.DDR_Alteration_Primary.fillna(df_patient.DDR_Alteration_Met)\n",
    "s_clin = 'DDR_Alteration' #_Primary\n",
    "s_cohort = 'Lung_Met_Resection'#'Cohort'\n",
    "s_fill = 'DNAseq_Patient'#'RNF43_Altered_Primary'\n",
    "\n",
    "df2 = df_patient.loc[df_patient.loc[:,s_fill].notna(),[s_clin,s_cohort]].copy()\n",
    "df2.loc[:,s_clin].fillna(0,inplace=True)\n",
    "df2 = df2.dropna()\n",
    "crosstab= pd.crosstab(df2[s_clin], df2[s_cohort]) \n",
    "annot = pd.crosstab(df2[s_clin], df2[s_cohort],normalize='columns') \n",
    "try:\n",
    "    statistic,pvalue, dof, expected_freq = stats.chi2_contingency(crosstab)\n",
    "    # if pvalue < 0.99:\n",
    "    fig, ax = plt.subplots(dpi=200)\n",
    "    sns.heatmap(crosstab/crosstab.sum(),ax=ax,annot=True,cbar_kws={'label':'Percent Pts.'}) #/crosstab.sum()\n",
    "    ax.set_title(f'Chi sq. p={pvalue:.3} n={len(df2)}')\n",
    "except:\n",
    "    print('error')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f30048e8-dab4-4dc8-bea9-2021899b77b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# s_time = 'Days from Diagnosis to FU'\n",
    "# s_censor = 'Survival'\n",
    "# sns.set_palette('tab10')\n",
    "# df_km = df_patient[(df_patient.TCRseq_Tumor_Patient) & (df_patient.Cohort_Met_Resection.notna())]\n",
    "# fig, ax, __ = util.km_plot(df_km,'Cohort_Met_Resection',s_time, s_censor)\n",
    "# ax.set_title('TCR patients')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d14d4247",
   "metadata": {},
   "source": [
    "## gene expresison versus liver lung"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73ad5cdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "Colorblind[8]\n",
    "pal_porg = ('#56B4E9','#E69F00')\n",
    "pal_liv = ('#0072B2','#D55E00')\n",
    "sns.set_palette(pal_liv)\n",
    "#sns.palplot(Colorblind[8])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebe37598",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# # add cohort\n",
    "# df_rna['Public_Patient_ID'] = df_rna.index\n",
    "# df_rna['Cohort'] = df_rna.index.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Cohort)))\n",
    "# #liver vs lung (t cell)\n",
    "# for s_samples in ['All Samples']:#'T cell samples',\n",
    "#     ls_marker = ['CD3E','CD4', 'CD8A', 'MS4A1','CD68','OLR1','MRC1','MX1','STAT1']\n",
    "#     if s_samples == 'T cell samples':\n",
    "#         df_plot = df_rna.loc[df_rna.Public_Patient_ID.isin(df_tcell.Public_Patient_ID),ls_marker].unstack().reset_index()\n",
    "#     else:\n",
    "#         df_plot = df_rna.loc[:,ls_marker].unstack().reset_index()\n",
    "#     df_plot['Public_Patient_ID'] = df_plot.level_1#[item[0:-2] for item in df_plot.level_1]\n",
    "#     df_plot['Cohort'] = df_plot.Public_Patient_ID.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Cohort)))\n",
    "#     df_plot.rename({0:'Expression','level_0':'Gene'},axis=1,inplace=True)\n",
    "#     fig,ax = plt.subplots(dpi=200,figsize=(5,3))\n",
    "#     sns.stripplot(data=df_plot,x='Gene',y='Expression',hue='Cohort',dodge=True,ax=ax,s=2)\n",
    "#     sns.boxplot(data=df_plot,x='Gene',y='Expression',hue='Cohort',ax=ax,showmeans=True,medianprops={'visible': False},\n",
    "#                            whiskerprops={'visible': False},meanline=True,showcaps=False,\n",
    "#                            meanprops={'color': 'k', 'ls': '-', 'lw': 2},showfliers=False,showbox=False)\n",
    "#     h, l = ax.get_legend_handles_labels()\n",
    "#     ax.legend(h[0:2],l[0:2],bbox_to_anchor=(1,1))#\n",
    "#     pairs = [((item,'Lung'),(item,'Liver')) for item in ls_marker]\n",
    "#     annot = Annotator(ax, pairs, data=df_plot,x='Gene',y='Expression',hue='Cohort',\n",
    "#                       order=ls_marker,hue_order=('Lung','Liver'))\n",
    "#     annot.configure(test='t-test_ind',comparisons_correction='fdr_bh')\n",
    "#     annot.apply_and_annotate()\n",
    "#     ax.set_title(f'Liver vs Lung: {s_samples}')\n",
    "#     plt.tight_layout()\n",
    "#     fig.savefig(f'figures/stripplot_cohort_genes.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07f2d281-3d31-4483-826b-0763e568365d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# s_time = 'Days from Diagnosis to FU'\n",
    "# short_long = 365 #\n",
    "# df_patient.loc[df_patient.loc[:,s_time] > short_long,'TCR_Lung_Median_Surv'] = 'Lung'\n",
    "# df_patient.loc[df_patient.loc[:,s_time] < short_long,'TCR_Lung_Median_Surv'] = 'lung_short'\n",
    "# df_patient.loc[(df_patient.Cohort!='Lung'),'TCR_Lung_Median_Surv'] = np.nan\n",
    "\n",
    "# df_patient['Cohort_Long'] = df_patient.Cohort.replace({'Lung':pd.NA}).copy()\n",
    "# df_patient['Cohort_Long'].fillna(df_patient.TCR_Lung_Median_Surv,inplace=True)\n",
    "# df_patient['Cohort_Long'] = df_patient.Cohort_Long.replace({'lung_short':pd.NA})\n",
    "# df_patient['Cohort_Long'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2541143c-41f6-4660-89f8-103dfd38987e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sorted(df_patient.loc[df_patient.TCR_Lung_Median_Surv == 'lung_short','Public_Patient_ID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b29483d-6944-4d7a-b384-005b07a96ba1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# viper\n",
    "df_viper = pd.read_excel(f'Supplemental_Data/Supplemental_Dataset_6.xlsx',index_col=0)\n",
    "df_viper.columns = [item[0:-2] for item in df_viper.columns]\n",
    "df_viper = df_viper.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e68c919f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# viper vs  cohort\n",
    "\n",
    "pal_liv_r = ('#D55E00','#0072B2',)\n",
    "sns.set_palette(pal_liv_r)\n",
    "\n",
    "df_viper['Public_Patient_ID'] = df_viper.index\n",
    "df_viper['Cohort'] = df_viper.index.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Cohort)))\n",
    "df_viper['Cohort_Long'] = df_viper.index.map(dict(zip(df_patient.Public_Patient_ID,df_patient.Cohort_Long)))\n",
    "\n",
    "hue_order = ('Liver','Lung')\n",
    "order = ls_marker\n",
    "x=\"Regulon\"\n",
    "y='VIPER Score'\n",
    "\n",
    "\n",
    "figsize=(4.5,2.5)\n",
    "b_correct =True\n",
    "\n",
    "for s_samples in ['All Samples']:#,'T cell samples',\n",
    "    for s_group in ['Cohort','Cohort_Long']:\n",
    "        hue=s_group\n",
    "        ls_marker = ['CD3G', 'MS4A1','MX1','STAT1','IFNAR1','IFNAR2'] #'CD274',\n",
    "        if s_samples == 'T cell samples':\n",
    "            df_plot = df_viper.loc[df_viper.Public_Patient_ID.isin(df_tcell.Public_Patient_ID),ls_marker].unstack().reset_index()\n",
    "        else:\n",
    "            df_plot = df_viper.loc[:,ls_marker].unstack().reset_index()\n",
    "        df_plot['Public_Patient_ID'] = df_plot.level_1#[item[0:-2] for item in df_plot.level_1]\n",
    "        df_plot[s_group] = df_plot.Public_Patient_ID.map(dict(zip(df_patient.Public_Patient_ID,df_patient.loc[:,s_group])))\n",
    "        df_plot.rename({0:'VIPER Score','level_0':'Regulon'},axis=1,inplace=True)\n",
    "        df_plot = df_plot[df_plot.loc[:,s_group].notna()]\n",
    "        fig, ax = util.annotated_stripplot_hue(df_plot,x,y,hue,order,hue_order,figsize=(6,3),b_correct=True)\n",
    "        ax.set_title(s_group)\n",
    "        ax.set_title(f'{s_group.replace(\"_\",\" \")}')\n",
    "        h, l = ax.get_legend_handles_labels()\n",
    "        ax.legend(h[0:2],l[0:2],bbox_to_anchor=(1,1))#\n",
    "        plt.tight_layout()\n",
    "        fig.savefig(f'figures/stripplot_VIPER_cohort.pdf')\n",
    "        break\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00f88061-b061-420c-ad8f-e76a5f7ef15c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_plot.to_csv(f'Source_Extended_Data_Figure6_F.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ce56879",
   "metadata": {},
   "source": [
    "## Section 7 <a name=\"tcell\"></a>\n",
    "\n",
    "re-analyze t cell data\n",
    "\n",
    "289 blood samples with matching 175 primary\n",
    "tumors (141 overlapping with the RNA-seq dataset) and 43 metastatic tumors (33 overlapping with\n",
    "the RNA-seq dataset). (218)\n",
    "\n",
    "**missing 2 tumor (have 216), all 289 blood there**\n",
    "\n",
    "290 unique patients (one primary tumor w/o blood: ST-00018360)\n",
    "\n",
    "215 blood and tumor are matched\n",
    "\n",
    "TOTAL = 174 primary, 42 met (216 total)\n",
    "\n",
    "Of the 290 patients, 284 are analyzed (some dropped surgery)\n",
    "\n",
    "we analyzed blood samples from 77 patients in the liver cohort and 16\n",
    "patients in the lung cohort, of which 60 and 16 were matched with tumor samples from the same\n",
    "patient, respectively\n",
    "\n",
    "TOTAL = 94 blood\n",
    "TOTAL = 76 tumor\n",
    "\n",
    "tumor distinct clones\n",
    "used data from 214 matched pairs of tumor and blood samples\n",
    "\n",
    "**213 are there, missing 1**\n",
    "\n",
    "(TCR tumor: 59 and 16)\n",
    "\n",
    "#### Exclude those who died of surgery\n",
    "\n",
    "yes      ST-00018963\n",
    "\n",
    "yes      ST-00020077\n",
    "\n",
    "yes     ST-00016968\n",
    "\n",
    "yes     ST-00006625\n",
    "\n",
    "yes     ST-00007146\n",
    "\n",
    "yes    ST-00018260\n",
    "\n",
    "#### Actually analyzed \n",
    "Liver    76\n",
    "Lung     16\n",
    "\n",
    "| Cohort  | Tumor Type  |  number pts.  |\n",
    "|---------|-------------|---------------|\n",
    "| Liver   | Met         | 17            |\n",
    "|         | Primary     | 42            |\n",
    "| Lung    | Met         | 3             |\n",
    "|         | Primary     | 13            |\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4ad9a5b-964c-4f66-93e8-3368f58bee29",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load tcr\n",
    "df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "                          index_col=0)\n",
    "df_tcr =  pd.read_csv('Supplemental_Data/Supplemental_Dataset_8.csv',index_col=0)\n",
    "df_merge = df_tcr.merge(df_patient,on='Public_Patient_ID',suffixes = ('','_x'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a51a578-04ee-40f7-b601-9a8891bbfee3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save source data\n",
    "ls_col = ['Public_Patient_ID','Templates_per_ng','Productive_Rearrangements',\n",
    "          \"Simpson's Evenness tumor\",'Shannon_Entropy_Tumor',\"Clonality_Tumor\",\n",
    "         'Tumor_Type','Cohort','pORG_Primary','pORG_Met',]\n",
    "df_merge.loc[df_merge.Tumor_Type.notna(),ls_col].to_excel('Souce_Data_Figure5.xlsx')\n",
    "\n",
    "ls_col = ['Public_Patient_ID','Clonality_Blood','pORG_Primary','pORG_Met','Cohort']\n",
    "df_merge.loc[df_merge.Blood_Type.notna(),ls_col].to_csv('Figure5_L.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5b280dc-ead8-444f-886d-d36b4afc1f6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "df_merge['pORG_All'] = df_merge.pORG_allPrimary.fillna(df_merge.pORG_allMet)\n",
    "df_merge.columns[df_merge.columns.str.contains('public')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e484fe67-1876-49f8-809c-d61575c6dcc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save source data\n",
    "ls_col = ['Public_Patient_ID','jaccard_Tumor','jaccard_Blood',\n",
    "          'public_Tumor','public_Blood',\n",
    "          'Days from Diagnosis to FU','Survival',\n",
    "         'Tumor_Type','Cohort','pORG_Primary','pORG_Met','pORG_All']\n",
    "df_merge.loc[df_merge.Tumor_Type.notna(),ls_col].to_excel('Souce_Data_Figure6.xlsx')\n",
    "\n",
    "ls_col = ['Public_Patient_ID','jaccard_Tumor_blood','public_Tumor_blood',\n",
    "          'pORG_All_quartiles','Cohort',\n",
    "         'Days from Diagnosis to FU','Survival',\n",
    "         'pORG_Primary','pORG_All']\n",
    "df_merge.loc[df_merge.Blood_Type.notna(),ls_col].to_csv('Figure6_C.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84e310bc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#add pORG quartiles, plot violins \n",
    "# Rosie chose this way\n",
    "%matplotlib inline\n",
    "importlib.reload(util)\n",
    "alpha = 0.05\n",
    "b_correct= False \n",
    "ls_foci = [\n",
    "    #     ####### tumor ##########\n",
    "     #'Percent Tumor Distinct Clones',\n",
    "               'Productive_Rearrangements',\n",
    "       'Templates_per_ng','Simpsons_Diversity_Tumor',\n",
    "    'Shannon_Entropy_Tumor', \n",
    "     'Clonality_Tumor',\n",
    "      \"Simpson's Evenness tumor\",#\n",
    "    #     ####### blood ##########\n",
    "              #  'Productive_Rearrangements_Blood',\n",
    "              #   \"Simpson's Evenness blood\",\n",
    "       #         'Shannon_Entropy_Blood',\n",
    "       # 'Simpsons_Diversity_Blood',\n",
    "                 'Clonality_Blood',\n",
    "    # overlap\n",
    "      # 'jaccard_Blood', 'jaccard_Blood_blood', 'public_Blood_blood',\n",
    "      #       'public_Tumor_blood','jaccard_Tumor_blood',\n",
    "      #       'public_Blood',\n",
    "      #       'public_Tumor','jaccard_Tumor',\n",
    "    ]\n",
    "ls_morisita = sorted(df_merge.loc[:,(df_merge.columns.str.contains('morisita'))])\n",
    "ls_jaccard = sorted(df_merge.loc[:,(df_merge.columns.str.contains('jaccard'))])\n",
    "ls_public = sorted(df_merge.loc[:,(df_merge.columns.str.contains('public'))])\n",
    "ls_foci =  ls_foci#+ls_jaccard + ls_public    #+ls_morisita\n",
    "d_colorblind = {'Liver':'#d55e00','Lung':'#0072b2',\n",
    "               'high':'#e69f00','low': '#56b4e9',\n",
    "               'basal-like':'#000000','classical':'#cc79a7',\n",
    "               'high pSUB': '#f0e442','low pSUB':'#009E73',\n",
    "               'Met':'black','Primary':'gray'}\n",
    "#non- parametric\n",
    "se_non_para = pd.Series( ['Percent Tumor Distinct Clones','Simpsons_Diversity_Tumor','Simpsons_Diversity_Blood', 'Simpsons_Evenness_Blood',\n",
    "        'Templates_per_ng','Productive_Rearrangements','Simpsons_Evenness_Tumor',\n",
    "         \"Simpson's Evenness tumor\",'Productive_Rearrangements_Blood',\n",
    "        \"Simpson's Evenness blood\"]+ls_public+ls_jaccard+ls_morisita)\n",
    "d_order =  {#'Blood_Type':['Primary','Met'],\n",
    "            'Cohort':['Liver','Lung'],#'PurIST_Subtype':['basal-like','classical'],\n",
    "   #'Tumor_Type':['Primary','Met'],\n",
    "    'quartiles':['high','low'],\n",
    "           }\n",
    "figsize=(2.5,2.7)\n",
    "s=4\n",
    "if (np.array([item.find('_Type') for item in  d_order.keys()]) > -1).any():\n",
    "    figsize=(2,2.2)\n",
    "    s=2\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    for s_porg in ['pORG_Met','pORG_All','pORG_Primary',]:#\n",
    "        for s_foci in ls_foci:#ls_public+ls_jaccard+ls_morisita:#\n",
    "            if s_porg.find('Primar') > -1:\n",
    "                print('Primaries')\n",
    "                df_pri = df_merge.loc[(df_merge.Tumor_Type=='Primary')].copy()\n",
    "                if s_foci.find('lood') > -1:\n",
    "                    df_pri = df_merge.loc[(df_merge.Blood_Type=='Primary')].copy()\n",
    "            elif s_porg.find('All') > -1: \n",
    "                print('Primaries and Mets')\n",
    "                df_whole = pd.DataFrame(columns=df_merge.columns)#\n",
    "                for s_all in ['pORG_allPrimary', 'pORG_allMet']:\n",
    "                    df_half = df_merge[(df_merge.loc[:,s_all].notna())]\n",
    "                    df_half = df_half[~df_half.Public_Patient_ID.isin(df_whole.Public_Patient_ID)].copy()\n",
    "                    df_half.rename({s_all:'pORG_All'},axis=1,inplace=True)\n",
    "                    #df_half = df_half.loc[:,~df_half.columns.duplicated()]\n",
    "                    #print(len(df_half))\n",
    "                    with warnings.catch_warnings():\n",
    "                        warnings.filterwarnings(\"ignore\",category=FutureWarning)\n",
    "                        df_whole = pd.concat([df_whole,df_half],axis=0,ignore_index=True)\n",
    "                        df_whole.reset_index(inplace=True, drop=True)\n",
    "                #s_foci not defined\n",
    "                df_pri = df_merge.merge(df_whole.loc[:,['Public_Patient_ID','pORG_All']],on='Public_Patient_ID',how='left')#.loc[:,s_foci].notna()\n",
    "                df_pri.loc[df_pri.Public_Patient_ID.duplicated(),'Cohort'] = np.nan\n",
    "            elif s_porg.find('Met') > -1:\n",
    "                print('Mets')\n",
    "                df_pri = df_merge.loc[df_merge.Tumor_Type=='Met'].copy()\n",
    "                if s_foci.find('lood') > -1:\n",
    "                    df_pri = df_merge.loc[(df_merge.Blood_Type=='Met')].copy()\n",
    "            df_pri = util.add_quartiles(df_pri,s_porg)#.drop(213) #p. rearrangements outlier \n",
    "            if s_foci.find('jaccard')>-1:\n",
    "                df_pri.loc[:,s_foci] = df_pri.loc[:,s_foci]/len(df_pri.loc[:,s_foci])\n",
    "            #get mean, not total\n",
    "            elif s_foci.find('public')>-1:\n",
    "                df_pri.loc[:,s_foci] = df_pri.loc[:,s_foci]/len(df_pri.loc[:,s_foci])\n",
    "                #break\n",
    "            for s_trans in ['none','log']:#\n",
    "                if se_non_para.isin([s_foci]).any():\n",
    "                    s_stats = 'non-parametric'\n",
    "                    if s_trans == 'log':#log transform\n",
    "                        if df_pri.loc[:,s_foci].min() <= 0:\n",
    "                            print('arcsin')\n",
    "                            df_pri.loc[:,s_foci] = np.arcsinh((df_pri.loc[:,s_foci]))\n",
    "                        else:\n",
    "                            df_pri.loc[:,s_foci] = np.log10(df_pri.loc[:,s_foci])\n",
    "                        s_stats = 'mean'\n",
    "                    else:\n",
    "                        if s_foci == \"Simpson's Evenness tumor\":\n",
    "                            s_stats = 'non-parametric'\n",
    "                        elif s_foci == \"Simpson's Evenness blood\":\n",
    "                            s_stats = 'non-parametric'\n",
    "                        elif s_foci == \"Percent Tumor Distinct Clones\":\n",
    "                            s_stats = 'non-parametric'\n",
    "                        elif s_foci == \"jaccard_Tumor\":\n",
    "                            s_stats = 'non-parametric'\n",
    "                        else:\n",
    "                            continue\n",
    "                else:\n",
    "                    s_stats = 'mean' \n",
    "                    if s_trans == 'log':\n",
    "                        continue\n",
    "                print(s_stats)\n",
    "                for key, items in d_order.items():\n",
    "                    a = df_pri.loc[df_pri.loc[:,key]==items[0],s_foci].dropna()\n",
    "                    b = df_pri.loc[df_pri.loc[:,key]==items[1],s_foci].dropna()\n",
    "                    __, pvalue_lev = stats.levene(a,b, center='mean')\n",
    "                    if pvalue_lev < 0.05:\n",
    "                        print(f'{s_foci} {key} Levene:{pvalue_lev:.02}')\n",
    "\n",
    "                df_both,d_pval,order = util.violin_stats2(df_pri,d_order,s_foci,s_stats)\n",
    "                #util.qq_plot_hist(df_pri,s_cat,s_foci)  #anova eval\n",
    "                fig,pvalues,corrected = util.plot_violins2(df_both,d_pval,d_order,s_stats,s_foci,\n",
    "                        order,d_colorblind,s_porg,b_correct=b_correct,figsize=figsize,s=s,b_title=False)#True#False\n",
    "                if s_trans == 'log':\n",
    "                    s_label = f'log {fig.get_axes()[0].get_ylabel().replace(\"_\",\" \")}'\n",
    "                else:\n",
    "                    s_label = f'{fig.get_axes()[0].get_ylabel().replace(\"_\",\" \")}'\n",
    "                fig.get_axes()[0].set_ylabel(s_label)\n",
    "                fig.get_axes()[0].set_xlabel(fig.get_axes()[0].get_xlabel().replace(\"_\",\" \"))\n",
    "                ticklabels = [item.get_text().capitalize() for item in fig.get_axes()[0].get_xticklabels()]\n",
    "                fig.get_axes()[0].set_xticklabels(ticklabels)\n",
    "                if len(fig.get_axes()[0].get_yticklabels()[0].get_text()) > 3:\n",
    "                    try:\n",
    "                        yticklabels = [f'{float(item.get_text()):.2f}' for item in fig.get_axes()[0].get_yticklabels()]\n",
    "                        fig.get_axes()[0].set_yticklabels(yticklabels)\n",
    "                    except:\n",
    "                        pass\n",
    "                fig.savefig(f'figures/violinplot_both_{s_foci}_{list(d_order.keys())[-1]}_{s_porg}_{s_stats}.pdf')\n",
    "                \n",
    "                #plt.close(fig)\n",
    "                # ax = fig.get_axes()\n",
    "                # #all combos, mann-whitney, no corrections\n",
    "                # plotting = {'data':df_both,'x':'color','y':s_foci}\n",
    "                # pairs = [item for item in combinations(df_both.color.unique(),2)]\n",
    "                # if s_foci != \"Simpson's Evenness tumor\":\n",
    "                #     fig,ax, test_results = util.annotated_boxplot(plotting,pairs,test='t-test_ind')\n",
    "                # else:\n",
    "                #     fig,ax, test_results = util.annotated_boxplot(plotting,pairs,test='Mann-Whitney')\n",
    "                # ax.set_title(s_foci)\n",
    "            #break\n",
    "            #'''\n",
    "            # SPEARMAN corr to Templates_per_ng\n",
    "            #for s_corr in ['morisita_Primary','morisita_Met','morisita_Blood']:\n",
    "            #'jaccard_Blood','jaccard_Tumor','public_Blood','public_Tumor'\n",
    "            #break\n",
    "            if s_foci == 'Percent Tumor Distinct Clones':\n",
    "                for s_corr in ['Clonality_Tumor']:\n",
    "                    if s_trans=='none':\n",
    "                        if s_porg.find('All') > -1:\n",
    "                            fig2, pvalues2 = util.plot_pearson(df_pri,s_corr,s_foci,s_stats='non-parametric',ls_plots=['Primaries','Mets','Both']) #='non-parametric'\n",
    "                            fig2.savefig(f'figures/scatterplot_spearman_all_{s_foci}_{list(d_order.keys())[-1]}_{s_porg}_{s_stats}.pdf')\n",
    "                        elif s_porg.find('Primar') > -1:\n",
    "                            fig2, pvalues2 = util.plot_pearson(df_pri,s_corr,s_foci,s_stats,ls_plots=['Primaries'])\n",
    "                            fig2.savefig(f'figures/scatterplot_pearson_primary_{s_foci}_{list(d_order.keys())[-1]}_{s_porg}_{s_stats}.pdf')\n",
    "                        elif s_porg.find('Met') > -1:\n",
    "                            fig2, pvalues2 = util.plot_pearson(df_pri,s_corr,s_foci,s_stats,ls_plots=['Mets'])\n",
    "                            fig2.savefig(f'figures/scatterplot_pearson_met_{s_foci}_{list(d_order.keys())[-1]}_{s_porg}_{s_stats}.pdf')\n",
    "                        #plt.close(fig2)\n",
    "        if (np.array([item.find('_Type') for item in  d_order.keys()]) > -1).any():\n",
    "             break\n",
    "        break\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12648092-85ac-4636-92bd-be60200a94ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ls_col = ['Public_Patient_ID',\n",
    "#  'Percent Tumor Distinct Clones',\n",
    "#           'Clonality_Tumor',\n",
    "#  'pORG_All_quartiles', 'pORG_Primary_quartiles',\n",
    "#  'pORG_Met_quartiles',\n",
    "#  'Cohort', 'Tumor_Type',\n",
    "#  'Days from Diagnosis to FU',\n",
    "#  'Survival']\n",
    "# df_merge.loc[:,ls_col].to_excel('Source_Data_Figure7.xlsx')\n",
    "# ls_col = ['Public_Patient_ID',\n",
    "# # 'Productive_Rearrangements',# 'Templates_per_ng',\n",
    "# #           \"Simpson's Evenness tumor\",    'Clonality_Tumor',#\n",
    "# #       'Shannon_Entropy_Tumor',  'Simpsons_Diversity_Tumor',\n",
    "# #    'Tumor_Type',\n",
    "#  # 'pORG_All_quartiles', 'pORG_Primary_quartiles',\n",
    "# # 'Clonality_Tumor',\n",
    "#           'pORG_Met_quartiles',\n",
    "#  'Cohort', # 'Tumor_Type',\n",
    "#  # 'Days from Diagnosis to FU',\n",
    "#  #'Survival','Alive_30_days_post_surgery'\n",
    "#           'Shannon_Entropy_Blood',\n",
    "#           'Clonality_Blood','Blood_Type',\n",
    "#          ]\n",
    "# df_merge.loc[:,ls_col].to_csv(f'Source_Data_Figure5_L.csv')\n",
    "#df_merge.loc[:,ls_col].to_csv(f'Source_Data_Figure5_I.csv')\n",
    "#df_merge.loc[:,ls_col].to_csv(f'Source_Data_Figure5_K.csv')\n",
    "\n",
    "# ls_col = ['Public_Patient_ID',\n",
    "# 'Productive_Rearrangements',\n",
    "#        'Templates_per_ng','Simpsons_Diversity_Tumor',\n",
    "#     'Shannon_Entropy_Tumor', \n",
    "#      'Clonality_Tumor',\n",
    "#        \"Simpson's Evenness tumor\",#\n",
    "#  'pORG_All_quartiles', 'pORG_Primary_quartiles',\n",
    "#  'pORG_Met_quartiles',\n",
    "#  'Cohort', 'Tumor_Type',\n",
    "#  'Days from Diagnosis to FU',\n",
    "#  'Survival','Stage','Grade','LV_Invasion','LN_Pos','Alive_30_days_post_surgery']\n",
    "# df_merge.loc[:,ls_col].to_csv(f'Source_Extended_Data_Figure7_D-M.csv')\n",
    "\n",
    "# ls_col = ['Public_Patient_ID',\n",
    "#                'Productive_Rearrangements_Blood',\n",
    "#                 \"Simpson's Evenness blood\",\n",
    "#                'Shannon_Entropy_Blood',\n",
    "#        'Simpsons_Diversity_Blood',\n",
    "#                 'Clonality_Blood',\n",
    "#  'pORG_All_quartiles', 'pORG_Primary_quartiles',\n",
    "#  'pORG_Met_quartiles',\n",
    "#  'Cohort', 'Blood_Type',\n",
    "#  'Days from Diagnosis to FU',\n",
    "#  'Survival','Alive_30_days_post_surgery']\n",
    "# df_merge.loc[:,ls_col].to_csv(f'Source_Extended_Data_Figure8_C-I.csv')\n",
    "\n",
    "# ls_col = ['Public_Patient_ID','Percent Tumor Distinct Clones',\n",
    "# 'jaccard_Blood_blood', 'public_Blood_blood',\n",
    "#     'public_Tumor_blood','jaccard_Tumor_blood',\n",
    "#  'pORG_All_quartiles', 'pORG_Primary_quartiles',\n",
    "#  'pORG_Met_quartiles',\n",
    "#  'Cohort', 'Blood_Type','Tumor_Type',\n",
    "#  'Days from Diagnosis to FU',\n",
    "#  'Survival','Alive_30_days_post_surgery']\n",
    "# df_merge.loc[:,ls_col].to_csv(f'Source_Extended_Data_Figure9_D-G_J.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4be5e55-cb3a-4b0a-8254-68e76ffad2a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# idx = 7\n",
    "# s_find = f'Source_Extended_Data_Figure{idx}'\n",
    "# d_fig = excel_dict_from_csvs(s_find,s_prefix=f'ED_Fig{idx}')\n",
    "# util.open_write_excel(d_fig,filename=f'Source_Extended_Data_Figure{idx}.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a9fdfde",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "## TCR survival <a name=\"tcells\"></a>\n",
    "\n",
    "- The goal is to find a rational method to set a pORG threshold selecting High/Low cohorts for use in Kaplan-Meier and pathway analysis.\n",
    "- Since we have survival data and its probably the best clinical endpoint we can use, we will start by selecting a survival threshold that seems relevant. (545 days)\n",
    "- Based on the survival threshold, we will divide the patient into short/long term survivor cohorts.\n",
    "- We can then generate rock curves testing the ability of our pORG score to predict these cohorts.\n",
    "- Using the Youdens Index, we can maximize the pORG threshold for making this prediction.\n",
    "- Finally, we can use this pORG threshold to generate Kaplan-Meier curves and test the significance of these using log-rank p-value.\n",
    "- We will try this with a few reasonable selections for our original survival threshold.\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f077e6c-8708-4bfa-a2f0-2ad5e99b0f4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load tcr\n",
    "df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "                          index_col=0)\n",
    "df_tcr =  pd.read_csv('Supplemental_Data/Supplemental_Dataset_8.csv',index_col=0)\n",
    "df_merge = df_tcr.merge(df_patient,on='Public_Patient_ID',suffixes = ('','_x'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38b5009b-5a7e-492f-8964-55215c20fbbf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "importlib.reload(util)\n",
    "sns.set_palette(['black','gray'])\n",
    "df_merge.loc[:,df_merge.columns.str.contains('TLS')]\n",
    "b_ = df_merge.TLS.notna() & df_merge.loc[:,'Percent Tumor Distinct Clones'].notna()##TLS_Met\n",
    "b_met = df_merge.TLS_Met.notna() & df_merge.loc[:,'Percent Tumor Distinct Clones'].notna()\n",
    "(b_ | b_met)\n",
    "df_merge['TLS_All'] = df_merge.TLS.fillna(df_merge.TLS_Met)\n",
    "ls_order = ['N','Y']\n",
    "y='Percent Tumor Distinct Clones'\n",
    "with warnings.catch_warnings():\n",
    "    warnings.filterwarnings(\"ignore\",category=FutureWarning)\n",
    "    for x in ['TLS_All','TLS','TLS_Met']:\n",
    "        #print(df_merge.loc[df_merge.loc[:,y].notna(),x].value_counts().sum())\n",
    "        df_plot = df_merge[df_merge.loc[:,y].notna()]\n",
    "        print(len(df_plot))\n",
    "        if x == 'TLS_All':\n",
    "            df_plot = pd.concat([df_plot,df_merge.loc[(b_ & b_met)]])\n",
    "        print(len(df_plot))\n",
    "        plotting = {'data':df_plot,'x':x,'y':y,'order':ls_order}\n",
    "        #util.annotated_stripplot(plotting,ls_order,figsize=(3,3))\n",
    "        fig, ax, __ = util.plot_violins(plotting,ls_order)\n",
    "        plt.tight_layout()\n",
    "        fig.savefig(f'figures/violinplot_{y}_vs_{x}.pdf')\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca51ca51-e8a0-4339-b0d4-55286142c4eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_plot.loc[:, ['Public_Patient_ID',x,y]].to_csv('Source_Data_Figure7_E.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7c5c8c8-5c11-44ac-a6de-ae09f84b49e5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "##CPH\n",
    "se_non_para = pd.Series( ['Percent Tumor Distinct Clones','Simpsons_Diversity_Tumor','Simpsons_Diversity_Blood', 'Simpsons_Evenness_Blood',\n",
    "        'Templates_per_ng','Productive_Rearrangements','Simpsons_Evenness_Tumor',\n",
    "         \"Simpson's Evenness tumor\",'Productive_Rearrangements_Blood',\n",
    "        \"Simpson's Evenness blood\",'public_Blood_blood','jaccard_Blood_blood',\n",
    "            'public_Tumor_blood','jaccard_Tumor_blood',\n",
    "            'public_Blood','jaccard_Blood',\n",
    "            'public_Tumor','jaccard_Tumor',])\n",
    "\n",
    "b_tcr_pts = (df_merge.Clonality_Blood.notna() | df_merge.Clonality_Tumor.notna())\n",
    "#print(b_tcr_pts.sum())\n",
    "df_km_samples = df_merge.loc[(df_merge.Alive_30_days_post_surgery) & b_tcr_pts]\n",
    "df_result = pd.DataFrame()\n",
    "se_recur = {'Other_Recurrence', 'No_Recurrence','Lung_Cohort','Liver_Cohort'} \n",
    "ls_cats = ['LV_Invasion','LN_Pos'] #categorical,'Neoadjuvant'\n",
    "ls_foci = [\n",
    "        ####### tumor ##########\n",
    "    # 'Templates_per_ng',\n",
    "    #  'Percent Tumor Distinct Clones',\n",
    "    #            'Productive_Rearrangements',\n",
    "    #    'Simpsons_Diversity_Tumor',\n",
    "    # 'Shannon_Entropy_Tumor', \n",
    "    # 'Clonality_Tumor',\n",
    "    #    \"Simpson's Evenness tumor\",#\n",
    "    #     ####### blood ##########\n",
    "               'Productive_Rearrangements_Blood',\n",
    "                 \"Simpson's Evenness blood\",\n",
    "       #        'Shannon_Entropy_Blood',\n",
    "       # 'Simpsons_Diversity_Blood',\n",
    "                'Clonality_Blood',\n",
    "    # ## overlap\n",
    "        # 'public_Blood_blood','jaccard_Blood_blood',\n",
    "        #     'public_Tumor_blood','jaccard_Tumor_blood',\n",
    "        #     'public_Blood','jaccard_Blood',\n",
    "        #     'public_Tumor','jaccard_Tumor',\n",
    "     ]\n",
    "\n",
    "tu_time_censor = (('Days from Diagnosis to FU','Survival'),\n",
    "    ('Days from Resection to Recurrence', 'Recurrence'),\n",
    "               )\n",
    "idx = 0\n",
    "s_time = tu_time_censor[idx][0]\n",
    "s_censor = tu_time_censor[idx][1]\n",
    "df_result_all = pd.DataFrame()\n",
    "df_boxplot = df_km_samples.loc[:,['Public_Patient_ID','pORG_Primary','pORG_Met']].copy()\n",
    "df_boxplot['pORG_All'] = df_km_samples.pORG_allPrimary.fillna(df_km_samples.pORG_allMet)\n",
    "for s_title_str in ['All','Primary','Met',]:\n",
    "    df_result = pd.DataFrame(columns=['exp(coef)','exp(coef) lower 95%','exp(coef) upper 95%','p','n'])\n",
    "    for s_col in ls_foci:\n",
    "            print(s_col)\n",
    "            ls_surv_col =[s_col,s_time,s_censor,'Public_Patient_ID']\n",
    "            if s_title_str == 'Primary':\n",
    "                if s_col.find('lood') > -1:\n",
    "                    print(f'Primaries only blood. n={sum(df_km_samples.Blood_Type==\"Primary\")}')\n",
    "                    df = df_km_samples.loc[(df_km_samples.Blood_Type=='Primary'),ls_surv_col].copy()\n",
    "                else:\n",
    "                    print(f'Primaries only tumor. n={sum(df_km_samples.Tumor_Type==\"Primary\")}')\n",
    "                    df = df_km_samples.loc[(df_km_samples.Tumor_Type=='Primary'),ls_surv_col].copy()\n",
    "            elif s_title_str == 'Met':\n",
    "                if s_col.find('lood') > -1:\n",
    "                    print(f'Met only blood. n={sum(df_km_samples.Blood_Type==\"Met\")}')\n",
    "                    df = df_km_samples.loc[(df_km_samples.Blood_Type=='Met'),ls_surv_col].copy()\n",
    "                else:\n",
    "                    print(f'Met only tumor. n={sum(df_km_samples.Tumor_Type==\"Met\")}')\n",
    "                    df = df_km_samples.loc[(df_km_samples.Tumor_Type=='Met'),ls_surv_col].copy()  \n",
    "            else:\n",
    "                df = df_km_samples.loc[:,ls_surv_col].dropna().copy()\n",
    "            df.set_index('Public_Patient_ID',inplace=True)\n",
    "            print(f'my test {len(df)}')\n",
    "            if se_non_para.isin([s_col]).any():\n",
    "                print(f'log {s_col}')\n",
    "                if df.loc[:,s_col].min() <= 0:\n",
    "                    print('arcsin')\n",
    "                    df.loc[:,s_col] = np.arcsinh((df.loc[:,s_col]))\n",
    "                else:\n",
    "                    df.loc[:,s_col] = np.log2(df.loc[:,s_col])\n",
    "            else:\n",
    "                print('')\n",
    "            df_boxplot[s_col] = df_boxplot.Public_Patient_ID.map(dict(zip(df.index,df.loc[:,s_col])))\n",
    "            if not df.loc[:,s_col].dtype== 'float64':\n",
    "                print('dummy')\n",
    "                df_dummy = pd.get_dummies(df.loc[:,[s_col]],drop_first=True)\n",
    "                df.drop(s_col,axis=1,inplace=True)\n",
    "                s_col = df_dummy.columns[0]\n",
    "                df[s_col] = df_dummy\n",
    "            if len(se_recur.intersection(set([s_col]))) > 0:\n",
    "                 df.loc[df_km_samples.set_index('Public_Patient_ID').Recurrence_Sites_4=='No_Resection',s_col] = np.nan\n",
    "            fig2, cph = util.cph_plot(df,s_col,s_time,s_censor,figsize=(4,1.5))\n",
    "            plt.tight_layout()\n",
    "            #plt.close(fig2)\n",
    "            fig2.savefig(f'figures/CPH_single_TCR_{s_col}_{s_title_str}.pdf')\n",
    "            df2 = cph.summary.assign(n=[cph._n_examples])\n",
    "            df_result = pd.concat([df_result, df2.loc[:,['exp(coef)','exp(coef) lower 95%','exp(coef) upper 95%','p','n']]])\n",
    "            #multi\n",
    "            if s_col.find('pORG')>-1:\n",
    "                print('multi')\n",
    "                df_multi = df_km_samples.loc[:,['Grade','Stage','Public_Patient_ID']].set_index('Public_Patient_ID')\n",
    "                df = df.merge(df_multi,left_index=True,right_index=True)\n",
    "                df_dummy = pd.get_dummies(df_km_samples.loc[:,ls_cats+['Public_Patient_ID']].dropna().set_index('Public_Patient_ID'),drop_first=True) \n",
    "                df = df.merge(df_dummy,left_index=True,right_index=True)\n",
    "                df = df.dropna()\n",
    "                fig, cph = util.cph_plot(df,s_col,s_time,s_censor) #figsize\n",
    "                plt.tight_layout()\n",
    "                fig.savefig(f'figures/cph_TCR_multi_{s_col}_clinical_covar.pdf')\n",
    "            #plt.close(fig)\n",
    "                #'''\n",
    "            # except:\n",
    "            #     print(f'error {s_col}')\n",
    "            break\n",
    "    df_result = df_result.iloc[:,0:5]\n",
    "    df_result['Samples_Included'] = s_title_str\n",
    "    df_result_all = pd.concat([df_result_all,df_result])\n",
    "    break\n",
    "#df_result_all.to_csv(f'results/results_single_CPH_TCR_{s_censor}.csv')\n",
    "df_boxplot.set_index('Public_Patient_ID',inplace=True)\n",
    "# youden + kaplan meier in TCR patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba4e4de9-8365-4960-a921-323ce3a2d5f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# order = sorted(df_boxplot.columns)\n",
    "# sns.boxplot(data=df_boxplot.stack().reset_index(),y='level_1',x=0,orient='h',order=order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42cf8ff6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#define high and low survivors\n",
    "s_bad = {\"Simpson's Evenness tumor\", \"Simpson's Evenness blood\",    'Simpsons_Diversity_Blood',\n",
    "    'Shannon_Entropy_Blood'}\n",
    "importlib.reload(util)\n",
    "# ls_foci = [\n",
    "#     #'jaccard_Tumor_blood','public_Tumor_blood',\n",
    "#     # 'public_Blood','jaccard_Blood',\n",
    "#     #         'public_Tumor','jaccard_Tumor',\n",
    "#     #  'Clonality_Tumor', \n",
    "#    # 'Clonality_Blood',\"Simpson's Evenness blood\",\n",
    "#     #      'Productive_Rearrangements','Simpsons_Diversity_Tumor',\n",
    "#     #          'Shannon_Entropy_Tumor','Templates_per_ng',\n",
    "#     #      \"Simpson's Evenness tumor\",\n",
    "#     # 'Simpsons_Diversity_Blood',\n",
    "#     # 'Shannon_Entropy_Blood','Productive_Rearrangements_Blood',\n",
    "#     'Percent Tumor Distinct Clones',\n",
    "\n",
    "#              ]\n",
    "s_time = 'Days from Diagnosis to FU'\n",
    "s_censor = 'Survival'\n",
    "df_patient = pd.read_csv('Supplemental_Data/Supplemental_Dataset_8.csv',index_col=0)#pd.read_csv('annotation/20240304_Patient_Metadata_TCR_Metrics.csv',index_col=0)\n",
    "#load tcr\n",
    "df_patient_all = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "                          index_col=0)\n",
    "df_patient = df_patient.merge(df_patient_all,on='Public_Patient_ID',suffixes = ('','_x'))\n",
    "print(len(df_patient))\n",
    "ls_primary = ['All','Primary',] #'Lung',,'Met',\n",
    "for b_primary in ls_primary:\n",
    "    print(b_primary)\n",
    "    for s_tcr in ls_foci:\n",
    "        print(s_tcr)\n",
    "        if b_primary == 'Primary':\n",
    "            if s_tcr.find('lood') > -1:\n",
    "                print(f'Primaries only blood. n={sum(df_patient.Blood_Type==\"Primary\")}')\n",
    "                df_patient2 = df_patient[(df_patient.Blood_Type=='Primary')& (df_patient.loc[:,s_tcr].notna()) & (df_patient.Alive_30_days_post_surgery)].copy()\n",
    "            else:\n",
    "                print(f'Primaries only tumor. n={sum(df_patient.Tumor_Type==\"Primary\")}')\n",
    "                df_patient2 = df_patient[(df_patient.Tumor_Type=='Primary') & (df_patient.loc[:,s_tcr].notna()) & (df_patient.Alive_30_days_post_surgery)].copy()\n",
    "        elif b_primary == 'Lung':\n",
    "            df_patient2 = df_patient[(df_patient.Cohort=='Lung')& (df_patient.Alive_30_days_post_surgery)].copy()\n",
    "        elif b_primary == 'Liver':\n",
    "            df_patient2 = df_patient[(df_patient.Cohort=='Liver')& (df_patient.Alive_30_days_post_surgery)].copy()\n",
    "        elif b_primary == 'Met':\n",
    "            if s_tcr.find('lood') > -1:\n",
    "                print(f'Met only blood. n={sum(df_patient.Blood_Type==\"Met\")}')\n",
    "                df_patient2 = df_patient[(df_patient.Blood_Type=='Met')& (df_patient.loc[:,s_tcr].notna()) & (df_patient.Alive_30_days_post_surgery)].copy()\n",
    "            else:\n",
    "                print(f'Met only tumor. n={sum(df_patient.Tumor_Type==\"Met\")}')\n",
    "                df_patient2 = df_patient[(df_patient.Tumor_Type=='Met')& (df_patient.loc[:,s_tcr].notna()) & (df_patient.Alive_30_days_post_surgery)].copy()\n",
    "            \n",
    "        else:\n",
    "            #df_patient2 = df_patient.copy()\n",
    "            df_patient2 = df_patient[(df_patient.loc[:,s_tcr].notna()) & (df_patient.Alive_30_days_post_surgery)].copy()\n",
    "            print(len(df_patient2))\n",
    "            print('using all')\n",
    "#             if s_tcr.find('lood') > -1:\n",
    "#                 df_patient2 = df_patient[df_patient.loc[:,s_foci].notna()].copy()\n",
    "        for SurvivalThreshold in [545]:# 365, #90, #, 730 #180,\n",
    "            GoodIdx = (df_patient2.loc[:,s_time] > SurvivalThreshold) & (df_patient2.loc[:,s_tcr].notna()) & (df_patient2.Alive_30_days_post_surgery)\n",
    "            BadIdx = (df_patient2.loc[:,s_time] <= SurvivalThreshold) & (df_patient2.loc[:,s_tcr].notna()) & (df_patient2.loc[:,s_censor] == 1) & (df_patient2.Alive_30_days_post_surgery)\n",
    "            df_patient2.loc[GoodIdx,f'{s_tcr}_{SurvivalThreshold}_day_survival'] = 'long'\n",
    "            df_patient2.loc[BadIdx,f'{s_tcr}_{SurvivalThreshold}_day_survival'] = 'short'\n",
    "        #plot\n",
    "        if len(set([s_tcr]).intersection(s_bad))>0:\n",
    "            d_fig = util.youden_low_good(df_patient2,b_primary,s_time,s_censor,s_tcr,figsize=(3,3.6))\n",
    "        else:\n",
    "            d_fig = util.youden_high_good(df_patient2,b_primary,s_time,s_censor,s_tcr,figsize=(3,3.6))#fontsize='small',\n",
    "        for key, item in d_fig.items():\n",
    "            if s_tcr.find('jaccard')>-1:\n",
    "                item.axes[0].set_title(f\"{s_tcr.replace('_',' ')} {b_primary}\",fontsize='large')\n",
    "            elif s_tcr .find('public')>-1:\n",
    "                item.axes[0].set_title(f\"{s_tcr.replace('_',' ')} {b_primary}\",fontsize='large')\n",
    "            else:\n",
    "                item.axes[0].set_title(f\"{s_tcr.replace('_',' ').title()}\",fontsize='large')\n",
    "            item.axes[0].set_ylabel(f\"Probability of {s_censor}\")\n",
    "            item.axes[0].set_xlabel(f\"{s_time}\")\n",
    "            plt.tight_layout()\n",
    "            item.savefig(f\"figures/KM_TCR_{s_tcr}_{s_censor}_{b_primary}.pdf\")\n",
    "        #break\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d50b8e6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# #Lung Cohort vs. Clinical Variables (chi squared)\n",
    "# ls_clin = [#'Grade',\n",
    "#            #'Age','Sex','Neoadjuvant',\n",
    "#            #'Stage',#'LV_Invasion','LN_Pos'\n",
    "#     'Tumor_Type'#'Blood_Type'\n",
    "#           ]\n",
    "# ls_cohort = [#'Recurrence_Sites_4',#'Lung_Cohort',\n",
    "# #              'No_Recurrence',\n",
    "# #              'Liver_Cohort',\n",
    "#     'Cohort',\n",
    "#              'pORG_All_quartiles'\n",
    "#              ]\n",
    "# b_resected = (df_merge.Recurrence_Sites_4!='No_Resection') & (df_merge.Recurrence_Sites_4.notna())\n",
    "# for s_group in ['TCR','All', 'TCR Met']:\n",
    "#     if s_group == 'TCR':\n",
    "#         #df = df_merge.loc[b_tcr_pts & b_resected,ls_clin + ls_cohort]\n",
    "#         df = df_merge.loc[b_tcr_pts ,ls_clin + ls_cohort]\n",
    "#     elif s_group == 'All':\n",
    "#         df = df_merge.loc[b_resected,ls_clin + ls_cohort]\n",
    "#         df = df_merge.loc[:,ls_clin + ls_cohort]\n",
    "#     elif s_group == 'TCR Met':\n",
    "#         df = df_merge.loc[df_merge.Tumor_Type=='Met',ls_clin + ls_cohort]\n",
    "#     df = df.astype('object')\n",
    "#     for s_clin in ls_clin:\n",
    "#         for s_cohort in ls_cohort:\n",
    "#             df2 = df.loc[:,[s_clin,s_cohort]].dropna()\n",
    "#             print(len(df2))\n",
    "#             crosstab= pd.crosstab(df2[s_clin], df2[s_cohort]) \n",
    "#             annot = pd.crosstab(df2[s_clin], df2[s_cohort],normalize='columns') \n",
    "#             try:\n",
    "#                 statistic,pvalue, dof, expected_freq = stats.chi2_contingency(crosstab)\n",
    "#                 if pvalue < 0.99:\n",
    "#                     fig, ax = plt.subplots()\n",
    "#                     sns.heatmap(crosstab,ax=ax,annot=True,cbar_kws={'label':'Percent Pts.'}) #/crosstab.sum()\n",
    "#                     ax.set_title(f'{s_group} p={pvalue:.3} n={len(df2)}')\n",
    "#     #             break\n",
    "#     #         break\n",
    "#             except:\n",
    "#                 continue\n",
    "#     break\n",
    "# #Lung Cohort TCR versus overall vs. Clinical Variables (chi squared)\n",
    "# ls_clin = [#'Grade',\n",
    "#            'Age','Sex','Neoadjuvant',#\n",
    "#            'Stage','LV_Invasion','LN_Pos'\n",
    "#           ]\n",
    "# ls_cohort = ['Lung_Cohort','No_Recurrence',\n",
    "#              'Liver_Cohort',\n",
    "#              ]\n",
    "\n",
    "# for s_clin in ls_clin:\n",
    "#     for s_cohort in ls_cohort:\n",
    "#         crosstab = pd.DataFrame()\n",
    "#         df1 = df_merge.loc[(b_tcr_pts) & (df_merge[s_cohort]),s_clin].dropna()\n",
    "#         crosstab['TCR'] = df1.value_counts()\n",
    "#         df2 = df_merge.loc[(df_merge[s_cohort]),s_clin].dropna()\n",
    "#         crosstab['All'] = df2.value_counts()\n",
    "#         statistic,pvalue, dof, expected_freq = stats.chi2_contingency(crosstab)\n",
    "#         if pvalue < 1.1:\n",
    "#             fig, ax = plt.subplots()\n",
    "#             #sns.heatmap(crosstab - expected_freq,ax=ax,annot=True,cbar_kws={'label':'Obs - Exp'})\n",
    "#             sns.heatmap(crosstab/crosstab.sum(),ax=ax,annot=True,cbar_kws={'label':'Percent Pts.'})\n",
    "#             ax.set_title(f'{s_clin} {s_cohort} p={pvalue:.3}')\n",
    "#             break\n",
    "#         break\n",
    "# #categorical correlation\n",
    "# sns.set_palette('tab10')\n",
    "# #df_all,s_group,s_marker,s_type,s_cell,alpha=0.05,s_propo='in',b_ttest=False):\n",
    "# ls_marker =  ['Shannon_Entropy_Tumor','Templates_per_ng','Shannon_Entropy_Blood',\n",
    "#  'Productive_Rearrangements','Simpsons_Diversity_Tumor','Simpsons_Diversity_Blood',\n",
    "#  'Clonality_Tumor','Clonality_Blood',\"Simpson's Evenness tumor\",\"Simpson's Evenness blood\",\n",
    "#  'Fraction Tumor Distinct TCRs','Percent Tumor Distinct Clones','Productive_Rearrangements_Blood']\n",
    "# ls_marker = ['Tumor_Cellularity_by_DNA_Primary',\n",
    "#        'Tumor_Cellularity_by_DNA_Met',]\n",
    "# ls_group = [#'Grade','Stage','LV_Invasion','LN_Pos',\n",
    "#     'pORG_Primary_quartiles', 'pORG_Met_quartiles',\n",
    "#     'Cohort',#'Age','Sex',\n",
    "#     'Neoadjuvant'\n",
    "#  ]\n",
    "# df_all = df_merge.loc[b_tcr_pts]\n",
    "# for s_marker in ls_marker:\n",
    "#     for s_group in ls_group:\n",
    "#         fig, pvalue,__,__ = util.categorical_correlation_boxplot(df_all,s_group,s_marker,\n",
    "#                                     s_type='',s_cell='',alpha=0.05,s_propo='',b_ttest=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c98b95b-e370-4339-95ad-c83ca7fea6ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load tcr\n",
    "df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "                          index_col=0)\n",
    "df_tcr =  pd.read_csv('Supplemental_Data/Supplemental_Dataset_8.csv',index_col=0)\n",
    "df_merge = df_tcr.merge(df_patient,on='Public_Patient_ID',suffixes = ('','_x'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "701208da",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# look as carl's youden cutoff for pORG\n",
    "pal_porg = ('#E69F00','#56B4E9',)\n",
    "pal_liv = ('#D55E00','#0072B2',)\n",
    "importlib.reload(util)\n",
    "sns.set_palette(pal_porg)\n",
    "cutoff = 0.02491\n",
    "df_merge['TCR_pORG_Primary'] = pd.NA\n",
    "df_merge.loc[df_merge.loc[:,'pORG_Primary'] > cutoff,'TCR_pORG_Primary'] = 'High' # 'pORG_0.2_Primary','pORG_0.2_Met','pORG_0.2_allPrimary','pORG_0.2_allMet',\n",
    "df_merge.loc[df_merge.loc[:,'pORG_Primary'] <= cutoff,'TCR_pORG_Primary'] = 'Low'\n",
    "\n",
    "cutoff = 0.02491\n",
    "df_merge['pORG_All'] = df_merge.loc[:,'pORG_allPrimary'].fillna(df_merge.loc[:,'pORG_allMet'])\n",
    "df_merge['TCR_pORG_All'] = pd.NA\n",
    "df_merge.loc[df_merge.loc[:,'pORG_All'] > cutoff,'TCR_pORG_All'] = 'High' # 'pORG_0.2_Primary','pORG_0.2_Met','pORG_0.2_allPrimary','pORG_0.2_allMet',\n",
    "df_merge.loc[df_merge.loc[:,'pORG_All'] <= cutoff,'TCR_pORG_All'] = 'Low'\n",
    "%matplotlib inline\n",
    "b_tcr = (df_merge.Clonality_Blood.notna() | df_merge.Clonality_Tumor.notna())\n",
    "#b_tcr = (df_merge.Shannon_Entropy_Tumor.notna())\n",
    "#b_tcr = (df_merge.Tumor_Type.notna())\n",
    "d_alive = df_merge.Alive_30_days_post_surgery\n",
    "\n",
    "#tcr\n",
    "#util.km_plot(df_patient[df_patient.Tumor_Type.notna()],'Cohort',s_time,s_censor)\n",
    "# fig, ax, __ = util.km_plot(df_merge[b_tcr & d_alive],'TCR_pORG_Primary',s_time,s_censor,fontsize='small',figsize=(3,2.8))\n",
    "# ax.set_title('TCR Pts: pORG Score (Primary)')\n",
    "# fig.savefig('figures/KM_TCR_pORG_Primary.pdf')\n",
    "fig, ax, __ = util.km_plot(df_merge[b_tcr & d_alive],'TCR_pORG_All',s_time,s_censor,fontsize='small',figsize=(3,2.8))\n",
    "ax.set_title('TCR Pts: pORG Score (All)')\n",
    "fig.savefig('figures/KM_TCR_pORG_All.pdf')\n",
    "#d_fig4.update({'Figure4_F_pORG':df_merge.loc[b_tcr & d_alive,['TCR_pORG_All',s_time,s_censor]]})\n",
    "\n",
    "sns.set_palette(pal_liv)\n",
    "fig, ax, __ = util.km_plot(df_merge[b_tcr & d_alive],'Cohort',s_time,s_censor,fontsize='small',figsize=(3,2.8))\n",
    "ax.set_title('TCR Pts: Cohort')\n",
    "fig.savefig('figures/KM_TCR_Cohort.pdf')\n",
    "#d_fig4.update({'Figure4_F_Cohort':df_merge.loc[b_tcr & d_alive,['Cohort',s_time,s_censor]]})\n",
    "\n",
    "#cox PH\n",
    "\n",
    "df_result = pd.DataFrame()\n",
    "for s_col in ['pORG_All',#'pORG_Primary',#'pORG_Met',#'pORG_allPrimary','pORG_allMet',\n",
    "              'Cohort']:\n",
    "    df = df_merge.loc[b_tcr & d_alive,[s_col,s_time,s_censor,'Public_Patient_ID']].dropna() #\n",
    "    \n",
    "    if not df.loc[:,s_col].dtype== 'float64':\n",
    "        df_dummy = pd.get_dummies(df.loc[:,[s_col]],drop_first=True)\n",
    "        df.drop(s_col,axis=1,inplace=True)\n",
    "        s_col = df_dummy.columns[0]\n",
    "        df[s_col] = df_dummy\n",
    "    print(df.Public_Patient_ID.duplicated().any())\n",
    "    print(len(df))\n",
    "    try: \n",
    "        fig2, cph = util.cph_plot(df.drop('Public_Patient_ID',axis=1),s_col,s_time,s_censor,figsize=(4,1.5))\n",
    "        plt.tight_layout()\n",
    "        fig2.savefig(f'figures/cph_TCR_{s_col}_{s_censor}.pdf')\n",
    "        df_result = pd.concat([df_result,cph.summary])\n",
    "        df_result.loc[s_col,'n'] = cph._n_examples\n",
    "    except:\n",
    "        print('error')\n",
    "    break\n",
    "#df_result.to_csv(f'results_TCR_cohorts_CPH_{s_censor}.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "957739cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #pORG vs survival within liver and lung TCR\n",
    "# df_patient['pORG_0.2_All'] = df_patient.loc[:,'pORG_0.2_allPrimary'].fillna(df_patient.loc[:,'pORG_0.2_allMet'])\n",
    "# for s_col in ['pORG_0.2_All']:\n",
    "#     for s_cohort in ['Liver','Lung']:\n",
    "#         print(s_cohort)\n",
    "#         try: #cox\n",
    "#             df = df_patient.loc[(df_patient.Tumor_Type.notna()) & (df_patient.Cohort==s_cohort),[s_col,s_time,s_censor]].dropna()\n",
    "#             fig2, cph = util.cph_plot(df,s_col,s_time,s_censor,figsize=(4,1.5))\n",
    "#             plt.tight_layout()\n",
    "#         except:\n",
    "#             print('error')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0014ae43-2fde-42c8-89fd-d174241242b9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sns.set_palette(pal_liv_r)\n",
    "importlib.reload(util)\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    ls_order=['Liver','Lung']\n",
    "    for y in ['pORG_Primary','pORG_Met']:\n",
    "        df_plot = df_merge.loc[:,['Cohort',y]].dropna()\n",
    "        print(len(df_plot))\n",
    "        plottting={'data':df_plot,'x':'Cohort','y':y,'order':ls_order}\n",
    "        fig, ax, __ = util.annotated_stripplot(plottting,ls_order,figsize=(2,2.3),s=3)\n",
    "        fig.savefig(f'figures/stripplot_TCRpts_{y}_vs_Cohort.pdf')\n",
    "        #d_fig4.update({f'Figure4_E_{y}':df_merge.loc[:,['Cohort',y]].dropna()})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5871ec57-c302-4d73-8aa0-619a63cc68a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merge.Cohort.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae2da4ef-c7b6-42d6-9762-a75c30311fb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merge.pORG_Primary.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a89f82bc-6bef-48f1-a4ef-c9d38d8afdb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merge.pORG_Met.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc5c21f4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# sns.set_palette(['black','gray'])\n",
    "# fig, __, __ = util.km_plot(df_merge,'Tumor_Type',s_time,s_censor,figsize=(3,3))\n",
    "# fig.savefig(f'figures/TCR_pts_primary_met_tum.pdf')\n",
    "# fig, __, __ = util.km_plot(df_merge,'Blood_Type',s_time,s_censor,figsize=(3,3))\n",
    "# fig.savefig(f'figures/TCR_pts_primary_met_blood.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7df27c68-dffb-4a58-af31-8ebe42a9e3a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# util.open_write_excel(d_fig4,filename='Source_Data_Figure4.xlsx')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f88315a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # define long and short lung\n",
    "# df_patient = pd.read_csv('Supplemental_Data/Supplemental_Dataset_8.csv',index_col=0)#pd.read_csv('annotation/20240304_Patient_Metadata_TCR_Metrics.csv',index_col=0)\n",
    "# #load tcr\n",
    "# df_patient_all = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "#                           index_col=0)\n",
    "# df_patient = df_patient.merge(df_patient_all,on='Public_Patient_ID',suffixes = ('','_x'))\n",
    "# print(len(df_patient))\n",
    "# short_long = 365 #df_patient.loc[(df_patient.Tumor_Type.notna()) & (df_patient.Cohort=='Lung'),s_time].median()\n",
    "# df_patient.loc[df_patient.loc[:,s_time] > short_long,'TCR_Lung_Median_Surv'] = 'lung_long'\n",
    "# df_patient.loc[df_patient.loc[:,s_time] < short_long,'TCR_Lung_Median_Surv'] = 'lung_short'\n",
    "# df_patient.loc[(df_patient.Tumor_Type.isna()) | (df_patient.Cohort!='Lung'),'TCR_Lung_Median_Surv'] = np.nan\n",
    "# # define long and short liver\n",
    "# short_long = df_patient.loc[(df_patient.Tumor_Type.notna()) & (df_patient.Cohort=='Liver'),s_time].median()\n",
    "# df_patient.loc[df_patient.loc[:,s_time] > short_long,'TCR_Liver_Median_Surv'] = 'liver_long'\n",
    "# df_patient.loc[df_patient.loc[:,s_time] < short_long,'TCR_Liver_Median_Surv'] = 'liver_short'\n",
    "# df_patient.loc[(df_patient.Tumor_Type.isna()) | (df_patient.Cohort!='Liver'),'TCR_Liver_Median_Surv'] = np.nan\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59bdea4c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# #util.km_plot(df_patient,'TCR_Lung_Median_Surv',s_time,s_censor)\n",
    "# # does is hold for the Do the conclusions remain valid for lung 365 days or longer\n",
    "# df_patient['Cohort_Long'] = df_patient.Cohort.replace({'Lung':pd.NA}).copy()\n",
    "# df_patient['Cohort_Long'].fillna(df_patient.TCR_Lung_Median_Surv,inplace=True)\n",
    "# df_patient['Cohort_Long'] = df_patient.Cohort_Long.replace({'lung_short':pd.NA})\n",
    "# df_patient['Cohort_Long'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "339b3b63",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#df_patient['TCR_Lung_Liver'] = df_patient.TCR_Lung_Median_Surv.fillna(df_patient.TCR_Liver_Median_Surv)\n",
    "#util.km_plot(df_patient,'TCR_Lung_Liver',s_time,s_censor)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "512a24a1",
   "metadata": {},
   "source": [
    "### TCR heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b13340f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# ls_foci_heatmap = ['public_Blood','jaccard_Blood',\n",
    "#           'jaccard_Tumor','public_Tumor','Clonality_Blood',\"Simpson's Evenness blood\",\n",
    "#             'Simpsons_Diversity_Blood','Shannon_Entropy_Blood','Productive_Rearrangements_Blood',\n",
    "#          'Productive_Rearrangements','Simpsons_Diversity_Tumor',\n",
    "#             'Shannon_Entropy_Tumor','Templates_per_ng',\n",
    "#         \"Simpson's Evenness tumor\",\n",
    "#     'Percent Tumor Distinct Clones',\n",
    "\n",
    "#              ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83fc85de",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# s_porg = 'pORG_All'\n",
    "# df_pri = df_patient.copy()\n",
    "# #heatmap\n",
    "# # make sure to end on pORG all\n",
    "# if s_porg == 'pORG_All':\n",
    "#     import matplotlib as mpl\n",
    "#     df_pri['Survival interval'] = pd.qcut(df_merge.loc[:,s_time],6)\n",
    "#     #df_pri.rename({'hue':f'{s_porg} (median)'},axis=1,inplace=True)\n",
    "#     ls_annot = ['Tumor_Type', 'Cohort',f'{s_porg}_quartiles' ,'Survival interval']\n",
    "#     cmap=plt.cm.get_cmap('Blues', 6)\n",
    "#     ls_color=[mpl.cm.Set1.colors,('#D55E00', '#0072B2'),('#E69F00', '#56B4E9'),\n",
    "#              [cmap(item) for item in np.arange(0,1,.17)]]\n",
    "#     df_annot = pd.DataFrame()\n",
    "#     dd_color = {}\n",
    "#     for idx, s_annot in enumerate(ls_annot):\n",
    "#         color_palette = ls_color[idx]\n",
    "#         d_color = dict(zip(sorted(df_pri.loc[:,s_annot].dropna().unique()),color_palette[0:len(df_pri.loc[:,s_annot].dropna().unique())]))\n",
    "#         d_color.update({'NA':'lightgray'})\n",
    "#         network_colors = df_pri.loc[:,s_annot].astype('object').fillna('NA').map(d_color)\n",
    "#         df_annot[s_annot] = pd.DataFrame(network_colors)\n",
    "#         dd_color.update({s_annot:d_color})\n",
    "#     from matplotlib.pyplot import gcf\n",
    "#     g = sns.clustermap(df_pri.loc[:,ls_foci_heatmap].dropna(),z_score=1,dendrogram_ratio=0.1,cmap='RdBu_r',vmin=-5,vmax=5,\n",
    "#           cbar_pos=(0.01, 0.94, 0.04, 0.08),figsize=(15, 10),row_colors=df_annot,cbar_kws={'label': 'z-score'},xticklabels=1,\n",
    "#                       method='complete')\n",
    "#     for idx, (s_annot, d_color) in enumerate(dd_color.items()):\n",
    "#         g.ax_col_dendrogram.bar(0, 0, color='w',label=' ', linewidth=0)\n",
    "#         for label,color in d_color.items():\n",
    "#             g.ax_col_dendrogram.bar(0, 0, color=color,label=label, linewidth=0)\n",
    "#     l1 = g.ax_col_dendrogram.legend(loc=\"right\", ncol=1,bbox_to_anchor=(0, 0.7),bbox_transform=gcf().transFigure)\n",
    "#     g.ax_heatmap.set_yticks([])\n",
    "#     g.ax_heatmap.set_title(s_porg)\n",
    "#     g.savefig(f'figures/heatmap_TCR_{s_porg}.png',dpi=300)\n",
    "#     #plt.close(g.fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "452b9b40",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# def heatmap_row_colors(df_meta,d_corr,ls_annot = ['Type','Cohort','pORG_0.2_All_quartiles']):\n",
    "#     ls_color=[mpl.cm.Set1.colors,('#D55E00', '#0072B2'),('#E69F00', '#56B4E9')]\n",
    "#     df_annot = pd.DataFrame()\n",
    "#     dd_color = {}\n",
    "#     for idx, s_annot in enumerate(ls_annot):\n",
    "#         color_palette = ls_color[idx]\n",
    "#         d_color = dict(zip(sorted(df_meta.loc[:,s_annot].dropna().unique()),color_palette[0:len(df_meta.loc[:,s_annot].dropna().unique())]))\n",
    "#         d_color.update({'NA':'lightgray'})\n",
    "#         d_corr[s_annot] = d_corr.index.map(dict(zip(df_meta.Sample,df_meta.loc[:,s_annot])))\n",
    "#         network_colors = d_corr.loc[:,s_annot].astype('object').fillna('NA').map(d_color)\n",
    "#         df_annot[s_annot] = pd.DataFrame(network_colors)\n",
    "#         dd_color.update({s_annot:d_color})\n",
    "#     return(df_annot,dd_color)\n",
    "\n",
    "# def row_colors_legend(g,dd_color):\n",
    "#     for idx, (s_annot, d_color) in enumerate(dd_color.items()):\n",
    "#         g.ax_col_dendrogram.bar(0, 0, color='w',label=' ', linewidth=0)\n",
    "#         for label,color in d_color.items():\n",
    "#             g.ax_col_dendrogram.bar(0, 0, color=color,label=label, linewidth=0)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "758c5f45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #overlap heatmaps\n",
    "# s_dir =f'{codedir.split(\"Liver_Lung_PDAC\")[0]}R'\n",
    "# df_meta = pd.read_csv(f'{s_dir}/raw_TCR_data_test/metadata.txt',sep='\\t')\n",
    "# df_meta['Type'] = df_meta.Site.copy()\n",
    "# df_meta['Type'] = df_meta.Type.replace('Blood',np.nan)\n",
    "# df_meta['Type'] = df_meta.Type.fillna(df_meta.Blood_Type)\n",
    "# ls_file=[#'results_TCR_public_overlap.csv','results_TCR_morisita_overlap.csv',\n",
    "#          'results_TCR_jaccard_overlap.csv']\n",
    "# for s_file in ls_file:\n",
    "#     df = pd.read_csv(f'{s_dir}/{s_file}',index_col=0)\n",
    "#     df.columns = [item.split('TCR_raw_data_')[1] for item in df.columns]\n",
    "#     df.index = [item.split('TCR_raw_data_')[1] for item in df.index]\n",
    "#     df_blood = np.log(df.loc[df.index.str.contains('-B'),df.columns.str.contains('-B')])\n",
    "#     df_tum = np.log(df.loc[~df.index.str.contains('-B'),~df.columns.str.contains('-B')])\n",
    "#     df_tum = df_tum.replace(-np.Inf, np.nan)\n",
    "#     d_dfs = {'Tumor':df_tum,'Blood':df_blood}\n",
    "#     for s_type, d_corr in d_dfs.items():\n",
    "#         corr = d_corr.fillna(0)\n",
    "#         df_annot,dd_color = heatmap_row_colors(df_meta,d_corr)\n",
    "#         g = sns.clustermap(corr, cmap=\"coolwarm\", figsize=(7, 6),row_colors=df_annot)\n",
    "#         row_colors_legend(g,dd_color)\n",
    "#         #mask = np.triu(np.ones_like(corr)) #triu\n",
    "#         mask = np.diag(np.diag(np.ones_like(corr)))\n",
    "#         values = g.ax_heatmap.collections[0].get_array().reshape(corr.shape)\n",
    "#         new_values = np.ma.array(values, mask=mask)\n",
    "#         g.ax_heatmap.collections[0].set_array(new_values)\n",
    "# #         g.ax_col_dendrogram.set_visible(False)\n",
    "#         g.ax_heatmap.set_yticks([])\n",
    "#         g.ax_heatmap.set_xticks([])\n",
    "#         s_title = f\"{s_type} {s_file.split('results_TCR_')[1].split('.csv')[0].replace('_',' ').title()}\"   \n",
    "#         g.ax_heatmap.set_title(s_title,pad=30)\n",
    "#         l1 = g.ax_col_dendrogram.legend(loc=\"right\", ncol=1,bbox_to_anchor=(0, 0.7),bbox_transform=gcf().transFigure)\n",
    "#         break\n",
    "#     break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad227f14-f1dd-40d3-a076-5d5cf8929b79",
   "metadata": {},
   "source": [
    "# TCR met site  <a name=\"TCRmet\"></a>\n",
    "\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "624810d9-6226-42b4-9354-72a0721c47b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load tcr\n",
    "df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "                          index_col=0)\n",
    "df_tcr =  pd.read_csv('Supplemental_Data/Supplemental_Dataset_8.csv',index_col=0)\n",
    "df_merge = df_tcr.merge(df_patient,on='Public_Patient_ID',suffixes = ('','_x'))\n",
    "df_merge[\"log_Simpsons_Evenness_tumor\"] = np.log(df_merge[\"Simpson's Evenness tumor\"] )\n",
    "ls_met_site = [#'pORG_Met','pORG_All',\n",
    "           #'pSUB_Met','PurIST_Met',\n",
    "     \"Simpson's Evenness tumor\",\n",
    "    #\"log_Simpsons_Evenness_tumor\",\n",
    "  'Shannon_Entropy_Tumor',#'Templates_per_ng',#'Shannon_Entropy_Blood',\n",
    " #\"Simpson's Evenness blood\",\n",
    " \n",
    " 'Productive_Rearrangements',#'Simpsons_Diversity_Tumor','Simpsons_Diversity_Blood',\n",
    " 'Clonality_Tumor',#'Clonality_Blood',\n",
    "  'Percent Tumor Distinct Clones',#'Productive_Rearrangements_Blood'\n",
    "         ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65f9b3a2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "##TCR met sites\n",
    "#color by met\n",
    "#add pORG quartiles, plot violins \n",
    "\n",
    "%matplotlib inline\n",
    "#update the lung error\n",
    "#df_merge.loc[df_merge.Public_Patient_ID=='ST-00015839','TCR_Met_Site'] = 'Lung'\n",
    "df_merge['TCR_Met_Site'] = df_merge.TCR_Met_Site.replace({np.NaN:pd.NA})\n",
    "d_replace = {'Gastric serosal':'Peritoneum', 'Lower quadrant mass':'Peritoneum',#'Mesocolon':'?',\n",
    "                 'Tissue near left hepatic artery':'Near hepatic artery',\n",
    "                'Pelvic nodule':'Peritoneum', 'Omentum':'Peritoneum',#'Gallbladder':'?', \n",
    "                 'Retroperitoneum':'Peritoneum',np.NaN:pd.NA}\n",
    "df_merge['TCR_Met_Sites'] = df_merge.TCR_Met_Site.replace(d_replace)\n",
    "hue_order = ['Lung','Peritoneum','Liver','Near hepatic artery','Mesocolon','Gallbladder','Lymph Node']\n",
    "sns.set_palette('tab10')\n",
    "importlib.reload(util)\n",
    "s_stats = 'mean'\n",
    "s_porg = 'pORG_Met'\n",
    "hue='TCR_Met_Sites'\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    #'''\n",
    "    for s_comp in ['Cohort','quartiles']:#, 'pORG_78_Primary',\n",
    "        df_pri = df_merge.loc[df_merge.Tumor_Type=='Met'].copy()\n",
    "        df_pri = util.add_quartiles(df_pri,s_porg)#.drop(219) #p. rearrangements outlier \n",
    "        for s_foci in ls_met_site:\n",
    "            #s_foci =  \"Productive_Rearrangements\"#\"Simpson's Evenness tumor\"#'Percent Tumor Distinct Clones'\n",
    "            print(s_foci)\n",
    "            fig, ax = util.plot_violins3(df_pri[(df_pri.loc[:,hue].notna()) & (df_pri.loc[:,s_comp].notna())],\n",
    "                        s_stats,s_foci,s_comp,s_porg,hue_order,hue=hue,figsize=(2,2.5))#True#False\n",
    "            ax.legend().remove()\n",
    "            ax.set_title('')#(f\"{s_foci.replace('_',' ')}\", loc='left',fontsize='medium',pad=10) #\n",
    "            if s_comp == 'quartiles':\n",
    "                ax.set_xlabel(s_porg.replace('_',' '),fontsize='small')\n",
    "            else:\n",
    "                ax.set_xlabel('')\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/violinplot_met_colors_{s_foci}_{s_porg}_{s_comp}.pdf')\n",
    "            #break\n",
    "        #break\n",
    "    \n",
    "    #group by met\n",
    "    #'''\n",
    "    d_replace = {'Mesocolon':'Other',\n",
    "                 'Near hepatic artery':'Other',\n",
    "                'Gallbladder':'Other',}\n",
    "    df_pri = df_merge\n",
    "    df_pri['TCR_Met_Site_S'] = df_pri.TCR_Met_Sites.replace(d_replace)\n",
    "    for s_foci in ls_met_site:#ls_foci:\n",
    "        df_plot=df_pri[(df_pri.loc[:,hue].notna()) & (df_pri.loc[:,s_foci].notna())]\n",
    "        fig, ax = util.plot_violins3(df_plot,s_stats,s_foci,\n",
    "                            s_comp='TCR_Met_Site_S',s_porg=s_porg,hue_order=hue_order,hue=hue,figsize=(5.5,2.5)) #True#False\n",
    "        ax.set_title('')#(f\"{s_foci.replace('_',' ')}\", loc='left',fontsize='large',pad=10) #\n",
    "        ax.set_xlabel('')\n",
    "        ax.legend(bbox_to_anchor=(1,1),fontsize='medium',loc='upper left')\n",
    "        plt.tight_layout()\n",
    "        fig.savefig(f'figures/violinplot_met_sites_{s_foci}.pdf')\n",
    "        #break\n",
    "    #df_pri.loc[~df_pri.Cohort.isna(),['Public_Patient_ID','Cohort','TCR_Met_Site']].sort_values(by=['Cohort','TCR_Met_Site']).to_csv('test.csv')\n",
    "    #'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5ce979d-7c89-4f5b-bf69-65c6df8f60c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_plot.loc[:,ls_met_site +['TCR_Met_Site_S','Public_Patient_ID']].to_csv(f'Source_Extended_Data_Figure7_N.csv')\n",
    "# df_plot.loc[:,ls_met_site +['TCR_Met_Site_S','Public_Patient_ID']].to_csv(f'Source_Extended_Data_Figure8_A-B.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8b802b9",
   "metadata": {},
   "source": [
    "## Section 8 <a name=\"violin2\"></a>\n",
    "\n",
    "\n",
    "GSEA plots. I think we should say we are showing any pathway with FDRq < 0.15 and NES>1.5 for all for datasets, so if a pathway doesnt meet these cutoffs, they are not shown as a bar in the two graphs: liver/lung and high/low pORG, and high/low PurIST and high/low pSUB. \n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79603856-bbee-4c63-b65f-53c82052d30e",
   "metadata": {},
   "source": [
    "## figure 2  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3cb8dac",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_colorblind_g ={'Liver': '#d55e00',\n",
    " 'Lung': '#0072b2',\n",
    " 'high': '#e69f00',\n",
    " 'low': '#56b4e9',\n",
    " 'basal-like': '#383838',# 'dimgray',# '#000000',#\n",
    " 'classical': '#cc79a7',\n",
    " 'high pSUB': '#f0e442',\n",
    " 'low pSUB': '#d55e00'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2a89b66",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# #lper patient\n",
    "# %matplotlib inline\n",
    "# #Can you redo this removing the non lung and non liver mets in their respective cohorts:\n",
    "# importlib.reload(util)\n",
    "# import warnings\n",
    "# # s_out = '20230828_Patient_Metadata.csv'\n",
    "# # s_out = '20230921_Patient_Metadata.csv'\n",
    "# # s_out = '20231206_Patient_Metadata.csv'\n",
    "# # df_patient= pd.read_csv(f'annotation/{s_out}',index_col=0)\n",
    "# #load \n",
    "# df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "#                           index_col=0)\n",
    "\n",
    "# s_stats = 'mean'\n",
    "# d_order =  {'Cohort':['Liver','Lung'],\n",
    "#             'PurIST Subtype':['basal-like','classical']}\n",
    "# lls_score = [['pORG_Primary','pSUB_Primary','PurIST_Primary'],\n",
    "#            #['pORG_0.2_All', 'pSUB1e-04_All','PurIST_All'],\n",
    "#           ['pORG_Met','pSUB_Met','PurIST_Met'],\n",
    "#           ]\n",
    "# with warnings.catch_warnings():\n",
    "#     warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "#     for ls_score in lls_score:\n",
    "#         if ls_score[0].find('All') > -1:\n",
    "#             ls_population = ['match','pORG_Met','pORG_Primary']\n",
    "#         elif ls_score[0].find('Met') > -1:\n",
    "#             ls_population = ['pORG_Met']\n",
    "#         else:\n",
    "#             ls_population = ['pORG_Primary']\n",
    "#         for s_population in ls_population:\n",
    "#             print(f'population {s_population}')\n",
    "#             if s_population.find('_Met')>-1:\n",
    "#                 lb_met = [False]#True,\n",
    "#             else:\n",
    "#                 lb_met = [False]\n",
    "#             for b_met in lb_met:\n",
    "#                 for s_score in ls_score:\n",
    "#                     print(f'score {s_score}')\n",
    "#                     if s_population == 'match':\n",
    "#                         s_population = s_score\n",
    "#                         print(f'population {s_population}')\n",
    "#                     if s_score.find('_All')> -1: #need to calculate and add \"all\" scores\n",
    "#                         print('Primaries and Mets')\n",
    "#                         df_pri = df_patient.copy()\n",
    "#                         df_pri['PurIST_allPrimary'] = df_pri.PurIST_Primary.copy()\n",
    "#                         df_pri['PurIST_allMet'] = df_pri.PurIST_Met.copy()\n",
    "#                         for s_score_all in ['PurIST','pORG','pSUB']:\n",
    "#                             df_whole = pd.DataFrame(columns=df_pri.columns)\n",
    "#                             for s_all in [f'{s_score_all}_allPrimary',f'{s_score_all}_allMet']:\n",
    "#                                 #select the mets or primaries (this will prioritze primary over met for matched)\n",
    "#                                 df_half = df_pri[df_pri.loc[:,s_all.replace(s_score_all,'pORG')].notna()].copy()\n",
    "#                                 #only fill in the non duplicate \n",
    "#                                 df_half = df_half[~df_half.Public_Patient_ID.isin(df_whole.Public_Patient_ID)].copy()\n",
    "#                                 df_half.rename({s_all:f'{s_score_all}_All'},axis=1,inplace=True)\n",
    "#                                 #df_whole.index = [str(item) + 'a' for item in df_whole.index]\n",
    "#                                 with warnings.catch_warnings():\n",
    "#                                     warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "#                                     df_whole = pd.concat([df_whole,df_half],axis=0,ignore_index=True)\n",
    "#                             #print(len(df_whole))\n",
    "#                             df_pri = df_pri.merge(df_whole.loc[:,['Public_Patient_ID',f'{s_score_all}_All']],on='Public_Patient_ID',how='left',suffixes=('','_'))\n",
    "#                     else:\n",
    "#                         df_pri = df_patient.copy()\n",
    "#                     if s_population.find('_Met')> -1:\n",
    "#                         print('mets')\n",
    "#                         if b_met: #select mets specific to cohort\n",
    "#                             continue\n",
    "#                             df_pri = df_pri[(df_pri.loc[:,s_population].notna()) & (df_patient.Specimen_Site_Met.isin(['Liver','Lung']))].copy()\n",
    "#                         else: #select mets\n",
    "#                             df_pri = df_pri[(df_pri.loc[:,s_population].notna())].copy()\n",
    "#                     elif s_population.find('_Primary')> -1: #select primaries\n",
    "#                         print('primaries')\n",
    "#                         df_pri = df_pri[df_pri.loc[:,s_population].notna()].copy()\n",
    "#                     df_pri['PurIST_All'] = df_pri.PurIST_Primary.fillna(df_pri.PurIST_Met)\n",
    "#                     df_pri.loc[df_pri.loc[:,f'PurIST_{s_score.split(\"_\")[-1]}'] > 0.5,'PurIST Subtype'] = 'basal-like'\n",
    "#                     df_pri.loc[df_pri.loc[:,f'PurIST_{s_score.split(\"_\")[-1]}'] <= 0.5,'PurIST Subtype'] = 'classical'\n",
    "#                     df_pri.loc[df_pri.Public_Patient_ID.duplicated(),'PurIST Subtype'] = np.nan\n",
    "#                     df_both,d_pval,order = util.violin_stats2(df_pri,d_order,s_score,s_stats)\n",
    "#                     fig,pvalues,corrected = util.plot_violins2(df_both,d_pval,d_order,s_stats,s_score,order,\n",
    "#                                                                d_colorblind_g,s_population,b_correct=True,figsize=(3.5,2.5))\n",
    "#                     plt.ylim(-0.7, 1.15)\n",
    "#                     if s_score.find('PurIST') > -1:\n",
    "#                         plt.ylim(-0.1, 1.3)\n",
    "#                     fig.savefig(f'figures/violinplot_fig2_{s_score}_{s_population.split(\"_\")[-1]}_mets_only{b_met}.png')\n",
    "#                 #break\n",
    "#             #break\n",
    "#         #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "affab052",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(df_patient.PurIST_Primary.notna().sum())\n",
    "# print(df_patient.loc[:,'pORG_Primary'].notna().sum())\n",
    "# print(df_patient.loc[:,'pSUB_Primary'].notna().sum())\n",
    "\n",
    "# print(df_patient.PurIST_Met.notna().sum()) #why only 71 mets have purist?\n",
    "# print(df_patient.loc[:, 'pSUB_Met'].notna().sum())\n",
    "# print(df_patient.loc[:, 'pORG_Met'].notna().sum())\n",
    "# print(216 + 72 - 10)\n",
    "\n",
    "# print(df_patient.loc[:,['PurIST_Primary','PurIST_Met']].notna().all(axis=1).sum())\n",
    "# print(df_patient.loc[:,['pSUB_Primary','pSUB_Met']].notna().all(axis=1).sum())\n",
    "# print(df_patient.loc[:,['pORG_Primary','pORG_Met']].notna().all(axis=1).sum())\n",
    "\n",
    "# print(df_patient.loc[:,'Cohort'].notna().sum())\n",
    "# print(df_patient.loc[df_patient.loc[:,'pORG_Primary'].notna(),'Cohort'].notna().sum())\n",
    "# print(df_patient.loc[df_patient.loc[:,'pORG_Met'].notna(),'Cohort'].notna().sum())\n",
    "# print(df_patient.loc[df_patient.loc[:,'PurIST_Met'].notna(),'Cohort'].notna().sum())\n",
    "# print(df_patient.loc[df_patient.loc[:,['pORG_Primary','pORG_Met']].notna().all(axis=1),'Cohort'].notna().sum())\n",
    "# 75 + 38 - 7"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "207770d8",
   "metadata": {},
   "source": [
    "## per specimen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a7d0788-cbd9-42b7-b6a0-ecc1f09b68ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_rename = {'pORG_Up_55_Primaries': 'pORG_Primary',\n",
    " 'pSUB_Up_51_Primaries': 'pSUB_Primary',\n",
    " 'pORG_Up_55_Mets': 'pORG_Met',\n",
    " 'pSUB_Up_51_Mets': 'pSUB_Met',\n",
    " 'PurIST_Subtype': 'PurIST Subtype',\n",
    "  'pORG_Up_55_All':'pORG_All',\n",
    " 'pSUB_Up_51_All': 'pSUB_All',\n",
    " 'PurIST_Score':'PurIST_All'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7cef8c9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#per specimen\n",
    "%matplotlib inline\n",
    "#Can you redo this removing the non lung and non liver mets in their respective cohorts:\n",
    "df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name='RNA_Specimen_Metadata',\n",
    "                          index_col=0)\n",
    "# add columns\n",
    "\n",
    "for s_site in ['Primary','Met']:\n",
    "    df_patient[f'PurIST_{s_site}'] = np.nan\n",
    "    df_patient.loc[df_patient.Tumor_Type==s_site,f'PurIST_{s_site}'] = df_patient.loc[df_patient.Tumor_Type==s_site,'PurIST_Score']\n",
    "    df_patient.loc[:,f'PurIST_{s_site}'] = df_patient.loc[:,f'PurIST_{s_site}'].astype('float')\n",
    "df_patient.rename(d_rename,axis=1,inplace=True)\n",
    "df_patient['Cohort'] = pd.NA\n",
    "df_patient.loc[df_patient.Lung_Met_Present==1,'Cohort'] = 'Lung'\n",
    "df_patient.loc[df_patient.Liver_Met_Present==1,'Cohort'] = 'Liver'\n",
    "\n",
    "importlib.reload(util)\n",
    "import warnings\n",
    "#primary, met\n",
    "s_stats = 'mean'\n",
    "d_order =  {'Cohort':['Liver','Lung'],\n",
    "            'PurIST Subtype':['basal-like','classical']}\n",
    "ls_score = ['pORG', 'pSUB','PurIST']\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    for s_score_name in ls_score:\n",
    "        for s_site in ['Primary','Met']:\n",
    "            s_score = f'{s_score_name}_{s_site}'\n",
    "            df_pri = df_patient[(df_patient.Tumor_Type==s_site) & (df_patient.loc[:,s_score].notna())].copy()\n",
    "            print(f'score {s_score}')\n",
    "            df_both,d_pval,order = util.violin_stats2(df_pri,d_order,s_score,s_stats)\n",
    "            fig,pvalues,corrected = util.plot_violins2(df_both,d_pval,d_order,s_stats,s_score,order,\n",
    "                                    d_colorblind_g,s_site,b_correct=True,figsize=(2.7,2.2),s=2,b_title=False,alpha=1)\n",
    "            if s_score_name == 'pORG':\n",
    "                plt.ylim(-0.7, 1.13)\n",
    "            elif s_score_name == 'pSUB':\n",
    "                plt.ylim(-0.7, 1.35)\n",
    "            elif s_score_name == 'PurIST':\n",
    "                plt.ylim(-0.1, 1.5)\n",
    "            newlabels = [item.get_text().replace('basal-like',f'basal-{linebreak}like') for item in fig.get_axes()[0].get_xticklabels()]\n",
    "            fig.get_axes()[0].set_xticklabels(newlabels,fontsize='small')\n",
    "            fig.get_axes()[0].set_ylabel(s_score.replace('_',' '),fontsize='small')\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/Fig1-2_GSVA_violoins_{s_score}_{s_site}.pdf')\n",
    "            if s_site=='Primary':\n",
    "                d_fig1.update({f'Figure1_I_{s_score}':df_both})\n",
    "            else:\n",
    "                d_fig1.update({f'Figure1_J_{s_score}':df_both})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52ec1dfa-b159-48cd-99e6-04e39c4f2c46",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_pri = df_patient.rename({'Tumor_Type':'Specimen Type'},axis=1)\n",
    "d_pm ={'Met': 'black','Primary': 'gray'}\n",
    "d_order_pm = {'Specimen Type':['Primary','Met']}\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    for s_score_name in ls_score:\n",
    "        for s_site in ['All']:\n",
    "            s_score = f'{s_score_name}_{s_site}'\n",
    "            df_both,d_pval,order = util.violin_stats2(df_pri,d_order_pm,s_score,s_stats)\n",
    "            fig,pvalues,corrected = util.plot_violins2(df_both,d_pval,d_order_pm,s_stats,s_score,order,\n",
    "                                            d_pm,s_site,b_correct=True,figsize=(2,1.8),s=2,b_title=False,alpha=1)\n",
    "            upper = fig.get_axes()[0].get_ylim()[1]*1.25\n",
    "            lower = fig.get_axes()[0].get_ylim()[0]\n",
    "            fig.get_axes()[0].set_ylim(lower,upper)\n",
    "            fig.get_axes()[0].set_ylabel(s_score.replace('_',' '),fontsize='small')\n",
    "            plt.tight_layout()\n",
    "            fig.savefig(f'figures/Fig1-2_GSVA_violoins_{s_score}_Pri_vs_Met.pdf')\n",
    "            d_fig1.update({f'Figure1_K_{s_score_name}':df_both})\n",
    "        #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6d3e3d4-8279-4bb5-8b43-094e5d14bfd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#util.open_write_excel(d_fig1,filename='Source_Data_Figure1.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "891fa00e-9fe0-4b6b-a660-29a674049db3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# figure 1 F\n",
    "sns.set_palette(['black','gray'])\n",
    "d_pm = {'Met': (0,0,0), 'Primary': (0.5,0.5,0.5)}\n",
    "ls_groups = ['Liver','Lung']\n",
    "color = 'Tumor_Type'\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "    plotting = {'data':df_patient[df_patient.Cohort.notna()],'y':'PurIST_All','x':'Cohort',\n",
    "                'order':ls_groups}\n",
    "    fig,ax,test_results = annotated_stripplot_color(plotting,ls_groups,color,b_correct=False,figsize=(3,2.5))\n",
    "    ax.legend(bbox_to_anchor=(1,1.1),title_fontsize='x-small',fontsize='x-small',markerscale=0.5,title='Specimen Type')\n",
    "    ax.set_ylabel('PurIST Score',fontsize='small')\n",
    "    ax.set_xlabel('Cohort',fontsize='small')\n",
    "    ax.set_xticklabels([f'Liver{linebreak}N=85',f'Lung{linebreak}N=28'],fontsize='small')\n",
    "    ax.set_yticklabels(ax.get_yticklabels(),fontsize='small')\n",
    "    ax.axhline(0.25,linestyle=':',linewidth=1)\n",
    "    ax.axhline(0.5,linestyle=':',linewidth=1)\n",
    "    ax.axhline(0.75,linestyle=':',linewidth=1)\n",
    "    ax.spines[['right', 'top']].set_visible(False)\n",
    "    plt.tight_layout()\n",
    "    fig.savefig(f'figures/Fig1F_Purist_vs_Site_vs_Cohort.pdf')\n",
    "    d_fig1.update({'Figure1_F':df_patient.loc[df_patient.Cohort.notna(),['Public_Specimen_ID','Cohort','PurIST_All']]})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acc0fc85",
   "metadata": {},
   "source": [
    "## Figure 3: GSEA barplots <a name=\"bars2\"></a> \n",
    "\n",
    "stacked, hatched\n",
    "\n",
    "[contents](#contents)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fe5bec4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load GSEA data\n",
    "ls_columns = ['NAME', 'SIZE', 'ES', 'NES', 'NOM.p.val', 'FDR.q.val', 'FWER.p.val' ]\n",
    "ls_gsea = ['Top4th_PurIST.Score_vs_Bottom4th_PurIST.Score_h.xlsx',\n",
    "     'basal-like_vs_classical_h.xlsx','LiverCohort_vs_LungNotLiverCohort_h.xlsx',\n",
    " 'Top4th_pSUB.1eNeg4_vs_Bottom4th_pSUB.1eNeg4_h.xlsx',\n",
    " 'Top4th_pSUB.1eNeg5_vs_Bottom4th_pSUB.1eNeg5_h.xlsx',\n",
    " 'Top4th_pORG.20_vs_Bottom4th_pORG.20_h.xlsx',\n",
    "           #mets\n",
    "    'Top4thMet_pORG_Up_55_vs_Bottom4thMet_JM_pORG_Up_55_h.xlsx',\n",
    "           'Top4thMet_pSUB_Up_51_vs_Bottom4thMet_JM_pSUB_Up_51_h.xlsx',\n",
    "           'LiverCohortMet_vs_LungNotLiverCohortMet_h.xlsx',\n",
    "'Top4thMet_PurIST_vs_Bottom4thMet_PurIST_h.xlsx',  \n",
    "           #tcr mets\n",
    " 'LiverTcr_pORG_Up_55_JM_vs_LungTcr_JM_pORG_Up_55_JM_h.xlsx',\n",
    " 'Top4thTcr_pORG_Up_55_JM_vs_Bot4thTcr_JM_pORG_Up_55_JM_h.xlsx',\n",
    "    ]\n",
    "d_en = {}\n",
    "for s_gsea in ls_gsea:\n",
    "    d_load =  pd.read_excel(f'data/{s_gsea}',sheet_name=None)#,index_col=0\n",
    "    df_up = d_load['GSEA_UP'].loc[:,ls_columns] #df[df.ES>0]\n",
    "    df_down = d_load['GSEA_DN'].loc[:,ls_columns]#df[df.ES<0]\n",
    "    d_en.update({f\"{s_gsea.split('_h.xlsx')[0]}_UP\":df_up})\n",
    "    d_en.update({f\"{s_gsea.split('_h.xlsx')[0]}_DN\":df_down})\n",
    "    #break\n",
    "\n",
    "d_colorblind = {'Liver': '#d55e00',\n",
    " 'Lung': '#0072b2',\n",
    " 'high pORG': '#e69f00',\n",
    " 'low pORG': '#56b4e9',\n",
    " 'basal-like': '#000000','high PurIST': '#000000',\n",
    " 'classical': '#cc79a7','low PurIST': '#cc79a7',\n",
    " 'high pSUB': '#f0e442',\n",
    " 'low pSUB': '#009E73'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "502c1013",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## USED:  bar plots of GSEA, hashmarks  \n",
    "%matplotlib inline\n",
    "importlib.reload(util)\n",
    "hatch='||'#'///'#\n",
    "my_cmap = plt.get_cmap(\"viridis_r\")\n",
    "my_cmap = plt.get_cmap(\"bwr\")\n",
    "my_cmap = plt.get_cmap(\"RdYlBu_r\")\n",
    "cmap = plt.get_cmap(\"RdYlBu_r\")\n",
    "height=0.35\n",
    "#rescale = lambda y: (y - np.min(y)) / (np.max(y) - np.min(y))\n",
    "#rescale = lambda y: (y - np.min(y)) / (np.max(y) - np.min(y))\n",
    "vmax=0.4\n",
    "# norm = mpl.colors.Normalize(vmin=0,vmax=vmax)\n",
    "# mappable = mpl.cm.ScalarMappable(norm=norm, cmap=my_cmap)\n",
    "# define the bins and normalize\n",
    "bounds = np.linspace(0, vmax, 9)\n",
    "norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "anchor=(0.3,-.09)\n",
    "figsize=(6,4.5)\n",
    "d_labels = {'Top4th_PurIST.Score_vs_Bottom4th_PurIST.Score':'Top vs Bottom\\nQuartile by PurIST',\n",
    "            'LiverCohort_vs_LungNotLiverCohort':'Liver Cohort vs.\\nLung Cohort',\n",
    "           'Top4th_pORG.20_vs_Bottom4th_pORG.20':'Top vs Bottom\\nQuartile by pORG',# 0.2\n",
    "           'Top4th_pSUB.1eNeg4_vs_Bottom4th_pSUB.1eNeg4':'Top vs Bottom\\nQuartile by pSUB',# 1e-4\n",
    "           #mets\n",
    "            'Top4thMet_pORG_Up_55_vs_Bottom4thMet_JM_pORG_Up_55':'Top vs Bottom\\nQuartile by pORG Mets',\n",
    "           'Top4thMet_pSUB_Up_51_vs_Bottom4thMet_JM_pSUB_Up_51':'Top vs Bottom\\nQuartile by pSUB Mets',\n",
    "            'LiverCohortMet_vs_LungNotLiverCohortMet':'Liver Cohort vs.\\nLung Cohort Mets',\n",
    "            'Top4thMet_PurIST_vs_Bottom4thMet_PurIST':'Top vs Bottom\\nQuartile by PurIST Mets'\n",
    "           } \n",
    "#subtype\n",
    "sorter_combined = ['HALLMARK_XENOBIOTIC_METABOLISM','HALLMARK_PEROXISOME',\n",
    "  'HALLMARK_FATTY_ACID_METABOLISM','HALLMARK_BILE_ACID_METABOLISM',\n",
    "  'HALLMARK_PANCREAS_BETA_CELLS','HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION',\n",
    "  'HALLMARK_APICAL_JUNCTION','HALLMARK_HYPOXIA','HALLMARK_GLYCOLYSIS']\n",
    "ls_plot_items = ['Top4th_pSUB.1eNeg4_vs_Bottom4th_pSUB.1eNeg4',\n",
    "    'Top4th_PurIST.Score_vs_Bottom4th_PurIST.Score']\n",
    "#generate dataframe with comparisons\n",
    "df_plot_bar,es_marker =util.compare_dataframe(ls_plot_items,d_en,sorter_combined,ls_columns)\n",
    "df_plot_bar['FDR_color'] = pd.cut(df_plot_bar['FDR.q.val'],bins=bounds,labels=bounds[0:-1])\n",
    "df_plot_bar['FDR_color'] = [np.where(bounds==(item))[0] for item in df_plot_bar['FDR_color'].astype('float')]\n",
    "\n",
    "#plot figure\n",
    "fig, ax = util.plot_double_bars_heat(ls_plot_items,df_plot_bar,d_labels,sorter_combined, cmap=cmap, norm=norm,#mappable,\n",
    "                                hatch=hatch,height=height,figsize=figsize,anchor=anchor)\n",
    "ax.set_title('Subtype')\n",
    "ax.set_yticklabels([item.replace('HALLMARK_','').replace('_',' ').title() for item in sorter_combined])\n",
    "ax.text(x=-0.3, y=-0.09, s='NES:',transform=ax.transAxes)\n",
    "ax.set_xlabel('')\n",
    "plt.tight_layout()\n",
    "d_result_fig.update({f'Figure3_B':df_plot_bar})\n",
    "fig.savefig(f'figures/fig3B.pdf')\n",
    "################################################################################\n",
    "# organo\n",
    "#'''\n",
    "sorter_combined =  ['HALLMARK_MYOGENESIS' ,'HALLMARK_CHOLESTEROL_HOMEOSTASIS',#'HALLMARK_KRAS_SIGNALING_UP' ,\n",
    "  'HALLMARK_ANDROGEN_RESPONSE','HALLMARK_OXIDATIVE_PHOSPHORYLATION',\n",
    "  'HALLMARK_DNA_REPAIR', 'HALLMARK_INTERFERON_ALPHA_RESPONSE',\n",
    "  'HALLMARK_E2F_TARGETS','HALLMARK_MYC_TARGETS_V1',\n",
    "  'HALLMARK_G2M_CHECKPOINT','HALLMARK_MITOTIC_SPINDLE',\n",
    "  'HALLMARK_GLYCOLYSIS','HALLMARK_MTORC1_SIGNALING','HALLMARK_PROTEIN_SECRETION',]\n",
    "\n",
    "ls_plot_items = ['Top4th_pORG.20_vs_Bottom4th_pORG.20',\n",
    "    'LiverCohort_vs_LungNotLiverCohort']\n",
    "\n",
    "#generate dataframe with comparisons\n",
    "df_plot_bar,es_marker = util.compare_dataframe(ls_plot_items,d_en,sorter_combined,ls_columns)\n",
    "df_plot_bar['FDR.Q.val'] = np.clip(df_plot_bar['FDR.q.val'],a_min=0,a_max=vmax)\n",
    "df_plot_bar['FDR_color'] = pd.cut(df_plot_bar['FDR.Q.val'],bins=bounds,labels=bounds[0:-1])\n",
    "df_plot_bar['FDR_color'] = [np.where(bounds==(item))[0] for item in df_plot_bar['FDR_color'].astype('float')]\n",
    "\n",
    "        \n",
    "fig, ax = util.plot_double_bars_heat(ls_plot_items,df_plot_bar,d_labels,sorter_combined,#mappable,\n",
    "                                 cmap=cmap, norm=norm,hatch=hatch,\n",
    "                                height=height,figsize=figsize,anchor=anchor)\n",
    "ax.set_title('Organotropism')\n",
    "labels = [item.replace('HALLMARK_','').replace('_',' ').title().replace('Mtorc1','mTORC1').replace('Myc','MYC').replace('Dna','DNA') for item in sorter_combined]\n",
    "__ = ax.set_yticklabels(labels)\n",
    "ax.text(x=-0.2, y=-0.07, s='NES:',transform=ax.transAxes)\n",
    "ax.set_xlabel('')\n",
    "plt.tight_layout()\n",
    "fig.savefig(f'figures/fig3A.pdf')\n",
    "d_result_fig.update({f'Figure3_A':df_plot_bar})\n",
    "#'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c21c620e-e96e-4754-9e17-95c5fec327bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#util.open_write_excel(d_result_fig,filename='Source_Data_Figure3.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc1ace71",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# pORG mets\n",
    "\n",
    "sorter_combined =  ['HALLMARK_MYOGENESIS','HALLMARK_WNT_BETA_CATENIN_SIGNALING',\n",
    " 'HALLMARK_OXIDATIVE_PHOSPHORYLATION','HALLMARK_G2M_CHECKPOINT','HALLMARK_MYC_TARGETS_V1',\n",
    " 'HALLMARK_PROTEIN_SECRETION','HALLMARK_E2F_TARGETS',\n",
    "                    'HALLMARK_GLYCOLYSIS','HALLMARK_MTORC1_SIGNALING']\n",
    "\n",
    "# sorter_combined = ['HALLMARK_UV_RESPONSE_DN','HALLMARK_ESTROGEN_RESPONSE_LATE',\n",
    "#  'HALLMARK_HYPOXIA','HALLMARK_BILE_ACID_METABOLISM','HALLMARK_MITOTIC_SPINDLE',\n",
    "#  'HALLMARK_MTORC1_SIGNALING','HALLMARK_UNFOLDED_PROTEIN_RESPONSE','HALLMARK_DNA_REPAIR',\n",
    "#  'HALLMARK_MYC_TARGETS_V1','HALLMARK_ANGIOGENESIS','HALLMARK_GLYCOLYSIS',\n",
    "#  'HALLMARK_COAGULATION','HALLMARK_XENOBIOTIC_METABOLISM','HALLMARK_G2M_CHECKPOINT',\n",
    "#  'HALLMARK_E2F_TARGETS','HALLMARK_MYC_TARGETS_V2',]\n",
    "ls_plot_items = ['Top4thMet_pORG_Up_55_vs_Bottom4thMet_JM_pORG_Up_55',\n",
    "   # 'LiverCohortMet_vs_LungNotLiverCohortMet'\n",
    " #'Top4thTcr_pORG_Up_55_JM_vs_Bot4thTcr_JM_pORG_Up_55_JM',\n",
    "   #'LiverTcr_pORG_Up_55_JM_vs_LungTcr_JM_pORG_Up_55_JM',\n",
    "                ]\n",
    "#generate dataframe with comparisons\n",
    "df_plot_bar,es_marker = util.compare_dataframe(ls_plot_items,d_en,sorter_combined,ls_columns)\n",
    "df_plot_bar['FDR.Q.val'] = np.clip(df_plot_bar['FDR.q.val'],a_min=0,a_max=vmax)\n",
    "df_plot_bar['FDR_color'] = pd.cut(df_plot_bar['FDR.Q.val'],bins=bounds,labels=bounds[0:-1])\n",
    "df_plot_bar['FDR_color'] = [np.where(bounds==(item))[0] for item in df_plot_bar['FDR_color'].astype('float')]\n",
    "fig, ax = util.plot_double_bars_heat(ls_plot_items,df_plot_bar,d_labels,sorter_combined,cmap=cmap, norm=norm,\n",
    "                                hatch=hatch,\n",
    "                                #height=0.4,figsize=(4,4),anchor=(.5,-.16))\n",
    "                                height=0.9,figsize=(4,4),anchor=(.3,-.07))\n",
    "ax.set_title('Organotropism Mets')\n",
    "labels = [item.replace('HALLMARK_','').replace('_',' ').title().replace('Mtorc1','mTORC1').replace('Myc','MYC').replace('Dna','DNA') for item in sorter_combined]\n",
    "__ = ax.set_yticklabels(labels)\n",
    "\n",
    "ax.text(x=-0.5, y=-0.07, s='NES:',transform=ax.transAxes)\n",
    "ax.set_xlabel('')\n",
    "plt.tight_layout()\n",
    "d_result_fig.update({f'Figure3_C':df_plot_bar})\n",
    "fig.savefig(f'figures/fig3C_mets.pdf')#'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94575b70-7a00-44ae-8341-0877453283ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "#util.open_write_excel(d_result_fig,filename='Source_Data_Figure3.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a49e65b6",
   "metadata": {},
   "source": [
    "## VIPER Barplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23272fde",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load viper\n",
    "df = pd.read_excel('data/ViperANOVA.xlsx',index_col=0)\n",
    "ls_gsea = ['MeanDiff_pORG_Up_55', 'MeanDiff_pSUB_Up_51',\n",
    "       'MeanDiff_LiverVsLungNotLiver', 'MeanDiff_PurIST.Score']\n",
    "d_en = {}\n",
    "for s_gsea in ls_gsea:\n",
    "    s_comp = s_gsea.split('MeanDiff_')[1]\n",
    "    ls_columns = df.columns[df.columns.str.contains(s_comp)]\n",
    "    #up\n",
    "    df_up = df.loc[df.loc[:,s_gsea]>0,ls_columns] \n",
    "    df_up.columns = [item.replace(s_comp,'').replace('TopBottomQuarter_','').replace('_','') for item in df_up.columns]\n",
    "    df_up.reset_index(inplace=True)\n",
    "    df_up.rename({'MeanDiff':'NES', 'qVal':'FDR.q.val','regulons':'NAME'},axis=1,inplace=True)\n",
    "    #down\n",
    "    df_down = df.loc[df.loc[:,s_gsea]<0,ls_columns]\n",
    "    df_down.columns = [item.replace(s_comp,'').replace('TopBottomQuarter_','').replace('_','') for item in df_down.columns]\n",
    "    df_down.reset_index(inplace=True)\n",
    "    df_down.rename({'MeanDiff':'NES', 'qVal':'FDR.q.val','regulons':'NAME'},axis=1,inplace=True)\n",
    "    #add to data dict\n",
    "    d_en.update({f\"{s_comp}_UP\":df_up})\n",
    "    d_en.update({f\"{s_comp}_DN\":df_down})\n",
    "    #break\n",
    "\n",
    "d_colorblind = {'Liver': '#d55e00','Lung': '#0072b2',\n",
    " 'high pORG': '#e69f00','low pORG': '#56b4e9',\n",
    " 'basal-like': '#000000','high PurIST': '#000000',\n",
    " 'classical': '#cc79a7','low PurIST': '#cc79a7',\n",
    " 'high pSUB': '#f0e442','low pSUB': '#009E73'}\n",
    "d_labels = {'pORG_Up_55':f'Top vs Bottom{linebreak}Qartile by pORG','LiverVsLungNotLiver':f'Liver vs.{linebreak}Lung Cohort',\n",
    "          'pSUB_Up_51':'pSUB','PurIST.Score':'PurIST'}\n",
    "\n",
    "#df.index[df.index.str.contains('CCN')]\n",
    "d_sorter_comb =  { 'DNA Repair':['ATM', 'ARID1A', 'ATR', 'BRCA1', 'BARD1', \n",
    "                                 'BRCA2', 'RAD51', 'CHEK1', 'PALB2'],\n",
    "'Replication': ['MCM3','MCM8', 'MCM6', 'MCM7', 'MCM2', 'ORC2'],#'MCM5',\n",
    "'Cell Cycle':['CDKN1B', 'CDK4', 'E2F2', 'CCND1', 'CDK2', 'CDK6', 'CCNA2','E2F1'],\n",
    " \"Immune\":['MS4A1', 'CD3G', 'IFNAR2', 'IFNAR1', 'MX1', 'STAT1'],\n",
    "                 }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8301a776",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#viper\n",
    "importlib.reload(util)\n",
    "hatch='||'\n",
    "ls_plot_items = ['pORG_Up_55','LiverVsLungNotLiver'] #\n",
    "#ls_plot_items = ['pSUB_Up_51', 'PurIST.Score']\n",
    "ls_columns = ['NAME', 'NES', 'pVal', 'FDR.q.val']\n",
    "\n",
    "for s_title, sorter_combined in d_sorter_comb.items():\n",
    "    #generate dataframe with comparisons\n",
    "    df_plot_bar,es_marker =util.compare_dataframe(ls_plot_items,d_en,sorter_combined,ls_columns)\n",
    "    # add color\n",
    "    df_plot_bar['FDR.Q.val'] = np.clip(df_plot_bar['FDR.q.val'],a_min=0,a_max=vmax)\n",
    "    df_plot_bar['FDR_color'] = pd.cut(df_plot_bar['FDR.Q.val'],bins=bounds,labels=bounds[0:-1])\n",
    "    df_plot_bar['FDR_color'] = [np.where(bounds==(item))[0] for item in df_plot_bar['FDR_color'].astype('float')]\n",
    "    #plot figure\n",
    "    fig, ax = util.plot_double_bars_heat(ls_plot_items,df_plot_bar,d_labels,\n",
    "                                         sorter_combined,cmap=cmap, norm=norm,#mappable,\n",
    "                                    hatch=hatch,height=height,figsize=(3.2,3.5),anchor=(0.35,-.09),\n",
    "                                        x_label='Mean Difference')\n",
    "    ax.set_title(s_title)\n",
    "    ax.text(x=-0.5, y=-0.07, s='Mean Diff.:',transform=ax.transAxes,fontsize='small')\n",
    "    ax.set_xlabel('')\n",
    "    plt.tight_layout()\n",
    "    d_result_fig.update({f'Figure3_D_{s_title}':df_plot_bar})\n",
    "    fig.savefig(f'figures/fig3D_{s_title}.pdf')#'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bc9a119-cafb-4f36-8d74-ed295f233a28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# util.open_write_excel(d_result_fig,filename='Source_Data_Figure3.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9cfed3c-d745-4571-91e0-f9f66d45ae8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "importlib.reload(util)\n",
    "df_meta = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name='Patient_Metadata',#'RNA_Specimen_Metadata',\n",
    "                          index_col=0)\n",
    "df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name='RNA_Specimen_Metadata',\n",
    "                          index_col=0)\n",
    "df_patient['Recurrence_Sites_4']=df_patient.index.map(dict(zip(df_meta.loc[:,'Public_Patient_ID'],df_meta.loc[:,'Recurrence_Sites_4'])))\n",
    "sns.set_palette(pal_four)\n",
    "ls_scores = ['pORG_Up_55_All']#['pORG_Primary','pORG_Met','PurIST_Primary','pSUB_Primary']\n",
    "ls_groups=['Liver', 'Lung', \n",
    "           'No_Doc_Recur', 'Other_site']\n",
    "for s_score in ls_scores:#df_patient.Recurrence_Sites_4.unique()\n",
    "    df = df_patient[df_patient.Recurrence_Sites_4.isin(ls_groups)]\n",
    "    plotting={'data':df,'x':'Recurrence_Sites_4','y':s_score,'order':ls_groups}\n",
    "    fig, ax, test_res = util.annotated_stripplot(plotting,ls_groups,b_correct=True,figsize=(3,3))\n",
    "    ax.set_xlabel('')\n",
    "    s_group ='Recurrence_Sites_4'\n",
    "    if len(ls_groups)==4:\n",
    "        xlabels = [f'{ls_groups[0]}{linebreak}(N={(df.loc[:,s_group].value_counts()[ls_groups[0]])})',\n",
    "               f'{ls_groups[1]}{linebreak}(N={(df.loc[:,s_group].value_counts()[ls_groups[1]])})',\n",
    "                  f'{ls_groups[2]}{linebreak}(N={(df.loc[:,s_group].value_counts()[ls_groups[2]])})',\n",
    "               f'{ls_groups[3]}{linebreak}(N={(df.loc[:,s_group].value_counts()[ls_groups[3]])})']\n",
    "        ax.set_xticklabels(xlabels,fontsize='x-small')\n",
    "    fig.savefig(f'figures/extended_data_5B_{s_score}.pdf')\n",
    "    #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0672e859-fa95-4854-98a5-4871ce24c6c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_patient.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6df50fa5-8e0f-4570-a087-3ea3f7aa1720",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_patient.loc[:,['Public_Specimen_ID', 'Recurrence_Sites_4', s_score]].to_csv('Source_Extended_Data_Figure5_B.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32bcfa60",
   "metadata": {},
   "source": [
    "## old code  <a name=\"split\"></a>\n",
    "\n",
    "\n",
    "[contents](#contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30a694c1-c8cd-455c-942a-0f3497b1398e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8cd7e51-6546-4666-ba72-9ce4ceb8a83e",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_cutoff\n",
    "cufoff = 0.02491"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78c0c786-58f5-487b-ada5-e38038621dd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_spec =  pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name='RNA_Specimen_Metadata',\n",
    "                          index_col=0)\n",
    "labels = ['low','high']\n",
    "for s_col in df_spec.columns[df_spec.columns.str.contains('pORG')]:\n",
    "    x = df_spec.loc[:,s_col]\n",
    "    min_val = x.min()\n",
    "    max_val = x.max()\n",
    "    bins = [min_val,cufoff,max_val]\n",
    "    high_low = pd.cut(x, bins, right=True, labels=labels,include_lowest=True)\n",
    "    df_spec[f\"{s_col.replace('_Up_55','')}_Category\"] = high_low\n",
    "    #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f199382-ac5d-424e-86c7-8287f53ca2b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ls_pt = df_spec[df_spec.Public_Specimen_ID.str.contains('T2')].index\n",
    "df_out[df_out.index.isin(ls_pt)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a95aed7b-cd5c-4d2b-8f19-b28da8e1c1e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_out = df_spec.loc[:,(df_spec.columns.str.contains('pORG')) | (df_spec.columns.str.contains('Specimen'))]\n",
    "#df_out.to_csv('pORG_Catgories_for_Dove.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a075cf8e-003e-49b0-bbbe-8127830bc2e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#break\n",
    "    #'''\n",
    "    #break\n",
    "#plot Cohort versus pORG\n",
    "'''\n",
    "pal_liv_r = ('#D55E00','#0072B2',)\n",
    "sns.set_palette(pal_liv_r)\n",
    "s_type = s_porg.split('_')[-1]\n",
    "if not s_type == 'All':\n",
    "    fig3, ax3 = plt.subplots(figsize=(3,3),dpi=200)\n",
    "    sns.stripplot(data=df_pri,x='Cohort',y=s_porg,ax=ax3)\n",
    "    sns.boxplot(data=df_pri,x='Cohort',y=s_porg,showmeans=True,medianprops={'visible': False},\n",
    "                       whiskerprops={'visible': False},meanline=True,showcaps=False,ax=ax3,\n",
    "                       meanprops={'color': 'k', 'ls': '-', 'lw': 2},showfliers=False,showbox=False)\n",
    "    a = df_pri.loc[(df_pri.Tumor_Type==s_type) & (df_pri.Cohort=='Liver'),s_porg].dropna()\n",
    "    b = df_pri.loc[(df_pri.Tumor_Type==s_type) & (df_pri.Cohort=='Lung'),s_porg].dropna()\n",
    "    stat, pvalue = stats.ttest_ind(a,b)\n",
    "    ax3.set_title(f'{s_type}\\np={pvalue:.3}')\n",
    "    plt.close()\n",
    "    #'''\n",
    "'''break\n",
    "#add pORG quartiles back to data (all, primary and met)\n",
    "df_merge[f'{s_porg}_quartiles'] = df_merge.Public_Patient_ID.map(dict(zip(df_pri.Public_Patient_ID,df_pri.quartiles)))\n",
    "print(df_merge[f'{s_porg}_quartiles'].value_counts())\n",
    "try:\n",
    "    df_merge['pORG_All'] = df_merge.Public_Patient_ID.map(dict(zip(df_pri.Public_Patient_ID,df_pri.loc[:,'pORG_All'])))\n",
    "except:\n",
    "    continue#'''\n",
    "# ####### get diagnosis dates ############\n",
    "# df_new_surv = pd.read_excel('../annotation/cancer_participant_overview_optr_2023-05-04_16-55-18.xlsx')\n",
    "# df_date = df_new_surv.loc[:,(df_new_surv.columns.str.contains('Date')) | (df_new_surv.columns=='Participant ID')].copy()\n",
    "# df_date.rename({'Participant ID':'OPTR ID'},axis=1,inplace=True)\n",
    "# df_date.set_index('OPTR ID',inplace=True)\n",
    "# df_date = df_date.astype('datetime64[ns]')\n",
    "# df_date['OPTR'] = df_date.index\n",
    "# ############ get adaptive TCR dates ############\n",
    "# df_adap = pd.read_excel('../annotation/Adaptive-collection dates_JE.xlsx') #297 rows\n",
    "# df_adap.rename({'OPTR ID':'OPTR'},axis=1,inplace=True)\n",
    "# df_adap['date tumor collection'] = df_adap['date tumor collection'].astype('datetime64[ns]')\n",
    "# b_blood = df_adap.loc[:,'tumor and blood same collection date?']!='yes'\n",
    "# ls_check = df_adap.loc[b_blood,'OPTR']\n",
    "# #number of unmatched bloods/ diff blood and tumor collection\n",
    "# print(len(ls_check))\n",
    "# #get IDs\n",
    "# d_ids = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/OLD Versions/Simplified_Public_IDs_Key.xlsx',sheet_name=None)\n",
    "# df_id = d_ids['RnaSeqKey']\n",
    "# # add patients w/o RNA seq\n",
    "# for s_key in ['TcrTumorKey','TcrBloodKey']:\n",
    "#     df_add = d_ids[s_key].loc[~d_ids[s_key].loc[:,'Public.Specimen.ID'].isin(ls_ids)]\n",
    "#     df_id = pd.concat([df_id,df_add])\n",
    "# ## get times from diagnosis, resection, recurrence\n",
    "# df_patient = pd.read_csv('annotation/20230921_Patient_Metadata_OPTR.csv',index_col=0) #not all OPTR\n",
    "# df_patient['OPTR'] = df_patient.Public_Patient_ID.map(dict(zip(df_id.loc[:,'Biolibrary.Subject.ID'],df_id.OPTR)))\n",
    "# df_patient['OPTR'] = np.int64(df_patient.OPTR)\n",
    "# df_patient['Diagnosis_Date'] = df_patient.OPTR.map(dict(zip(df_date.OPTR,df_date.loc[:,'Diagnosis Date'])))\n",
    "# #df_check = df_patient[df_patient.OPTR.isin(ls_check)].copy()\n",
    "# df_check = df_adap[df_adap.OPTR.isin(ls_check)].copy()\n",
    "# df_check['OPTR'] = np.int64(df_check.OPTR)\n",
    "# #number of not same day patients with diagnosis date\n",
    "# print(len(df_check))\n",
    "# df_patient.rename({'Vital Status at FU':'Vital_Status_at_FU'},axis=1, inplace=True)\n",
    "# #convert to date time\n",
    "# ls_col = ['OPTR', 'Days from Diagnosis to FU', 'Days from Resection to Recurrence',\n",
    "#        'Days from Resection to FU', 'Days from Earliest Recur to FU','Diagnosis_Date','Vital_Status_at_FU']\n",
    "# df_pt_dates = df_patient.loc[:,ls_col].copy()\n",
    "\n",
    "# for s_col in ls_col:\n",
    "#     if s_col.find(' ') > -1:\n",
    "#         df_pt_dates[s_col.replace(' ',\"_\")] = pd.to_timedelta(df_pt_dates[s_col], unit='D')\n",
    "# #calculate dates\n",
    "# import datetime as dt\n",
    "# df_pt_dates['FU_date'] = df_pt_dates.Diagnosis_Date + df_pt_dates.Days_from_Diagnosis_to_FU\n",
    "# df_pt_dates['Recur_date'] = df_pt_dates.FU_date - df_pt_dates.Days_from_Earliest_Recur_to_FU\n",
    "# df_pt_dates['Resect_date'] = df_pt_dates.FU_date - df_pt_dates.Days_from_Resection_to_FU\n",
    "# #merge\n",
    "# df_check = df_check.merge(df_pt_dates,on='OPTR',how='left')\n",
    "# #classify bloods\n",
    "# se_blood = df_check.loc[:,'date blood collection']\n",
    "# df_check.loc[se_blood > df_check.Recur_date,'Blood_Status'] = 'Blood_after_recurrence'\n",
    "# df_check.loc[se_blood <= df_check.Recur_date,'Blood_Status'] = 'Blood_after_resection_before_recurrence'\n",
    "# df_check.loc[se_blood <= df_check.Resect_date,'Blood_Status'] = 'Blood_before_resection'\n",
    "# df_check.loc[se_blood < df_check.Diagnosis_Date,'Blood_Status'] = 'Blood_before_diagnosis'\n",
    "# #save out blood type\n",
    "# ls_drop = ['Days from Diagnosis to FU', 'Days from Resection to Recurrence',\n",
    "#        'Days from Resection to FU', 'Days from Earliest Recur to FU',]\n",
    "# df_check.drop(ls_drop,axis=1).to_csv('Blood_dates_to_review.csv')\n",
    "# old non parametric\n",
    " # 'Hill_1', 'Hill_2', 'Hill_3',\n",
    " #           'Hill_4', 'Hill_5', 'Hill_6', 'true_Value',#'ginisimp_Value',\n",
    " #                'd50_Clones','chao_Estimator',     'Hill_1_blood','Hill_2_blood', 'Hill_3_blood',\n",
    " #       'Number Tumor Distinct Clones','Hill_4_blood', 'chao_Estimator_blood','d50_Clones_blood',\n",
    " #          'true_Value_blood', 'morisita_Liver_Primary_Bld', 'morisita_Liver_Met_Bld',\n",
    " #           'morisita_Lung_Met_Bld', 'morisita_Lung_Primary_Bld',\n",
    " #           'morisita_Liver_Primary', 'morisita_Liver_Met', 'morisita_Lung_Primary',\n",
    " #           'morisita_Lung_Met', 'morisita_low_Primary',\n",
    " #           'morisita_high_Primary_Bld', 'morisita_high_Primary',\n",
    " #           'morisita_low_Primary_Bld', 'morisita_low_Met', 'morisita_high_Met_Bld',\n",
    " #           'morisita_high_Met', 'morisita_low_Met_Bld', 'morisita_Liver_Primary_Bld_blood','morisita_Liver_Primary_blood', 'morisita_Lung_Primary_Bld_blood',\n",
    " #     'morisita_Lung_Primary_blood', 'morisita_high_Primary_Bld_blood',\n",
    " #     'morisita_high_Primary_blood', 'morisita_low_Primary_Bld_blood','morisita_low_Primary_blood' ,\n",
    " #     'morisita_Liver_Met_Bld_blood','morisita_Liver_Met_blood', 'morisita_Lung_Met_Bld_blood',\n",
    " #     'morisita_Lung_Met_blood','morisita_high_Met_Bld_blood','morisita_high_Met_blood',\n",
    " #      'morisita_low_Met_Bld_blood','morisita_low_Met_blood',]\n",
    "# ##save out metadata with TCR\n",
    "# #ls_col = df_merge.columns[~df_merge.columns.isin(df_patient.columns)].tolist() + ['Public_Patient_ID']\n",
    "# ls_merge = ['pORG_All_quartiles','pORG_Met_quartiles',\n",
    "#             'pORG_Primary_quartiles','Blood_Type','Tumor_Type','TCR_Met_Site']\n",
    "# ls_col = ls_foci + ls_merge \n",
    "# #print(len(ls_col))\n",
    "# df_out = df_patient.loc[:,~df_patient.columns.isin(ls_col)].merge(df_merge.loc[:,ls_col+['Public_Patient_ID']],on='Public_Patient_ID',how='left')\n",
    "# df_out = df_merge.loc[:,ls_col+['Public_Patient_ID']].merge(df_patient.loc[:,~df_patient.columns.isin(ls_col)],on='Public_Patient_ID',how='left')\n",
    "# df_out.to_csv('annotation/20240304_Patient_Metadata_TCR_Metrics.csv') #20231108 reclassify bloods\n",
    "# df_out.Public_Patient_ID.duplicated().any()\n",
    "# #add the all Cohort vs pORG\n",
    "# if s_porg == 'pORG_0.2_All':\n",
    "#     for s_type in ['Primary','Met']:\n",
    "#         fig3, ax3 = plt.subplots(figsize=(3,3),dpi=300)\n",
    "#         sns.stripplot(data=df_pri[df_pri.Tumor_Type==s_type],x='Cohort',y=s_porg,ax=ax3)\n",
    "#         sns.boxplot(data=df_pri[df_pri.Tumor_Type==s_type],x='Cohort',y=s_porg,showmeans=True,medianprops={'visible': False},\n",
    "#                            whiskerprops={'visible': False},meanline=True,showcaps=False,ax=ax3,\n",
    "#                            meanprops={'color': 'k', 'ls': '-', 'lw': 2},showfliers=False,showbox=False)\n",
    "#         a = df_pri.loc[(df_pri.Tumor_Type==s_type) & (df_pri.Cohort=='Liver'),s_porg].dropna()\n",
    "#         b = df_pri.loc[(df_pri.Tumor_Type==s_type) & (df_pri.Cohort=='Lung'),s_porg].dropna()\n",
    "#         stat, pvalue = stats.ttest_ind(a,b)\n",
    "#         ax3.set_title(f'{s_type}\\np={pvalue:.3}')\n",
    "# #Lung cohort: low pORG vs high pORG\n",
    "# for tu_plot in [('Lung','medians'),('Liver','medians'),('Lung','quartiles'),('Liver','quartiles')]:\n",
    "#     fig,ax = plt.subplots(figsize=(3,3),dpi=200)\n",
    "#     df_pri.loc[df_pri.Cohort==tu_plot[0],tu_plot[1]].value_counts(dropna=False).plot(kind='bar',ax=ax)\n",
    "#     ax.set_title(f'{tu_plot[0]}: {s_porg} {tu_plot[1]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2b44bd7-9033-458c-9fd6-71394627b01a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #load full patient data\n",
    "# # s_out = '20230921_Patient_Metadata.csv' #'20231206_Patient_Metadata.csv'#\n",
    "# # df_patient= pd.read_csv(f'annotation/{s_out}',index_col=0)\n",
    "\n",
    "# s_out = 'skip'#'20231222_Patient_Metadata.csv'#'20230921_Patient_Metadata.csv'\n",
    "# #df_patient= pd.read_csv(f'annotation/{s_out}',index_col=0)\n",
    "# df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "#                           index_col=0)\n",
    "# if s_out == '20230921_Patient_Metadata.csv':\n",
    "#     d_rename = {'pORG_0.2_Primary':'pORG_Primary', 'pORG_0.2_Met':'pORG_Met', \n",
    "#            'pORG_0.2_allPrimary':'pORG_allPrimary', 'pORG_0.2_allMet':'pORG_allMet' }\n",
    "#     df_patient.rename(d_rename,axis=1,inplace=True)\n",
    "# elif s_out == '20231206_Patient_Metadata.csv':\n",
    "#     df_patient.drop('pORG_All',axis=1,inplace=True)\n",
    "# elif s_out == '20231222_Patient_Metadata.csv':\n",
    "#     df_patient.drop('pORG_All',axis=1,inplace=True)\n",
    "# else:\n",
    "#     print(s_out)\n",
    "\n",
    "# # old patient data\n",
    "# df_patient_old = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset1.xlsx',\n",
    "#                            sheet_name='Patients - Tab 1')\n",
    "# #load T cell data - tumor\n",
    "# df_tcell = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset6.xlsx',\n",
    "#                          sheet_name='Tumor Samples')\n",
    "# print(f'T cell tumor {len(df_tcell)}')\n",
    "# df_tcell.rename({'Tumor Type':'Tumor_Type'},axis=1,inplace=True)\n",
    "# #blood\n",
    "# df_tcell_blood = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/SupplementalDataset6.xlsx',\n",
    "#                          sheet_name='Blood Samples')\n",
    "# #df_tcell_blood['Tumor Type'] = 'Blood'\n",
    "# print(f'T cell blood {len(df_tcell_blood)}')\n",
    "\n",
    "# #both\n",
    "# df_tcell = df_tcell.merge(df_tcell_blood,on='Patient ID',suffixes=(' tumor',' blood'),how='outer')\n",
    "# df_tcell.rename({'Productive Rearrangements (Observed Richness)':'Productive_Rearrangements',\n",
    "#                 'Templates per ng':'Templates_per_ng','Patient ID':'Public_Patient_ID'},axis=1,inplace=True)\n",
    "# print(f'T cell both {len(df_tcell)}')\n",
    "# #df_merge = df_patient[df_patient.Public_Patient_ID.isin(df_tcell.Public_Patient_ID)]\n",
    "# df_merge = df_patient.merge(df_tcell,on='Public_Patient_ID',how='inner')\n",
    "# # missing_ids: some were dropped surgery due to death from surgery\n",
    "# missing_ids = df_tcell[~df_tcell.Public_Patient_ID.isin(df_merge.Public_Patient_ID)].Public_Patient_ID\n",
    "# df_not_matched = df_patient_old[df_patient_old.loc[:,'Patient ID'].isin(missing_ids)]\n",
    "# print(f'dropped {len(df_not_matched)} patients; days surv post surgery: {df_not_matched.loc[:,\"Days from Resection to FU\"].max()}')\n",
    "\n",
    "# print(f'Unique patients t cell {(df_merge.Public_Patient_ID.nunique())}')\n",
    "# print(df_merge.Cohort.value_counts())\n",
    "# print(df_merge.groupby(['Cohort','Tumor_Type']).count().Public_Patient_ID)\n",
    "# print(len(df_merge))\n",
    "# ## add calculated Entropy, simpsons \n",
    "# #load data\n",
    "# ls_clones = ['TCR_Blood_Shared_Clones_rare0to2_all.csv', 'TCR_Blood_Shared_Clones_rare2to5_all.csv',\n",
    "#               'TCR_Blood_Shared_Clones_rare5to10_all.csv','TCR_Blood_Shared_Clones_rare10to665800_all.csv',#'TCR_Blood_Shared_Clones_rare1_all.csv',\n",
    "#               'TCR_Tumor_Distinct_Clones_rare0to2_tumbld.csv','TCR_Tumor_Distinct_Clones_rare2to5_tumbld.csv',\n",
    "#               'TCR_Tumor_Distinct_Clones_rare5to10_tumbld.csv', 'TCR_Tumor_Distinct_Clones_rare10to665800_tumbld.csv', #  'TCR_Blood_Shared_Clones_rare2_all.csv',\n",
    "# #  'TCR_Blood_Shared_Clones_rare5_all.csv',\n",
    "# #  'TCR_Blood_Shared_Clones_rare10_all.csv',\n",
    "# #  'TCR_Tumor_Distinct_Clones_rare10_tumor.csv',\n",
    "# #  'TCR_Tumor_Distinct_Clones_rare1_tumor.csv',\n",
    "# #  'TCR_Tumor_Distinct_Clones_rare2_tumor.csv',\n",
    "# #  'TCR_Tumor_Distinct_Clones_rare5_tumor.csv'\n",
    "#              ]\n",
    "# d_load = {#'TCR_Tumor_Distinct_Clones_no_rare_denom.csv':['Public_Patient_ID',\n",
    "#            #     'Number Tumor Distinct Clones', 'Fraction Tumor Distinct Clones','Prod. Freq. Tumor Distinct Clones'],\n",
    "#           #'TCR_Tumor_Distinct_Rearrangements_no_rare.csv':['Public_Patient_ID',\n",
    "#        #'Number Tumor Distinct Rearrangements','Fraction Tumor Distinct Rearrangements','Prod. Freq. Tumor Distinct Rearrangements'],\n",
    "#         #  'TCR_Tumor_Distinct_Clones.csv':['Public_Patient_ID','Fraction_Tumor_Distinct_Clones'],#'Number Tumor Distinct Clones (rare)',\n",
    "#     'TCR_Simpsons_Evenness_templates.csv':['Public_Patient_ID','Simpsons_Evenness','Simpsons_Evenness_no_CMV'],\n",
    "#           'TCR_Simpsons_D_templates.csv':['Public_Patient_ID','Simpsons_D'],\n",
    "#           'TCR_Shannon_entropy_productive_frequency.csv':['Public_Patient_ID','Shannon_Entropy','Normalized_Shannon_Entropy'],\n",
    "#          }\n",
    "# for s_en, ls_markers in d_load.items():\n",
    "#     print(s_en)\n",
    "#     df_en = pd.read_csv(f'data/{s_en}')\n",
    "#     df_en.rename({'Unnamed: 0':'Patient_Specimen_ID','Shannon Entropy':'Shannon_Entropy',#'Percent Tumor Distinct Rearrangements in Tumor':'Percent Tumor Distinct Clones (JL)',   \n",
    "#              'Normalized Shannon Entropy':'Normalized_Shannon_Entropy'},axis=1,inplace=True)\n",
    "#     if s_en == 'TCR_Tumor_Distinct_Clones.csv':\n",
    "#         df_en.rename({'Unnamed: 0':'Patient_Specimen_ID','Fraction Tumor Distinct Clones in Tumor':'Fraction_Tumor_Distinct_Clones'},axis=1,inplace=True)\n",
    "#     df_en['dash_end'] = [item.split('-')[-1] for item in df_en.Patient_Specimen_ID]\n",
    "#     df_en['Public_Patient_ID'] = [item.split('-')[0] + '-' + item.split('-')[1] for item in df_en.Patient_Specimen_ID]\n",
    "#     #patient rows with columns for blood and tumor\n",
    "#     df_en_pat = pd.DataFrame(index=df_en.Public_Patient_ID.unique())\n",
    "#     df_tum = df_en[df_en.dash_end!='B'].loc[:,ls_markers].set_index('Public_Patient_ID')\n",
    "#     print(df_tum.index.duplicated().any())\n",
    "#     df_en_pat = df_en_pat.merge(df_tum,left_index=True,right_index=True,how='left')\n",
    "#     if s_en.find('Tumor_Distinct') == -1:\n",
    "#         df_bld = df_en[df_en.dash_end=='B'].loc[:,ls_markers].set_index('Public_Patient_ID')\n",
    "#         print(df_bld.index.duplicated().any())\n",
    "#         df_en_pat = df_en_pat.merge(df_bld,left_index=True,right_index=True,how='left',suffixes=('_Tumor','_Blood'))\n",
    "#     df_merge = df_merge.merge(df_en_pat.reset_index().rename({'index':'Public_Patient_ID'},axis=1),on='Public_Patient_ID',how='left')\n",
    "#     #break\n",
    "# for s_en in ls_clones:\n",
    "#     print(s_en)\n",
    "#     df_en = pd.read_csv(f'data/{s_en}')\n",
    "#     s_col_name = df_en.columns[df_en.columns.str.contains('Fraction')][0]\n",
    "#     s_type = s_en.split('TCR_')[1].split('_Clones')[0]\n",
    "#     s_rare = s_en.split('rare')[1].split('_')[0]\n",
    "#     s_new = f'{s_col_name.replace(\"Blood\",\"\").replace(\" in Tumor\",\"\")} {s_rare}'\n",
    "#     s_new = s_new.replace(\"Clones 1\",\"TCRs\").replace(\"TCRs0\",\"Clones 10\").replace(\"  \",\" \").replace('10to665800','>=10')\n",
    "#     print(s_new)\n",
    "#     df_en.rename({'Unnamed: 0':'Patient_Specimen_ID',s_col_name:s_new},axis=1,inplace=True)\n",
    "#     df_en['dash_end'] = [item.split('-')[-1] for item in df_en.Patient_Specimen_ID]\n",
    "#     df_en['Public_Patient_ID'] = [item.split('-')[0] + '-' + item.split('-')[1] for item in df_en.Patient_Specimen_ID]\n",
    "#     #patient rows with columns for blood and tumor\n",
    "#     df_en_pat = pd.DataFrame(index=df_en.Public_Patient_ID.unique())\n",
    "#     df_tum = df_en[df_en.dash_end!='B'].loc[:,[s_new,'Public_Patient_ID']].set_index('Public_Patient_ID')\n",
    "#     print(df_tum.index.duplicated().any())\n",
    "#     df_en_pat = df_en_pat.merge(df_tum,left_index=True,right_index=True,how='left')\n",
    "#     df_merge = df_merge.merge(df_en_pat.reset_index().rename({'index':'Public_Patient_ID'},axis=1),on='Public_Patient_ID',how='left')\n",
    "    \n",
    "# df_merge['Simpsons_Diversity_Tumor'] = 1/df_merge.Simpsons_D_Tumor#1-df_merge.Simpsons_D_Tumor#\n",
    "# df_merge['Simpsons_Diversity_Blood'] = 1/df_merge.Simpsons_D_Blood#1-df_merge.Simpsons_D_Blood#\n",
    "# df_merge['Clonality_Tumor'] = 1 - df_merge.Normalized_Shannon_Entropy_Tumor\n",
    "# df_merge['Clonality_Blood'] = 1 - df_merge.Normalized_Shannon_Entropy_Blood\n",
    "\n",
    "# #met or primary\n",
    "# df_merge['Percent Tumor Distinct Clones'] = 100 - df_merge.loc[:,'Percentage Tumor-Distinct Clones in Paired Tumor Sample']\n",
    "# df_merge.rename({'Patient Specimen ID':'Patient_Specimen_ID','Tumor Type':'Tumor_Type'},axis=1,inplace=True)\n",
    "# print((df_merge.Public_Patient_ID.nunique()))\n",
    "\n",
    "# # df_merge.loc[df_merge.loc[:,'PurIST_Met'] > 0.5,'PurIST_Subtype'] = 'basal-like'\n",
    "# # df_merge.loc[df_merge.loc[:,'PurIST_Met'] <= 0.5,'PurIST_Subtype'] = 'classical'\n",
    "\n",
    "# # define long and short lung\n",
    "# s_time = 'Days from Diagnosis to FU'\n",
    "# short_long = df_merge.loc[(df_merge.Tumor_Type.notna()) & (df_merge.Cohort=='Lung'),s_time].median()\n",
    "# df_merge.loc[df_merge.loc[:,s_time] > short_long,'TCR_Lung_Median_Surv'] = 'lung_long'\n",
    "# df_merge.loc[df_merge.loc[:,s_time] < short_long,'TCR_Lung_Median_Surv'] = 'lung_short'\n",
    "# df_merge.loc[(df_merge.Tumor_Type.isna()) | (df_merge.Cohort!='Lung'),'TCR_Lung_Median_Surv'] = np.nan\n",
    "# # define long and short liver\n",
    "# short_long = df_merge.loc[(df_merge.Tumor_Type.notna()) & (df_merge.Cohort=='Liver'),s_time].median()\n",
    "# df_merge.loc[df_merge.loc[:,s_time] > short_long,'TCR_Liver_Median_Surv'] = 'liver_long'\n",
    "# df_merge.loc[df_merge.loc[:,s_time] < short_long,'TCR_Liver_Median_Surv'] = 'liver_short'\n",
    "# df_merge.loc[(df_merge.Tumor_Type.isna()) | (df_merge.Cohort!='Liver'),'TCR_Liver_Median_Surv'] = np.nan\n",
    "\n",
    "# #add blood productive rearrangements\n",
    "# df = pd.read_csv('results/results_Productive_rearrangements.csv',index_col=0)\n",
    "# df_blood_productive = df[df.Site=='Blood']\n",
    "# #df_blood_productive.Public_Patient_ID.duplicated().any()\n",
    "# df_merge['Productive_Rearrangements_Blood'] = df_merge.Public_Patient_ID.map(dict(zip(df_blood_productive.Public_Patient_ID,df_blood_productive.amino_acid)))\n",
    "# ## did I make a mistake here by commenting out \n",
    "# #'Y - but include in all',\n",
    "# # add blood data\n",
    "# ls_blood = ['Shannon_Entropy_Blood','Simpsons_Diversity_Blood', \n",
    "#             'Clonality_Blood', 'Simpsons_Evenness_Blood',\n",
    "#             \"Simpson's Evenness blood\",'Productive_Rearrangements_Blood',\n",
    "#             'Fraction Tumor Distinct TCRs',\n",
    "#            'Percent Tumor Distinct Clones',\n",
    "#             'Fraction Shared Clones 10',\n",
    "#             'Fraction Shared Clones 0to2',\n",
    "#        'Fraction Shared Clones 2to5', 'Fraction Shared Clones 5to10',\n",
    "#        'Fraction Shared Clones >=10', 'Fraction Tumor Distinct Clones 0to2',\n",
    "#        'Fraction Tumor Distinct Clones 2to5',\n",
    "#        'Fraction Tumor Distinct Clones 5to10',\n",
    "#        'Fraction Tumor Distinct Clones >=10', \n",
    "#         ]\n",
    "# df_merge['Blood_Type'] = df_merge.Tumor_Type.copy()\n",
    "# d_ids = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/OLD Versions/Simplified_Public_IDs_Key.xlsx',sheet_name=None)\n",
    "# df_id = d_ids['RnaSeqKey']\n",
    "# ls_ids = df_id.loc[:,'Public.Specimen.ID']\n",
    "# # # add patients w/o RNA seq\n",
    "# for s_key in ['TcrTumorKey','TcrBloodKey']:\n",
    "#     df_add = d_ids[s_key].loc[~d_ids[s_key].loc[:,'Public.Specimen.ID'].isin(ls_ids)]\n",
    "#     df_id = pd.concat([df_id,df_add])\n",
    "# df_blood_dates = pd.read_excel(f'FINAL_Blood_dates_to_review_JE.xlsx',index_col=0)\n",
    "# df_blood_dates['Public_Patient_ID'] = df_blood_dates.OPTR.map(dict(zip(df_id.OPTR,df_id.loc[:,'Biolibrary.Subject.ID'])))\n",
    "# print(len(df_blood_dates))\n",
    "# #df_dk = pd.read_excel('Blood_dates_to_review_JE-dk.xlsx',index_col=0)\n",
    "# #df_blood_dates = df_blood_dates.merge(df_dk.loc[:,['Potential calls','chart notes','OPTR']],on='OPTR',how='left',suffixes=('','_x'))\n",
    "# #print(len(df_blood_dates))\n",
    "# ls_types = ['Y - but include in all', 'unknown', # no categorization of met blood\n",
    "#             'can be considered a primary blood',\n",
    "#         'can be considered a met blood',\n",
    "#        'exclude from analysis']\n",
    "# #df_blood_dates.loc[:,'exclude blood primary or met?'] = df_blood_dates.loc[:,'exclude blood primary or met?'].fillna(df_blood_dates.loc[:,'Potential calls'])\n",
    "# for s_type in ls_types: #df_blood_dates.loc[:,'exclude blood primary or met?'].unique()\n",
    "#     print(s_type)\n",
    "#     df_type = df_blood_dates[df_blood_dates.loc[:,'exclude blood primary or met?']==s_type]\n",
    "#     b_type = df_merge.Public_Patient_ID.isin(df_type.Public_Patient_ID)\n",
    "#     print(b_type.sum())\n",
    "# #     if s_type == 'N - change tumor to met, blood is primary': #old\n",
    "# #         df_merge.loc[b_type,'Tumor_Type'] = 'Met'\n",
    "# #         df_merge.loc[b_type,'Blood_Type'] = 'Primary'\n",
    "#     if s_type == 'exclude from analysis': #Y - exclude from analysis?\n",
    "#         df_merge.loc[b_type,'Blood_Type'] = np.nan\n",
    "#         df_merge.loc[b_type,ls_blood] = np.nan\n",
    "#     elif s_type == 'can be considered a primary blood':\n",
    "#         df_merge.loc[b_type,'Blood_Type'] = 'Primary'\n",
    "#     elif s_type == 'can be considered a met blood': #likely met\n",
    "#         df_merge.loc[b_type,'Blood_Type'] = 'Met'\n",
    "#     elif s_type == 'Y - but include in all':\n",
    "#         df_merge.loc[b_type,'Blood_Type'] = np.nan\n",
    "#     elif s_type == 'unknown':\n",
    "#         df_merge.loc[b_type,'Blood_Type'] = np.nan\n",
    "#     #break\n",
    "# #save\n",
    "# #ls_col = ['Stage','Lung_Cohort','Liver_Cohort','pORG_Primary', 'pORG_Met','Public_Patient_ID']\n",
    "# #df_blood_dates.merge(df_merge.loc[:,ls_col],on='Public_Patient_ID',how='left').to_csv('Blood_dates_to_review.csv')\n",
    "# # removed blood from analysi because it was collected 417 days before the PDAC diagnosis\n",
    "# # df_id[df_id.loc[:,'Biolibrary.Subject.ID']=='ST-00021096']\n",
    "# # df_tcell_blood[df_tcell_blood.loc[:,'Patient ID'] == 'ST-00021096']\n",
    "# # #within how many days\n",
    "# # b_match = df_merge.Clonality_Tumor.notna() & df_merge.Clonality_Blood.notna()\n",
    "# # print(len(df_merge.loc[b_match]))\n",
    "# # df_test = df_blood_dates[df_blood_dates.loc[:,'tumor and blood same collection date?'] == 'no'].copy()\n",
    "# # df_test['difference'] = df_test['date tumor collection'] - df_test['date blood processed']\n",
    "# # print( 1 - (19/ 215))\n",
    "# # # df_test.dtypes\n",
    "# # ls_col = ['tumor specimens','Public_Patient_ID','difference']\n",
    "# # df_test.loc[df_test.difference!='0 days',ls_col].sort_values(by='difference').to_csv('Tumor-blood_in_days.csv')\n",
    "# #add immunarch metrics\n",
    "# importlib.reload(util)\n",
    "# s_dir =f'{codedir.split(\"Liver_Lung_PDAC\")[0]}R'\n",
    "# df_meta = pd.read_csv(f'{s_dir}/raw_TCR_data_test/metadata.txt',sep='\\t')\n",
    "# df_meta['Sample_Type'] = df_meta.Site.replace({'Primary':'Tumor','Met':'Tumor'})\n",
    "# for s_file in os.listdir(s_dir):\n",
    "#     if s_file.find('results_TCR') > -1:\n",
    "#         #s_file = 'results_TCR_jaccard_overlap.csv'\n",
    "#         print(s_file)\n",
    "#         df = pd.read_csv(f'{s_dir}/{s_file}',index_col=0)\n",
    "#         if s_file =='results_TCR_hill_diversity.csv':\n",
    "#             df.set_index('Sample',inplace=True)\n",
    "#             df = df.pivot(columns='Q')\n",
    "#             df.columns = [f'Hill_{item[1]}' for item in df.columns]\n",
    "#             df.index.name = 'INDEX'\n",
    "#         elif s_file.find('overlap.csv') > -1:\n",
    "#             df.columns = [item.split('TCR_raw_data_')[1] for item in df.columns]\n",
    "#             df.index = [item.split('TCR_raw_data_')[1] for item in df.index]\n",
    "#             #break\n",
    "#             df = util.process_overlap2(df,df_meta,ls_site=['Sample_Type'])\n",
    "#             prefix = f\"{s_file.split('_')[-2]}_\"\n",
    "#             df.rename({item:prefix+item for item in df.columns[df.columns!='Sample']},axis=1,inplace=True)\n",
    "#             print(df.columns)\n",
    "#             #break\n",
    "#         else:\n",
    "#             prefix = f\"{s_file.split('_')[-2]}_\"\n",
    "#             df.rename({item:prefix+item for item in df.columns[df.columns!='Sample']},axis=1,inplace=True)\n",
    "#         try:\n",
    "#             df['Sample'] = [item.split('TCR_raw_data_')[1] for item in df.Sample]\n",
    "#         except:\n",
    "#             try: \n",
    "#                 df['Sample'] = [item.split('TCR_raw_data_')[1] for item in df.index]\n",
    "#             except:\n",
    "#                 df['Sample'] = df.index\n",
    "#         df['Public_Patient_ID'] = ['ST-' + item.split('-')[1] for item in df.Sample]\n",
    "#         df_blood = df.loc[df.Sample.str.contains('-B'),~df.columns.str.contains('Sample')]\n",
    "#         df_tum = df.loc[~df.Sample.str.contains('-B'),~df.columns.str.contains('Sample')]\n",
    "#         df_merge = df_merge.merge(df_tum,on='Public_Patient_ID',suffixes=('','_tumor'),how='left')\n",
    "#         df_merge = df_merge.merge(df_blood,on='Public_Patient_ID',suffixes=('','_blood'),how='left')\n",
    "#         #break\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5f27afc-e89f-4e9b-b185-8980f4540d7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # #add predictors\n",
    "# # df_patient['Liver_Cohort'] = pd.NA\n",
    "# # df_patient.loc[df_patient.Cohort=='Liver','Liver_Cohort'] = 1\n",
    "# # df_patient.loc[df_patient.Cohort=='Lung','Liver_Cohort'] = 0\n",
    "# # print(df_patient.Liver_Cohort.value_counts())\n",
    "# # df_patient['Basal_Like'] = pd.NA\n",
    "# # df_patient.loc[df_patient.PurIST_Subtype=='basal-like','Basal_Like'] = 1\n",
    "# # df_patient.loc[df_patient.PurIST_Subtype=='classical','Basal_Like'] = 0\n",
    "# # df_patient.Basal_Like.value_counts()\n",
    "# #add LOO scores\n",
    "# from sklearn.preprocessing import minmax_scale\n",
    "# d_ids = pd.read_excel(f'{codedir.split(\"Liver_Lung_PDAC\")[0]}MethodsAndReferencesSupplementalData/OLD Versions/Simplified_Public_IDs_Key.xlsx',sheet_name=None)\n",
    "\n",
    "# df_id = d_ids['RnaSeqKey']\n",
    "# ls_ids = df_id.loc[:,'Public.Specimen.ID']\n",
    "\n",
    "# # add patients w/o RNA seq\n",
    "# for s_key in ['TcrTumorKey','TcrBloodKey','SuppTable2Key','DnaPanelKey']:\n",
    "#     df_add = d_ids[s_key].loc[~d_ids[s_key].loc[:,'Public.Specimen.ID'].isin(ls_ids)]\n",
    "#     df_id = pd.concat([df_id,df_add])\n",
    "# d_id = dict(zip(df_id.OPTR,df_id.loc[:,'Biolibrary.Subject.ID']))\n",
    "# for s_type in ['pSUB','pORG']:\n",
    "#     df_porg = pd.read_csv(f'../R/results_loo_{s_type}.csv',index_col=0)\n",
    "#     df_porg['OPTR'] = [int(item.split('-')[0]) for item in df_porg.SampleID]\n",
    "#     df_porg['Public_Patient_ID'] = df_porg['OPTR'].map(d_id)\n",
    "#     d_loo = dict(zip(df_porg['Public_Patient_ID'],df_porg.loc[:,s_type]))\n",
    "#     df_patient[f'LOO_{s_type}'] = df_patient.Public_Patient_ID.map(d_loo)\n",
    "#     if s_type == 'pORG':\n",
    "#         d_loo = dict(zip(df_porg['Public_Patient_ID'],df_porg.loc[:,'pSUB']))\n",
    "#         df_patient[f'LOO_pSUB_0.2'] = df_patient.Public_Patient_ID.map(d_loo)\n",
    "#     #break\n",
    "# df_patient['PurIST_pORG'] = (minmax_scale(df_patient.PurIST_Primary) + minmax_scale(df_patient.LOO_pORG))\n",
    "# df_patient['PurIST_pSUB'] = (minmax_scale(df_patient.PurIST_Primary) + minmax_scale(df_patient.LOO_pSUB))\n",
    "# # # df = df_porg.loc[:,['pORG_Length','pSUB_Length']].astype('int64')\n",
    "# # # df = df.astype('int64')\n",
    "# # # #sns.stripplot(data=df)\n",
    "# # # df.dtypes\n",
    "# # from sklearn.metrics import roc_curve, RocCurveDisplay, roc_auc_score\n",
    "# # %matplotlib inline \n",
    "# # #'PurIST_pORG','PurIST_pSUB','LOO_pSUB_0.2',\n",
    "# # for s_col in ['Liver_Cohort','Basal_Like',]:\n",
    "# #     for s_tcr in ['LOO_pSUB','LOO_pORG','PurIST_Primary','pORG_Primary','pSUB_Primary',]: #'pORG_Met','pSUB_Met',\n",
    "# #         df_test = df_patient[(df_patient.loc[:,s_col].notna()) & (df_patient.loc[:,s_tcr].notna())].copy()\n",
    "# #         df_test.loc[:,s_col] = df_test.loc[:,s_col].astype('int')\n",
    "# #         #print(df_test.loc[:,s_col].value_counts())\n",
    "# #         #auc\n",
    "# #         pos_label = 1\n",
    "# #         y_true=df_test.loc[:,s_col]\n",
    "# #         #print(len(y_true))\n",
    "# #         y_score=minmax_scale(df_test.loc[:,s_tcr])\n",
    "# #         #print(len(y_score))\n",
    "# #         y_true = y_true.astype(int)\n",
    "# #         #print(len(y_true))\n",
    "# #         fpr, tpr, thresholds = roc_curve(y_true=y_true,\n",
    "# #                                          y_score=y_score,\n",
    "# #                                          pos_label=pos_label,drop_intermediate=False)\n",
    "# #         #print(len(thresholds) + y_score.duplicated().sum())\n",
    "# #         auc = roc_auc_score(y_true=y_true, y_score=y_score)\n",
    "# #         fig,ax = plt.subplots(dpi=300)\n",
    "# #         display = RocCurveDisplay.from_predictions(y_true, y_score,ax=ax)#y_true=y_true, y_score=y_score\n",
    "# #         #display.plot(ax=ax)\n",
    "# #         ax.plot([0, 1], [0, 1], transform=ax.transAxes,ls='--',color='lightgray')\n",
    "# #         ax.set_title(f'{s_tcr} vs. {s_col}\\nAUC={auc:.2} ') #N={len(df_test)}\n",
    "# #         fig.savefig(f'ROC_{s_tcr}_vs_{s_col}.pdf')\n",
    "# #         break\n",
    "# #     #break\n",
    "# importlib.reload(util)\n",
    "# tu_time_censor = (('Days from Diagnosis to FU','Survival'),\n",
    "#                   ('Days from Resection to Recurrence', 'Recurrence')\n",
    "#                  )\n",
    "# df_patient['pORG_binary_LOO'] = pd.NA\n",
    "# df_patient.loc[df_patient.LOO_pORG > 0.0249,'pORG_binary_LOO'] = 'High'\n",
    "# df_patient.loc[df_patient.LOO_pORG <= 0.0249,'pORG_binary_LOO'] = 'Low'\n",
    "# for (s_time, s_censor) in tu_time_censor:\n",
    "#     util.km_plot(df_patient[df_patient.pORG_binary_LOO.notna()], 'pORG_binary_LOO',s_time,s_censor)\n",
    "#     break\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66dcfdb2-8dd5-416a-b4cc-cc21c4d52598",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# #high low pORG quartile versus gene expresison\n",
    "# #bug \n",
    "# pal_porg = ('#56B4E9','#E69F00')\n",
    "# pal_liv = ('#0072B2','#D55E00')\n",
    "# sns.set_palette(pal_porg)\n",
    "# #if s_porg == 'pORG_0.2_Primary':\n",
    "# #if s_porg == 'pORG_0.2_Met':\n",
    "# #df_rna['Public_Patient_ID'] = [item.split('-T')[0] for item in df_rna.index]\n",
    "# #liver vs lung (run t cell section to load t cell data!)\n",
    "# with warnings.catch_warnings():\n",
    "#     warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "#     for s_porg in ['pORG_Primary','pORG_Met']:\n",
    "#         ls_marker = ['CD3E','CD4', 'CD8A', 'MS4A1','CD68','OLR1','MRC1','MX1','STAT1',]\n",
    "#         df_plot = df_rna.loc[df_rna.index.isin(df_merge.Public_Patient_ID),ls_marker].unstack().reset_index()\n",
    "#         df_plot['Public_Patient_ID'] = df_plot.level_1 #[item[0:-2] for item in df_plot.level_1]\n",
    "#         df_plot['pORG'] = df_plot.Public_Patient_ID.map(dict(zip(df_merge.Public_Patient_ID,df_merge.loc[:,s_porg])))\n",
    "#         df_plot.rename({0:'Expression','level_0':'Gene'},axis=1,inplace=True)\n",
    "#         fig,ax = plt.subplots(dpi=200,figsize=(5,3))\n",
    "#         sns.stripplot(data=df_plot,x='Gene',y='Expression',hue='pORG',dodge=True,ax=ax,s=2)\n",
    "#         sns.boxplot(data=df_plot,x='Gene',y='Expression',hue='pORG',ax=ax,showmeans=True,medianprops={'visible': False},\n",
    "#                                whiskerprops={'visible': False},meanline=True,showcaps=False,\n",
    "#                                meanprops={'color': 'k', 'ls': '-', 'lw': 2},showfliers=False,showbox=False)\n",
    "#         h, l = ax.get_legend_handles_labels()\n",
    "#         ax.legend(h[0:2],l[0:2],loc='lower left')\n",
    "#         pairs = [((item,'low'),(item,'high')) for item in ls_marker]\n",
    "#         annot = Annotator(ax, pairs, data=df_plot,x='Gene',y='Expression',hue='pORG',\n",
    "#                           order=ls_marker,hue_order=('low','high'))\n",
    "#         annot.configure(test='t-test_ind', text_format='simple',fontsize=7,comparisons_correction='fdr_bh') #\n",
    "#         annot.apply_test()\n",
    "#         d_pval = dict([(res.data.group1[0],res.data.pvalue) for res in annot.annotations])\n",
    "#         pvalues = [d_pval[item] for item in ls_marker]\n",
    "#         reject, corrected, __, __ = statsmodels.stats.multitest.multipletests(pvalues,method='fdr_bh')\n",
    "#         formatted_pvalues = [f'p={pvalue:.2}' for pvalue in list(corrected)]\n",
    "#         annot.set_custom_annotations(formatted_pvalues)\n",
    "#         annot.annotate()\n",
    "#         ax.set_title(f'{s_porg} Quartiles') \n",
    "#         plt.tight_layout()\n",
    "#         fig.savefig(f'figures/gene_expression_binary_{s_porg}_Quartiles.png')\n",
    "#         #plt.close(fig)\n",
    "#         break\n",
    "\n",
    "\n",
    "# #pORG vs survival within liver and lung (all patients)\n",
    "# df_patient['pORG_0.2_All'] = df_patient.loc[:,'pORG_0.2_allPrimary'].fillna(df_patient.loc[:,'pORG_0.2_allMet']).fillna(df_patient.loc[:,'pORG_0.2_All_T2'])\n",
    "# df_patient['pSUB_1e-04_All'] = df_patient.loc[:,'pSUB1e-04_allPrimary'].fillna(df_patient.loc[:,'pSUB1e-04_allMet'])\n",
    "# df_patient['PurIST_All'] = df_patient.loc[:,'PurIST_Primary'].fillna(df_patient.loc[:,'PurIST_Met']).fillna(df_patient.loc[:,'PurIST_Primary_T2'])\n",
    "\n",
    "# for s_col in ['PurIST_All','pORG_0.2_All','pSUB_1e-04_All']:\n",
    "#     for s_cohort in ['Liver','Lung','both']:\n",
    "#         print(s_cohort)\n",
    "#         try: #cox\n",
    "#             df = df_patient.loc[(df_patient.Cohort==s_cohort),[s_col,s_time,s_censor]].dropna()\n",
    "#             fig2, cph = util.cph_plot(df,s_col,s_time,s_censor,figsize=(4,1.5))\n",
    "#             plt.tight_layout()\n",
    "#         except:\n",
    "#             df = df_patient.loc[:,[s_col,s_time,s_censor]].dropna()\n",
    "#             fig2, cph = util.cph_plot(df,s_col,s_time,s_censor,figsize=(4,1.5))\n",
    "#             plt.tight_layout()\n",
    "#     break\n",
    "\n",
    "# 'Fraction Shared Clones 0to2',\n",
    "#  'Fraction Shared Clones 2to5',\n",
    "#  'Fraction Shared Clones 5to10',\n",
    "#  'Fraction Shared Clones >=10',\n",
    "#  'Fraction Tumor Distinct Clones 0to2',\n",
    "#  'Fraction Tumor Distinct Clones 2to5',\n",
    "#  'Fraction Tumor Distinct Clones 5to10',\n",
    "#  'Fraction Tumor Distinct Clones >=10',\n",
    "#         'Hill_1', 'Hill_2', 'Hill_3',\n",
    "#        'Hill_4', 'Hill_5', 'Hill_6',\n",
    "#     'propshared_proportion_shared_clones',\n",
    "#       'propshared_proportion_shared_templates',\n",
    "#     'propshared_proportion_shared_clones_blood',\n",
    "#        'propshared_proportion_shared_templates_blood',\n",
    "#     'Fraction Shared Clones 10',\n",
    "#       'Fraction Shared Clones 5',\n",
    "#       'Fraction Shared Clones 2',\n",
    "#            'Hill_1_blood','Hill_2_blood', 'Hill_3_blood',\n",
    "#         'Hill_4_blood', 'Hill_5_blood','Hill_6_blood',\n",
    "#     'd50_Clones', 'd50_Clones_blood',\n",
    "#            'chao_Estimator','chao_Estimator_blood',\n",
    "#            'true_Value','true_Value_blood',\n",
    "#            'ginisimp_Value','ginisimp_Value_blood',\n",
    "# # km - old - using predetermined cutoffs\n",
    "# alpha = 0.1\n",
    "# savedir = 'figures'\n",
    "# d_cut = {4:['low','med-low','med-high','high'],\n",
    "#          3: ['low','highX','high'], #['low','med','high'],#\n",
    "#          2:['low','high']\n",
    "#         }\n",
    "# for i_cut, labels in d_cut.items():\n",
    "#     for s_title_str in  ['All','Primary','Met',]:\n",
    "#         df_result = pd.DataFrame()\n",
    "#         if s_title_str == 'All':\n",
    "#             df_km_samples = df_merge.loc[df_merge.Alive_30_days_post_surgery]\n",
    "#         else:\n",
    "#             df_km_samples = df_merge.loc[(df_merge.loc[:,'Tumor_Type']==s_title_str) & (df_merge.Alive_30_days_post_surgery),[s_time,s_censor,'Public_Patient_ID']+ls_foci]\n",
    "#         for s_col in ls_foci: #[2::]\n",
    "#             try:\n",
    "#                 df_km, pvalue = util.quartile_km(df_km_samples,s_col,s_title_str,savedir,alpha,i_cut,\n",
    "#                                                 labels,s_time=s_time,s_censor='Survival')\n",
    "#                 print(f'{s_col} {s_title_str} {i_cut}: {pvalue}')\n",
    "#             except:\n",
    "#                 print(f'no KM {s_col} {s_title_str} {i_cut}')\n",
    "\n",
    "# save out results\n",
    "\n",
    "#df_km_pts\n",
    "#         df_result_model = cph.summary.loc[:,['exp(coef)','p']].reset_index()\n",
    "#         df_result=pd.concat([df_result,df_result_model])\n",
    "#         #plt.close(fig2)\n",
    "    #print(s_title_str)\n",
    "#break\n",
    "#df_result.sort_values(by='p').set_index('covariate').to_csv(f'results_CPH_TCR_{s_title_str}.csv')\n",
    "\n",
    " # ls_foci_heatmap = ['Fraction Shared Clones 0to2','Fraction Shared Clones 2to5',\n",
    " # 'Fraction Shared Clones 5to10','Fraction Shared Clones >=10',\n",
    " # 'Fraction Tumor Distinct Clones 0to2','Fraction Tumor Distinct Clones 2to5',\n",
    " # 'Fraction Tumor Distinct Clones 5to10','Fraction Tumor Distinct Clones >=10',\n",
    " # 'Hill_1','Hill_2','Hill_3','Hill_4','Hill_5','Hill_6',\n",
    " # 'Hill_1_blood','Hill_2_blood','Hill_3_blood','Hill_4_blood','Hill_5_blood','Hill_6_blood',\n",
    " # 'd50_Clones','d50_Clones_blood',\n",
    " # 'chao_Estimator','chao_Estimator_blood',\n",
    " # 'true_Value','true_Value_blood',\n",
    " # 'ginisimp_Value','ginisimp_Value_blood',\n",
    " # 'Simpsons_Diversity_Tumor','Productive_Rearrangements',\n",
    " # 'Templates_per_ng','Shannon_Entropy_Tumor',\n",
    " # 'Clonality_Tumor',\"Simpson's Evenness tumor\",\n",
    " # 'Productive_Rearrangements_Blood',\"Simpson's Evenness blood\",'Shannon_Entropy_Blood',\n",
    " # 'Simpsons_Diversity_Blood','Clonality_Blood',\n",
    " # 'jaccard_Primary','jaccard_Blood','jaccard_Met',\n",
    " # 'public_Blood','public_Met','public_Primary',\n",
    " # 'morisita_Blood','morisita_Met','morisita_Primary',\n",
    " # 'public_Met_Bld_blood','public_Primary_Bld_blood',\n",
    " # 'jaccard_Primary_Bld_blood','jaccard_Met_Bld_blood',\n",
    " # 'morisita_Met_Bld_blood','morisita_Primary_Bld_blood']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ab55c61-89b3-4c2a-ae86-71d47b9c6d3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#old data loading for survival analysis\n",
    "# s_out = 'annotation/20231222_Patient_Metadata.csv'#'annotation/20230921_Patient_Metadata.csv'\n",
    "# df_patient = pd.read_csv(s_out,index_col=0)\n",
    "# which patients are in which dataset\n",
    "# #add recurrence censor\n",
    "# df_patient['Recurrence'] = np.nan\n",
    "# df_patient.loc[~df_patient.loc[:,'Days from Resection to Recurrence'].isna(),'Recurrence'] = 1\n",
    "# df_patient.loc[df_patient.No_Recurrence,'Recurrence'] = 0\n",
    "# print(len(df_patient))\n",
    "# %matplotlib inline\n",
    "# # add purist subtype\n",
    "# df_patient.loc[(df_patient.PurIST_Primary <= 0.5),'PurIST_Subtype'] = 'classical'\n",
    "# df_patient.loc[(df_patient.PurIST_Met <= 0.5),'PurIST_Subtype_Met'] = 'classical'\n",
    "# df_patient.loc[(df_patient.PurIST_Met <= 0.5),'PurIST_Subtype'] = 'classical'\n",
    "# df_patient.loc[(df_patient.PurIST_Primary > 0.5),'PurIST_Subtype'] = 'basal-like'\n",
    "# df_patient.loc[(df_patient.PurIST_Met > 0.5),'PurIST_Subtype_Met'] = 'basal-like'\n",
    "# df_patient.loc[(df_patient.PurIST_Met > 0.5),'PurIST_Subtype'] = 'basal-like'\n",
    "# #df_patient['PurIST_Subtype'] = df_patient.PurIST_Subtype.fillna(df_patient.PurIST_Subtype_Met)\n",
    "# df_patient.loc[df_patient.PurIST_Subtype=='classical','Classical_Cohort'] = df_patient.loc[df_patient.PurIST_Subtype=='classical','Cohort']\n",
    "# print(df_patient.Classical_Cohort.value_counts())\n",
    "\n",
    "# #fix no recur and other recur - out of resected patients\n",
    "# df_patient.loc[:,'No_Doc_Recur'] = pd.NA #np.nan #False\n",
    "# df_patient.loc[df_patient.Recurrence_Sites_4!='No_Resection','No_Doc_Recur'] = False#0 #False\n",
    "# df_patient.loc[df_patient.Recurrence_Sites_4=='No_Doc_Recur','No_Doc_Recur'] = True#1 #True\n",
    "# print(df_patient.No_Recurrence.value_counts())\n",
    "\n",
    "# df_patient.loc[:,'Other_Recurrence'] = pd.NA #np.nan #False\n",
    "# df_patient.loc[df_patient.Recurrence_Sites_4!='No_Resection','Other_Recurrence'] = False #0#\n",
    "# df_patient.loc[df_patient.Recurrence_Sites_4=='Other_site','Other_Recurrence'] = True #1#\n",
    "# print(df_patient.Other_Recurrence.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ba67e8f-c5cd-4010-ae68-1cba8eb9390f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #for letter to reviewers\n",
    "# which patients are in which dataset\n",
    "# df_patient = pd.read_excel('Supplemental_Data/Supplemental_Dataset_1.xlsx',sheet_name=0,\n",
    "#                           index_col=0)\n",
    "# df_tcr = pd.read_csv('Supplemental_Data/Supplemental_Dataset_8.csv',\n",
    "#                           index_col=0)\n",
    "# # #lung wedge resection, liver met resection versus survival\n",
    "# df_patient['No_Met_Resections'] = df_patient.Cohort_Met_Resection.replace({'Lung_Met_Res':pd.NA,'Liver_Met_Res':pd.NA})\n",
    "# df_patient['Met_Resections'] = df_patient.Cohort_Met_Resection.replace({'Lung':pd.NA,'Liver':pd.NA})\n",
    "# df_patient['Liver_Met_Resection'] = df_patient.Cohort_Met_Resection.replace({'Lung_Met_Res':pd.NA,'Lung':pd.NA})\n",
    "\n",
    "# s_time = 'Days from Diagnosis to FU'\n",
    "# s_censor = 'Survival'\n",
    "# util.km_plot(df_patient[df_patient.Alive_30_days_post_surgery],'Lung_Met_Resection',s_time,s_censor)\n",
    "# util.km_plot(df_patient[df_patient.Alive_30_days_post_surgery],'Cohort_Met_Resection',s_time,s_censor)\n",
    "# util.km_plot(df_patient[df_patient.Alive_30_days_post_surgery],'No_Met_Resections',s_time,s_censor)\n",
    "# util.km_plot(df_patient[df_patient.Alive_30_days_post_surgery],'Met_Resections',s_time,s_censor)\n",
    "# util.km_plot(df_patient[df_patient.Alive_30_days_post_surgery],'Liver_Met_Resection',s_time,s_censor)\n",
    "# #df_patient.loc[df_patient.Public_Patient_ID.isin(se_lung_pub),ls_col]\n",
    "# #df_patient.loc[df_patient.Public_Patient_ID.isin(se_liver),ls_col]\n",
    "# df_patient['RNAseq_Patient'] = df_patient.loc[:,df_patient.columns.str.contains('PurIST_')].notna().any(axis=1)\n",
    "# df_patient['DNAseq_Patient'] = df_patient.TP53_Altered.notna()\n",
    "# #TCR\n",
    "# ls_tum = df_tcr[df_tcr.Shannon_Entropy_Tumor.notna()].Public_Patient_ID\n",
    "# ls_blood = df_tcr[df_tcr.Shannon_Entropy_Blood.notna()].Public_Patient_ID\n",
    "# df_patient['TCRseq_Tumor_Patient'] = df_patient.Public_Patient_ID.isin(ls_tum)\n",
    "# df_patient['TCRseq_Blood_Patient'] = df_patient.Public_Patient_ID.isin(ls_blood)\n",
    "# ls_data = ['RNAseq_Patient','DNAseq_Patient','TCRseq_Tumor_Patient','TCRseq_Blood_Patient']\n",
    "# df_patient['Any_omics'] = df_patient.loc[:,ls_data].any(axis=1)\n",
    "# #df_patient.to_csv('Supplemental_Data/Supplemental_Dataset_1.csv')\n",
    "# ## CT scan patients (tempus only)\n",
    "# df_lesion = pd.read_excel('../Lesions/LesionSizeList-FMP-19July2021.xlsx')\n",
    "# #df_lesion.loc[df_lesion['Lesion Site Description'] == 'Retroperitoneal','Lesion Site'] = 'Retroperitoneal'\n",
    "# #df_lesion.loc[df_lesion['Lesion Site Description'] == 'Common hepatic artery node','Lesion Site'] = 'Common hepatic artery'\n",
    "# df_lesion['Lesion Site'] = df_lesion['Lesion Site'].fillna(pd.NA)\n",
    "# df_lesion['Public_Patient_ID'] = df_lesion.loc[:,'OPTR ID'].map(dict(zip(df_id.OPTR,df_id.loc[:,'Biolibrary.Subject.ID'])))\n",
    "# df_patient['Lesion_Reviewed'] = df_patient.Public_Patient_ID.isin(df_lesion.Public_Patient_ID)\n",
    "# #distant met\n",
    "# ls_local = ['Pancreas','Local','Ampulla of Vater',pd.NA]\n",
    "# ls_dist = df_lesion[~df_lesion['Lesion Site'].isin(ls_local)].Public_Patient_ID\n",
    "# df_patient['Distant_Lesion_Reviewed'] = df_patient.Public_Patient_ID.isin(ls_dist)\n",
    "# #df_patient.groupby(['Recurrence_Sites_4','Lesion_Reviewed']).Public_Patient_ID.count()\n",
    "# #df_patient.groupby(['Cohort','Distant_Lesion_Reviewed']).Public_Patient_ID.count()\n",
    "# #df_patient.groupby(['Distant_Lesion_Reviewed','Cohort']).Public_Patient_ID.count()\n",
    "# #ls_data = ['RNAseq_Patient','DNAseq_Patient','TCRseq_Tumor_Patient','TCRseq_Blood_Patient']\n",
    "# sum(df_patient.loc[:,['Any_omics','Lesion_Reviewed']].sum(axis=1)==2)\n",
    "# df_patient[~df_patient.Any_omics].Cohort.value_counts()\n",
    "# df_patient.groupby(['Cohort','Any_omics']).Public_Patient_ID.count()\n",
    "# FYI #none of histology features are prognostic\n",
    "# ls_hist = [ 'TLS',\n",
    "#  'Inflammation',\n",
    "#  'PNI',\n",
    "#  'ALI (Y/N)',\n",
    "#  'Desmoplasia (Y/N)']\n",
    "# for s_hist in ls_hist:\n",
    "#     df_patient[f'{s_hist}_All'] = df_patient.loc[:,s_hist].fillna(df_patient.loc[:,f'{s_hist}_Met'])\n",
    "#     fig,ax,ls_order = util.km_plot(df_patient,s_hist,s_time,s_censor)\n",
    "#     ax.set_xlabel(s_censor)\n",
    "#     fig,ax,ls_order = util.km_plot(df_patient,f'{s_hist}_Met',s_time,s_censor)\n",
    "#     ax.set_xlabel(s_censor)\n",
    "#     fig,ax,ls_order = util.km_plot(df_patient,f'{s_hist}_All',s_time,s_censor)\n",
    "#     ax.set_xlabel(s_censor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7c83cab-7f13-49e4-98e9-3d5d4d03b524",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pORG quartiles vs gene alterations\n",
    "# # #pORG primary versus TMB\n",
    "# d_fdr = {'pORG_Primary':{'KRAS' : 0.38287593347963944,\n",
    "# 'GATA1' : 0.1855045586524062,\n",
    "# 'CDKN2A' : 0.06609676055955013,\n",
    "# 'TP53' : 1.038674992908113e-09,\n",
    "# 'KDM5C' : 0.4251464077159879,},\n",
    "#         'pSUB_Primary':{'TSC2' : 1.0,},'Cohort':{'ATM' : 0.8990812819973347,\n",
    "# 'GATA1' : 0.8990812819973347,},\n",
    "#         'PurIST_Primary':{'TGFBR2' : 0.284384813956165,\n",
    "# 'ATM' : 0.284384813956165,\n",
    "# 'GNAS' : 0.284384813956165,}}\n",
    "\n",
    "# df_patient['PurIST_All'] = df_patient.loc[:,'PurIST_Primary'].fillna(df_patient.loc[:,'PurIST_Met'])\n",
    "# import warnings\n",
    "# with warnings.catch_warnings():\n",
    "#     warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "#     d_all = {}\n",
    "#     d_all_p = {}\n",
    "#     cutoff = 0.1\n",
    "#     for s_gsva_name in ['pORG','Cohort','pSUB','PurIST']:\n",
    "#         print(s_gsva_name)\n",
    "#         for s_type in ['Primary','Met',]: #'All',\n",
    "#             if s_gsva_name == 'Cohort':\n",
    "#                 s_gsva = 'Cohort'\n",
    "#             elif s_type == 'All' and s_gsva_name !='PurIST':\n",
    "#                 ls = df_patient.columns[(df_patient.columns.str.contains(s_gsva_name)) & (df_patient.columns.str.contains(f'_all'))]\n",
    "#                 s_gsva = f\"{ls[0].split('_all')[0]}_All\"\n",
    "#                 df_patient[s_gsva] = df_patient.loc[:,ls[0]].fillna(df_patient.loc[:,ls[1]])\n",
    "#             else:\n",
    "#                 s_gsva = df_patient.columns[(df_patient.columns.str.contains(s_gsva_name)) & (df_patient.columns.str.contains(f'_{s_type}'))][0]\n",
    "#             b_rna = df_patient.loc[:,s_gsva].notna()\n",
    "            \n",
    "#             ls_pvalue = []\n",
    "#             ls_gene_correct = []\n",
    "#             for s_col in ls_gene:\n",
    "#                 s_gene = f'{s_col}_Altered'\n",
    "#                 try:\n",
    "#                     b_dna = df_patient.loc[:,s_gene].notna()\n",
    "#                 except:\n",
    "#                     continue\n",
    "#                 df_plot = df_patient.loc[b_rna & b_dna,['Public_Patient_ID',s_gene,s_gsva]].copy()\n",
    "#                 if not s_gsva == 'Cohort':\n",
    "#                     #df_plot = df_patient.loc[b_rna & b_dna,['Public_Patient_ID',s_gene,s_gsva]].copy()\n",
    "#                     i_quart = df_plot.loc[:,s_gsva].notna().sum()//4\n",
    "#                     se_low = df_plot.loc[:,['Public_Patient_ID',s_gsva]].sort_values(by=s_gsva)[0:i_quart].Public_Patient_ID\n",
    "#                     se_high = df_plot.loc[:,['Public_Patient_ID',s_gsva]].sort_values(by=s_gsva)[-i_quart::].Public_Patient_ID\n",
    "#                     #df_plot['quartiles'] = pd.qcut(df_plot.loc[:,s_gsva],q=4,labels=['low','medX','med','high'])\n",
    "#                     #df_long = df_plot.loc[~df_plot.quartiles.str.contains('med'),[s_gene,'quartiles']]\n",
    "#                     df_plot['quartiles'] = pd.NA\n",
    "#                     df_plot.loc[df_plot.Public_Patient_ID.isin(se_low),'quartiles'] = 'low'\n",
    "#                     df_plot.loc[df_plot.Public_Patient_ID.isin(se_high),'quartiles'] = 'high'\n",
    "#                     #print(df_plot.quartiles.value_counts())\n",
    "#                     df_long = df_plot.dropna()\n",
    "#                     break\n",
    "#                 '''\n",
    "#                     try:\n",
    "#                         stat, pvalue = scipy.stats.fisher_exact(pd.crosstab(df_long.quartiles, df_long.loc[:,s_gene]))\n",
    "#                     except:\n",
    "#                         print('error')\n",
    "#                         pvalue = 1\n",
    "#                 else:\n",
    "#                     df_long = df_patient.loc[b_rna & b_dna,[s_gene,s_gsva]].copy()\n",
    "#                     try:\n",
    "#                         stat, pvalue = scipy.stats.fisher_exact(pd.crosstab(df_long.Cohort, df_long.loc[:,s_gene]))\n",
    "#                     except:\n",
    "#                         pvalue = 1\n",
    "                \n",
    "#                 ls_gene_correct.append(s_col)\n",
    "#                 ls_pvalue.append(pvalue)\n",
    "#                 #plot\n",
    "#                 fig,ax=plt.subplots(figsize=(2,2),dpi=200)\n",
    "#                 if not s_gsva == 'Cohort':\n",
    "#                     sns.pointplot(data=df_long,x='quartiles',y=s_gene,ax=ax,errorbar=None,order=['low','high'])\n",
    "#                 else:\n",
    "#                     sns.pointplot(data=df_long,x='Cohort',y=s_gene,ax=ax,errorbar=None,order=['Lung','Liver'])\n",
    "#                 ax.set_xlabel(s_gsva)\n",
    "#                 try:\n",
    "#                     ax.set_title(f\"{s_col} {s_type}\\nFisher's exact\\np={pvalue:.2}\",fontsize=10)\n",
    "#                 except:\n",
    "#                     continue\n",
    "#                 ax.set_ylabel(f'Fraction Pts n={len(df_long)}')\n",
    "#                 ax.set_ylim(-0.1,1.1)\n",
    "#                 plt.tight_layout()\n",
    "#                 if pvalue < cutoff:\n",
    "#                     try:\n",
    "#                         ax.set_title(f\"{s_col} {s_type}\\nFisher's p={pvalue:.2}\\nFDR={d_fdr[s_gsva][s_col]:.2}\",fontsize=10)\n",
    "#                     except:\n",
    "#                         continue\n",
    "#                     fig.savefig(f'figures/Fraction_{s_col}_{s_type}_{s_gsva}.png')\n",
    "#                 else:\n",
    "#                     plt.close(fig)\n",
    "#             #correct pvalues\n",
    "#             __, corrected = statsmodels.stats.multitest.fdrcorrection(np.array(ls_pvalue))\n",
    "#             d_pval = dict(zip(ls_gene_correct,ls_pvalue))\n",
    "#             d_all_p.update({f'{s_gsva}':d_pval})\n",
    "#             d_qval = dict(zip(ls_gene_correct,corrected))\n",
    "#             d_all.update({f'{s_gsva}':d_qval})\n",
    "#             '''\n",
    "#             #pORG primary versus TMB\n",
    "#             df_high_low = df_patient.merge(df_plot.loc[:,['Public_Patient_ID','quartiles']],\n",
    "#                              on='Public_Patient_ID',how='inner').copy()\n",
    "#             df_high_low = df_high_low[df_high_low.quartiles.notna()]\n",
    "#             ls_groups = ['low','high']\n",
    "#             if s_type == 'Met':\n",
    "#                 y = 'Tumor_Mutation_Burden_Met'\n",
    "#             elif s_type == 'Primary':\n",
    "#                 y = 'Tumor_Mutation_Burden_Primary'\n",
    "#             plotting = {\"data\": df_high_low,\"x\":\"quartiles\",\"y\":y,\"order\":ls_groups}\n",
    "#             fig,ax,test_results = util.annotated_stripplot(plotting,ls_groups,y)\n",
    "#             for res in test_results:\n",
    "#                 pvalue=res.data.pvalue\n",
    "#             ax.set_title(f'TMB {y.split(\"_\")[-1]} vs {s_gsva_name}\\np={pvalue:.2}')\n",
    "#         break\n",
    "# # #load carl's gene level analysis\n",
    "# # ls_pvalue = []\n",
    "# # ls_gene_correct = []\n",
    "# # cutoff = 0.1\n",
    "# # ls_file = [ 'SigGenesBySelectCohorts_AllSamples_MH_MinAltered10.xlsx',\n",
    "# #  'SigGenesBySelectCohorts_MetSamples_MH_MinAltered10.xlsx',\n",
    "# #  'SigGenesBySelectCohorts_PrimarySamples_MH_MinAltered10.xlsx',]\n",
    "# # for s_file in ls_file:\n",
    "# #     s_type = s_file.split('_')[1].split('Samples')[0]\n",
    "# #     df = pd.read_excel(f'data/{s_file}',index_col=0)\n",
    "# #     ls_col = df.columns[df.columns.str.contains('qVal')]\n",
    "# #     for s_col in ls_col:\n",
    "# #         df_sig = df.loc[df.loc[:,s_col] < cutoff,s_col]\n",
    "# #         if len(df_sig) > 0:\n",
    "# #             print(s_col)\n",
    "# #             s_score = s_col.split('_')[0]\n",
    "# #             for (index, value) in df_sig.items():\n",
    "# #                 print(index)\n",
    "# #                 s_gene = f'{index}_Altered_{s_type}'\n",
    "# #                 if s_type == 'All':\n",
    "# #                     ls = df_patient.columns[(df_patient.columns.str.contains(s_score)) & (df_patient.columns.str.contains(f'_all'))]\n",
    "# #                     s_gsva = f\"{ls[0].split('_all')[0]}_All\"\n",
    "# #                     df_patient[s_gsva] = df_patient.loc[:,ls[0]].fillna(df_patient.loc[:,ls[1]])\n",
    "# #                     s_gene = f'{index}_Altered'\n",
    "# #                 else:\n",
    "# #                     s_gsva = df_patient.columns[(df_patient.columns.str.contains(s_score)) & (df_patient.columns.str.contains(f'_{s_type}'))][0]\n",
    "# #                 b_rna = df_patient.loc[:,s_gsva].notna()\n",
    "# #                 try:\n",
    "# #                     b_dna = df_patient.loc[:,s_gene].notna()\n",
    "# #                 except:\n",
    "# #                     continue\n",
    "# #                 df_plot = df_patient.loc[b_rna & b_dna,[s_gene,s_gsva]].copy()\n",
    "# #                 df_plot['quartiles'] = pd.qcut(df_plot.loc[:,s_gsva],q=4,labels=['low','medX','med','high'])\n",
    "# #                 df_long = df_plot.loc[~df_plot.quartiles.str.contains('med'),[s_gene,'quartiles']]\n",
    "# #                 # average\n",
    "# #                 stat, pvalue = scipy.stats.fisher_exact(pd.crosstab(df_long.quartiles, df_long.loc[:,s_gene]))\n",
    "# #                 ls_gene_correct.append(s_gene)\n",
    "# #                 ls_pvalue.append(pvalue)\n",
    "# #                 #plot\n",
    "# #                 fig,ax=plt.subplots(figsize=(2,2),dpi=200)\n",
    "# #                 sns.pointplot(data=df_long,x='quartiles',y=s_gene,ax=ax,errorbar=None,order=['low','high'])\n",
    "# #                 ax.set_xlabel(s_score)\n",
    "# #                 #ax.set_title(f\"{index} {s_type}\\nFisher's exact\\nFDR={value:.2}\",fontsize=10)\n",
    "# #                 ax.set_ylabel(f'Fraction Pts n={len(df_long)}')\n",
    "# #                 ax.set_ylim(-0.1,1.1)\n",
    "# #                 plt.tight_layout()\n",
    "# #                 fig.savefig(f'figures/Fraction_{index}_{s_type}_{s_score}.png')\n",
    "# #                 break\n",
    "# #             break\n",
    "# #         break\n",
    "# #     break\n",
    "# # #correct pvalues\n",
    "# # __, corrected = statsmodels.stats.multitest.fdrcorrection(np.array(ls_pvalue))\n",
    "# # dict(zip(ls_gene_correct,corrected))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4f61928-9947-4044-885c-9e2c114a84ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "#dot plots - size of dot = FDR.Q\n",
    "# #generate more GSEA\n",
    "# ls_columns = ['NAME', 'SIZE', 'ES', 'NES', 'NOM.p.val', 'FDR.q.val', 'FWER.p.val',#'RANK.AT.MAX'\n",
    "#        ]\n",
    "# ls_gsea = ['Top4th_PurIST.Score_vs_Bottom4th_PurIST.Score_h.xlsx',\n",
    "#      'basal-like_vs_classical_h.xlsx','LiverCohort_vs_LungNotLiverCohort_h.xlsx',\n",
    "#  'Top4th_pSUB.1eNeg4_vs_Bottom4th_pSUB.1eNeg4_h.xlsx',\n",
    "#  'Top4th_pSUB.1eNeg5_vs_Bottom4th_pSUB.1eNeg5_h.xlsx',\n",
    "#  'Top4th_pORG.20_vs_Bottom4th_pORG.20_h.xlsx'\n",
    "#     ]\n",
    "# d_en = {}\n",
    "# for s_gsea in ls_gsea:\n",
    "#     d_load =  pd.read_excel(f'data/{s_gsea}',sheet_name=None)#,index_col=0\n",
    "#     df_up = d_load['GSEA_UP'].loc[:,ls_columns] #df[df.ES>0]\n",
    "#     df_down = d_load['GSEA_DN'].loc[:,ls_columns]#df[df.ES<0]\n",
    "#     d_en.update({f\"{s_gsea.split('_h.xlsx')[0]}_UP\":df_up})\n",
    "#     d_en.update({f\"{s_gsea.split('_h.xlsx')[0]}_DN\":df_down})\n",
    "#     #break\n",
    "\n",
    "# #subtype\n",
    "# #bubble plots\n",
    "# %matplotlib inline\n",
    "# sorter_combined = ['HALLMARK_XENOBIOTIC_METABOLISM','HALLMARK_PEROXISOME',\n",
    "#   'HALLMARK_FATTY_ACID_METABOLISM','HALLMARK_BILE_ACID_METABOLISM',\n",
    "#   'HALLMARK_PANCREAS_BETA_CELLS', 'HALLMARK_GLYCOLYSIS',\n",
    "#   'HALLMARK_HYPOXIA','HALLMARK_APICAL_JUNCTION',\n",
    "#     'HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION',]\n",
    "# sns.set_palette(('#000000','#CC79A7','#F0E442','#009E73'))\n",
    "# palette ={'PurIST_UP': '#000000',\n",
    "#  'PurIST_DN': '#CC79A7',\n",
    "#  'pSUB_UP': '#F0E442',\n",
    "#  'pSUB_DN': '#009E73'}\n",
    "# ls_plot_items = [ 'Top4th_PurIST.Score_vs_Bottom4th_PurIST.Score',\n",
    "#     'Top4th_pSUB.1eNeg4_vs_Bottom4th_pSUB.1eNeg4']\n",
    "# marker = 'o'\n",
    "# df_plot_long = pd.DataFrame()\n",
    "# for s_comp in ls_plot_items:\n",
    "#     for s_direction in ['UP','DN']:\n",
    "#         s_compare =f'{s_comp}_{s_direction}'\n",
    "#         df_plot_o = d_en[s_compare]\n",
    "#         df_plot_o['comparison'] = s_compare\n",
    "#         df_plot_long = pd.concat([df_plot_long,df_plot_o],ignore_index=True)\n",
    "# #plot\n",
    "# df_plot = df_plot_long[df_plot_long.NAME.isin(sorter_combined)]\n",
    "# d_rename = {'Top4th_pSUB.1eNeg4_vs_Bottom4th_pSUB.1eNeg4_UP':'pSUB_UP','Top4th_pSUB.1eNeg4_vs_Bottom4th_pSUB.1eNeg4_DN':'pSUB_DN',\n",
    "#        'Top4th_PurIST.Score_vs_Bottom4th_PurIST.Score_UP':'PurIST_UP','Top4th_PurIST.Score_vs_Bottom4th_PurIST.Score_DN':'PurIST_DN'}\n",
    "# df_plot['comparison'] = df_plot.comparison.replace(d_rename)\n",
    "# size_order =  ['<0.05','<0.1','<0.25','0.25+']\n",
    "# sizes = [100,50,25,10]\n",
    "# bins=[0,0.05,0.1,0.25,1]\n",
    "# d_size = dict(zip(size_order,sizes))\n",
    "# df_plot['FDR.Q'] = pd.cut(df_plot.loc[:,'FDR.q.val'],bins=bins,\n",
    "#                           right=True,labels=size_order)\n",
    "# fig, ax = plt.subplots(dpi=300,figsize=(3.2,3))\n",
    "# sns.scatterplot(data=df_plot, y='NAME',x='NES',hue='comparison',size='FDR.Q',ax=ax,linewidth=0,\n",
    "#                 marker=marker,alpha=0.8,size_order=size_order)\n",
    "# ax.set_title('Subtype vs. Hallmarks',fontweight='bold')\n",
    "# handles, labels = zip(*[(plt.scatter([], [],s=color,ec='k',linewidth=0,fc='k',marker=marker), key) for key, color in d_size.items()]) #,\n",
    "# legend1 = plt.legend(handles, labels, title=\"FDR.Q\",loc=2,bbox_to_anchor=(1,0.54)) #,markerscale=0.9\n",
    "# handles, labels = zip(*[(plt.scatter([], [], fc=color, marker=marker), key) for key, color in palette.items()])\n",
    "# legend2 = plt.legend(handles, labels, title=\"comparison\",bbox_to_anchor=(1,1.01),fontsize='small',loc=2)\n",
    "# ax.add_artist(legend1)\n",
    "# ax.add_artist(legend2)\n",
    "# ax.set_title('Subtype vs. Hallmarks',fontweight='bold')\n",
    "# ax.set_ylabel('')#HALLMARK\n",
    "# labels_long = ax.get_yticklabels().copy()\n",
    "# ticks = ax.get_yticks()\n",
    "# labels_ = [item.replace('HALLMARK_','').replace('_',' ').title().replace('Epithelial Mesenchymal Transition','EMT') for item in reversed(sorter_combined)]\n",
    "# ax.set_yticks(ticks=ticks,labels=labels_)\n",
    "# ax.set_xlim(-2.3,2.3)\n",
    "# plt.tight_layout()\n",
    "# fig.savefig('figures/GSEA_Subtype.png')\n",
    "\n",
    "\n",
    "# # organo\n",
    "# sorter_combined =  ['HALLMARK_MYOGENESIS' ,'HALLMARK_CHOLESTEROL_HOMEOSTASIS','HALLMARK_ANDROGEN_RESPONSE',\n",
    "#   'HALLMARK_OXIDATIVE_PHOSPHORYLATION', 'HALLMARK_DNA_REPAIR','HALLMARK_INTERFERON_ALPHA_RESPONSE',\n",
    "#   'HALLMARK_E2F_TARGETS','HALLMARK_MYC_TARGETS_V1','HALLMARK_G2M_CHECKPOINT','HALLMARK_MITOTIC_SPINDLE',\n",
    "#   'HALLMARK_GLYCOLYSIS','HALLMARK_MTORC1_SIGNALING','HALLMARK_PROTEIN_SECRETION',\n",
    "#                   ]\n",
    "# sns.set_palette(('#E69F00','#56B4E9','#D55E00','#0072B2'))\n",
    "# d_palette = {'Liver_UP': '#D55E00','Liver_DN': '#0072B2','pORG_UP': '#E69F00',\n",
    "#     'pORG_DN': '#56B4E9'}\n",
    "# ls_plot_items = ['Top4th_pORG.20_vs_Bottom4th_pORG.20',\n",
    "#     'LiverCohort_vs_LungNotLiverCohort']\n",
    "# df_plot_long = pd.DataFrame()\n",
    "# for s_comp in ls_plot_items:\n",
    "#     for s_direction in ['UP','DN']:\n",
    "#         s_compare =f'{s_comp}_{s_direction}'\n",
    "#         df_plot_o = d_en[s_compare]\n",
    "#         df_plot_o['comparison'] = s_compare\n",
    "#         df_plot_long = pd.concat([df_plot_long,df_plot_o])\n",
    "# #plot\n",
    "# df_plot = df_plot_long[df_plot_long.NAME.isin(sorter_combined)].copy()\n",
    "# d_rename = {'Top4th_pORG.20_vs_Bottom4th_pORG.20_UP':'pORG_UP','Top4th_pORG.20_vs_Bottom4th_pORG.20_DN':'pORG_DOWN',\n",
    "#        'LiverCohort_vs_LungNotLiverCohort_UP':'Liver_UP','LiverCohort_vs_LungNotLiverCohort_DN':'Liver_DN'}\n",
    "# df_plot['comparison'] = df_plot.comparison.replace(d_rename)\n",
    "# df_plot['FDR.Q'] = pd.cut(df_plot.loc[:,'FDR.q.val'],bins=bins,right=True,labels=size_order)\n",
    "# fig, ax = plt.subplots(dpi=300,figsize=(3.3,3))\n",
    "# sns.scatterplot(data=df_plot, y='NAME',x='NES',hue='comparison',size='FDR.Q',\n",
    "#                 ax=ax,size_order=size_order,sizes=sizes,marker=marker,linewidth=0)\n",
    "# handles1, labels1 = zip(*[(plt.scatter([], [],s=color,ec='k', fc='k',marker=marker,linewidth=0), key) for key, color in d_size.items()])\n",
    "# legend1 = plt.legend(handles1, labels1, title=\"FDR.Q\",bbox_to_anchor=(1,0.56))#,markerscale=0.95\n",
    "# handles2, labels2 = zip(*[(plt.scatter([], [], fc=color, marker=marker), key) for key, color in d_palette.items()])\n",
    "# legend2 = plt.legend(handles2, labels2, title=\"comparison\",bbox_to_anchor=(1,1.02),fontsize='small')\n",
    "# ax.add_artist(legend1)\n",
    "# ax.add_artist(legend2)\n",
    "# ax.set_title('Organotropism vs. Hallmarks',fontweight='bold')\n",
    "# ax.set_ylabel('')\n",
    "# ticks = ax.get_yticks()\n",
    "# ax.set_yticks(ticks=ticks,labels=[item.replace('HALLMARK_','').replace('_',' ').title().replace('Dna','DNA').replace('Myc','MYC') for item in reversed(sorter_combined)])\n",
    "# ax.set_xlim(-2.2,2.35)\n",
    "# plt.tight_layout()\n",
    "# fig.savefig('figures/GSEA_Organotropism.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d51df8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #organo\n",
    "# sorter_combined = ['ATM', 'ARID1A', 'ATR', 'BRCA1', 'BARD1', 'BRCA2', 'RAD51', 'CHEK1', 'PALB2']\n",
    "# sorter_combined = ['MCM3', 'MCM6', 'MCM7', 'MCM2', 'ORC2']\n",
    "# sorter_combined = ['CDKN1B', 'CDK4', 'E2F2', 'CCND1', 'CDK2', 'CDK6', 'E2F1']\n",
    "# sorter_combined = ['CDKN1B', 'CDK4', 'E2F2', 'CCND1', 'CDK2', 'CDK6', 'E2F1','MCM3', 'MCM6', 'MCM7', 'MCM2', 'ORC2']\n",
    "# ls_plot_items = ['pORG_Up_55','LiverVsLungNotLiver'] #['pSUB_Up_51', 'PurIST.Score']\n",
    "# ls_columns = ['NAME', 'NES', 'pVal', 'FDR.q.val']\n",
    "# #generate dataframe with comparisons\n",
    "# df_plot_bar,es_marker = util.compare_dataframe(ls_plot_items,d_en,sorter_combined,ls_columns)\n",
    "# #add colors\n",
    "# b_purist = df_plot_bar.comparison==ls_plot_items[1]#'PurIST.Score'\n",
    "# b_psub = df_plot_bar.comparison==ls_plot_items[0]#'pSUB_Up_51'\n",
    "# df_plot_bar.loc[b_purist & (df_plot_bar.direction=='UP'),'color'] = f'high {d_label[ls_plot_items[1]]}'\n",
    "# df_plot_bar.loc[b_purist & (df_plot_bar.direction=='DN'),'color'] = f'low {d_label[ls_plot_items[1]]}'\n",
    "# df_plot_bar.loc[b_psub & (df_plot_bar.direction=='UP'),'color'] = f'high {d_label[ls_plot_items[0]]}'\n",
    "# df_plot_bar.loc[b_psub & (df_plot_bar.direction=='DN'),'color'] = f'low {d_label[ls_plot_items[0]]}'\n",
    "# df_plot_bar['color'] = df_plot_bar.color.replace({'high Cohort':'Liver','low Cohort':'Lung'})\n",
    "# # plot figure\n",
    "# fig, ax = util.plot_double_bars(df_plot_bar,ls_plot_items,d_colorblind,\n",
    "#                                 sorter_combined,hatchbar,figsize=(4,3))\n",
    "# #custom legend\n",
    "# handles = []\n",
    "# for s_color in df_plot_bar[df_plot_bar.comparison==ls_plot_items[0]].color.unique():\n",
    "#     facecolor = d_colorblind[s_color]\n",
    "#     handles.append(mpl.patches.Patch(facecolor=facecolor, edgecolor='black',label=s_color,alpha=0.8,))\n",
    "# #hatch\n",
    "# for s_color in df_plot_bar[df_plot_bar.comparison==ls_plot_items[1]].color.unique():\n",
    "#     facecolor = d_colorblind[s_color]\n",
    "#     handles.append(mpl.patches.Patch(facecolor=facecolor, edgecolor='black',label=s_color,\n",
    "#                                      alpha=0.8,hatch=hatchbar))\n",
    "# ax.legend(handles=handles,bbox_to_anchor = (1.01,1.01),markerscale=1,title='NES',loc='upper left')\n",
    "# ax.set_yticklabels([item.replace('HALLMARK_','').replace('EPITHELIAL_MESENCHYMAL_TRANSITION','EMT') for item in sorter_combined])\n",
    "# #add pvalue axis\n",
    "# fig, ax2 = util.twin_pvalue_axis(df_plot_bar,ls_plot_items,sorter_combined,fig,ax,\n",
    "#                                  height=height,hatch=hatch)\n",
    "# ax2.set_xlim((-0.0092, 0.21))\n",
    "# plt.tight_layout()\n",
    "# #add FDRQ\n",
    "# util.add_fdrq_legend(texts= ['pORG','Cohort'],hatch=hatch,anchor=(1.01,0.4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bee194c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #pSUB and purist (side by side barplot and dot plot - motoyuki suggestion - looks okay, not used)\n",
    "# importlib.reload(util)\n",
    "# hatch = '//'#''#\n",
    "# hatchbar = '//'#''\n",
    "# height=0.4\n",
    "# sorter_combined = ['HALLMARK_XENOBIOTIC_METABOLISM','HALLMARK_PEROXISOME', 'HALLMARK_FATTY_ACID_METABOLISM',\n",
    "#   'HALLMARK_BILE_ACID_METABOLISM', 'HALLMARK_PANCREAS_BETA_CELLS','HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION',\n",
    "#   'HALLMARK_APICAL_JUNCTION', 'HALLMARK_HYPOXIA', 'HALLMARK_GLYCOLYSIS']\n",
    "# ls_plot_items = ['Top4th_pSUB.1eNeg4_vs_Bottom4th_pSUB.1eNeg4', 'Top4th_PurIST.Score_vs_Bottom4th_PurIST.Score']\n",
    "# #generate dataframe with comparisons\n",
    "# df_plot_bar,es_marker = util.compare_dataframe(ls_plot_items,d_en,sorter_combined,ls_columns)\n",
    "# #add colors\n",
    "# b_purist = df_plot_bar.comparison=='Top4th_PurIST.Score_vs_Bottom4th_PurIST.Score'\n",
    "# b_psub = df_plot_bar.comparison=='Top4th_pSUB.1eNeg4_vs_Bottom4th_pSUB.1eNeg4'\n",
    "# df_plot_bar.loc[b_purist & (df_plot_bar.direction=='UP'),'color'] = 'high PurIST'\n",
    "# df_plot_bar.loc[b_purist & (df_plot_bar.direction=='DN'),'color'] = 'low PurIST'\n",
    "# df_plot_bar.loc[b_psub & (df_plot_bar.direction=='UP'),'color'] = 'high pSUB'\n",
    "# df_plot_bar.loc[b_psub & (df_plot_bar.direction=='DN'),'color'] = 'low pSUB'\n",
    "# # plot figure\n",
    "# fig, ax1 = util.plot_double_bars_grid(df_plot_bar,ls_plot_items,d_colorblind,sorter_combined,\n",
    "#                                  hatchbar,figsize=(5.5,4))\n",
    "# #custom legend\n",
    "# handles = []\n",
    "# for s_color in ['high pSUB', 'low pSUB']:\n",
    "#     facecolor = d_colorblind[s_color]\n",
    "#     handles.append(mpl.patches.Patch(facecolor=facecolor, edgecolor='black',label=s_color,alpha=0.8,))\n",
    "# #hatch\n",
    "# for s_color in [ 'high PurIST', 'low PurIST']:\n",
    "#     facecolor = d_colorblind[s_color]\n",
    "#     handles.append(mpl.patches.Patch(facecolor=facecolor, edgecolor='black',label=s_color,\n",
    "#                                      alpha=0.8,hatch=hatchbar))\n",
    "\n",
    "# ax1.set_yticklabels([item.replace('HALLMARK_','').replace('EPITHELIAL_MESENCHYMAL_TRANSITION','EMT') for item in sorter_combined])\n",
    "# #add pvalue axis\n",
    "# fig, ax2 = util.add_pvalue_axis(df_plot_bar,ls_plot_items,sorter_combined,fig,ax1,\n",
    "#                                  height=height,hatch=hatch)\n",
    "# ax2.set_xlim((-0.1, 0.5))\n",
    "# legend1 = ax2.legend(handles=handles,bbox_to_anchor = (1.1,1),markerscale=1,title='NES',frameon=False)\n",
    "# ax2.axvline(0.2,linestyle='--',color='gray')\n",
    "# plt.tight_layout()\n",
    "# legend2 = util.add_fdrq_legend_ax2(texts= ['pSUB', 'PurIST'],hatch=hatch,ax2=ax2)\n",
    "# ax2.add_artist(legend1)\n",
    "# ax2.add_artist(legend2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba5219d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #per rosie GSVA question\n",
    "# what is the difference between pORG All and pORG Primary of Met\n",
    "# for s_met in ['Met','Primary']:\n",
    "#     df1 = df_patient.sort_values(by=f'pORG_0.2_{s_met}').loc[:,['Public_Patient_ID',f'pORG_0.2_{s_met}']].dropna()\n",
    "#     df2 = df_patient.sort_values(by=f'pORG_0.2_all{s_met}').loc[:,['Public_Patient_ID',f'pORG_0.2_all{s_met}']].dropna()\n",
    "#     df1.reset_index(inplace=True)\n",
    "#     df2.reset_index(inplace=True)\n",
    "#     df1.reset_index(inplace=True)\n",
    "#     df2.reset_index(inplace=True)\n",
    "#     df1.rename({'index':'INDEX','level_0':s_met},axis=1,inplace=True)\n",
    "#     df2.rename({'index':'INDEX'},axis=1,inplace=True)\n",
    "#     df1['All'] = df1.INDEX.map(dict(zip(df2.INDEX,df2.level_0)))\n",
    "#     df1[f'{s_met} quartiles'] = pd.qcut(df1.loc[:,s_met],4,labels = ['low','medx','med','high'])\n",
    "#     df1[f'{s_met} quartiles'] = df1.loc[:,f'{s_met} quartiles'].replace({'medx':'med',})#'med':np.nan\n",
    "#     #rank\n",
    "# #     fig, ax = plt.subplots()\n",
    "# #     sns.scatterplot(data=df1, x=s_met, y='All', hue=f'{s_met} quartiles',ax=ax)\n",
    "# #     ax.set_title(f'{s_met} pORG Rank')\n",
    "# #     ax.set_ylabel('Rank Order with GSVA All')\n",
    "# #     ax.set_xlabel(f'Rank Order with GSVA {s_met}')\n",
    "# #     for i_quantile in np.quantile(df1.All, q=[0.25,0.75]):\n",
    "# #         ax.axhline(i_quantile)\n",
    "#     #scatter\n",
    "#     fig, ax =plt.subplots()\n",
    "#     df_test = df1.merge(df2,on='Public_Patient_ID')\n",
    "#     sns.scatterplot(data=df_test,x=f'pORG_0.2_{s_met}',y=f'pORG_0.2_all{s_met}',hue=f'{s_met} quartiles',ax=ax)    \n",
    "#     for i_quantile in np.quantile(df_test.loc[:,f'pORG_0.2_all{s_met}'], q=[0.25,0.75]):\n",
    "#             ax.axhline(i_quantile)\n",
    "#     #kde plot\n",
    "#     fig,ax=plt.subplots()\n",
    "#     sns.kdeplot(df1.loc[:,f'pORG_0.2_{s_met}'],ax=ax,label=s_met,cut=0)\n",
    "#     sns.kdeplot(df2.loc[:,f'pORG_0.2_all{s_met}'],ax=ax,label='All',cut=0)\n",
    "#     plt.legend()\n",
    "#     ax.set_xlabel('pORG')\n",
    "#     #break\n",
    "# #all\n",
    "# fig, ax = plt.subplots()\n",
    "# sns.kdeplot(np.concatenate([df_patient.loc[:,f'pORG_0.2_allMet'].dropna().values,\n",
    "#            df_patient.loc[:,f'pORG_0.2_allPrimary'].dropna().values]),cut=0,ax=ax,\n",
    "#            label='pORG All')\n",
    "# plt.legend()\n",
    "# #add pORG quartiles, plot violins \n",
    "# # use pORG All\n",
    "# %matplotlib inline\n",
    "# sns.set_palette('tab10')\n",
    "# importlib.reload(util)\n",
    "# alpha = 0.05\n",
    "# b_correct= False#False #True #\n",
    "# ls_foci = ['Shannon_Entropy_Tumor','Templates_per_ng', 'Shannon_Entropy_Blood',\n",
    "#            'Productive_Rearrangements',\n",
    "#         'Simpsons_Diversity_Tumor','Simpsons_Diversity_Blood',\n",
    "#            'Clonality_Tumor', 'Clonality_Blood',\n",
    "#         \"Simpson's Evenness tumor\", \"Simpson's Evenness blood\",\n",
    "#             'Fraction Tumor Distinct TCRs',\n",
    "#            'Percent Tumor Distinct Clones',\n",
    "#             'Fraction Shared Clones 10'\n",
    "#         ]\n",
    "# d_colorblind = {'Liver':'#d55e00','Lung':'#0072b2',\n",
    "#                 'lung_short':'#d55e00','lung_long':'#0072b2',\n",
    "#                'high':'#e69f00','low': '#56b4e9',\n",
    "#                'basal-like':'#000000','classical':'#cc79a7',\n",
    "#                'high pSUB': '#f0e442','low pSUB':'#009E73'} #\n",
    "# #non- parametric\n",
    "# se_non_para = pd.Series(['Simpsons_Diversity_Blood', \n",
    "#     'Templates_per_ng','Productive_Rearrangements',\n",
    "#      \"Simpson's Evenness tumor\",\n",
    "#     \"Simpson's Evenness blood\",\n",
    "#    'Number Tumor Distinct Clones'])\n",
    "# d_order =  {#'Cohort':['Liver','Lung'],    \n",
    "#         'TCR_Lung_Median_Surv':['lung_short','lung_long'],\n",
    "#     'quartiles':['high','low']}\n",
    "# for s_porg in ['pORG_0.2_All', 'pORG_0.2_Met','pORG_0.2_Primary', ]:#, 'pORG_78_Primary',\n",
    "#     print('Primaries and Mets')\n",
    "#     df_whole = pd.DataFrame(columns=df_merge.columns)\n",
    "#     for s_all in ['pORG_0.2_allPrimary', 'pORG_0.2_allMet']:\n",
    "#         df_half = df_merge[(df_merge.loc[:,s_all].notna())]\n",
    "#         df_half = df_half[~df_half.Public_Patient_ID.isin(df_whole.Public_Patient_ID)].copy()\n",
    "#         df_half.rename({s_all:'pORG_0.2_All'},axis=1,inplace=True)\n",
    "#         print(len(df_half))\n",
    "#         df_whole = pd.concat([df_whole,df_half],axis=0,ignore_index=True)\n",
    "#     #s_foci not defined\n",
    "#     df_pri = df_merge.merge(df_whole.loc[df_whole.loc[:,['Simpsons_Diversity_Blood','Simpsons_Diversity_Tumor']].notna().any(axis=1),['Public_Patient_ID','pORG_0.2_All']],on='Public_Patient_ID',how='left')\n",
    "#     df_pri.loc[df_pri.Public_Patient_ID.duplicated(),'Cohort'] = np.nan\n",
    "#     if s_porg.find('Primar') > -1:\n",
    "#         print('Primaries')\n",
    "#         df_pri = df_pri.loc[(df_pri.Tumor_Type=='Primary')].copy()\n",
    "#     elif s_porg.find('Met') > -1:\n",
    "#         df_pri = df_pri.loc[df_pri.Tumor_Type=='Met'].copy()\n",
    "#     df_pri = util.add_quartiles(df_pri,'pORG_0.2_All')#s_porg#.drop(219) #p. rearrangements outlier \n",
    "#     for s_foci in ls_foci:\n",
    "#         print(s_foci)\n",
    "#         if se_non_para.isin([s_foci]).any():\n",
    "#             s_stats = 'non-parametric'\n",
    "#         else:\n",
    "#             s_stats = 'mean' \n",
    "#         df_both,d_pval,order = util.violin_stats2(df_pri,d_order,s_foci,s_stats)\n",
    "#         #util.qq_plot_hist(df_pri,s_cat,s_foci)  #anova eval\n",
    "#         fig,pvalues,corrected = util.plot_violins2(df_both,d_pval,d_order,s_stats,s_foci,order,d_colorblind,s_porg,b_correct=b_correct)#True#False\n",
    "#         #fig.savefig(f'figures/violinplot_both_{s_foci}_{list(d_order.keys())[-1]}_{s_porg}_{s_stats}.png')\n",
    "#         break\n",
    "#     break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c7dcadc-a9b1-44cd-80d8-08df0e7ba8e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# OLD #add pORG \n",
    "# ls_pORG = ['trim_padj_0.2_pORG_Up_55_Genes','Original_pORG_Up_78_Genes']\n",
    "# s_pSUB = 'Kallisto55_pSUB1e-04'\n",
    "# df_pORG = pd.read_csv(f'../data/20230608_GSVA_Scores.csv',index_col=0)\n",
    "# df_pORG.reset_index(drop=True,inplace=True)\n",
    "# df_pORG['Specimen_ID'] = [item.split('-')[-1] for item in df_pORG.Public_Specimen_ID]\n",
    "# df_pORG['Public_Patient_ID'] = [item.split('-')[0] + '-' + item.split('-')[1] for item in df_pORG.Public_Specimen_ID]\n",
    "# #drop 'T2'\n",
    "# df_t2 = df_pORG[df_pORG.Public_Specimen_ID.str.contains('-T2')]\n",
    "# ls_drop = df_pORG[df_pORG.Public_Specimen_ID.str.contains('-T2')].index\n",
    "# df_pORG = df_pORG.drop(ls_drop)\n",
    "# #add pri and met\n",
    "# for s_pORG in ls_pORG:\n",
    "#     for s_group in ['GSVA_Primary', 'GSVA_Met']: \n",
    "#         df_group = df_pORG.loc[df_pORG.Group==s_group,[s_pORG,'Public_Patient_ID']]\n",
    "#         df_group.rename({'trim_padj_0.2_pORG_Up_55_Genes':'pORG_0.2',\n",
    "#                         'Original_pORG_Up_78_Genes':'pORG_78'},axis=1,inplace=True)\n",
    "#         s_suffix = s_group.replace('GSVA','')\n",
    "#         df_patient = df_patient.merge(df_group,on='Public_Patient_ID',suffixes=('',s_suffix),how='left')\n",
    "# df_patient.rename({'pORG_0.2':'pORG_0.2_Primary',\n",
    "#                       'pORG_78':'pORG_78_Primary'},axis=1,inplace=True)\n",
    "\n",
    "# #add 'GSVA_All': must pivot\n",
    "# df_group = df_pORG.loc[df_pORG.Group=='GSVA_All',[s_pORG,'Public_Patient_ID','Specimen_ID']]\n",
    "# df_group = df_group.pivot(index='Public_Patient_ID',columns='Specimen_ID',values=s_pORG)\n",
    "# df_group['pORG_0.2_allPrimary'] = df_group.loc[:,'T'].fillna(df_group.F)\n",
    "# df_group['pORG_0.2_allMet'] = df_group.loc[:,'M'].fillna(df_group.M2)\n",
    "# df_patient = df_patient.merge(df_group.reset_index().loc[:,['Public_Patient_ID','pORG_0.2_allPrimary','pORG_0.2_allMet']],on='Public_Patient_ID',how='left')\n",
    "# print(len(df_patient))\n",
    "# #add T2 specimens, primary, all\n",
    "# d_t2 = dict(zip(df_t2[df_t2.Group=='GSVA_Primary'].Public_Patient_ID,df_t2.loc[df_t2.Group=='GSVA_Primary',s_pORG]))\n",
    "# df_patient['pORG_0.2_Primary_T2'] = df_patient.Public_Patient_ID.map(d_t2)\n",
    "# d_t2_all = dict(zip(df_t2[df_t2.Group=='GSVA_All'].Public_Patient_ID,df_t2.loc[df_t2.Group=='GSVA_All',s_pORG]))\n",
    "# df_patient['pORG_0.2_All_T2'] = df_patient.Public_Patient_ID.map(d_t2)\n",
    "\n",
    "# #label 10 matched samples\n",
    "# df_patient.loc[(~df_patient['pORG_0.2_allPrimary'].isna()) | (~df_patient['pORG_0.2_allMet'].isna()),'Matched_Primary_Met_RNA'] = False \n",
    "# df_patient.loc[(~df_patient['pORG_0.2_allPrimary'].isna()) & (~df_patient['pORG_0.2_allMet'].isna()),'Matched_Primary_Met_RNA'] = True\n",
    "# df_patient.Matched_Primary_Met_RNA.value_counts()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
